---
title: "SCENIC for Gene Regulatory Networks Analysis"
author: "Yichao Hua"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{SCENIC for Gene Regulatory Networks Analysis}
  %\VignetteEngine{knitr::rmarkdown}
  \usepackage[utf8]{inputenc}
---

## Table of Contents

1. [Importing SCENIC Loom Files into Seurat](#importing-scenic-loom-files-into-seurat)
2. [Visualizing SCENIC Results](#visualizing-scenic-results) 

## Importing SCENIC Loom Files into Seurat {#importing-scenic-loom-files-into-seurat}

[SCENIC](https://www.nature.com/articles/nmeth.4463) (Single-Cell Regulatory Network Inference and Clustering) is a computational method that provides deep insights into the regulatory networks governing gene expression in single cells. It is highly recommended to use the Nextflow pipeline to run SCENIC, which can be found [here](https://github.com/aertslab/SCENICprotocol/). This process requires a loom file as input, which can be generated directly using the `Seu2Loom()` function. Currently, the `SeuratExtend` package does not integrate the `RunScenic` functionality directly (requiring the use of Nextflow command line), but if there is user demand, this could be considered for future updates. The workflow results in a file named "pyscenic_integrated-output.loom," which includes a list of transcription factors (TFs) and their regulated genes, as well as a TF-cell matrix representing the AUCell values. The AUCell score represents the enrichment score of all genes regulated by a TF, indicating regulon activity. The `ImportPyscenicLoom()` function allows for the import of SCENIC-generated loom files into Seurat objects, facilitating further analysis and visualization within the Seurat framework.

As an example, we use a pre-computed SCENIC loom file which can be downloaded as follows:

```{r, message=FALSE, warning=FALSE}
library(SeuratExtend)
scenic_loom_path <- file.path(tempdir(), "pyscenic_integrated-output.loom")
download.file("https://zenodo.org/records/10944066/files/pbmc3k_small_pyscenic_integrated-output.loom",
              scenic_loom_path, mode = "wb")  # Use binary mode for Windows compatibility

# Importing SCENIC Loom Files into Seurat
pbmc <- ImportPyscenicLoom(scenic_loom_path, seu = pbmc)
```

If you prefer to import SCENIC results without specifying a Seurat object:

```{r, message=FALSE, warning=FALSE}
# Importing SCENIC results without an existing Seurat object
scenic_output <- ImportPyscenicLoom(scenic_loom_path)
```

### Examining SCENIC Outputs

SCENIC results are stored in `seu@misc$SCENIC`, which includes `seu@misc$SCENIC$Regulons`, a list where each element's name is a TF name and the value is a list of genes it regulates. Additionally, `seu@misc$SCENIC$RegulonsAUC`, a TF-cell AUCell matrix, is also loaded into the "TF" assay of the Seurat object, making the manipulation of TF regulon activity as straightforward as handling gene expression data.

```{r}
# Viewing the outputs
tf_auc <- pbmc@misc$SCENIC$RegulonsAUC
head(tf_auc, 4:5)

tf_gene_list <- pbmc@misc$SCENIC$Regulons
head(tf_gene_list, 5)
```

## Visualizing SCENIC Results {#visualizing-scenic-results}

Once SCENIC data is integrated into a Seurat object, users can leverage a variety of visualization tools provided in the **Enhanced Visualization** section to explore and interpret these regulatory networks. Both the extracted `tf_auc` matrix or the Seurat object itself can be used as inputs. Here are some practical examples:

### Identifying Top Activated TFs in Each Cluster

```{r, fig.width=5, fig.height=7, message=FALSE, warning=FALSE}
tf_zscore <- CalcStats(tf_auc, f = pbmc$cluster, order = "p", n = 4, t = TRUE)
Heatmap(tf_zscore, lab_fill = "zscore")
```

### Comparing TF Gene Expression Levels and Regulon Activity (AUCell)

Since we have imported SCENIC results into the "TF" assay, we can easily access the corresponding AUCell values by prefixing "tf_" to the TF name:

```{r, fig.width=7, fig.height=5.5}
DimPlot2(
  pbmc,
  features = c("ETS1", "ATF3", "tf_ETS1", "tf_ATF3"),
  cols = list("tf_ETS1" = "D", "tf_ATF3" = "D"),
  theme = NoAxes()
)
```

### Simplifying Regulon Activity Access by Setting Default Assay

If you find manually adding "tf_" to each transcription factor cumbersome, you can set the default assay to "TF", which simplifies operations involving regulon activity. For example, to create a waterfall plot that compares the regulon activity between two cell types, you can do the following:

```{r, fig.width=8, fig.height=4}
# Setting the default assay to "TF" for easier access to regulon activity
DefaultAssay(pbmc) <- "TF"

# Creating a waterfall plot to compare regulon activity between monocytes and CD8 T cells
WaterfallPlot(
  pbmc,
  features = rownames(pbmc),  # Using all available TFs in the "TF" assay
  ident.1 = "Mono CD14",      # First group of cells
  ident.2 = "CD8 T cell",     # Second group of cells
  exp.transform = FALSE,      # Disable transformation of expression data
  top.n = 20                  # Display the top 20 most differentially active TFs
)
```

These examples illustrate how to integrate and utilize SCENIC analysis within the Seurat framework, providing a comprehensive approach to understanding gene regulatory mechanisms at the single-cell level.

```{r}
sessionInfo()
```
