<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Frequently Asked Questions (FAQ) • SeuratExtend</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Frequently Asked Questions (FAQ)">
<meta property="og:description" content="SeuratExtend">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">SeuratExtend</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/News.html">What's New in v1.1.0</a>
    </li>
    <li class="divider">
    </li>
<li>
      <a href="../articles/Visualization.html">Enhanced Visualization</a>
    </li>
    <li>
      <a href="../articles/GSEA.html">Geneset Enrichment Analysis (GSEA) </a>
    </li>
    <li>
      <a href="../articles/Trajectory.html">Trajectory and Pseudotime Analysis </a>
    </li>
    <li>
      <a href="../articles/SCENIC.html">SCENIC for Gene Regulatory Networks Analysis </a>
    </li>
    <li>
      <a href="../articles/Utilities.html">Utility Tools and Functions</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    scRNA-seq Course
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/single-cell-course/1.basic-R.html">1. Introduction to R</a>
    </li>
    <li>
      <a href="../articles/single-cell-course/2.Seurat.html">2. Basic Single-Cell Analysis Workflow</a>
    </li>
    <li>
      <a href="../articles/single-cell-course/3.Visualization.html">3. Advanced Visualization Techniques</a>
    </li>
    <li>
      <a href="../articles/single-cell-course/4.GSEA.html">4. Gene Set Enrichment Analysis</a>
    </li>
    <li>
      <a href="../articles/single-cell-course/5.Core-Enhancement.html">5. Core scRNA-seq Workflow Enhancements</a>
    </li>
    <li>
      <a href="../articles/single-cell-course/6.Advanced.html">6. Advanced Methods (Part 1)</a>
    </li>
    <li>
      <a href="../articles/single-cell-course/6.Advanced-2.html">7. Advanced Methods (Part 2)</a>
    </li>
  </ul>
</li>
<li>
  <a href="../articles/FAQ.html">FAQ</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Frequently Asked Questions (FAQ)</h1>
                        <h4 data-toc-skip class="author">Yichao Hua</h4>
            
            <h4 data-toc-skip class="date">2024-11-28</h4>
      
      
      <div class="hidden name"><code>FAQ.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="table-of-contents">Table of Contents<a class="anchor" aria-label="anchor" href="#table-of-contents"></a>
</h2>
<ol style="list-style-type: decimal">
<li><a href="#scvelo-linux">Running scVelo Functions in RStudio on
Linux</a></li>
<li><a href="#conda-env">Creating ‘seuratextend’ Conda
Environment</a></li>
<li><a href="#seurat-v5-merge">Merging Multiple Samples with Seurat
v5.0</a></li>
</ol>
</div>
<div class="section level2">
<h2 id="scvelo-linux">Running scVelo Functions in RStudio on Linux<a class="anchor" aria-label="anchor" href="#scvelo-linux"></a>
</h2>
<div class="section level3">
<h3 id="q-im-getting-an-error-when-running-scvelo-related-functions-in-rstudio-on-linux--what-should-i-do">Q: I’m getting an error when running scVelo-related functions in
RStudio on Linux. What should I do?<a class="anchor" aria-label="anchor" href="#q-im-getting-an-error-when-running-scvelo-related-functions-in-rstudio-on-linux--what-should-i-do"></a>
</h3>
<p>When running scVelo-related functions in RStudio on a Linux system,
you might encounter an error similar to this:</p>
<pre><code>Error in py_run_string_impl(code, local, convert) : 
  ImportError: /usr/lib/x86_64-linux-gnu/libstdc++.so.6: version `GLIBCXX_3.4.29' not found (required by /home/username/.local/share/r-miniconda/envs/seuratextend/lib/python3.10/site-packages/pandas/_libs/window/aggregations.cpython-310-x86_64-linux-gnu.so)</code></pre>
<p>This error occurs because RStudio can’t find the correct shared
library ‘libstdc++’. Even though a higher version of libstdc++.so.6.0.32
exists in the ‘seuratextend’ conda environment, RStudio can’t locate
it.</p>
</div>
<div class="section level3">
<h3 id="solution">Solution:<a class="anchor" aria-label="anchor" href="#solution"></a>
</h3>
<p>RStudio searches for shared libraries in the paths specified by the
<code>LD_LIBRARY_PATH</code> environment variable. You can check these
paths with:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Sys.getenv.html" class="external-link">Sys.getenv</a></span><span class="op">(</span><span class="st">"LD_LIBRARY_PATH"</span><span class="op">)</span></span></code></pre></div>
<p>To resolve this issue, manually copy the following files from
<code>/home/username/.local/share/r-miniconda/envs/seuratextend/lib/</code>
to one of the paths in <code>LD_LIBRARY_PATH</code> (e.g.,
<code>/usr/lib/R/lib</code>, which may require sudo permissions):</p>
<ul>
<li>libR.so</li>
<li>libstdc++.so</li>
<li>libstdc++.so.6</li>
<li>libstdc++.so.6.0.32</li>
</ul>
<p>After copying these files, restart your R session, and the issue
should be resolved.</p>
</div>
</div>
<div class="section level2">
<h2 id="conda-env">Creating ‘seuratextend’ Conda Environment<a class="anchor" aria-label="anchor" href="#conda-env"></a>
</h2>
<div class="section level3">
<h3 id="q-the-create_condaenv_seuratextend-function-is-failing-to-create-the-seuratextend-conda-environment--what-could-be-the-problem">Q: The <code>create_condaenv_seuratextend()</code> function is
failing to create the ‘seuratextend’ conda environment. What could be
the problem?<a class="anchor" aria-label="anchor" href="#q-the-create_condaenv_seuratextend-function-is-failing-to-create-the-seuratextend-conda-environment--what-could-be-the-problem"></a>
</h3>
<p>If <code><a href="../reference/create_condaenv_seuratextend.html">create_condaenv_seuratextend()</a></code> is failing to create
the ‘seuratextend’ conda environment, it might be due to the system not
finding conda or git.</p>
</div>
<div class="section level3">
<h3 id="solutions">Solutions:<a class="anchor" aria-label="anchor" href="#solutions"></a>
</h3>
<ol style="list-style-type: decimal">
<li>
<p>If conda is not found, you can install miniconda using:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">reticulate</span><span class="fu">::</span><span class="fu"><a href="https://rstudio.github.io/reticulate/reference/install_miniconda.html" class="external-link">install_miniconda</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
</li>
<li><p>If git is not found, download and install it from the official
website: <a href="https://git-scm.com/" class="external-link uri">https://git-scm.com/</a>. Choose the appropriate version for
your operating system.</p></li>
</ol>
<p>After installing conda or git, restart your R session and try running
<code><a href="../reference/create_condaenv_seuratextend.html">create_condaenv_seuratextend()</a></code> again.</p>
</div>
</div>
<div class="section level2">
<h2 id="seurat-v5-merge">Merging Multiple Samples with Seurat v5.0<a class="anchor" aria-label="anchor" href="#seurat-v5-merge"></a>
</h2>
<div class="section level3">
<h3 id="q-after-merging-multiple-samples-using-the-merge-function-in-seurat-v5-0-im-getting-an-error-when-running-certain-functions--how-can-i-fix-this">Q: After merging multiple samples using the <code>merge()</code>
function in Seurat v5.0, I’m getting an error when running certain
functions. How can I fix this?<a class="anchor" aria-label="anchor" href="#q-after-merging-multiple-samples-using-the-merge-function-in-seurat-v5-0-im-getting-an-error-when-running-certain-functions--how-can-i-fix-this"></a>
</h3>
<p>When using Seurat v5.0, you might encounter the following error after
merging multiple samples and trying to run certain functions:</p>
<pre><code>Error in `GetAssayData()`:
! GetAssayData doesn't work for multiple layers in v5 assay.</code></pre>
<p>This error occurs because in Seurat v5.0, the <code><a href="https://rdrr.io/r/base/merge.html" class="external-link">merge()</a></code>
function creates separate count layers for each sample by default. This
prevents <code>GetAssayData()</code> from extracting the matrix.</p>
</div>
<div class="section level3">
<h3 id="solution-1">Solution:<a class="anchor" aria-label="anchor" href="#solution-1"></a>
</h3>
<p>To resolve this issue, you need to join the layers before using
<code>GetAssayData()</code>. Use the following code:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seu</span> <span class="op">&lt;-</span> <span class="fu">JoinLayers</span><span class="op">(</span><span class="va">seu</span><span class="op">)</span></span></code></pre></div>
<p>After joining the layers, you should be able to use
<code>GetAssayData()</code> without errors.</p>
<p>Note that this issue does not occur in Seurat v4 and earlier
versions.</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Yichao Hua.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
