<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Geneset Enrichment Analysis (GSEA) • SeuratExtend</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Geneset Enrichment Analysis (GSEA)">
<meta property="og:description" content="SeuratExtend">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">SeuratExtend</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/News.html">What's New in v1.1.0</a>
    </li>
    <li class="divider">
    </li>
<li>
      <a href="../articles/Visualization.html">Enhanced Visualization</a>
    </li>
    <li>
      <a href="../articles/GSEA.html">Geneset Enrichment Analysis (GSEA) </a>
    </li>
    <li>
      <a href="../articles/Trajectory.html">Trajectory and Pseudotime Analysis </a>
    </li>
    <li>
      <a href="../articles/SCENIC.html">SCENIC for Gene Regulatory Networks Analysis </a>
    </li>
    <li>
      <a href="../articles/Utilities.html">Utility Tools and Functions</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    scRNA-seq Course
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/single-cell-course/1.basic-R.html">1. Introduction to R</a>
    </li>
    <li>
      <a href="../articles/single-cell-course/2.Seurat.html">2. Basic Single-Cell Analysis Workflow</a>
    </li>
    <li>
      <a href="../articles/single-cell-course/3.Visualization.html">3. Advanced Visualization Techniques</a>
    </li>
    <li>
      <a href="../articles/single-cell-course/4.GSEA.html">4. Gene Set Enrichment Analysis</a>
    </li>
    <li>
      <a href="../articles/single-cell-course/5.Core-Enhancement.html">5. Core scRNA-seq Workflow Enhancements</a>
    </li>
    <li>
      <a href="../articles/single-cell-course/6.Advanced.html">6. Advanced Methods (Part 1)</a>
    </li>
    <li>
      <a href="../articles/single-cell-course/6.Advanced-2.html">7. Advanced Methods (Part 2)</a>
    </li>
  </ul>
</li>
<li>
  <a href="../articles/FAQ.html">FAQ</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Geneset Enrichment Analysis (GSEA)</h1>
                        <h4 data-toc-skip class="author">Yichao Hua</h4>
            
            <h4 data-toc-skip class="date">2024-11-28</h4>
      
      
      <div class="hidden name"><code>GSEA.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="table-of-contents">Table of Contents<a class="anchor" aria-label="anchor" href="#table-of-contents"></a>
</h2>
<ol style="list-style-type: decimal">
<li><a href="#conduct-gsea-using-the-go-or-reactome-database">Conduct
GSEA using the GO or Reactome database</a></li>
<li><a href="#perform-gsea-using-customized-genesets">Perform GSEA using
customized genesets</a></li>
<li><a href="#find-pathways-in-the-goreactome-database-or-customized-genesets">Find
pathways in the GO/Reactome database or customized genesets</a></li>
<li><a href="#convert-goreactome-pathway-ids-to-pathway-names">Convert
GO/Reactome pathway IDs to pathway names</a></li>
<li><a href="#filter-the-goreactome-pathway-list-based-on-certain-criteria">Filter
the GO/Reactome pathway list based on certain criteria</a></li>
<li><a href="#create-a-gsea-plot-emulating-the-broad-institute-analysis">Create
a GSEA plot emulating the Broad Institute analysis</a></li>
</ol>
</div>
<div class="section level2">
<h2 id="conduct-gsea-using-the-go-or-reactome-database">Conduct GSEA using the GO or Reactome database<a class="anchor" aria-label="anchor" href="#conduct-gsea-using-the-go-or-reactome-database"></a>
</h2>
<p>The <code>SeuratExtend</code> package integrates both the GO and
Reactome databases, streamlining the GSEA analysis process. This is
primarily facilitated through the <code>GeneSetAnalysisGO</code> and
<code>GeneSetAnalysisReactome</code> functions, among other
supplementary functions. In this section, we’ll delve into the usage and
features of these functions.</p>
<div class="section level3">
<h3 id="gene-ontology-go-database">Gene Ontology (GO) Database<a class="anchor" aria-label="anchor" href="#gene-ontology-go-database"></a>
</h3>
<p>Performing GSEA using the GO database can be resource-intensive due
to its extensive size. To make the analysis more feasible, you might
consider evaluating pathways under specific categories. For instance, in
the example below, only the pathways under the “immune_system_process”
category are evaluated. The results from this analysis are saved in the
location: <code>seu@misc$AUCell$GO[[title]]</code>.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SeuratExtend</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">options</a></span><span class="op">(</span>max.print <span class="op">=</span> <span class="fl">12</span>, spe <span class="op">=</span> <span class="st">"human"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/GeneSetAnalysis.html">GeneSetAnalysisGO</a></span><span class="op">(</span><span class="va">pbmc</span>, parent <span class="op">=</span> <span class="st">"immune_system_process"</span>, nCores <span class="op">=</span> <span class="fl">4</span><span class="op">)</span> <span class="co"># calculating with 4 cores</span></span>
<span><span class="va">matr</span> <span class="op">&lt;-</span> <span class="va">pbmc</span><span class="op">@</span><span class="va">misc</span><span class="op">$</span><span class="va">AUCell</span><span class="op">$</span><span class="va">GO</span><span class="op">$</span><span class="va">immune_system_process</span></span>
<span><span class="va">matr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RenameGO.html">RenameGO</a></span><span class="op">(</span><span class="va">matr</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">matr</span>, <span class="fl">2</span><span class="op">:</span><span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                                           cells</span></span>
<span><span class="co">## gene sets                                  CTATAAGATCGTTT-1 GTGATTCTGGTTCA-1</span></span>
<span><span class="co">##   GO:0002376 immune system process (3213g)       0.25540633       0.29013621</span></span>
<span><span class="co">##   GO:0001776 leukocyte homeostasis (84g)         0.06705023       0.05466948</span></span>
<span><span class="co">##                                           cells</span></span>
<span><span class="co">## gene sets                                  ACGTTGGACCGTAA-1</span></span>
<span><span class="co">##   GO:0002376 immune system process (3213g)       0.30090373</span></span>
<span><span class="co">##   GO:0001776 leukocyte homeostasis (84g)         0.05634452</span></span></code></pre>
<p>For the “parent” argument, you can input any term from the GO
database, be it a GO ID or a pathway name. To get a glimpse of commonly
used GO categories, you can run <code><a href="../reference/GeneSetAnalysis.html">GeneSetAnalysisGO()</a></code> without
any arguments:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/GeneSetAnalysis.html">GeneSetAnalysisGO</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                 immune_system_process                  response_to_stimulus </span></span>
<span><span class="co">##                          "GO:0002376"                          "GO:0050896" </span></span>
<span><span class="co">##                             signaling                     metabolic_process </span></span>
<span><span class="co">##                          "GO:0023052"                          "GO:0008152" </span></span>
<span><span class="co">## regulation_of_vasculature_development                   signal_transduction </span></span>
<span><span class="co">##                          "GO:1901342"                          "GO:0007165"</span></span></code></pre>
<p>Here are some suggested visualization methods:</p>
<ul>
<li>
<strong>Heatmap</strong>: Useful for comparing multiple groups with
numerous features but provides a less detailed representation.</li>
</ul>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/Heatmap.html">Heatmap</a></span><span class="op">(</span><span class="fu"><a href="../reference/CalcStats.html">CalcStats</a></span><span class="op">(</span><span class="va">matr</span>, f <span class="op">=</span> <span class="va">pbmc</span><span class="op">$</span><span class="va">cluster</span>, order <span class="op">=</span> <span class="st">"p"</span>, n <span class="op">=</span> <span class="fl">4</span><span class="op">)</span>, lab_fill <span class="op">=</span> <span class="st">"zscore"</span><span class="op">)</span></span></code></pre></div>
<p><img src="GSEA_files/figure-html/unnamed-chunk-3-1.png" width="1104"></p>
<ul>
<li>
<strong>Violin Plot</strong>: Best suited for comparing multiple
groups with fewer features while presenting more details for individual
data points.</li>
</ul>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/VlnPlot2.html">VlnPlot2</a></span><span class="op">(</span><span class="va">matr</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span>,<span class="op">]</span>, f <span class="op">=</span> <span class="va">pbmc</span><span class="op">$</span><span class="va">cluster</span>, ncol <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<p><img src="GSEA_files/figure-html/unnamed-chunk-4-1.png" width="480"></p>
<ul>
<li>
<strong>Waterfall Plot</strong>: Ideal for comparing two
groups.</li>
</ul>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/WaterfallPlot.html">WaterfallPlot</a></span><span class="op">(</span><span class="va">matr</span>, f <span class="op">=</span> <span class="va">pbmc</span><span class="op">$</span><span class="va">cluster</span>, ident.1 <span class="op">=</span> <span class="st">"B cell"</span>, ident.2 <span class="op">=</span> <span class="st">"CD8 T cell"</span>, top.n <span class="op">=</span> <span class="fl">20</span><span class="op">)</span></span></code></pre></div>
<p><img src="GSEA_files/figure-html/unnamed-chunk-5-1.png" width="1008"></p>
</div>
<div class="section level3">
<h3 id="reactome-database">Reactome Database<a class="anchor" aria-label="anchor" href="#reactome-database"></a>
</h3>
<p>For GSEA using the Reactome database, consider assessing pathways
under certain categories to make the process more manageable. The
example below evaluates pathways under the “Immune System” category.
Results from this analysis are saved under:
<code>seu@misc$AUCell$Reactome[[title]]</code>.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/GeneSetAnalysis.html">GeneSetAnalysisReactome</a></span><span class="op">(</span><span class="va">pbmc</span>, parent <span class="op">=</span> <span class="st">"Immune System"</span><span class="op">)</span></span>
<span><span class="va">matr</span> <span class="op">&lt;-</span> <span class="va">pbmc</span><span class="op">@</span><span class="va">misc</span><span class="op">$</span><span class="va">AUCell</span><span class="op">$</span><span class="va">Reactome</span><span class="op">$</span><span class="va">`Immune System`</span></span>
<span><span class="va">matr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RenameGO.html">RenameReactome</a></span><span class="op">(</span><span class="va">matr</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/Heatmap.html">Heatmap</a></span><span class="op">(</span><span class="fu"><a href="../reference/CalcStats.html">CalcStats</a></span><span class="op">(</span><span class="va">matr</span>, f <span class="op">=</span> <span class="va">pbmc</span><span class="op">$</span><span class="va">cluster</span>, order <span class="op">=</span> <span class="st">"p"</span>, n <span class="op">=</span> <span class="fl">4</span><span class="op">)</span>, lab_fill <span class="op">=</span> <span class="st">"zscore"</span><span class="op">)</span></span></code></pre></div>
<p><img src="GSEA_files/figure-html/unnamed-chunk-6-1.png" width="1008"></p>
<p>Similar to the GO database, running
<code><a href="../reference/GeneSetAnalysis.html">GeneSetAnalysisReactome()</a></code> without any arguments lets you
view commonly used categories in the Reactome database:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/GeneSetAnalysis.html">GeneSetAnalysisReactome</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                           R-HSA-109582                           R-HSA-112316 </span></span>
<span><span class="co">##                           "Hemostasis"                      "Neuronal System" </span></span>
<span><span class="co">##                          R-HSA-1266738                          R-HSA-1430728 </span></span>
<span><span class="co">##                "Developmental Biology"                           "Metabolism" </span></span>
<span><span class="co">##                          R-HSA-1474165                          R-HSA-1474244 </span></span>
<span><span class="co">##                         "Reproduction"    "Extracellular matrix organization" </span></span>
<span><span class="co">##                          R-HSA-1500931                           R-HSA-162582 </span></span>
<span><span class="co">##              "Cell-Cell communication"                  "Signal Transduction" </span></span>
<span><span class="co">##                          R-HSA-1640170                          R-HSA-1643685 </span></span>
<span><span class="co">##                           "Cell Cycle"                              "Disease" </span></span>
<span><span class="co">##                           R-HSA-168256                          R-HSA-1852241 </span></span>
<span><span class="co">##                        "Immune System" "Organelle biogenesis and maintenance" </span></span>
<span><span class="co">##  [ reached getOption("max.print") -- omitted 15 entries ]</span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="perform-gsea-using-customized-genesets">Perform GSEA using customized genesets<a class="anchor" aria-label="anchor" href="#perform-gsea-using-customized-genesets"></a>
</h2>
<p>To conduct a Gene Set Enrichment Analysis (GSEA) with custom gene
sets, the <code>GeneSetAnalysis</code> function is the tool of choice.
For instance, one might consider utilizing the <a href="https://www.gsea-msigdb.org/gsea/msigdb/human/genesets.jsp?collection=H" class="external-link">Hallmark
50</a> gene set, commonly employed for general screening. This set can
be accessed via the <code>hall50</code> object. Upon execution, the
resulting AUCell matrix will be stored under the path:
<code>seu@misc$AUCell[[title]]</code>.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/GeneSetAnalysis.html">GeneSetAnalysis</a></span><span class="op">(</span><span class="va">pbmc</span>, genesets <span class="op">=</span> <span class="va">hall50</span><span class="op">$</span><span class="va">human</span><span class="op">)</span></span>
<span><span class="va">matr</span> <span class="op">&lt;-</span> <span class="va">pbmc</span><span class="op">@</span><span class="va">misc</span><span class="op">$</span><span class="va">AUCell</span><span class="op">$</span><span class="va">genesets</span></span>
<span><span class="fu"><a href="../reference/Heatmap.html">Heatmap</a></span><span class="op">(</span><span class="fu"><a href="../reference/CalcStats.html">CalcStats</a></span><span class="op">(</span><span class="va">matr</span>, f <span class="op">=</span> <span class="va">pbmc</span><span class="op">$</span><span class="va">cluster</span><span class="op">)</span>, lab_fill <span class="op">=</span> <span class="st">"zscore"</span><span class="op">)</span></span></code></pre></div>
<p><img src="GSEA_files/figure-html/unnamed-chunk-8-1.png" width="720"></p>
<p>For those seeking a plethora of other gene sets, the
<code>SeuratExtendData::Genesets_data</code> offers an expansive
collection sourced from the <a href="https://www.gsea-msigdb.org/gsea/msigdb/human/collections.jsp" class="external-link">GSEA
MSigDB website</a>. Here’s how you can view the available
collections:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu">SeuratExtendData</span><span class="fu">::</span><span class="va">Genesets_data</span><span class="op">$</span><span class="va">human</span><span class="op">$</span><span class="va">GSEA</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] "positional gene sets"                </span></span>
<span><span class="co">##  [2] "all curated gene sets"               </span></span>
<span><span class="co">##  [3] "chemical and genetic perturbations"  </span></span>
<span><span class="co">##  [4] "BioCarta gene sets"                  </span></span>
<span><span class="co">##  [5] "KEGG gene sets"                      </span></span>
<span><span class="co">##  [6] "PID gene sets"                       </span></span>
<span><span class="co">##  [7] "all canonical pathways"              </span></span>
<span><span class="co">##  [8] "all motif gene sets"                 </span></span>
<span><span class="co">##  [9] "transcription factor targets"        </span></span>
<span><span class="co">## [10] "all computational gene sets"         </span></span>
<span><span class="co">## [11] "all immunologic signatures gene sets"</span></span>
<span><span class="co">## [12] "hallmark gene sets"</span></span></code></pre>
<p>Furthermore, for cluster annotations, the
<code>SeuratExtend::PanglaoDB_data</code> contains a valuable resource:
marker lists for 178 distinct cell types, curated from <a href="https://panglaodb.se/markers.html" class="external-link">PanglaoDB</a>. To explore these
marker lists:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu">SeuratExtend</span><span class="fu">::</span><span class="va">PanglaoDB_data</span><span class="op">$</span><span class="va">marker_list_human</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] "Acinar cells"                   "Adipocyte progenitor cells"    </span></span>
<span><span class="co">##  [3] "Adipocytes"                     "Adrenergic neurons"            </span></span>
<span><span class="co">##  [5] "Airway epithelial cells"        "Airway goblet cells"           </span></span>
<span><span class="co">##  [7] "Airway smooth muscle cells"     "Alpha cells"                   </span></span>
<span><span class="co">##  [9] "Alveolar macrophages"           "Anterior pituitary gland cells"</span></span>
<span><span class="co">## [11] "Astrocytes"                     "B cells"                       </span></span>
<span><span class="co">##  [ reached getOption("max.print") -- omitted 166 entries ]</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="find-pathways-in-the-goreactome-database-or-customized-genesets">Find pathways in the GO/Reactome database or customized
genesets<a class="anchor" aria-label="anchor" href="#find-pathways-in-the-goreactome-database-or-customized-genesets"></a>
</h2>
<p>Navigating the plethora of pathways in databases like GO and Reactome
can be overwhelming. The <code>SearchDatabase</code> function simplifies
this process by offering a wide array of customizable search
parameters.</p>
<div class="section level3">
<h3 id="general-search">General Search<a class="anchor" aria-label="anchor" href="#general-search"></a>
</h3>
<p>The ‘item’ parameter is highly versatile, allowing you to search by
gene name, pathway ID, or even keywords within pathway names. The
following example demonstrates how to find pathways containing the gene
“CD3D” or pathways with names including “metabolic.”</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">result</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SearchDatabase.html">SearchDatabase</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD3D"</span>, <span class="st">"metabolic"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">result</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] "GO:0000023 maltose metabolic process (3g)"                                  </span></span>
<span><span class="co">##  [2] "GO:0000038 very long-chain fatty acid metabolic process (31g)"              </span></span>
<span><span class="co">##  [3] "GO:0000052 citrulline metabolic process (8g)"                               </span></span>
<span><span class="co">##  [4] "GO:0000053 argininosuccinate metabolic process (1g)"                        </span></span>
<span><span class="co">##  [5] "GO:0000096 sulfur amino acid metabolic process (43g)"                       </span></span>
<span><span class="co">##  [6] "GO:0000255 allantoin metabolic process (2g)"                                </span></span>
<span><span class="co">##  [7] "GO:0000270 peptidoglycan metabolic process (6g)"                            </span></span>
<span><span class="co">##  [8] "GO:0000820 regulation of glutamine family amino acid metabolic process (6g)"</span></span>
<span><span class="co">##  [9] "GO:0000821 regulation of arginine metabolic process (2g)"                   </span></span>
<span><span class="co">## [10] "GO:0000959 mitochondrial RNA metabolic process (48g)"                       </span></span>
<span><span class="co">## [11] "GO:0001523 retinoid metabolic process (103g)"                               </span></span>
<span><span class="co">## [12] "GO:0001573 ganglioside metabolic process (21g)"                             </span></span>
<span><span class="co">##  [ reached getOption("max.print") -- omitted 1029 entries ]</span></span></code></pre>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://pillar.r-lib.org/reference/glimpse.html" class="external-link">glimpse</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">result</span>, <span class="fl">3</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## List of 3</span></span>
<span><span class="co">##  $ GO:0000023 maltose metabolic process (3g)                    :List of 3</span></span>
<span><span class="co">##   ..$ SetID  : chr "GO:0000023"</span></span>
<span><span class="co">##   ..$ SetName: chr "maltose metabolic process"</span></span>
<span><span class="co">##   ..$ Genes  : chr [1:3] "MGAM" "GAA" "GANC"</span></span>
<span><span class="co">##  $ GO:0000038 very long-chain fatty acid metabolic process (31g):List of 3</span></span>
<span><span class="co">##   ..$ SetID  : chr "GO:0000038"</span></span>
<span><span class="co">##   ..$ SetName: chr "very long-chain fatty acid metabolic process"</span></span>
<span><span class="co">##   ..$ Genes  : chr [1:31] "ELOVL7" "HACD1" "SLC27A2" "ACAA1" ...</span></span>
<span><span class="co">##  $ GO:0000052 citrulline metabolic process (8g)                 :List of 3</span></span>
<span><span class="co">##   ..$ SetID  : chr "GO:0000052"</span></span>
<span><span class="co">##   ..$ SetName: chr "citrulline metabolic process"</span></span>
<span><span class="co">##   ..$ Genes  : chr [1:8] "DDAH1" "DDAH2" "OTC" "ASS1" ...</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="type-specific-search">Type-Specific Search<a class="anchor" aria-label="anchor" href="#type-specific-search"></a>
</h3>
<p>If you wish to limit your search to specific types of items such as
gene names, you can utilize the ‘type’ parameter as shown below.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">result</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SearchDatabase.html">SearchDatabase</a></span><span class="op">(</span><span class="st">"CD3D"</span>, type <span class="op">=</span> <span class="st">"gene"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">result</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] "GO:0001775 cell activation (1458g)"                                                  </span></span>
<span><span class="co">##  [2] "GO:0002250 adaptive immune response (719g)"                                          </span></span>
<span><span class="co">##  [3] "GO:0002253 activation of immune response (738g)"                                     </span></span>
<span><span class="co">##  [4] "GO:0002376 immune system process (3213g)"                                            </span></span>
<span><span class="co">##  [5] "GO:0002429 immune response-activating cell surface receptor signaling pathway (493g)"</span></span>
<span><span class="co">##  [6] "GO:0002520 immune system development (967g)"                                         </span></span>
<span><span class="co">##  [7] "GO:0002521 leukocyte differentiation (517g)"                                         </span></span>
<span><span class="co">##  [8] "GO:0002682 regulation of immune system process (1662g)"                              </span></span>
<span><span class="co">##  [9] "GO:0002684 positive regulation of immune system process (1213g)"                     </span></span>
<span><span class="co">## [10] "GO:0002757 immune response-activating signal transduction (656g)"                    </span></span>
<span><span class="co">## [11] "GO:0002764 immune response-regulating signaling pathway (690g)"                      </span></span>
<span><span class="co">## [12] "GO:0002768 immune response-regulating cell surface receptor signaling pathway (526g)"</span></span>
<span><span class="co">##  [ reached getOption("max.print") -- omitted 168 entries ]</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="database-specific-search">Database-Specific Search<a class="anchor" aria-label="anchor" href="#database-specific-search"></a>
</h3>
<p>To focus your search within a particular database, specify the
database name using the ‘database’ parameter.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">result</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SearchDatabase.html">SearchDatabase</a></span><span class="op">(</span><span class="st">"CD3D"</span>, database <span class="op">=</span> <span class="st">"Reactome"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">result</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] "R-HSA-1280218 Adaptive Immune System (817g)"                                                 </span></span>
<span><span class="co">##  [2] "R-HSA-168256 Immune System (2190g)"                                                          </span></span>
<span><span class="co">##  [3] "R-HSA-198933 Immunoregulatory interactions between a Lymphoid and a non-Lymphoid cell (191g)"</span></span>
<span><span class="co">##  [4] "R-HSA-199991 Membrane Trafficking (628g)"                                                    </span></span>
<span><span class="co">##  [5] "R-HSA-202403 TCR signaling (124g)"                                                           </span></span>
<span><span class="co">##  [6] "R-HSA-202424 Downstream TCR signaling (102g)"                                                </span></span>
<span><span class="co">##  [7] "R-HSA-202427 Phosphorylation of CD3 and TCR zeta chains (27g)"                               </span></span>
<span><span class="co">##  [8] "R-HSA-202430 Translocation of ZAP-70 to Immunological synapse (24g)"                         </span></span>
<span><span class="co">##  [9] "R-HSA-202433 Generation of second messenger molecules (39g)"                                 </span></span>
<span><span class="co">## [10] "R-HSA-388841 Costimulation by the CD28 family (74g)"                                         </span></span>
<span><span class="co">## [11] "R-HSA-389948 PD-1 signaling (28g)"                                                           </span></span>
<span><span class="co">## [12] "R-HSA-5653656 Vesicle-mediated transport (723g)"                                             </span></span>
<span><span class="co">##  [ reached getOption("max.print") -- omitted 2 entries ]</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="species-specific-search">Species-Specific Search<a class="anchor" aria-label="anchor" href="#species-specific-search"></a>
</h3>
<p>You can specify either ‘human’ or ‘mouse’ using the ‘spe’
parameter.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">result</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SearchDatabase.html">SearchDatabase</a></span><span class="op">(</span><span class="st">"Cd3d"</span>, spe <span class="op">=</span> <span class="st">"mouse"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://pillar.r-lib.org/reference/glimpse.html" class="external-link">glimpse</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">result</span>, <span class="fl">3</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## List of 3</span></span>
<span><span class="co">##  $ GO:0001775 cell activation (1169g)        :List of 3</span></span>
<span><span class="co">##   ..$ SetID  : chr "GO:0001775"</span></span>
<span><span class="co">##   ..$ SetName: chr "cell activation"</span></span>
<span><span class="co">##   ..$ Genes  : chr [1:1169] "Cd86" "Cd79a" "Cd80" "Serpine2" ...</span></span>
<span><span class="co">##  $ GO:0002250 adaptive immune response (640g):List of 3</span></span>
<span><span class="co">##   ..$ SetID  : chr "GO:0002250"</span></span>
<span><span class="co">##   ..$ SetName: chr "adaptive immune response"</span></span>
<span><span class="co">##   ..$ Genes  : chr [1:640] "Cd86" "Cd79a" "Msh2" "Traf2" ...</span></span>
<span><span class="co">##  $ GO:0002376 immune system process (2850g)  :List of 3</span></span>
<span><span class="co">##   ..$ SetID  : chr "GO:0002376"</span></span>
<span><span class="co">##   ..$ SetName: chr "immune system process"</span></span>
<span><span class="co">##   ..$ Genes  : chr [1:2850] "Gm42543" "Cd86" "Cd79a" "Cd80" ...</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="customizing-return-types">Customizing Return Types<a class="anchor" aria-label="anchor" href="#customizing-return-types"></a>
</h3>
<p>The function also offers flexibility in output types. For example, if
you require a list of pathway IDs for downstream analysis, you can use
the ‘return’ parameter as follows.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">result</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SearchDatabase.html">SearchDatabase</a></span><span class="op">(</span><span class="st">"CD3D"</span>, return <span class="op">=</span> <span class="st">"ID"</span><span class="op">)</span></span>
<span><span class="va">result</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] "GO:0001775" "GO:0002250" "GO:0002253" "GO:0002376" "GO:0002429"</span></span>
<span><span class="co">##  [6] "GO:0002520" "GO:0002521" "GO:0002682" "GO:0002684" "GO:0002757"</span></span>
<span><span class="co">## [11] "GO:0002764" "GO:0002768"</span></span>
<span><span class="co">##  [ reached getOption("max.print") -- omitted 168 entries ]</span></span></code></pre>
<p>Alternatively, if you need the output as a gene list formatted for
<code>GeneSetAnalysis</code>, adjust the ‘return’ parameter like so:</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">result</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SearchDatabase.html">SearchDatabase</a></span><span class="op">(</span><span class="st">"CD3D"</span>, return <span class="op">=</span> <span class="st">"genelist"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://pillar.r-lib.org/reference/glimpse.html" class="external-link">glimpse</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">result</span>, <span class="fl">5</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## List of 5</span></span>
<span><span class="co">##  $ GO:0001775: chr [1:1458] "IGHV3-64" "IGHV4-4" "IGHV4OR15-8" "IGHV3OR16-12" ...</span></span>
<span><span class="co">##  $ GO:0002250: chr [1:719] "IGKV3-7" "IGKV1D-42" "IGLV4-69" "IGLV8-61" ...</span></span>
<span><span class="co">##  $ GO:0002253: chr [1:738] "IGKV2-28" "IGHV3-64" "IGHV4-4" "IGKV2D-30" ...</span></span>
<span><span class="co">##  $ GO:0002376: chr [1:3213] "IGKV3-7" "IGKV1D-42" "IGLV4-69" "IGLV8-61" ...</span></span>
<span><span class="co">##  $ GO:0002429: chr [1:493] "IGKV2-28" "IGHV3-64" "IGHV4-4" "IGKV2D-30" ...</span></span></code></pre>
<p>To export the result as a data frame, suitable for formats like Excel
or CSV, set the ‘export.to.data.frame’ parameter to TRUE.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">result</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SearchDatabase.html">SearchDatabase</a></span><span class="op">(</span><span class="st">"CD3D"</span>, export.to.data.frame <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://pillar.r-lib.org/reference/glimpse.html" class="external-link">glimpse</a></span><span class="op">(</span><span class="va">result</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Rows: 180</span></span>
<span><span class="co">## Columns: 3</span></span>
<span><span class="co">## $ SetID   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> "GO:0001775", "GO:0002250", "GO:0002253", "GO:0002376", "GO:00…</span></span>
<span><span class="co">## $ SetName <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> "cell activation", "adaptive immune response", "activation of …</span></span>
<span><span class="co">## $ Genes   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> "IGHV3-64,IGHV4-4,IGHV4OR15-8,IGHV3OR16-12,IGHV1OR15-1,IGHV3OR…</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="filtering-a-customized-gene-set">Filtering a Customized Gene Set<a class="anchor" aria-label="anchor" href="#filtering-a-customized-gene-set"></a>
</h3>
<p>Lastly, you can also filter a given gene set list with the
<code>SearchPathways</code> function. For instance, within the “Hallmark
50” database, you can find pathways that include the gene “CD3D” or have
names that contain “interferon.”</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/SearchPathways.html">SearchPathways</a></span><span class="op">(</span>genesets <span class="op">=</span> <span class="va">hall50</span><span class="op">$</span><span class="va">human</span>, item <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD3D"</span>, <span class="st">"interferon"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## $HALLMARK_INTERFERON_ALPHA_RESPONSE</span></span>
<span><span class="co">##  [1] "MX1"        "ISG15"      "AC004551.1" "IFIT3"      "IFI44"     </span></span>
<span><span class="co">##  [6] "IFI35"      "IRF7"       "RSAD2"      "IFI44L"     "IFITM1"    </span></span>
<span><span class="co">## [11] "IFI27"      "IRF9"      </span></span>
<span><span class="co">##  [ reached getOption("max.print") -- omitted 85 entries ]</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $HALLMARK_INTERFERON_GAMMA_RESPONSE</span></span>
<span><span class="co">##  [1] "STAT1"   "ISG15"   "IFIT1"   "MX1"     "IFIT3"   "IFI35"   "IRF7"   </span></span>
<span><span class="co">##  [8] "IFIT2"   "OAS2"    "TAP1"    "EIF2AK2" "RSAD2"  </span></span>
<span><span class="co">##  [ reached getOption("max.print") -- omitted 188 entries ]</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $HALLMARK_ALLOGRAFT_REJECTION</span></span>
<span><span class="co">##  [1] "PTPRC" "IL12B" "TGFB1" "IL12A" "CD3E"  "CD3D"  "CD28"  "LYN"   "HCLS1"</span></span>
<span><span class="co">## [10] "IL18"  "CRTAM" "IFNG" </span></span>
<span><span class="co">##  [ reached getOption("max.print") -- omitted 188 entries ]</span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="convert-goreactome-pathway-ids-to-pathway-names">Convert GO/Reactome pathway IDs to pathway names<a class="anchor" aria-label="anchor" href="#convert-goreactome-pathway-ids-to-pathway-names"></a>
</h2>
<p>During the course of analyses, researchers often encounter pathway
IDs from databases such as GO and Reactome. While these IDs are great
for computational tasks, they can be cryptic when it comes to
interpretability. <code>RenameGO</code> and <code>RenameReactome</code>
functions provide a convenient means to transform these IDs into their
more descriptive pathway names.</p>
<p>The primary parameter these functions require is “item”. This can
either be:</p>
<ol style="list-style-type: decimal">
<li>A character vector of GO or Reactome IDs. For example, this could be
the output from functions like <code>FilterGOTerms</code> or
<code>FilterReactomeTerms</code>.</li>
<li>A matrix where the IDs are stored in the rownames, such as the
output of <code>GeneSetAnalysisGO</code> or
<code>GeneSetAnalysisReactome</code>.</li>
</ol>
<p>Convert GO IDs to their respective pathway names for human:</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/RenameGO.html">RenameGO</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"GO:0002376"</span>,<span class="st">"GO:0050896"</span><span class="op">)</span>, spe <span class="op">=</span> <span class="st">"human"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "GO:0002376 immune system process (3213g)"</span></span>
<span><span class="co">## [2] "GO:0050896 response to stimulus (9342g)"</span></span></code></pre>
<p>Similarly, for Reactome IDs:</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/RenameGO.html">RenameReactome</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"R-HSA-109582"</span>,<span class="st">"R-HSA-112316"</span><span class="op">)</span>, spe <span class="op">=</span> <span class="st">"human"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "R-HSA-109582 Hemostasis (679g)"      "R-HSA-112316 Neuronal System (411g)"</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="filter-the-goreactome-pathway-list-based-on-certain-criteria">Filter the GO/Reactome pathway list based on certain criteria<a class="anchor" aria-label="anchor" href="#filter-the-goreactome-pathway-list-based-on-certain-criteria"></a>
</h2>
<p>Both GO and Reactome databases contain thousands of pathways, but not
all of which may be relevant to your study. To streamline the analysis,
you can use the <code>FilterGOTerms</code> and
<code>FilterReactomeTerms</code> functions to subset and refine the list
of GO or Reactome pathways based on specific criteria.</p>
<div class="section level3">
<h3 id="filtering-go-pathways">Filtering GO Pathways<a class="anchor" aria-label="anchor" href="#filtering-go-pathways"></a>
</h3>
<p>Let’s start by looking at how you can filter GO pathways:</p>
<ul>
<li>
<strong>By Parent Term:</strong> To select pathways under a specific
category, use the <code>parent</code> parameter. For example, to get
pathways related to the immune system process:</li>
</ul>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">terms</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/FilterGOTerms.html">FilterGOTerms</a></span><span class="op">(</span>parent <span class="op">=</span> <span class="st">"GO:0002376"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/RenameGO.html">RenameGO</a></span><span class="op">(</span><span class="va">terms</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] "GO:0001773 myeloid dendritic cell activation (28g)"      </span></span>
<span><span class="co">##  [2] "GO:0001774 microglial cell activation (45g)"             </span></span>
<span><span class="co">##  [3] "GO:0001776 leukocyte homeostasis (84g)"                  </span></span>
<span><span class="co">##  [4] "GO:0001777 T cell homeostatic proliferation (4g)"        </span></span>
<span><span class="co">##  [5] "GO:0001779 natural killer cell differentiation (21g)"    </span></span>
<span><span class="co">##  [6] "GO:0001780 neutrophil homeostasis (16g)"                 </span></span>
<span><span class="co">##  [7] "GO:0001782 B cell homeostasis (27g)"                     </span></span>
<span><span class="co">##  [8] "GO:0001787 natural killer cell proliferation (11g)"      </span></span>
<span><span class="co">##  [9] "GO:0001788 antibody-dependent cellular cytotoxicity (2g)"</span></span>
<span><span class="co">## [10] "GO:0001794 type IIa hypersensitivity (4g)"               </span></span>
<span><span class="co">## [11] "GO:0001802 type III hypersensitivity (2g)"               </span></span>
<span><span class="co">## [12] "GO:0001806 type IV hypersensitivity (4g)"                </span></span>
<span><span class="co">##  [ reached getOption("max.print") -- omitted 395 entries ]</span></span></code></pre>
<ul>
<li>
<strong>By Gene Count:</strong> If you’re interested in pathways of
a specific size (i.e., containing a certain number of genes), you can
further refine your previously filtered list using the
<code>n.min</code> and <code>n.max</code> parameters. Building upon the
pathways we selected under the “immune system process”
(<code>terms</code>), to keep only those pathways that contain between
10 and 1000 genes:</li>
</ul>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">terms2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/FilterGOTerms.html">FilterGOTerms</a></span><span class="op">(</span>term <span class="op">=</span> <span class="va">terms</span>, n.min <span class="op">=</span> <span class="fl">10</span>, n.max <span class="op">=</span> <span class="fl">1000</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/RenameGO.html">RenameGO</a></span><span class="op">(</span><span class="va">terms2</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] "GO:0001773 myeloid dendritic cell activation (28g)"                                        </span></span>
<span><span class="co">##  [2] "GO:0001774 microglial cell activation (45g)"                                               </span></span>
<span><span class="co">##  [3] "GO:0001776 leukocyte homeostasis (84g)"                                                    </span></span>
<span><span class="co">##  [4] "GO:0001779 natural killer cell differentiation (21g)"                                      </span></span>
<span><span class="co">##  [5] "GO:0001780 neutrophil homeostasis (16g)"                                                   </span></span>
<span><span class="co">##  [6] "GO:0001782 B cell homeostasis (27g)"                                                       </span></span>
<span><span class="co">##  [7] "GO:0001787 natural killer cell proliferation (11g)"                                        </span></span>
<span><span class="co">##  [8] "GO:0001865 NK T cell differentiation (10g)"                                                </span></span>
<span><span class="co">##  [9] "GO:0001867 complement activation, lectin pathway (11g)"                                    </span></span>
<span><span class="co">## [10] "GO:0001913 T cell mediated cytotoxicity (46g)"                                             </span></span>
<span><span class="co">## [11] "GO:0002200 somatic diversification of immune receptors (77g)"                              </span></span>
<span><span class="co">## [12] "GO:0002204 somatic recombination of immunoglobulin genes involved in immune response (50g)"</span></span>
<span><span class="co">##  [ reached getOption("max.print") -- omitted 202 entries ]</span></span></code></pre>
<ul>
<li>
<strong>End-Level Pathways:</strong> If you prefer to look at only
the end-level (most specific) pathways, set the
<code>only.end.terms</code> parameter to <code>TRUE</code>.</li>
</ul>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">terms3</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/FilterGOTerms.html">FilterGOTerms</a></span><span class="op">(</span>term <span class="op">=</span> <span class="va">terms</span>, only.end.terms <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/RenameGO.html">RenameGO</a></span><span class="op">(</span><span class="va">terms3</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] "GO:0001777 T cell homeostatic proliferation (4g)"                      </span></span>
<span><span class="co">##  [2] "GO:0001780 neutrophil homeostasis (16g)"                               </span></span>
<span><span class="co">##  [3] "GO:0001787 natural killer cell proliferation (11g)"                    </span></span>
<span><span class="co">##  [4] "GO:0001788 antibody-dependent cellular cytotoxicity (2g)"              </span></span>
<span><span class="co">##  [5] "GO:0001802 type III hypersensitivity (2g)"                             </span></span>
<span><span class="co">##  [6] "GO:0001806 type IV hypersensitivity (4g)"                              </span></span>
<span><span class="co">##  [7] "GO:0001865 NK T cell differentiation (10g)"                            </span></span>
<span><span class="co">##  [8] "GO:0001866 NK T cell proliferation (7g)"                               </span></span>
<span><span class="co">##  [9] "GO:0001867 complement activation, lectin pathway (11g)"                </span></span>
<span><span class="co">## [10] "GO:0001905 activation of membrane attack complex (3g)"                 </span></span>
<span><span class="co">## [11] "GO:0001922 B-1 B cell homeostasis (3g)"                                </span></span>
<span><span class="co">## [12] "GO:0002223 stimulatory C-type lectin receptor signaling pathway (112g)"</span></span>
<span><span class="co">##  [ reached getOption("max.print") -- omitted 188 entries ]</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="filtering-reactome-pathways">Filtering Reactome Pathways<a class="anchor" aria-label="anchor" href="#filtering-reactome-pathways"></a>
</h3>
<p>The process for Reactome pathways is analogous. For instance, to
select pathways related to the Immune System:</p>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">terms</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/FilterGOTerms.html">FilterReactomeTerms</a></span><span class="op">(</span>parent <span class="op">=</span> <span class="st">"R-HSA-168256"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/RenameGO.html">RenameReactome</a></span><span class="op">(</span><span class="va">terms</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] "R-HSA-1059683 Interleukin-6 signaling (11g)"                                         </span></span>
<span><span class="co">##  [2] "R-HSA-109703 PKB-mediated events (2g)"                                               </span></span>
<span><span class="co">##  [3] "R-HSA-109704 PI3K Cascade (44g)"                                                     </span></span>
<span><span class="co">##  [4] "R-HSA-1168372 Downstream signaling events of B Cell Receptor (BCR) (81g)"            </span></span>
<span><span class="co">##  [5] "R-HSA-1169091 Activation of NF-kappaB in B cells (67g)"                              </span></span>
<span><span class="co">##  [6] "R-HSA-1169092 Activation of RAS in B cells (5g)"                                     </span></span>
<span><span class="co">##  [7] "R-HSA-1169408 ISG15 antiviral mechanism (74g)"                                       </span></span>
<span><span class="co">##  [8] "R-HSA-1169410 Antiviral mechanism by IFN-stimulated genes (82g)"                     </span></span>
<span><span class="co">##  [9] "R-HSA-1170546 Prolactin receptor signaling (15g)"                                    </span></span>
<span><span class="co">## [10] "R-HSA-1222556 ROS and RNS production in phagocytes (36g)"                            </span></span>
<span><span class="co">## [11] "R-HSA-1236973 Cross-presentation of particulate exogenous antigens (phagosomes) (8g)"</span></span>
<span><span class="co">## [12] "R-HSA-1236974 ER-Phagosome pathway (83g)"                                            </span></span>
<span><span class="co">##  [ reached getOption("max.print") -- omitted 185 entries ]</span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="create-a-gsea-plot-emulating-the-broad-institute-analysis">Create a GSEA plot emulating the Broad Institute analysis<a class="anchor" aria-label="anchor" href="#create-a-gsea-plot-emulating-the-broad-institute-analysis"></a>
</h2>
<p>The <code>GSEAplot</code> function is designed to generate plots that
emulate the Gene Set Enrichment Analysis (GSEA) as developed by the
Broad Institute. This function provides a way to visualize the
enrichment of specific gene sets within different biological states or
conditions.</p>
<p>Here’s how you can create a GSEA plot for the
“INTERFERON_GAMMA_RESPONSE” gene set within the “Naive CD4 T” cell
population of the <code>pbmc</code> dataset:</p>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/GSEAplot.html">GSEAplot</a></span><span class="op">(</span></span>
<span>  <span class="va">pbmc</span>, </span>
<span>  ident.1 <span class="op">=</span> <span class="st">"CD4 T Naive"</span>, </span>
<span>  title <span class="op">=</span> <span class="st">"INTERFERON_GAMMA_RESPONSE"</span>,</span>
<span>  geneset <span class="op">=</span> <span class="va">hall50</span><span class="op">$</span><span class="va">human</span><span class="op">$</span><span class="va">HALLMARK_INTERFERON_GAMMA_RESPONSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="GSEA_files/figure-html/unnamed-chunk-25-1.png" width="528"></p>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.4.0 (2024-04-24)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">## Running under: Ubuntu 20.04.6 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.9.0 </span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.9.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">##  [3] LC_TIME=de_BE.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">##  [5] LC_MONETARY=de_BE.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=de_BE.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">## [11] LC_MEASUREMENT=de_BE.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: Europe/Brussels</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] parallel  stats4    stats     graphics  grDevices utils     datasets </span></span>
<span><span class="co">## [8] methods   base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] ggpubr_0.6.0       tidyr_1.3.1        rlist_0.4.6.2      RColorBrewer_1.1-3</span></span>
<span><span class="co">##  [5] viridis_0.6.5      viridisLite_0.4.2  mosaic_1.9.1       mosaicData_0.20.4 </span></span>
<span><span class="co">##  [9] ggformula_0.12.0   lattice_0.22-6     rlang_1.1.4        scales_1.3.0      </span></span>
<span><span class="co">##  [ reached getOption("max.print") -- omitted 21 entries ]</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##  [1] RcppAnnoy_0.0.22  splines_4.4.0     later_1.3.2       tibble_3.2.1     </span></span>
<span><span class="co">##  [5] polyclip_1.10-6   fastDummies_1.7.3 lifecycle_1.0.4   rstatix_0.7.2    </span></span>
<span><span class="co">##  [9] globals_0.16.3    MASS_7.3-61       backports_1.5.0   magrittr_2.0.3   </span></span>
<span><span class="co">##  [ reached getOption("max.print") -- omitted 115 entries ]</span></span></code></pre>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Yichao Hua.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
