<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>SCENIC for Gene Regulatory Networks Analysis • SeuratExtend</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="SCENIC for Gene Regulatory Networks Analysis">
<meta property="og:description" content="SeuratExtend">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">SeuratExtend</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/News.html">What's New in v1.1.0</a>
    </li>
    <li class="divider">
    </li>
<li>
      <a href="../articles/Visualization.html">Enhanced Visualization</a>
    </li>
    <li>
      <a href="../articles/GSEA.html">Geneset Enrichment Analysis (GSEA) </a>
    </li>
    <li>
      <a href="../articles/Trajectory.html">Trajectory and Pseudotime Analysis </a>
    </li>
    <li>
      <a href="../articles/SCENIC.html">SCENIC for Gene Regulatory Networks Analysis </a>
    </li>
    <li>
      <a href="../articles/Utilities.html">Utility Tools and Functions</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    scRNA-seq Course
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/single-cell-course/1.basic-R.html">1. Introduction to R</a>
    </li>
    <li>
      <a href="../articles/single-cell-course/2.Seurat.html">2. Basic Single-Cell Analysis Workflow</a>
    </li>
    <li>
      <a href="../articles/single-cell-course/3.Visualization.html">3. Advanced Visualization Techniques</a>
    </li>
    <li>
      <a href="../articles/single-cell-course/4.GSEA.html">4. Gene Set Enrichment Analysis</a>
    </li>
    <li>
      <a href="../articles/single-cell-course/5.Core-Enhancement.html">5. Core scRNA-seq Workflow Enhancements</a>
    </li>
    <li>
      <a href="../articles/single-cell-course/6.Advanced.html">6. Advanced Methods (Part 1)</a>
    </li>
    <li>
      <a href="../articles/single-cell-course/6.Advanced-2.html">7. Advanced Methods (Part 2)</a>
    </li>
  </ul>
</li>
<li>
  <a href="../articles/FAQ.html">FAQ</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>SCENIC for Gene Regulatory Networks
Analysis</h1>
                        <h4 data-toc-skip class="author">Yichao Hua</h4>
            
            <h4 data-toc-skip class="date">2024-11-28</h4>
      
      
      <div class="hidden name"><code>SCENIC.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="table-of-contents">Table of Contents<a class="anchor" aria-label="anchor" href="#table-of-contents"></a>
</h2>
<ol style="list-style-type: decimal">
<li><a href="#importing-scenic-loom-files-into-seurat">Importing SCENIC
Loom Files into Seurat</a></li>
<li><a href="#visualizing-scenic-results">Visualizing SCENIC
Results</a></li>
</ol>
</div>
<div class="section level2">
<h2 id="importing-scenic-loom-files-into-seurat">Importing SCENIC Loom Files into Seurat<a class="anchor" aria-label="anchor" href="#importing-scenic-loom-files-into-seurat"></a>
</h2>
<p><a href="https://www.nature.com/articles/nmeth.4463" class="external-link">SCENIC</a>
(Single-Cell Regulatory Network Inference and Clustering) is a
computational method that provides deep insights into the regulatory
networks governing gene expression in single cells. It is highly
recommended to use the Nextflow pipeline to run SCENIC, which can be
found <a href="https://github.com/aertslab/SCENICprotocol/" class="external-link">here</a>.
This process requires a loom file as input, which can be generated
directly using the <code><a href="../reference/adata.Load.html">Seu2Loom()</a></code> function. Currently, the
<code>SeuratExtend</code> package does not integrate the
<code>RunScenic</code> functionality directly (requiring the use of
Nextflow command line), but if there is user demand, this could be
considered for future updates. The workflow results in a file named
“pyscenic_integrated-output.loom,” which includes a list of
transcription factors (TFs) and their regulated genes, as well as a
TF-cell matrix representing the AUCell values. The AUCell score
represents the enrichment score of all genes regulated by a TF,
indicating regulon activity. The <code><a href="../reference/ImportPyscenicLoom.html">ImportPyscenicLoom()</a></code>
function allows for the import of SCENIC-generated loom files into
Seurat objects, facilitating further analysis and visualization within
the Seurat framework.</p>
<p>As an example, we use a pre-computed SCENIC loom file which can be
downloaded as follows:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SeuratExtend</span><span class="op">)</span></span>
<span><span class="va">scenic_loom_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">"pyscenic_integrated-output.loom"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/download.file.html" class="external-link">download.file</a></span><span class="op">(</span><span class="st">"https://zenodo.org/records/10944066/files/pbmc3k_small_pyscenic_integrated-output.loom"</span>,</span>
<span>              <span class="va">scenic_loom_path</span>, mode <span class="op">=</span> <span class="st">"wb"</span><span class="op">)</span>  <span class="co"># Use binary mode for Windows compatibility</span></span>
<span></span>
<span><span class="co"># Importing SCENIC Loom Files into Seurat</span></span>
<span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ImportPyscenicLoom.html">ImportPyscenicLoom</a></span><span class="op">(</span><span class="va">scenic_loom_path</span>, seu <span class="op">=</span> <span class="va">pbmc</span><span class="op">)</span></span></code></pre></div>
<p>If you prefer to import SCENIC results without specifying a Seurat
object:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Importing SCENIC results without an existing Seurat object</span></span>
<span><span class="va">scenic_output</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ImportPyscenicLoom.html">ImportPyscenicLoom</a></span><span class="op">(</span><span class="va">scenic_loom_path</span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="examining-scenic-outputs">Examining SCENIC Outputs<a class="anchor" aria-label="anchor" href="#examining-scenic-outputs"></a>
</h3>
<p>SCENIC results are stored in <code>seu@misc$SCENIC</code>, which
includes <code>seu@misc$SCENIC$Regulons</code>, a list where each
element’s name is a TF name and the value is a list of genes it
regulates. Additionally, <code>seu@misc$SCENIC$RegulonsAUC</code>, a
TF-cell AUCell matrix, is also loaded into the “TF” assay of the Seurat
object, making the manipulation of TF regulon activity as
straightforward as handling gene expression data.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Viewing the outputs</span></span>
<span><span class="va">tf_auc</span> <span class="op">&lt;-</span> <span class="va">pbmc</span><span class="op">@</span><span class="va">misc</span><span class="op">$</span><span class="va">SCENIC</span><span class="op">$</span><span class="va">RegulonsAUC</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">tf_auc</span>, <span class="fl">4</span><span class="op">:</span><span class="fl">5</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                         AHR     ARID3A        ARNT     ARNTL        ATF1</span></span>
<span><span class="co">## CTATAAGATCGTTT-1 0.01406902 0.03347861 0.000000000 0.1144568 0.016730526</span></span>
<span><span class="co">## GTGATTCTGGTTCA-1 0.00000000 0.00000000 0.000000000 0.1730939 0.017191920</span></span>
<span><span class="co">## ACGTTGGACCGTAA-1 0.00000000 0.03189382 0.000000000 0.1463286 0.003671087</span></span>
<span><span class="co">## GGATACTGCAGCTA-1 0.02483910 0.01347068 0.006867406 0.0945589 0.016670344</span></span></code></pre>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tf_gene_list</span> <span class="op">&lt;-</span> <span class="va">pbmc</span><span class="op">@</span><span class="va">misc</span><span class="op">$</span><span class="va">SCENIC</span><span class="op">$</span><span class="va">Regulons</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">tf_gene_list</span>, <span class="fl">5</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## $AHR</span></span>
<span><span class="co">##  [1] "NFYC-AS1" "MYSM1"    "ZZZ3"     "FUBP1"    "WDR77"    "TMEM183A"</span></span>
<span><span class="co">##  [7] "TGOLN2"   "SEC22C"   "ATXN7"    "RBPJ"     "C5orf24"  "CREBRF"  </span></span>
<span><span class="co">## [13] "POLR1C"   "YIPF3"    "RUNX2"    "BCLAF1"   "HOXA9"    "HOXA10"  </span></span>
<span><span class="co">## [19] "CREB5"    "STAG3"    "POP7"     "ASAP1"    "RMI1"     "PSMD5"   </span></span>
<span><span class="co">## [25] "LRSAM1"   "ABI1"     "HNRNPH3"  "IFT46"    "ST3GAL4"  "BAZ2A"   </span></span>
<span><span class="co">## [31] "BRI3BP"   "MICU2"    "TLE3"     "PARP6"    "NPTN"     "PPCDC"   </span></span>
<span><span class="co">## [37] "FAM103A1" "CHMP1A"   "TMEM107"  "PHF12"    "TLK2"     "NOL11"   </span></span>
<span><span class="co">## [43] "MBP"      "DIDO1"    "SH3GL1"   "SPINT2"   "RBFOX2"   "GABPA"   </span></span>
<span><span class="co">## [49] "CCT8"    </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $ARID3A</span></span>
<span><span class="co">##  [1] "CDC7"      "IL6R"      "LAMC1"     "LINC01136" "CDC42EP3"  "HNMT"     </span></span>
<span><span class="co">##  [7] "GPBAR1"    "OARD1"     "GNB2"      "ATP6V1F"   "YWHAZ"     "XPA"      </span></span>
<span><span class="co">## [13] "ANKRD22"   "APLP2"     "PSMB5"     "LGALS3"    "RPH3AL"    "EMILIN2"  </span></span>
<span><span class="co">## [19] "ME2"       "MAFB"      "GNAS"      "ARID3A"    "UQCRFS1"   "CECR1"    </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $ARNT</span></span>
<span><span class="co">## [1] "EPHB3"   "SMIM14"  "GPR68"   "ANXA2"   "MORF4L1" "ZFP3"   </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $ARNTL</span></span>
<span><span class="co">## [1] "SLC4A10" "AP2M1"   "EMC2"    "PHF20L1" "SLC43A1" "CFL1"    "SPG21"  </span></span>
<span><span class="co">## [8] "CRB3"    "RPS28"  </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $ATF1</span></span>
<span><span class="co">##  [1] "CAMK2N1"        "PAFAH2"         "BCAS2"          "S100A13"       </span></span>
<span><span class="co">##  [5] "ZNF281"         "LIN9"           "THUMPD2"        "CALM2"         </span></span>
<span><span class="co">##  [9] "ACVR1"          "NAB1"           "ZDBF2"          "SNRK"          </span></span>
<span><span class="co">## [13] "IFRD2"          "DUSP7"          "MITF"           "FOXP1"         </span></span>
<span><span class="co">## [17] "DHX36"          "KIAA0232"       "TEC"            "G3BP2"         </span></span>
<span><span class="co">## [21] "ZNF330"         "ADAMTS6"        "PPARGC1B"       "RREB1"         </span></span>
<span><span class="co">## [25] "FAM65B"         "GNL1"           "SYNGAP1"        "PAQR8"         </span></span>
<span><span class="co">## [29] "MOSPD3"         "GNB2"           "FIS1"           "ARF5"          </span></span>
<span><span class="co">## [33] "RPS6KA3"        "CCDC120"        "PLS3"           "TMEM187"       </span></span>
<span><span class="co">## [37] "SGK3"           "PKIA"           "PAG1"           "RAD54B"        </span></span>
<span><span class="co">## [41] "UBAP2"          "TESK1"          "MRRF"           "BAG3"          </span></span>
<span><span class="co">## [45] "CARS"           "PGAP2"          "BTBD10"         "ITFG2"         </span></span>
<span><span class="co">## [49] "ESYT1"          "C12orf73"       "MLXIP"          "SPATA7"        </span></span>
<span><span class="co">## [53] "C15orf41"       "TLE3"           "KIAA0556"       "DNAJA2"        </span></span>
<span><span class="co">## [57] "CYLD"           "PDP2"           "CENPN"          "RPH3AL"        </span></span>
<span><span class="co">## [61] "SMG6"           "TMEM256-PLSCR3" "ARL4D"          "B3GNTL1"       </span></span>
<span><span class="co">## [65] "GPCPD1"         "SPAG4"          "MYBL2"          "ADNP"          </span></span>
<span><span class="co">## [69] "SBNO2"          "CIRBP"          "MVB12A"         "PLEKHF1"       </span></span>
<span><span class="co">## [73] "TMEM147"        "SPHK2"          "EMC10"          "ZNF787"        </span></span>
<span><span class="co">## [77] "DGCR8"          "CRKL"           "PDE9A"</span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="visualizing-scenic-results">Visualizing SCENIC Results<a class="anchor" aria-label="anchor" href="#visualizing-scenic-results"></a>
</h2>
<p>Once SCENIC data is integrated into a Seurat object, users can
leverage a variety of visualization tools provided in the
<strong>Enhanced Visualization</strong> section to explore and interpret
these regulatory networks. Both the extracted <code>tf_auc</code> matrix
or the Seurat object itself can be used as inputs. Here are some
practical examples:</p>
<div class="section level3">
<h3 id="identifying-top-activated-tfs-in-each-cluster">Identifying Top Activated TFs in Each Cluster<a class="anchor" aria-label="anchor" href="#identifying-top-activated-tfs-in-each-cluster"></a>
</h3>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tf_zscore</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/CalcStats.html">CalcStats</a></span><span class="op">(</span><span class="va">tf_auc</span>, f <span class="op">=</span> <span class="va">pbmc</span><span class="op">$</span><span class="va">cluster</span>, order <span class="op">=</span> <span class="st">"p"</span>, n <span class="op">=</span> <span class="fl">4</span>, t <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/Heatmap.html">Heatmap</a></span><span class="op">(</span><span class="va">tf_zscore</span>, lab_fill <span class="op">=</span> <span class="st">"zscore"</span><span class="op">)</span></span></code></pre></div>
<p><img src="SCENIC_files/figure-html/unnamed-chunk-4-1.png" width="480"></p>
</div>
<div class="section level3">
<h3 id="comparing-tf-gene-expression-levels-and-regulon-activity-aucell">Comparing TF Gene Expression Levels and Regulon Activity
(AUCell)<a class="anchor" aria-label="anchor" href="#comparing-tf-gene-expression-levels-and-regulon-activity-aucell"></a>
</h3>
<p>Since we have imported SCENIC results into the “TF” assay, we can
easily access the corresponding AUCell values by prefixing “tf_” to the
TF name:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/DimPlot2.html">DimPlot2</a></span><span class="op">(</span></span>
<span>  <span class="va">pbmc</span>,</span>
<span>  features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ETS1"</span>, <span class="st">"ATF3"</span>, <span class="st">"tf_ETS1"</span>, <span class="st">"tf_ATF3"</span><span class="op">)</span>,</span>
<span>  cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">"tf_ETS1"</span> <span class="op">=</span> <span class="st">"D"</span>, <span class="st">"tf_ATF3"</span> <span class="op">=</span> <span class="st">"D"</span><span class="op">)</span>,</span>
<span>  theme <span class="op">=</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SeuratTheme.html" class="external-link">NoAxes</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Loading required package: cowplot</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Attaching package: 'cowplot'</span></span></code></pre>
<pre><code><span><span class="co">## The following object is masked from 'package:mosaic':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     theme_map</span></span></code></pre>
<p><img src="SCENIC_files/figure-html/unnamed-chunk-5-1.png" width="672"></p>
</div>
<div class="section level3">
<h3 id="simplifying-regulon-activity-access-by-setting-default-assay">Simplifying Regulon Activity Access by Setting Default Assay<a class="anchor" aria-label="anchor" href="#simplifying-regulon-activity-access-by-setting-default-assay"></a>
</h3>
<p>If you find manually adding “tf_” to each transcription factor
cumbersome, you can set the default assay to “TF”, which simplifies
operations involving regulon activity. For example, to create a
waterfall plot that compares the regulon activity between two cell
types, you can do the following:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Setting the default assay to "TF" for easier access to regulon activity</span></span>
<span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/DefaultAssay.html" class="external-link">DefaultAssay</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"TF"</span></span>
<span></span>
<span><span class="co"># Creating a waterfall plot to compare regulon activity between monocytes and CD8 T cells</span></span>
<span><span class="fu"><a href="../reference/WaterfallPlot.html">WaterfallPlot</a></span><span class="op">(</span></span>
<span>  <span class="va">pbmc</span>,</span>
<span>  features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span>,  <span class="co"># Using all available TFs in the "TF" assay</span></span>
<span>  ident.1 <span class="op">=</span> <span class="st">"Mono CD14"</span>,      <span class="co"># First group of cells</span></span>
<span>  ident.2 <span class="op">=</span> <span class="st">"CD8 T cell"</span>,     <span class="co"># Second group of cells</span></span>
<span>  exp.transform <span class="op">=</span> <span class="cn">FALSE</span>,      <span class="co"># Disable transformation of expression data</span></span>
<span>  top.n <span class="op">=</span> <span class="fl">20</span>                  <span class="co"># Display the top 20 most differentially active TFs</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Attaching package: 'tidyr'</span></span></code></pre>
<pre><code><span><span class="co">## The following object is masked from 'package:reshape2':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     smiths</span></span></code></pre>
<pre><code><span><span class="co">## The following objects are masked from 'package:Matrix':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     expand, pack, unpack</span></span></code></pre>
<p><img src="SCENIC_files/figure-html/unnamed-chunk-6-1.png" width="768"></p>
<p>These examples illustrate how to integrate and utilize SCENIC
analysis within the Seurat framework, providing a comprehensive approach
to understanding gene regulatory mechanisms at the single-cell
level.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.4.0 (2024-04-24)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">## Running under: Ubuntu 20.04.6 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.9.0 </span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.9.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">##  [3] LC_TIME=de_BE.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">##  [5] LC_MONETARY=de_BE.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=de_BE.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">## [11] LC_MEASUREMENT=de_BE.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: Europe/Brussels</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] tidyr_1.3.1            rlist_0.4.6.2          cowplot_1.1.3         </span></span>
<span><span class="co">##  [4] RColorBrewer_1.1-3     viridis_0.6.5          viridisLite_0.4.2     </span></span>
<span><span class="co">##  [7] rlang_1.1.4            scales_1.3.0           reshape2_1.4.4        </span></span>
<span><span class="co">## [10] mosaic_1.9.1           mosaicData_0.20.4      ggformula_0.12.0      </span></span>
<span><span class="co">## [13] Matrix_1.7-0           ggplot2_3.5.1          lattice_0.22-6        </span></span>
<span><span class="co">## [16] dplyr_1.1.4            loomR_0.2.0            itertools_0.1-3       </span></span>
<span><span class="co">## [19] iterators_1.0.14       hdf5r_1.3.11           R6_2.5.1              </span></span>
<span><span class="co">## [22] Seurat_5.1.0           SeuratExtend_1.1.0     SeuratObject_5.0.2    </span></span>
<span><span class="co">## [25] sp_2.1-4               SeuratExtendData_0.2.1</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] rstudioapi_0.16.0      jsonlite_1.8.8         magrittr_2.0.3        </span></span>
<span><span class="co">##   [4] spatstat.utils_3.0-5   farver_2.1.2           rmarkdown_2.29        </span></span>
<span><span class="co">##   [7] fs_1.6.4               ragg_1.3.2             vctrs_0.6.5           </span></span>
<span><span class="co">##  [10] ROCR_1.0-11            memoise_2.0.1          spatstat.explore_3.2-7</span></span>
<span><span class="co">##  [13] forcats_1.0.0          htmltools_0.5.8.1      haven_2.5.4           </span></span>
<span><span class="co">##  [16] sass_0.4.9             sctransform_0.4.1      parallelly_1.37.1     </span></span>
<span><span class="co">##  [19] KernSmooth_2.23-24     bslib_0.4.2            htmlwidgets_1.6.4     </span></span>
<span><span class="co">##  [22] desc_1.4.3             ica_1.0-3              plyr_1.8.9            </span></span>
<span><span class="co">##  [25] plotly_4.10.4          zoo_1.8-12             cachem_1.1.0          </span></span>
<span><span class="co">##  [28] igraph_2.0.3           mime_0.12              lifecycle_1.0.4       </span></span>
<span><span class="co">##  [31] pkgconfig_2.0.3        fastmap_1.2.0          fitdistrplus_1.2-1    </span></span>
<span><span class="co">##  [34] future_1.33.2          shiny_1.8.1.1          digest_0.6.36         </span></span>
<span><span class="co">##  [37] colorspace_2.1-0       patchwork_1.2.0        tensor_1.5            </span></span>
<span><span class="co">##  [40] RSpectra_0.16-1        irlba_2.3.5.1          textshaping_0.4.0     </span></span>
<span><span class="co">##  [43] labeling_0.4.3         progressr_0.14.0       fansi_1.0.6           </span></span>
<span><span class="co">##  [46] spatstat.sparse_3.1-0  httr_1.4.7             polyclip_1.10-6       </span></span>
<span><span class="co">##  [49] abind_1.4-5            compiler_4.4.0         withr_3.0.0           </span></span>
<span><span class="co">##  [52] bit64_4.0.5            fastDummies_1.7.3      highr_0.11            </span></span>
<span><span class="co">##  [55] MASS_7.3-61            tools_4.4.0            lmtest_0.9-40         </span></span>
<span><span class="co">##  [58] httpuv_1.6.15          future.apply_1.11.2    goftest_1.2-3         </span></span>
<span><span class="co">##  [61] glue_1.7.0             nlme_3.1-165           promises_1.3.0        </span></span>
<span><span class="co">##  [64] grid_4.4.0             Rtsne_0.17             cluster_2.1.6         </span></span>
<span><span class="co">##  [67] generics_0.1.3         gtable_0.3.5           spatstat.data_3.1-2   </span></span>
<span><span class="co">##  [70] labelled_2.13.0        hms_1.1.3              data.table_1.15.4     </span></span>
<span><span class="co">##  [73] utf8_1.2.4             spatstat.geom_3.2-9    RcppAnnoy_0.0.22      </span></span>
<span><span class="co">##  [76] ggrepel_0.9.5          RANN_2.6.1             pillar_1.9.0          </span></span>
<span><span class="co">##  [79] stringr_1.5.1          spam_2.10-0            RcppHNSW_0.6.0        </span></span>
<span><span class="co">##  [82] later_1.3.2            splines_4.4.0          survival_3.7-0        </span></span>
<span><span class="co">##  [85] bit_4.0.5              deldir_2.0-4           tidyselect_1.2.1      </span></span>
<span><span class="co">##  [88] miniUI_0.1.1.1         pbapply_1.7-2          knitr_1.48            </span></span>
<span><span class="co">##  [91] gridExtra_2.3          scattermore_1.2        xfun_0.45             </span></span>
<span><span class="co">##  [94] mosaicCore_0.9.4.0     matrixStats_1.3.0      stringi_1.8.4         </span></span>
<span><span class="co">##  [97] lazyeval_0.2.2         yaml_2.3.9             evaluate_0.24.0       </span></span>
<span><span class="co">## [100] codetools_0.2-20       tibble_3.2.1           cli_3.6.3             </span></span>
<span><span class="co">## [103] uwot_0.2.2             xtable_1.8-4           reticulate_1.38.0     </span></span>
<span><span class="co">## [106] systemfonts_1.1.0      munsell_0.5.1          jquerylib_0.1.4       </span></span>
<span><span class="co">## [109] Rcpp_1.0.13            globals_0.16.3         spatstat.random_3.2-3 </span></span>
<span><span class="co">## [112] png_0.1-8              parallel_4.4.0         pkgdown_2.0.7         </span></span>
<span><span class="co">## [115] dotCall64_1.1-1        listenv_0.9.1          ggridges_0.5.6        </span></span>
<span><span class="co">## [118] leiden_0.4.3.1         purrr_1.0.2</span></span></code></pre>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Yichao Hua.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
