<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Utility Tools and Functions • SeuratExtend</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Utility Tools and Functions">
<meta property="og:description" content="SeuratExtend">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">SeuratExtend</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Visualization.html">Enhanced Visualization</a>
    </li>
    <li>
      <a href="../articles/GSEA.html">Geneset Enrichment Analysis (GSEA) </a>
    </li>
    <li>
      <a href="../articles/Trajectory.html">Trajectory and Pseudotime Analysis </a>
    </li>
    <li>
      <a href="../articles/SCENIC.html">SCENIC for Gene Regulatory Networks Analysis </a>
    </li>
    <li>
      <a href="../articles/Utilities.html">Utility Tools and Functions</a>
    </li>
  </ul>
</li>
<li>
  <a href="../articles/FAQ.html">FAQ</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Utility Tools and Functions</h1>
                        <h4 data-toc-skip class="author">Yichao Hua</h4>
            
            <h4 data-toc-skip class="date">2024-07-11</h4>
      
      
      <div class="hidden name"><code>Utilities.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="table-of-contents">Table of Contents<a class="anchor" aria-label="anchor" href="#table-of-contents"></a>
</h2>
<ol style="list-style-type: decimal">
<li><a href="#facilitate-gene-naming-conversions">Facilitate Gene Naming
Conversions</a></li>
<li><a href="#compute-statistics-grouped-by-clusters">Compute Statistics
Grouped by Clusters</a></li>
<li><a href="#assess-proportion-of-positive-cells-in-clusters">Assess
Proportion of Positive Cells in Clusters</a></li>
<li><a href="#run-standard-seurat-pipeline">Run Standard Seurat
Pipeline</a></li>
</ol>
</div>
<div class="section level2">
<h2 id="facilitate-gene-naming-conversions">Facilitate Gene Naming Conversions<a class="anchor" aria-label="anchor" href="#facilitate-gene-naming-conversions"></a>
</h2>
<div class="section level3">
<h3 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h3>
<p>Gene naming conventions can vary significantly between organisms and
databases, presenting a common challenge in scRNA-seq data analysis.
<code>SeuratExtend</code> includes several functions to facilitate the
conversion between human and mouse gene symbols and Ensembl IDs, as well
as conversions between human and mouse homologous gene symbols. These
functions leverage the <code>biomaRt</code> database for conversions but
improve on reliability and performance by localizing the most commonly
used databases, thus eliminating the need for internet connectivity and
addressing the frequent instability issues with
<code>biomaRt</code>.</p>
</div>
<div class="section level3">
<h3 id="functions-for-gene-naming-conversions">Functions for Gene Naming Conversions<a class="anchor" aria-label="anchor" href="#functions-for-gene-naming-conversions"></a>
</h3>
<p>The functions provided for these conversions are:</p>
<ul>
<li><code>HumanToMouseGenesymbol</code></li>
<li><code>MouseToHumanGenesymbol</code></li>
<li><code>EnsemblToGenesymbol</code></li>
<li><code>GenesymbolToEnsembl</code></li>
</ul>
<p>These functions share a similar usage pattern, as detailed below
using <code>HumanToMouseGenesymbol</code> as an example.</p>
<div class="section level4">
<h4 id="getting-started-with-examples">Getting Started with Examples<a class="anchor" aria-label="anchor" href="#getting-started-with-examples"></a>
</h4>
<p>First, let’s retrieve a few human gene symbols from a dataset as an
example:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Loading required package: SeuratObject</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: sp</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Attaching package: 'SeuratObject'</span></span></code></pre>
<pre><code><span><span class="co">## The following objects are masked from 'package:base':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     intersect, t</span></span></code></pre>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SeuratExtend</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Loading required package: SeuratExtendData</span></span></code></pre>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">human_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/VariableFeatures.html" class="external-link">VariableFeatures</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">human_genes</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "PPBP"   "LYZ"    "S100A9" "IGLL5"  "GNLY"   "FTL"</span></span></code></pre>
</div>
<div class="section level4">
<h4 id="default-usage">Default Usage<a class="anchor" aria-label="anchor" href="#default-usage"></a>
</h4>
<p>By default, <code>HumanToMouseGenesymbol</code> returns a data frame
showing how human gene symbols (HGNC) match with mouse gene symbols
(MGI):</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/gene-naming-conversions.html">HumanToMouseGenesymbol</a></span><span class="op">(</span><span class="va">human_genes</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##          MGI.symbol HGNC.symbol</span></span>
<span><span class="co">## 227   9530003J23Rik         LYZ</span></span>
<span><span class="co">## 6304           Ftl1         FTL</span></span>
<span><span class="co">## 6305       Ftl1-ps1         FTL</span></span>
<span><span class="co">## 8244         Gm5849      S100A9</span></span>
<span><span class="co">## 15496          Ppbp        PPBP</span></span></code></pre>
<p>This table indicates that not all human genes have direct mouse
homologs, and some human genes may correspond to multiple mouse
genes.</p>
</div>
<div class="section level4">
<h4 id="simplified-output">Simplified Output<a class="anchor" aria-label="anchor" href="#simplified-output"></a>
</h4>
<p>If you prefer a simpler vector output without the matching
details:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/gene-naming-conversions.html">HumanToMouseGenesymbol</a></span><span class="op">(</span><span class="va">human_genes</span>, match <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "9530003J23Rik" "Ftl1"          "Ftl1-ps1"      "Gm5849"       </span></span>
<span><span class="co">## [5] "Ppbp"</span></span></code></pre>
</div>
<div class="section level4">
<h4 id="one-to-one-correspondence">One-to-One Correspondence<a class="anchor" aria-label="anchor" href="#one-to-one-correspondence"></a>
</h4>
<p>For cases where you require a one-to-one correspondence:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/gene-naming-conversions.html">HumanToMouseGenesymbol</a></span><span class="op">(</span><span class="va">human_genes</span>, keep.seq <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##            PPBP             LYZ          S100A9           IGLL5            GNLY </span></span>
<span><span class="co">##          "Ppbp" "9530003J23Rik"        "Gm5849"              NA              NA </span></span>
<span><span class="co">##             FTL </span></span>
<span><span class="co">##          "Ftl1"</span></span></code></pre>
</div>
<div class="section level4">
<h4 id="converting-gene-expression-matrices">Converting Gene Expression Matrices<a class="anchor" aria-label="anchor" href="#converting-gene-expression-matrices"></a>
</h4>
<p>These functions can also directly convert human gene expression
matrices to their mouse counterparts:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Create an example gene expression matrix</span></span>
<span><span class="va">human_matr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/AssayData.html" class="external-link">GetAssayData</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span><span class="op">[</span><span class="va">human_genes</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">human_matr</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 6 x 4 sparse Matrix of class "dgCMatrix"</span></span>
<span><span class="co">##        CTATAAGATCGTTT-1 GTGATTCTGGTTCA-1 ACGTTGGACCGTAA-1 GGATACTGCAGCTA-1</span></span>
<span><span class="co">## PPBP           .                .                .                .       </span></span>
<span><span class="co">## LYZ            .                .                5.203442         2.145600</span></span>
<span><span class="co">## S100A9         .                .                3.615756         .       </span></span>
<span><span class="co">## IGLL5          .                .                .                .       </span></span>
<span><span class="co">## GNLY           .                5.301497         .                .       </span></span>
<span><span class="co">## FTL            3.804611         3.441956         6.284984         3.656766</span></span></code></pre>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Convert to a mouse gene expression matrix</span></span>
<span><span class="fu"><a href="../reference/gene-naming-conversions.html">HumanToMouseGenesymbol</a></span><span class="op">(</span><span class="va">human_matr</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 4 x 4 sparse Matrix of class "dgCMatrix"</span></span>
<span><span class="co">##               CTATAAGATCGTTT-1 GTGATTCTGGTTCA-1 ACGTTGGACCGTAA-1</span></span>
<span><span class="co">## Ppbp                  .                .                .       </span></span>
<span><span class="co">## 9530003J23Rik         .                .                5.203442</span></span>
<span><span class="co">## Gm5849                .                .                3.615756</span></span>
<span><span class="co">## Ftl1                  3.804611         3.441956         6.284984</span></span>
<span><span class="co">##               GGATACTGCAGCTA-1</span></span>
<span><span class="co">## Ppbp                  .       </span></span>
<span><span class="co">## 9530003J23Rik         2.145600</span></span>
<span><span class="co">## Gm5849                .       </span></span>
<span><span class="co">## Ftl1                  3.656766</span></span></code></pre>
</div>
<div class="section level4">
<h4 id="other-gene-naming-conversion-functions">Other Gene Naming Conversion Functions<a class="anchor" aria-label="anchor" href="#other-gene-naming-conversion-functions"></a>
</h4>
<p>The usage patterns for the other conversion functions in
<code>SeuratExtend</code>, such as <code>MouseToHumanGenesymbol</code>,
<code>GenesymbolToEnsembl</code>, and <code>EnsemblToGenesymbol</code>,
are similar to those already discussed. These functions also leverage
local databases to enhance performance and reliability but provide
options to use online databases via <code>biomaRt</code> if
necessary.</p>
<p>Here are some examples demonstrating the use of other gene naming
conversion functions:</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Converting mouse gene symbols to human</span></span>
<span><span class="va">mouse_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Cd14"</span>, <span class="st">"Cd3d"</span>, <span class="st">"Cd79a"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/gene-naming-conversions.html">MouseToHumanGenesymbol</a></span><span class="op">(</span><span class="va">mouse_genes</span>, match <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "CD14"  "CD3D"  "CD79A"</span></span></code></pre>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Converting human gene symbols to Ensembl IDs</span></span>
<span><span class="va">human_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"PPBP"</span>, <span class="st">"LYZ"</span>, <span class="st">"S100A9"</span>, <span class="st">"IGLL5"</span>, <span class="st">"GNLY"</span>, <span class="st">"FTL"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/gene-naming-conversions.html">GenesymbolToEnsembl</a></span><span class="op">(</span><span class="va">human_genes</span>, spe <span class="op">=</span> <span class="st">"human"</span>, keep.seq <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##              PPBP               LYZ            S100A9             IGLL5 </span></span>
<span><span class="co">## "ENSG00000163736" "ENSG00000090382" "ENSG00000163220" "ENSG00000254709" </span></span>
<span><span class="co">##              GNLY               FTL </span></span>
<span><span class="co">## "ENSG00000115523" "ENSG00000087086"</span></span></code></pre>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Converting mouse gene symbols to Ensembl IDs</span></span>
<span><span class="fu"><a href="../reference/gene-naming-conversions.html">GenesymbolToEnsembl</a></span><span class="op">(</span><span class="va">mouse_genes</span>, spe <span class="op">=</span> <span class="st">"mouse"</span>, keep.seq <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                 Cd14                 Cd3d                Cd79a </span></span>
<span><span class="co">## "ENSMUSG00000051439" "ENSMUSG00000032094" "ENSMUSG00000003379"</span></span></code></pre>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Converting Ensembl IDs to human gene symbols</span></span>
<span><span class="fu"><a href="../reference/gene-naming-conversions.html">EnsemblToGenesymbol</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ENSG00000163736"</span>, <span class="st">"ENSG00000090382"</span><span class="op">)</span>, spe <span class="op">=</span> <span class="st">"human"</span>, keep.seq <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## ENSG00000163736 ENSG00000090382 </span></span>
<span><span class="co">##          "PPBP"           "LYZ"</span></span></code></pre>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Converting Ensembl IDs to mouse gene symbols</span></span>
<span><span class="fu"><a href="../reference/gene-naming-conversions.html">EnsemblToGenesymbol</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ENSMUSG00000051439"</span>, <span class="st">"ENSMUSG00000032094"</span><span class="op">)</span>, spe <span class="op">=</span> <span class="st">"mouse"</span>, keep.seq <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## ENSMUSG00000051439 ENSMUSG00000032094 </span></span>
<span><span class="co">##             "Cd14"             "Cd3d"</span></span></code></pre>
</div>
<div class="section level4">
<h4 id="using-online-resources">Using Online Resources<a class="anchor" aria-label="anchor" href="#using-online-resources"></a>
</h4>
<p>While <code>SeuratExtend</code> typically uses localized databases
for conversions, you have the option to directly fetch results from
<code>biomaRt</code> databases if required. This can be useful when
working with less common genes or newer annotations not yet available in
the local database:</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Fetching Ensembl IDs for human genes directly from biomaRt</span></span>
<span><span class="fu"><a href="../reference/gene-naming-conversions.html">GenesymbolToEnsembl</a></span><span class="op">(</span><span class="va">human_genes</span>, spe <span class="op">=</span> <span class="st">"human"</span>, local.mode <span class="op">=</span> <span class="cn">FALSE</span>, keep.seq <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="converting-uniprot-ids-to-gene-symbols">Converting UniProt IDs to Gene Symbols<a class="anchor" aria-label="anchor" href="#converting-uniprot-ids-to-gene-symbols"></a>
</h4>
<p>In addition to facilitating gene symbol and Ensembl ID conversions
between human and mouse, <code>SeuratExtend</code> also includes
functionality to convert UniProt IDs, which are widely used in proteomic
databases, to gene symbols. This can be particularly useful when
integrating proteomic and genomic data or when working with databases
that use UniProt identifiers.</p>
<p>The function <code>UniprotToGenesymbol</code> in
<code>SeuratExtend</code> provides a straightforward way to translate
UniProt IDs into gene symbols. This function supports both human and
mouse species, accommodating research that spans multiple types of
biological data. Here’s how you can convert UniProt IDs to gene symbols
for both human and mouse:</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Converting UniProt IDs to human gene symbols</span></span>
<span><span class="fu"><a href="../reference/UniprotToGenesymbol.html">UniprotToGenesymbol</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Q8NF67"</span>, <span class="st">"Q9NPB9"</span><span class="op">)</span>, spe <span class="op">=</span> <span class="st">"human"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "ANKRD20A12P" "ACKR4"</span></span></code></pre>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Converting UniProt IDs to mouse gene symbols</span></span>
<span><span class="fu"><a href="../reference/UniprotToGenesymbol.html">UniprotToGenesymbol</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Q9R1C8"</span>, <span class="st">"Q9QY84"</span><span class="op">)</span>, spe <span class="op">=</span> <span class="st">"mouse"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "Htr6"   "Actl7a"</span></span></code></pre>
</div>
</div>
</div>
<div class="section level2">
<h2 id="compute-statistics-grouped-by-clusters">Compute Statistics Grouped by Clusters<a class="anchor" aria-label="anchor" href="#compute-statistics-grouped-by-clusters"></a>
</h2>
<p>The <code>CalcStats</code> function from the
<code>SeuratExtend</code> package provides a comprehensive approach to
compute various statistics, such as mean, median, z-scores, or LogFC,
for genomic data. This function can handle data stored in Seurat objects
or standard matrices, allowing for versatile analyses tailored to
single-cell datasets.</p>
<p>Whether you’re analyzing genes or pathways, <code>CalcStats</code>
simplifies the task by computing statistics for selected features across
different cell groups or clusters.</p>
<div class="section level3">
<h3 id="using-a-seurat-object">Using a Seurat Object<a class="anchor" aria-label="anchor" href="#using-a-seurat-object"></a>
</h3>
<p>Begin by selecting a subset of features, such as genes. For this
example, let’s pick the first 20 variable features from a Seurat
object:</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SeuratExtend</span><span class="op">)</span></span>
<span></span>
<span><span class="va">genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/VariableFeatures.html" class="external-link">VariableFeatures</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">20</span><span class="op">]</span></span></code></pre></div>
<p>Using <code>CalcStats</code>, compute your desired metric, like
z-scores, for each feature across different cell clusters:</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">genes.zscore</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/CalcStats.html">CalcStats</a></span><span class="op">(</span><span class="va">pbmc</span>, features <span class="op">=</span> <span class="va">genes</span>, method <span class="op">=</span> <span class="st">"zscore"</span>, group.by <span class="op">=</span> <span class="st">"cluster"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">genes.zscore</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##            B cell CD4 T Memory CD4 T Naive CD8 T cell          DC  Mono CD14</span></span>
<span><span class="co">## PPBP   -0.3371314   -0.3304297  -0.3504324 -0.3241762 -0.35043243 -0.3026097</span></span>
<span><span class="co">## LYZ    -0.7086608   -0.5312264  -0.7989052 -0.6494664  1.25840432  1.8734207</span></span>
<span><span class="co">## S100A9 -0.6585180   -0.5422840  -0.5670929 -0.5281999  0.06906373  2.3873809</span></span>
<span><span class="co">## IGLL5   2.6061006   -0.2791647  -0.3973560 -0.4712923 -0.47129230 -0.4712923</span></span>
<span><span class="co">## GNLY   -0.4437914   -0.3985860  -0.4761490  0.2209286 -0.38933003 -0.2937568</span></span>
<span><span class="co">## FTL    -0.4544019   -0.8814705  -0.7728842 -0.7232855  0.05204804  1.6055627</span></span>
<span><span class="co">##        Mono FCGR3A    NK cell   Platelet</span></span>
<span><span class="co">## PPBP    -0.3206744 -0.3504324  2.6663186</span></span>
<span><span class="co">## LYZ      0.6437754 -0.7119758 -0.3753658</span></span>
<span><span class="co">## S100A9   0.7169325 -0.6237505 -0.2535318</span></span>
<span><span class="co">## IGLL5   -0.2258753 -0.4712923  0.1814646</span></span>
<span><span class="co">## GNLY    -0.4197184  2.6058127 -0.4054095</span></span>
<span><span class="co">## FTL      1.6557236 -0.7193185  0.2380262</span></span></code></pre>
<p>Display the computed statistics using a heatmap:</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/Heatmap.html">Heatmap</a></span><span class="op">(</span><span class="va">genes.zscore</span>, lab_fill <span class="op">=</span> <span class="st">"zscore"</span><span class="op">)</span></span></code></pre></div>
<p><img src="Utilities_files/figure-html/unnamed-chunk-11-1.png" width="480"></p>
<p>Select more genes and retain the top 4 genes of each cluster, sorted
by p-value. This can be a convenient method to display the top marker
genes of each cluster:</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/VariableFeatures.html" class="external-link">VariableFeatures</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span></span>
<span><span class="va">genes.zscore</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/CalcStats.html">CalcStats</a></span><span class="op">(</span></span>
<span>  <span class="va">pbmc</span>, features <span class="op">=</span> <span class="va">genes</span>, method <span class="op">=</span> <span class="st">"zscore"</span>, group.by <span class="op">=</span> <span class="st">"cluster"</span>, </span>
<span>  order <span class="op">=</span> <span class="st">"p"</span>, n <span class="op">=</span> <span class="fl">4</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/Heatmap.html">Heatmap</a></span><span class="op">(</span><span class="va">genes.zscore</span>, lab_fill <span class="op">=</span> <span class="st">"zscore"</span><span class="op">)</span></span></code></pre></div>
<p><img src="Utilities_files/figure-html/unnamed-chunk-12-1.png" width="480"></p>
</div>
<div class="section level3">
<h3 id="using-matrices-as-input">Using Matrices as Input<a class="anchor" aria-label="anchor" href="#using-matrices-as-input"></a>
</h3>
<p>For instance, you might perform Enrichment Analysis (GSEA) using the
Hallmark 50 geneset and obtain the AUCell matrix (rows represent
pathways, columns represent cells):</p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/GeneSetAnalysis.html">GeneSetAnalysis</a></span><span class="op">(</span><span class="va">pbmc</span>, genesets <span class="op">=</span> <span class="va">hall50</span><span class="op">$</span><span class="va">human</span><span class="op">)</span></span>
<span><span class="va">matr</span> <span class="op">&lt;-</span> <span class="va">pbmc</span><span class="op">@</span><span class="va">misc</span><span class="op">$</span><span class="va">AUCell</span><span class="op">$</span><span class="va">genesets</span></span></code></pre></div>
<p>Using the matrix, compute the z-scores for the genesets across
various cell clusters:</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gsea.zscore</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/CalcStats.html">CalcStats</a></span><span class="op">(</span><span class="va">matr</span>, f <span class="op">=</span> <span class="va">pbmc</span><span class="op">$</span><span class="va">cluster</span>, method <span class="op">=</span> <span class="st">"zscore"</span><span class="op">)</span></span></code></pre></div>
<p>Present the z-scores using a heatmap:</p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/Heatmap.html">Heatmap</a></span><span class="op">(</span><span class="va">gsea.zscore</span>, lab_fill <span class="op">=</span> <span class="st">"zscore"</span><span class="op">)</span></span></code></pre></div>
<p><img src="Utilities_files/figure-html/unnamed-chunk-15-1.png" width="720"></p>
</div>
</div>
<div class="section level2">
<h2 id="assess-proportion-of-positive-cells-in-clusters">Assess Proportion of Positive Cells in Clusters<a class="anchor" aria-label="anchor" href="#assess-proportion-of-positive-cells-in-clusters"></a>
</h2>
<p>This section describes how to utilize the
<code>feature_percent</code> function in the <code>SeuratExtend</code>
package to determine the proportion of positive cells within specified
clusters or groups based on defined criteria. This function is
particularly useful for identifying the expression levels of genes or
other features within subpopulations of cells in scRNA-seq datasets.</p>
<div class="section level3">
<h3 id="basic-usage">Basic Usage<a class="anchor" aria-label="anchor" href="#basic-usage"></a>
</h3>
<p>To calculate the proportion of positive cells for the top 5 variable
features in a Seurat object:</p>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SeuratExtend</span><span class="op">)</span></span>
<span></span>
<span><span class="va">genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/VariableFeatures.html" class="external-link">VariableFeatures</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># Default usage</span></span>
<span><span class="va">proportions</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/feature_percent.html">feature_percent</a></span><span class="op">(</span><span class="va">pbmc</span>, feature <span class="op">=</span> <span class="va">genes</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Attaching package: 'rlist'</span></span></code></pre>
<pre><code><span><span class="co">## The following object is masked from 'package:S4Vectors':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     List</span></span></code></pre>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">proportions</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##            B cell CD4 T Memory CD4 T Naive CD8 T cell        DC  Mono CD14</span></span>
<span><span class="co">## PPBP   0.01492537   0.02247191  0.00000000 0.02173913 0.0000000 0.05050505</span></span>
<span><span class="co">## LYZ    0.43283582   0.61797753  0.37931034 0.50000000 0.9333333 1.00000000</span></span>
<span><span class="co">## S100A9 0.05970149   0.16853933  0.12931034 0.17391304 0.4666667 1.00000000</span></span>
<span><span class="co">## IGLL5  0.20895522   0.02247191  0.00862069 0.00000000 0.0000000 0.00000000</span></span>
<span><span class="co">## GNLY   0.05970149   0.08988764  0.02586207 0.32608696 0.1333333 0.15151515</span></span>
<span><span class="co">##        Mono FCGR3A    NK cell   Platelet</span></span>
<span><span class="co">## PPBP    0.03333333 0.00000000 1.00000000</span></span>
<span><span class="co">## LYZ     1.00000000 0.45833333 0.50000000</span></span>
<span><span class="co">## S100A9  0.86666667 0.08333333 0.28571429</span></span>
<span><span class="co">## IGLL5   0.03333333 0.00000000 0.07142857</span></span>
<span><span class="co">## GNLY    0.10000000 0.95833333 0.07142857</span></span></code></pre>
<p>This will return a matrix where rows are features and columns are
clusters, showing the proportion of cells in each cluster where the
feature’s expression is above the default threshold (0).</p>
</div>
<div class="section level3">
<h3 id="adjusting-the-expression-threshold">Adjusting the Expression Threshold<a class="anchor" aria-label="anchor" href="#adjusting-the-expression-threshold"></a>
</h3>
<p>To count a cell as positive only if its expression is above a value
of 2:</p>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">proportions_above_2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/feature_percent.html">feature_percent</a></span><span class="op">(</span><span class="va">pbmc</span>, feature <span class="op">=</span> <span class="va">genes</span>, above <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">proportions_above_2</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##            B cell CD4 T Memory CD4 T Naive CD8 T cell        DC  Mono CD14</span></span>
<span><span class="co">## PPBP   0.00000000   0.00000000  0.00000000 0.02173913 0.0000000 0.02020202</span></span>
<span><span class="co">## LYZ    0.20895522   0.23595506  0.14655172 0.15217391 0.8666667 1.00000000</span></span>
<span><span class="co">## S100A9 0.00000000   0.01123596  0.03448276 0.02173913 0.2666667 1.00000000</span></span>
<span><span class="co">## IGLL5  0.08955224   0.00000000  0.00000000 0.00000000 0.0000000 0.00000000</span></span>
<span><span class="co">## GNLY   0.00000000   0.00000000  0.00862069 0.23913043 0.0000000 0.07070707</span></span>
<span><span class="co">##        Mono FCGR3A   NK cell   Platelet</span></span>
<span><span class="co">## PPBP     0.0000000 0.0000000 1.00000000</span></span>
<span><span class="co">## LYZ      0.9000000 0.1250000 0.42857143</span></span>
<span><span class="co">## S100A9   0.6333333 0.0000000 0.14285714</span></span>
<span><span class="co">## IGLL5    0.0000000 0.0000000 0.00000000</span></span>
<span><span class="co">## GNLY     0.0000000 0.9583333 0.07142857</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="targeting-specific-clusters">Targeting Specific Clusters<a class="anchor" aria-label="anchor" href="#targeting-specific-clusters"></a>
</h3>
<p>To calculate proportions for only a subset of clusters:</p>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">proportions_subset</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/feature_percent.html">feature_percent</a></span><span class="op">(</span><span class="va">pbmc</span>, feature <span class="op">=</span> <span class="va">genes</span>, ident <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"B cell"</span>, <span class="st">"CD8 T cell"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">proportions_subset</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##            B cell CD8 T cell</span></span>
<span><span class="co">## PPBP   0.01492537 0.02173913</span></span>
<span><span class="co">## LYZ    0.43283582 0.50000000</span></span>
<span><span class="co">## S100A9 0.05970149 0.17391304</span></span>
<span><span class="co">## IGLL5  0.20895522 0.00000000</span></span>
<span><span class="co">## GNLY   0.05970149 0.32608696</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="grouping-by-different-metadata">Grouping by Different Metadata<a class="anchor" aria-label="anchor" href="#grouping-by-different-metadata"></a>
</h3>
<p>If you wish to group cells by a different variable other than the
default cluster identities:</p>
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">proportions_by_ident</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/feature_percent.html">feature_percent</a></span><span class="op">(</span><span class="va">pbmc</span>, feature <span class="op">=</span> <span class="va">genes</span>, group.by <span class="op">=</span> <span class="st">"orig.ident"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">proportions_by_ident</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##           sample1    sample2</span></span>
<span><span class="co">## PPBP   0.03571429 0.05421687</span></span>
<span><span class="co">## LYZ    0.60119048 0.63554217</span></span>
<span><span class="co">## S100A9 0.35119048 0.36445783</span></span>
<span><span class="co">## IGLL5  0.03571429 0.03915663</span></span>
<span><span class="co">## GNLY   0.13690476 0.15361446</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="proportions-relative-to-total-cell-numbers">Proportions Relative to Total Cell Numbers<a class="anchor" aria-label="anchor" href="#proportions-relative-to-total-cell-numbers"></a>
</h3>
<p>To also check the proportion of expressed cells in total across
selected clusters:</p>
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">proportions_total</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/feature_percent.html">feature_percent</a></span><span class="op">(</span><span class="va">pbmc</span>, feature <span class="op">=</span> <span class="va">genes</span>, total <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">proportions_total</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##            B cell CD4 T Memory CD4 T Naive CD8 T cell        DC  Mono CD14</span></span>
<span><span class="co">## PPBP   0.01492537   0.02247191  0.00000000 0.02173913 0.0000000 0.05050505</span></span>
<span><span class="co">## LYZ    0.43283582   0.61797753  0.37931034 0.50000000 0.9333333 1.00000000</span></span>
<span><span class="co">## S100A9 0.05970149   0.16853933  0.12931034 0.17391304 0.4666667 1.00000000</span></span>
<span><span class="co">## IGLL5  0.20895522   0.02247191  0.00862069 0.00000000 0.0000000 0.00000000</span></span>
<span><span class="co">## GNLY   0.05970149   0.08988764  0.02586207 0.32608696 0.1333333 0.15151515</span></span>
<span><span class="co">##        Mono FCGR3A    NK cell   Platelet total</span></span>
<span><span class="co">## PPBP    0.03333333 0.00000000 1.00000000 0.048</span></span>
<span><span class="co">## LYZ     1.00000000 0.45833333 0.50000000 0.624</span></span>
<span><span class="co">## S100A9  0.86666667 0.08333333 0.28571429 0.360</span></span>
<span><span class="co">## IGLL5   0.03333333 0.00000000 0.07142857 0.038</span></span>
<span><span class="co">## GNLY    0.10000000 0.95833333 0.07142857 0.148</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="logical-output-for-expression">Logical Output for Expression<a class="anchor" aria-label="anchor" href="#logical-output-for-expression"></a>
</h3>
<p>For scenarios where you need a logical output indicating whether a
significant proportion of cells are expressing the feature above a
certain level (e.g., 20%):</p>
<div class="sourceCode" id="cb56"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">expressed_logical</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/feature_percent.html">feature_percent</a></span><span class="op">(</span><span class="va">pbmc</span>, feature <span class="op">=</span> <span class="va">genes</span>, if.expressed <span class="op">=</span> <span class="cn">TRUE</span>, min.pct <span class="op">=</span> <span class="fl">0.2</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">expressed_logical</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##        B cell CD4 T Memory CD4 T Naive CD8 T cell    DC Mono CD14 Mono FCGR3A</span></span>
<span><span class="co">## PPBP    FALSE        FALSE       FALSE      FALSE FALSE     FALSE       FALSE</span></span>
<span><span class="co">## LYZ      TRUE         TRUE        TRUE       TRUE  TRUE      TRUE        TRUE</span></span>
<span><span class="co">## S100A9  FALSE        FALSE       FALSE      FALSE  TRUE      TRUE        TRUE</span></span>
<span><span class="co">## IGLL5    TRUE        FALSE       FALSE      FALSE FALSE     FALSE       FALSE</span></span>
<span><span class="co">## GNLY    FALSE        FALSE       FALSE       TRUE FALSE     FALSE       FALSE</span></span>
<span><span class="co">##        NK cell Platelet</span></span>
<span><span class="co">## PPBP     FALSE     TRUE</span></span>
<span><span class="co">## LYZ       TRUE     TRUE</span></span>
<span><span class="co">## S100A9   FALSE     TRUE</span></span>
<span><span class="co">## IGLL5    FALSE    FALSE</span></span>
<span><span class="co">## GNLY      TRUE    FALSE</span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="run-standard-seurat-pipeline">Run Standard Seurat Pipeline<a class="anchor" aria-label="anchor" href="#run-standard-seurat-pipeline"></a>
</h2>
<p>The <code>RunBasicSeurat</code> function in the
<code>SeuratExtend</code> package automates the execution of a standard
Seurat pipeline for single-cell RNA sequencing data analysis. This
comprehensive function includes steps such as normalization, PCA,
clustering, and optionally integrates batch effects using Harmony. This
automation is designed to streamline the analysis process, making it
more efficient and reproducible.</p>
<div class="section level3">
<h3 id="overview-of-the-pipeline">Overview of the Pipeline<a class="anchor" aria-label="anchor" href="#overview-of-the-pipeline"></a>
</h3>
<p>The Seurat pipeline typically includes the following steps, which are
all encapsulated within the <code>RunBasicSeurat</code> function:</p>
<ol style="list-style-type: decimal">
<li>
<strong>Calculating Percent Mitochondrial Content</strong>:
Identifying and filtering cells based on the proportion of mitochondrial
genes, which is a common quality control metric.</li>
<li>
<strong>Normalization</strong>: Scaling data to account for
cell-specific differences in library size.</li>
<li>
<strong>PCA</strong>: Performing principal component analysis to
reduce dimensionality and highlight the major sources of variation.</li>
<li>
<strong>Clustering</strong>: Grouping cells based on their gene
expression profiles to identify distinct cell types or states.</li>
<li>
<strong>UMAP Visualization</strong>: Projecting the high-dimensional
data into two dimensions for visualization.</li>
<li>
<strong>Batch Integration (Optional)</strong>: Using Harmony to
correct for batch effects, ensuring that variations driven by
experimental conditions are minimized.</li>
</ol>
<p>For a comprehensive tutorial on the standard Seurat workflow, refer
to the official <a href="https://satijalab.org/seurat/articles/pbmc3k_tutorial" class="external-link">Seurat PBMC
tutorial</a>.</p>
</div>
<div class="section level3">
<h3 id="using-runbasicseurat">Using <code>RunBasicSeurat</code><a class="anchor" aria-label="anchor" href="#using-runbasicseurat"></a>
</h3>
<p>Below are examples demonstrating how to use the
<code>RunBasicSeurat</code> function to process scRNA-seq data:</p>
<div class="sourceCode" id="cb58"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SeuratExtend</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Run the full pipeline with forced normalization and default parameters</span></span>
<span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RunBasicSeurat.html">RunBasicSeurat</a></span><span class="op">(</span><span class="va">pbmc</span>, force.Normalize <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Centering and scaling data matrix</span></span></code></pre>
<pre><code><span><span class="co">## PC_ 1 </span></span>
<span><span class="co">## Positive:  CST3, TYROBP, FTH1, LST1, AIF1, FCER1G, FTL, CFD, TYMP, LYZ </span></span>
<span><span class="co">##     S100A9, LGALS1, FCN1, SPI1, CD68, COTL1, PSAP, CTSS, SERPINA1, SAT1 </span></span>
<span><span class="co">##     S100A11, IFITM3, AP1S2, IFI30, S100A8, LGALS2, NPC2, LGALS3, GPX1, OAZ1 </span></span>
<span><span class="co">## Negative:  MALAT1, RPS27A, LTB, IL32, TPT1, CXCR4, IL7R, B2M, CTSW, RARRES3 </span></span>
<span><span class="co">##     GZMA, TRAF3IP3, NOSIP, CST7, PRDX2, MYL12A, AQP3, RPL34, FAIM3, GIMAP5 </span></span>
<span><span class="co">##     PPP2R5C, GIMAP7, MAL, PRF1, CD8B, ITM2A, CCL5, HOPX, SAMD3, OPTN </span></span>
<span><span class="co">## PC_ 2 </span></span>
<span><span class="co">## Positive:  AP001189.4, GP9, ACRBP, TMEM40, CLDN5, PTCRA, CLEC1B, LY6G6F, AC147651.3, TUBA8 </span></span>
<span><span class="co">##     PCP2, SDPR, HIST1H2AC, TSC22D1, C2orf88, PF4, CMTM5, PPBP, GNG11, ESAM </span></span>
<span><span class="co">##     MMD, SPOCD1, GP6, TMCC2, ENKUR, ASAP2, AC137932.6, LGALSL, MYLK, LCN2 </span></span>
<span><span class="co">## Negative:  LYZ, S100A9, LGALS2, FCN1, TYROBP, AIF1, IFI30, CST3, S100A8, LST1 </span></span>
<span><span class="co">##     MS4A6A, CTSS, CD14, NCF2, LGALS1, AP1S2, FTL, S100A6, S100A11, RGS2 </span></span>
<span><span class="co">##     TYMP, PYCARD, IFITM3, FCER1G, CD68, FTH1, ALDH2, PSAP, CYBA, CTSB </span></span>
<span><span class="co">## PC_ 3 </span></span>
<span><span class="co">## Positive:  NKG7, GZMA, CST7, PRF1, B2M, CTSW, S100A4, GNLY, FGFBP2, GZMB </span></span>
<span><span class="co">##     KLRD1, SPON2, GZMH, CCL4, CCL5, XCL2, PFN1, FCGR3A, HOPX, RARRES3 </span></span>
<span><span class="co">##     IL32, CLIC3, TMSB4X, XCL1, AKR1C3, S1PR5, TPST2, GIMAP7, SRGN, ITGB2 </span></span>
<span><span class="co">## Negative:  CD79A, HLA-DQA1, TCL1A, HLA-DQB1, MS4A1, LINC00926, HLA-DRA, CD79B, CD74, HLA-DPB1 </span></span>
<span><span class="co">##     VPREB3, HLA-DPA1, FCER2, HLA-DRB5, HLA-DQA2, HLA-DRB1, HLA-DMA, CD37, TSPAN13, KIAA0125 </span></span>
<span><span class="co">##     HLA-DOB, BLNK, SPIB, PKIG, FCRLA, BLK, BTK, PNOC, CD180, PDLIM1 </span></span>
<span><span class="co">## PC_ 4 </span></span>
<span><span class="co">## Positive:  GZMB, SERPINF1, CLIC3, PLD4, LILRA4, NKG7, FGFBP2, GNLY, CLEC4C, CST7 </span></span>
<span><span class="co">##     PRF1, MZB1, KLRD1, SPON2, GZMH, HLA-DQA1, GZMA, PLAC8, FCGR3A, IRF8 </span></span>
<span><span class="co">##     TIFAB, CD74, PTGDS, IL3RA, CCL4, TSPAN13, XCL2, IGJ, HLA-DPA1, C12orf75 </span></span>
<span><span class="co">## Negative:  IL7R, S100A8, VIM, S100A9, MAL, S100A10, CD40LG, CD14, NOSIP, C6orf48 </span></span>
<span><span class="co">##     LGALS2, GIMAP5, RGS10, AQP3, ANP32B, LTB, FLT3LG, GIMAP4, IL32, RBP7 </span></span>
<span><span class="co">##     NGFRAP1, TMSB4X, LGALS3BP, FHIT, NDFIP1, FOLR3, FCN1, AIF1, AC013264.2, GIMAP7 </span></span>
<span><span class="co">## PC_ 5 </span></span>
<span><span class="co">## Positive:  FCGR3A, CDKN1C, CKB, SIGLEC10, HES4, MS4A7, RHOC, CD79B, CTD-2006K23.1, LILRA3 </span></span>
<span><span class="co">##     RP11-290F20.3, CSF1R, MS4A4A, LRRC25, IFITM2, PAPSS2, LILRB1, FAM110A, BATF3, VMO1 </span></span>
<span><span class="co">##     PPM1N, EMR2, CXCL16, TESC, MTSS1, INSIG1, CEACAM3, ZNF703, GSTA4, EMR1 </span></span>
<span><span class="co">## Negative:  SERPINF1, LILRA4, GPX1, CLEC4C, PPP1R14B, GAS6, TIFAB, LGALS2, GRN, SCT </span></span>
<span><span class="co">##     CUEDC1, LRRC26, S100A8, MS4A6A, IL3RA, APP, SMPD3, ALDH2, GSN, GSTP1 </span></span>
<span><span class="co">##     RPS6KA4, CD14, CD33, FAM213A, ZNF789, ZFAT, LYZ, ASGR1, LAMP5, VIM</span></span></code></pre>
<pre><code><span><span class="co">## Using 'pca' as the reduction method for FindNeighbors, FindClusters, and RunUMAP.</span></span></code></pre>
<pre><code><span><span class="co">## Computing nearest neighbor graph</span></span></code></pre>
<pre><code><span><span class="co">## Computing SNN</span></span></code></pre>
<pre><code><span><span class="co">## Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Number of nodes: 452</span></span>
<span><span class="co">## Number of edges: 14235</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Running Louvain algorithm...</span></span>
<span><span class="co">## Maximum modularity in 10 random starts: 0.8333</span></span>
<span><span class="co">## Number of communities: 5</span></span>
<span><span class="co">## Elapsed time: 0 seconds</span></span></code></pre>
<pre><code><span><span class="co">## Warning: The default method for RunUMAP has changed from calling Python UMAP via reticulate to the R-native UWOT using the cosine metric</span></span>
<span><span class="co">## To use Python UMAP via reticulate, set umap.method to 'umap-learn' and metric to 'correlation'</span></span>
<span><span class="co">## This message will be shown once per session</span></span></code></pre>
<pre><code><span><span class="co">## 10:22:00 UMAP embedding parameters a = 0.9922 b = 1.112</span></span></code></pre>
<pre><code><span><span class="co">## 10:22:00 Read 452 rows and found 10 numeric columns</span></span></code></pre>
<pre><code><span><span class="co">## 10:22:00 Using Annoy for neighbor search, n_neighbors = 30</span></span></code></pre>
<pre><code><span><span class="co">## 10:22:00 Building Annoy index with metric = cosine, n_trees = 50</span></span></code></pre>
<pre><code><span><span class="co">## 0%   10   20   30   40   50   60   70   80   90   100%</span></span></code></pre>
<pre><code><span><span class="co">## [----|----|----|----|----|----|----|----|----|----|</span></span></code></pre>
<pre><code><span><span class="co">## **************************************************|</span></span>
<span><span class="co">## 10:22:00 Writing NN index file to temp file /tmp/RtmpCvLAoI/file34781171227952</span></span>
<span><span class="co">## 10:22:00 Searching Annoy index using 1 thread, search_k = 3000</span></span>
<span><span class="co">## 10:22:00 Annoy recall = 100%</span></span>
<span><span class="co">## 10:22:01 Commencing smooth kNN distance calibration using 1 thread with target n_neighbors = 30</span></span>
<span><span class="co">## 10:22:02 Initializing from normalized Laplacian + noise (using RSpectra)</span></span>
<span><span class="co">## 10:22:02 Commencing optimization for 500 epochs, with 16790 positive edges</span></span>
<span><span class="co">## 10:22:02 Optimization finished</span></span></code></pre>
<div class="sourceCode" id="cb73"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Visualize the clusters using DimPlot</span></span>
<span><span class="fu"><a href="../reference/DimPlot2.html">DimPlot2</a></span><span class="op">(</span><span class="va">pbmc</span>, group.by <span class="op">=</span> <span class="st">"cluster"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## The 'I want hue' color presets were generated from: https://medialab.github.io/iwanthue/</span></span>
<span><span class="co">## This message is shown once per session</span></span>
<span><span class="co">## Loading required package: cowplot</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Attaching package: 'cowplot'</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## The following object is masked from 'package:mosaic':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     theme_map</span></span></code></pre>
<p><img src="Utilities_files/figure-html/unnamed-chunk-22-1.png" width="624"></p>
<div class="section level4">
<h4 id="parameters-and-customization">Parameters and Customization<a class="anchor" aria-label="anchor" href="#parameters-and-customization"></a>
</h4>
<p>The function allows for extensive customization of each step through
various parameters:</p>
<ul>
<li>
<strong><code>spe</code></strong>: Specifies the species (human or
mouse) for mitochondrial calculations.</li>
<li>
<strong><code>nFeature_RNA.min</code> and
<code>nFeature_RNA.max</code></strong>: Define the range of RNA features
considered for each cell.</li>
<li>
<strong><code>percent.mt.max</code></strong>: Sets the maximum
allowed mitochondrial gene expression percentage.</li>
<li>
<strong><code>dims</code></strong>: Determines the number of
dimensions used in PCA and neighbor finding.</li>
<li>
<strong><code>resolution</code></strong>: Adjusts the granularity of
the clustering algorithm.</li>
<li>
<strong><code>reduction</code></strong>: Chooses the dimensional
reduction technique, with options for PCA or Harmony.</li>
<li>
<strong><code>harmony.by</code></strong>: Specifies the metadata
column for batch correction when using Harmony.</li>
</ul>
</div>
</div>
<div class="section level3">
<h3 id="conditional-execution">Conditional Execution<a class="anchor" aria-label="anchor" href="#conditional-execution"></a>
</h3>
<p><code>RunBasicSeurat</code> intelligently decides whether to re-run
certain steps based on parameter changes or previous executions:</p>
<ul>
<li>
<strong><code>force.*</code> Parameters</strong>: Each
<code>force</code> parameter (e.g., <code>force.Normalize</code>,
<code>force.RunPCA</code>) overrides the function’s internal checks,
ensuring that specific steps are executed regardless of prior results.
This feature is particularly useful when parameters are adjusted or when
updates to the dataset require reanalysis.</li>
</ul>
</div>
<div class="section level3">
<h3 id="conclusion">Conclusion<a class="anchor" aria-label="anchor" href="#conclusion"></a>
</h3>
<p>The <code>RunBasicSeurat</code> function simplifies the execution of
a comprehensive scRNA-seq data analysis pipeline, incorporating advanced
features such as conditional execution and batch effect integration.
This function ensures that users can efficiently process their data
while maintaining flexibility to adapt the analysis to specific
requirements.</p>
<div class="sourceCode" id="cb75"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.4.0 (2024-04-24)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">## Running under: Ubuntu 20.04.6 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.9.0 </span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.9.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">##  [3] LC_TIME=de_BE.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">##  [5] LC_MONETARY=de_BE.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=de_BE.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">## [11] LC_MEASUREMENT=de_BE.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: Europe/Brussels</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">## [8] base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] cowplot_1.1.3             rlist_0.4.6.2            </span></span>
<span><span class="co">##  [3] DelayedMatrixStats_1.26.0 DelayedArray_0.30.1      </span></span>
<span><span class="co">##  [5] SparseArray_1.4.8         S4Arrays_1.4.1           </span></span>
<span><span class="co">##  [7] abind_1.4-5               IRanges_2.38.0           </span></span>
<span><span class="co">##  [9] S4Vectors_0.42.0          MatrixGenerics_1.16.0    </span></span>
<span><span class="co">## [11] matrixStats_1.3.0         BiocGenerics_0.50.0      </span></span>
<span><span class="co">## [13] rlang_1.1.4               scales_1.3.0             </span></span>
<span><span class="co">## [15] reshape2_1.4.4            mosaic_1.9.1             </span></span>
<span><span class="co">## [17] mosaicData_0.20.4         ggformula_0.12.0         </span></span>
<span><span class="co">## [19] Matrix_1.7-0              ggplot2_3.5.1            </span></span>
<span><span class="co">## [21] lattice_0.22-6            dplyr_1.1.4              </span></span>
<span><span class="co">## [23] SeuratExtend_1.0.0        SeuratExtendData_0.2.0   </span></span>
<span><span class="co">## [25] Seurat_5.1.0              SeuratObject_5.0.2       </span></span>
<span><span class="co">## [27] sp_2.1-4                 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] RColorBrewer_1.1-3       rstudioapi_0.16.0        jsonlite_1.8.8          </span></span>
<span><span class="co">##   [4] magrittr_2.0.3           spatstat.utils_3.0-4     farver_2.1.2            </span></span>
<span><span class="co">##   [7] rmarkdown_2.27           zlibbioc_1.50.0          fs_1.6.4                </span></span>
<span><span class="co">##  [10] ragg_1.3.2               vctrs_0.6.5              ROCR_1.0-11             </span></span>
<span><span class="co">##  [13] memoise_2.0.1            spatstat.explore_3.2-7   forcats_1.0.0           </span></span>
<span><span class="co">##  [16] htmltools_0.5.8.1        haven_2.5.4              sass_0.4.9              </span></span>
<span><span class="co">##  [19] sctransform_0.4.1        parallelly_1.37.1        KernSmooth_2.23-24      </span></span>
<span><span class="co">##  [22] bslib_0.7.0              htmlwidgets_1.6.4        desc_1.4.3              </span></span>
<span><span class="co">##  [25] ica_1.0-3                plyr_1.8.9               plotly_4.10.4           </span></span>
<span><span class="co">##  [28] zoo_1.8-12               cachem_1.1.0             igraph_1.3.4            </span></span>
<span><span class="co">##  [31] mime_0.12                lifecycle_1.0.4          pkgconfig_2.0.3         </span></span>
<span><span class="co">##  [34] R6_2.5.1                 fastmap_1.2.0            fitdistrplus_1.1-11     </span></span>
<span><span class="co">##  [37] future_1.33.2            shiny_1.8.1.1            digest_0.6.35           </span></span>
<span><span class="co">##  [40] colorspace_2.1-0         patchwork_1.2.0          tensor_1.5              </span></span>
<span><span class="co">##  [43] RSpectra_0.16-1          irlba_2.3.5.1            textshaping_0.4.0       </span></span>
<span><span class="co">##  [46] labeling_0.4.3           progressr_0.14.0         fansi_1.0.6             </span></span>
<span><span class="co">##  [49] spatstat.sparse_3.0-3    httr_1.4.7               polyclip_1.10-6         </span></span>
<span><span class="co">##  [52] compiler_4.4.0           withr_3.0.0              fastDummies_1.7.3       </span></span>
<span><span class="co">##  [55] highr_0.11               MASS_7.3-60.2            tools_4.4.0             </span></span>
<span><span class="co">##  [58] lmtest_0.9-40            httpuv_1.6.15            future.apply_1.11.2     </span></span>
<span><span class="co">##  [61] goftest_1.2-3            glue_1.7.0               nlme_3.1-165            </span></span>
<span><span class="co">##  [64] promises_1.3.0           grid_4.4.0               Rtsne_0.17              </span></span>
<span><span class="co">##  [67] cluster_2.1.6            generics_0.1.3           gtable_0.3.5            </span></span>
<span><span class="co">##  [70] spatstat.data_3.0-4      labelled_2.13.0          tidyr_1.3.1             </span></span>
<span><span class="co">##  [73] hms_1.1.3                data.table_1.15.4        XVector_0.44.0          </span></span>
<span><span class="co">##  [76] utf8_1.2.4               spatstat.geom_3.2-9      RcppAnnoy_0.0.22        </span></span>
<span><span class="co">##  [79] ggrepel_0.9.5            RANN_2.6.1               pillar_1.9.0            </span></span>
<span><span class="co">##  [82] stringr_1.5.1            spam_2.10-0              RcppHNSW_0.6.0          </span></span>
<span><span class="co">##  [85] later_1.3.2              splines_4.4.0            survival_3.7-0          </span></span>
<span><span class="co">##  [88] deldir_2.0-4             tidyselect_1.2.1         miniUI_0.1.1.1          </span></span>
<span><span class="co">##  [91] pbapply_1.7-2            knitr_1.47               gridExtra_2.3           </span></span>
<span><span class="co">##  [94] scattermore_1.2          xfun_0.44                mosaicCore_0.9.4.0      </span></span>
<span><span class="co">##  [97] stringi_1.8.4            lazyeval_0.2.2           yaml_2.3.8              </span></span>
<span><span class="co">## [100] evaluate_0.23            codetools_0.2-20         tibble_3.2.1            </span></span>
<span><span class="co">## [103] cli_3.6.2                uwot_0.2.2               xtable_1.8-4            </span></span>
<span><span class="co">## [106] reticulate_1.37.0        systemfonts_1.1.0        munsell_0.5.1           </span></span>
<span><span class="co">## [109] jquerylib_0.1.4          Rcpp_1.0.12              globals_0.16.3          </span></span>
<span><span class="co">## [112] spatstat.random_3.2-3    png_0.1-8                parallel_4.4.0          </span></span>
<span><span class="co">## [115] pkgdown_2.0.9            dotCall64_1.1-1          sparseMatrixStats_1.16.0</span></span>
<span><span class="co">## [118] listenv_0.9.1            viridisLite_0.4.2        ggridges_0.5.6          </span></span>
<span><span class="co">## [121] crayon_1.5.2             leiden_0.4.3.1           purrr_1.0.2</span></span></code></pre>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Yichao Hua.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
