<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Enhanced Visualization • SeuratExtend</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Enhanced Visualization">
<meta property="og:description" content="SeuratExtend">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">SeuratExtend</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Visualization.html">Enhanced Visualization</a>
    </li>
    <li>
      <a href="../articles/GSEA.html">Geneset Enrichment Analysis (GSEA) </a>
    </li>
    <li>
      <a href="../articles/Trajectory.html">Trajectory and Pseudotime Analysis </a>
    </li>
    <li>
      <a href="../articles/SCENIC.html">SCENIC for Gene Regulatory Networks Analysis </a>
    </li>
    <li>
      <a href="../articles/Utilities.html">Utility Tools and Functions</a>
    </li>
  </ul>
</li>
<li>
  <a href="../articles/FAQ.html">FAQ</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Enhanced Visualization</h1>
                        <h4 data-toc-skip class="author">Yichao Hua</h4>
            
            <h4 data-toc-skip class="date">2024-07-11</h4>
      
      
      <div class="hidden name"><code>Visualization.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="table-of-contents">Table of Contents<a class="anchor" aria-label="anchor" href="#table-of-contents"></a>
</h2>
<ol style="list-style-type: decimal">
<li><a href="#generate-a-heatmap-plot">Generate a Heatmap Plot</a></li>
<li><a href="#create-an-enhanced-dimensional-reduction-plot">Create an
Enhanced Dimensional Reduction Plot</a></li>
<li><a href="#create-an-enhanced-violin-plot">Create an Enhanced Violin
Plot</a></li>
<li><a href="#visualize-cluster-distribution-in-samples">Visualize
Cluster Distribution in Samples</a></li>
<li><a href="#generate-a-waterfall-plot">Generate a Waterfall
Plot</a></li>
<li><a href="#explore-color-functions">Explore Color Functions</a></li>
</ol>
</div>
<div class="section level2">
<h2 id="generate-a-heatmap-plot">Generate a Heatmap Plot<a class="anchor" aria-label="anchor" href="#generate-a-heatmap-plot"></a>
</h2>
<div class="section level3">
<h3 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h3>
<p>The <code>Heatmap</code> function provides a flexible and
comprehensive way to visualize matrices, especially those produced by
the <code>CalcStats</code> function. This vignette provides a quick
overview of how to utilize the various features and capabilities of the
<code>Heatmap</code> function to generate customized visualizations.</p>
</div>
<div class="section level3">
<h3 id="basic-usage">Basic Usage<a class="anchor" aria-label="anchor" href="#basic-usage"></a>
</h3>
<p>First, let’s generate a sample matrix using the
<code>CalcStats</code> function:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SeuratExtend</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Assuming pbmc data and VariableFeatures function are available</span></span>
<span><span class="va">genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/VariableFeatures.html" class="external-link">VariableFeatures</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span></span>
<span><span class="va">toplot</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/CalcStats.html">CalcStats</a></span><span class="op">(</span><span class="va">pbmc</span>, features <span class="op">=</span> <span class="va">genes</span>, method <span class="op">=</span> <span class="st">"zscore"</span>, order <span class="op">=</span> <span class="st">"p"</span>, n <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span></code></pre></div>
<p>Now, we can produce a basic heatmap:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/Heatmap.html">Heatmap</a></span><span class="op">(</span><span class="va">toplot</span>, lab_fill <span class="op">=</span> <span class="st">"zscore"</span><span class="op">)</span></span></code></pre></div>
<p><img src="Visualization_files/figure-html/unnamed-chunk-2-1.png" width="480"></p>
</div>
<div class="section level3">
<h3 id="customizing-the-heatmap">Customizing the Heatmap<a class="anchor" aria-label="anchor" href="#customizing-the-heatmap"></a>
</h3>
<div class="section level4">
<h4 id="adjusting-color-schemes">Adjusting Color Schemes<a class="anchor" aria-label="anchor" href="#adjusting-color-schemes"></a>
</h4>
<p>The <code>color_scheme</code> parameter allows for flexibility in
visualizing data. Here are some ways to change the color theme of your
heatmap:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># White to dark green</span></span>
<span><span class="fu"><a href="../reference/Heatmap.html">Heatmap</a></span><span class="op">(</span><span class="va">toplot</span>, lab_fill <span class="op">=</span> <span class="st">"zscore"</span>, color_scheme <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"white"</span>, <span class="fu"><a href="https://scales.r-lib.org/reference/muted.html" class="external-link">muted</a></span><span class="op">(</span><span class="st">"green"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="Visualization_files/figure-html/unnamed-chunk-3-1.png" width="480"></p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Dark blue to light yellow (centered at 0) to dark red</span></span>
<span><span class="fu"><a href="../reference/Heatmap.html">Heatmap</a></span><span class="op">(</span><span class="va">toplot</span>, lab_fill <span class="op">=</span> <span class="st">"zscore"</span>, color_scheme <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>  low <span class="op">=</span> <span class="fu"><a href="https://scales.r-lib.org/reference/muted.html" class="external-link">muted</a></span><span class="op">(</span><span class="st">"blue"</span><span class="op">)</span>,</span>
<span>  mid <span class="op">=</span> <span class="st">"lightyellow"</span>,</span>
<span>  high <span class="op">=</span> <span class="fu"><a href="https://scales.r-lib.org/reference/muted.html" class="external-link">muted</a></span><span class="op">(</span><span class="st">"red"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="Visualization_files/figure-html/unnamed-chunk-3-2.png" width="480"></p>
<p>You can also use predefined color schemes, such as those from the
viridis package:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/Heatmap.html">Heatmap</a></span><span class="op">(</span><span class="va">toplot</span>, lab_fill <span class="op">=</span> <span class="st">"zscore"</span>, color_scheme <span class="op">=</span> <span class="st">"A"</span><span class="op">)</span></span></code></pre></div>
<p><img src="Visualization_files/figure-html/unnamed-chunk-4-1.png" width="480"></p>
</div>
<div class="section level4">
<h4 id="modifying-axis-and-labels">Modifying Axis and Labels<a class="anchor" aria-label="anchor" href="#modifying-axis-and-labels"></a>
</h4>
<p>Sometimes, the first name on the x-axis might be too long and exceed
the left boundary of the plot. To prevent this issue and ensure all
labels are fully visible, you can increase the space on the left side of
the plot by adjusting the <code>plot.margin</code> parameter. For
example, to add more space, you can specify a larger value for the left
margin (<code>l</code>) like this:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/Heatmap.html">Heatmap</a></span><span class="op">(</span><span class="va">toplot</span>, lab_fill <span class="op">=</span> <span class="st">"zscore"</span>, plot.margin <span class="op">=</span> <span class="fu">margin</span><span class="op">(</span>l <span class="op">=</span> <span class="fl">30</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="Visualization_files/figure-html/unnamed-chunk-5-1.png" width="480"></p>
<p>For denser matrices, you may wish to only show a subset of gene
names:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">toplot2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/CalcStats.html">CalcStats</a></span><span class="op">(</span><span class="va">pbmc</span>, features <span class="op">=</span> <span class="va">genes</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">500</span><span class="op">]</span>, method <span class="op">=</span> <span class="st">"zscore"</span>, order <span class="op">=</span> <span class="st">"p"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/Heatmap.html">Heatmap</a></span><span class="op">(</span><span class="va">toplot2</span>, lab_fill <span class="op">=</span> <span class="st">"zscore"</span>, feature_text_subset <span class="op">=</span> <span class="va">genes</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">20</span><span class="op">]</span>, expand_limits_x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">0.5</span>, <span class="fl">11</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="Visualization_files/figure-html/unnamed-chunk-6-1.png" width="480"></p>
</div>
<div class="section level4">
<h4 id="faceting-the-heatmap">Faceting the Heatmap<a class="anchor" aria-label="anchor" href="#faceting-the-heatmap"></a>
</h4>
<p>You can also split the heatmap based on gene groups:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gene_groups</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"group1"</span>, <span class="st">"group2"</span>, <span class="st">"group3"</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">toplot2</span><span class="op">)</span>, replace <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/Heatmap.html">Heatmap</a></span><span class="op">(</span><span class="va">toplot2</span>, lab_fill <span class="op">=</span> <span class="st">"zscore"</span>, facet_row <span class="op">=</span> <span class="va">gene_groups</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme</span><span class="op">(</span>axis.text.y <span class="op">=</span> <span class="fu">element_blank</span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="Visualization_files/figure-html/unnamed-chunk-7-1.png" width="480"></p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="create-an-enhanced-dimensional-reduction-plot">Create an Enhanced Dimensional Reduction Plot<a class="anchor" aria-label="anchor" href="#create-an-enhanced-dimensional-reduction-plot"></a>
</h2>
<div class="section level3">
<h3 id="introduction-1">Introduction<a class="anchor" aria-label="anchor" href="#introduction-1"></a>
</h3>
<p>In Seurat, dimension reduction plots such as UMAP are typically
created using <code>DimPlot</code> for discrete variables and
<code>FeaturePlot</code> for continuous variables.
<code>SeuratExtend</code> simplifies this process with
<code>DimPlot2</code>, which does not require differentiation between
variable types. This function automatically recognizes the type of input
parameters, whether discrete or continuous. <code>DimPlot2</code>
retains most of the usage conventions of both <code>DimPlot</code> and
<code>FeaturePlot</code>, allowing for an easy transition if you are
accustomed to the original Seurat functions. Additionally,
<code>DimPlot2</code> introduces numerous extra parameters to enrich the
customization of the plots.</p>
</div>
<div class="section level3">
<h3 id="basic-usage-1">Basic Usage<a class="anchor" aria-label="anchor" href="#basic-usage-1"></a>
</h3>
<p>To generate a basic dimension reduction plot, simply call
<code>DimPlot2</code> with your Seurat object:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SeuratExtend</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/DimPlot2.html">DimPlot2</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span></span></code></pre></div>
<p><img src="Visualization_files/figure-html/unnamed-chunk-8-1.png" width="624"></p>
</div>
<div class="section level3">
<h3 id="visualizing-different-variables">Visualizing Different Variables<a class="anchor" aria-label="anchor" href="#visualizing-different-variables"></a>
</h3>
<p><code>DimPlot2</code> can handle both discrete and continuous
variables seamlessly. Here’s how to input different variables into the
plot:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/DimPlot2.html">DimPlot2</a></span><span class="op">(</span><span class="va">pbmc</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"cluster"</span>, <span class="st">"orig.ident"</span>, <span class="st">"CD14"</span>, <span class="st">"CD3D"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="Visualization_files/figure-html/unnamed-chunk-9-1.png" width="864"></p>
</div>
<div class="section level3">
<h3 id="splitting-by-variable">Splitting by Variable<a class="anchor" aria-label="anchor" href="#splitting-by-variable"></a>
</h3>
<p>You can also split the visualization by a specific variable, which is
particularly useful for comparative analysis across conditions or
identities:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/DimPlot2.html">DimPlot2</a></span><span class="op">(</span><span class="va">pbmc</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"cluster"</span>, <span class="st">"CD14"</span><span class="op">)</span>, split.by <span class="op">=</span> <span class="st">"orig.ident"</span>, ncol <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<p><img src="Visualization_files/figure-html/unnamed-chunk-10-1.png" width="672"></p>
</div>
<div class="section level3">
<h3 id="highlighting-specific-cells">Highlighting Specific Cells<a class="anchor" aria-label="anchor" href="#highlighting-specific-cells"></a>
</h3>
<p>To highlight cells of interest, such as a specific cluster, you can
define the cells explicitly and use them in your plot:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">b_cells</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span><span class="op">[</span><span class="va">pbmc</span><span class="op">$</span><span class="va">cluster</span> <span class="op">==</span> <span class="st">"B cell"</span><span class="op">]</span></span>
<span><span class="fu"><a href="../reference/DimPlot2.html">DimPlot2</a></span><span class="op">(</span><span class="va">pbmc</span>, cells.highlight <span class="op">=</span> <span class="va">b_cells</span><span class="op">)</span></span></code></pre></div>
<p><img src="Visualization_files/figure-html/unnamed-chunk-11-1.png" width="624"></p>
</div>
<div class="section level3">
<h3 id="advanced-customization">Advanced Customization<a class="anchor" aria-label="anchor" href="#advanced-customization"></a>
</h3>
<p>For each variable, you can specify custom colors, adjust themes, and
more. For detailed information on color customization, refer to the <a href="#explore-color-functions">Explore Color Functions</a> section:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/DimPlot2.html">DimPlot2</a></span><span class="op">(</span></span>
<span>  <span class="va">pbmc</span>,</span>
<span>  features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"cluster"</span>, <span class="st">"orig.ident"</span>, <span class="st">"CD14"</span>, <span class="st">"CD3D"</span><span class="op">)</span>,</span>
<span>  cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>    <span class="st">"cluster"</span> <span class="op">=</span> <span class="st">"pro_blue"</span>,</span>
<span>    <span class="st">"CD14"</span> <span class="op">=</span> <span class="st">"D"</span>,</span>
<span>    <span class="st">"CD3D"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"#EEEEEE"</span>, <span class="st">"black"</span><span class="op">)</span></span>
<span>  <span class="op">)</span>,</span>
<span>  theme <span class="op">=</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SeuratTheme.html" class="external-link">NoAxes</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="Visualization_files/figure-html/unnamed-chunk-12-1.png" width="864"></p>
</div>
<div class="section level3">
<h3 id="adding-labels-and-boxes">Adding Labels and Boxes<a class="anchor" aria-label="anchor" href="#adding-labels-and-boxes"></a>
</h3>
<p>To further enhance the plot, you can add labels and bounding boxes to
clearly delineate different groups or points of interest:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/DimPlot2.html">DimPlot2</a></span><span class="op">(</span><span class="va">pbmc</span>, label <span class="op">=</span> <span class="cn">TRUE</span>, box <span class="op">=</span> <span class="cn">TRUE</span>, label.color <span class="op">=</span> <span class="st">"black"</span>, repel <span class="op">=</span> <span class="cn">TRUE</span>, theme <span class="op">=</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SeuratTheme.html" class="external-link">NoLegend</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="Visualization_files/figure-html/unnamed-chunk-13-1.png" width="480"></p>
</div>
<div class="section level3">
<h3 id="simplifying-labels-with-indices">Simplifying Labels with Indices<a class="anchor" aria-label="anchor" href="#simplifying-labels-with-indices"></a>
</h3>
<p>Sometimes, cluster names are too lengthy and can make the plot appear
cluttered when displayed with labels. To address this, consider using
indices to replace the cluster names in the plot, which helps make the
visualization cleaner. For instance, you can label clusters as ‘C1’,
‘C2’, etc., on the plot itself, while detailing what each index stands
for (e.g., ‘C1: B cell’, ‘C2: CD4 T Memory’) in the figure legend:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/DimPlot2.html">DimPlot2</a></span><span class="op">(</span><span class="va">pbmc</span>, index.title <span class="op">=</span> <span class="st">"C"</span>, box <span class="op">=</span> <span class="cn">TRUE</span>, label.color <span class="op">=</span> <span class="st">"black"</span><span class="op">)</span></span></code></pre></div>
<p><img src="Visualization_files/figure-html/unnamed-chunk-14-1.png" width="624"></p>
<p>This approach ensures that the plot remains legible and aesthetically
pleasing, even when dealing with numerous or complex labels.</p>
</div>
<div class="section level3">
<h3 id="simultaneous-display-of-three-features-on-a-dimension-reduction-plot">Simultaneous Display of Three Features on a Dimension Reduction
Plot<a class="anchor" aria-label="anchor" href="#simultaneous-display-of-three-features-on-a-dimension-reduction-plot"></a>
</h3>
<p>In <code>SeuratExtend</code>, a unique visualization method allows
for the simultaneous display of three features on the same dimension
reduction plot. The functions <code>FeaturePlot3</code> and
<code>FeaturePlot3.grid</code> employ a color mixing system (either RYB
or RGB) to represent three different genes (or other continuous
variables). This method uses the principles of color mixing to
quantitatively display the expression levels or intensities of these
three features in each cell.</p>
<div class="section level4">
<h4 id="ryb-and-rgb-color-systems">RYB and RGB Color Systems<a class="anchor" aria-label="anchor" href="#ryb-and-rgb-color-systems"></a>
</h4>
<p>In the RGB system, black represents no or low expression, and
brighter colors indicate higher levels: <img src="Visualization_files/RGB_mixing.png" alt="RGB System Example"></p>
<p>In the RYB system, white represents no expression, and deeper colors
indicate higher expression levels: <img src="Visualization_files/RYB_mixing.png" alt="RYB System Example"></p>
</div>
<div class="section level4">
<h4 id="examples-using-ryb-and-rgb-systems">Examples Using RYB and RGB Systems<a class="anchor" aria-label="anchor" href="#examples-using-ryb-and-rgb-systems"></a>
</h4>
<p>Here’s how to display three markers using the RYB system, with red
for CD3D, yellow for CD14, and blue for CD79A:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/FeaturePlot3.html">FeaturePlot3</a></span><span class="op">(</span><span class="va">pbmc</span>, color <span class="op">=</span> <span class="st">"ryb"</span>, feature.1 <span class="op">=</span> <span class="st">"CD3D"</span>, feature.2 <span class="op">=</span> <span class="st">"CD14"</span>, feature.3 <span class="op">=</span> <span class="st">"CD79A"</span>, pt.size <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></span></code></pre></div>
<p><img src="Visualization_files/figure-html/unnamed-chunk-15-1.png" width="576"></p>
<p>For the RGB system, with red for CD3D, green for CD14, and blue for
CD79A:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/FeaturePlot3.html">FeaturePlot3</a></span><span class="op">(</span><span class="va">pbmc</span>, color <span class="op">=</span> <span class="st">"rgb"</span>, feature.1 <span class="op">=</span> <span class="st">"CD3D"</span>, feature.2 <span class="op">=</span> <span class="st">"CD14"</span>, feature.3 <span class="op">=</span> <span class="st">"CD79A"</span>, pt.size <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<p><img src="Visualization_files/figure-html/unnamed-chunk-16-1.png" width="576"></p>
</div>
<div class="section level4">
<h4 id="batch-visualization-with-featureplot3-grid">Batch Visualization with <code>FeaturePlot3.grid</code><a class="anchor" aria-label="anchor" href="#batch-visualization-with-featureplot3-grid"></a>
</h4>
<p><code>FeaturePlot3.grid</code> extends <code>FeaturePlot3</code> by
allowing multiple plots to be generated in one go. The
<code>features</code> parameter requires a vector where every three
values are assigned a color (RYB or RGB) and placed together in one
plot. If you wish to skip a color, use <code>NA</code> as a
placeholder.</p>
<p>For instance, to place the following five genes into two plots using
the RYB system, and skip yellow in the second plot:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/FeaturePlot3-grid.html">FeaturePlot3.grid</a></span><span class="op">(</span><span class="va">pbmc</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD3D"</span>, <span class="st">"CD14"</span>, <span class="st">"CD79A"</span>, <span class="st">"FCGR3A"</span>, <span class="cn">NA</span>, <span class="st">"LYZ"</span><span class="op">)</span>, pt.size <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></span></code></pre></div>
<p><img src="Visualization_files/figure-html/unnamed-chunk-17-1.png" width="624"></p>
<p>Using the RGB system:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/FeaturePlot3-grid.html">FeaturePlot3.grid</a></span><span class="op">(</span><span class="va">pbmc</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD3D"</span>, <span class="st">"CD14"</span>, <span class="st">"CD79A"</span>, <span class="st">"FCGR3A"</span>, <span class="cn">NA</span>, <span class="st">"LYZ"</span><span class="op">)</span>, color <span class="op">=</span> <span class="st">"rgb"</span>, pt.size <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<p><img src="Visualization_files/figure-html/unnamed-chunk-18-1.png" width="624"></p>
</div>
<div class="section level4">
<h4 id="tips-on-point-size">Tips on Point Size<a class="anchor" aria-label="anchor" href="#tips-on-point-size"></a>
</h4>
<p>The background is usually white, so the choice of color system and
point size can significantly affect visual perception. In the RYB
system, where higher expression results in darker colors, a smaller
<code>pt.size</code> is preferable to prevent overlapping points. In
contrast, in the RGB system, higher expressions result in lighter
colors, potentially leading to visibility issues for highly expressed
cells that may blend into the white background. Here, a larger
<code>pt.size</code> is recommended so that the darker, low-expression
points can form a “background” to highlight the lighter, high-expression
points.</p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="create-an-enhanced-violin-plot">Create an Enhanced Violin Plot<a class="anchor" aria-label="anchor" href="#create-an-enhanced-violin-plot"></a>
</h2>
<div class="section level3">
<h3 id="introduction-2">Introduction<a class="anchor" aria-label="anchor" href="#introduction-2"></a>
</h3>
<p>The <code>VlnPlot2</code> function from the <code>SeuratExtend</code>
package offers a revamped version of the traditional violin plot,
designed to be more space-efficient while introducing a wide array of
additional visualization features. Unlike the original
<code>VlnPlot</code> in Seurat, the enhanced <code>VlnPlot2</code>
integrates functionalities to superimpose boxplots, easily add
statistical annotations, and offers greater flexibility in the plot
presentation.</p>
<p>This function has been optimized for visualizing multiple variables
and can handle both Seurat objects and matrices.</p>
</div>
<div class="section level3">
<h3 id="usage">Usage<a class="anchor" aria-label="anchor" href="#usage"></a>
</h3>
<p>Depending on your input, whether it’s a Seurat object or a matrix,
the method to employ <code>VlnPlot2</code> will differ.</p>
<div class="section level4">
<h4 id="using-a-seurat-object">Using a Seurat Object<a class="anchor" aria-label="anchor" href="#using-a-seurat-object"></a>
</h4>
<p><strong>Basic violin plot with box plot and points:</strong> To begin
with, select the genes you intend to analyze. Here’s an example using
three genes:</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SeuratExtend</span><span class="op">)</span></span>
<span></span>
<span><span class="va">genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD3D"</span>,<span class="st">"CD14"</span>,<span class="st">"CD79A"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/VlnPlot2.html">VlnPlot2</a></span><span class="op">(</span><span class="va">pbmc</span>, features <span class="op">=</span> <span class="va">genes</span>, ncol <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<p><img src="Visualization_files/figure-html/unnamed-chunk-19-1.png" width="480"></p>
<p><strong>Customizing plot elements:</strong> The function allows for
versatile visual alterations. For instance, one might want to omit the
violin plot while retaining the box plot, using a quasirandom style for
point adjustment.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/VlnPlot2.html">VlnPlot2</a></span><span class="op">(</span><span class="va">pbmc</span>, features <span class="op">=</span> <span class="va">genes</span>, violin <span class="op">=</span> <span class="cn">F</span>, pt.style <span class="op">=</span> <span class="st">"quasirandom"</span>, ncol <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<p><img src="Visualization_files/figure-html/unnamed-chunk-20-1.png" width="480"></p>
<p><strong>Hiding data points but retaining outliers:</strong></p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/VlnPlot2.html">VlnPlot2</a></span><span class="op">(</span><span class="va">pbmc</span>, features <span class="op">=</span> <span class="va">genes</span>, pt <span class="op">=</span> <span class="cn">FALSE</span>, ncol <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<p><img src="Visualization_files/figure-html/unnamed-chunk-21-1.png" width="480"></p>
<p><strong>Hide points and outliers for a cleaner
appearance:</strong></p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/VlnPlot2.html">VlnPlot2</a></span><span class="op">(</span><span class="va">pbmc</span>, features <span class="op">=</span> <span class="va">genes</span>, pt <span class="op">=</span> <span class="cn">FALSE</span>, hide.outlier <span class="op">=</span> <span class="cn">T</span>, ncol <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<p><img src="Visualization_files/figure-html/unnamed-chunk-22-1.png" width="480"></p>
<p><strong>Grouping by cluster and splitting each cluster by
samples:</strong></p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/VlnPlot2.html">VlnPlot2</a></span><span class="op">(</span><span class="va">pbmc</span>, features <span class="op">=</span> <span class="va">genes</span>, group.by <span class="op">=</span> <span class="st">"cluster"</span>, split.by <span class="op">=</span> <span class="st">"orig.ident"</span><span class="op">)</span></span></code></pre></div>
<p><img src="Visualization_files/figure-html/unnamed-chunk-23-1.png" width="864"></p>
<p><strong>Filtering for certain subtypes and arranging plots in
columns:</strong></p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cells</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span><span class="op">[</span><span class="va">pbmc</span><span class="op">$</span><span class="va">cluster</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"B cell"</span>, <span class="st">"Mono CD14"</span>, <span class="st">"CD8 T cell"</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="fu"><a href="../reference/VlnPlot2.html">VlnPlot2</a></span><span class="op">(</span><span class="va">pbmc</span>, features <span class="op">=</span> <span class="va">genes</span>, group.by <span class="op">=</span> <span class="st">"cluster"</span>, cells <span class="op">=</span> <span class="va">cells</span><span class="op">)</span></span></code></pre></div>
<p><img src="Visualization_files/figure-html/unnamed-chunk-24-1.png" width="576"></p>
<p><strong>Adding statistical annotations using the wilcoxon
test:</strong></p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/VlnPlot2.html">VlnPlot2</a></span><span class="op">(</span><span class="va">pbmc</span>, features <span class="op">=</span> <span class="va">genes</span>, group.by <span class="op">=</span> <span class="st">"cluster"</span>, cell <span class="op">=</span> <span class="va">cells</span>, </span>
<span>         stat.method <span class="op">=</span> <span class="st">"wilcox.test"</span>, hide.ns <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="Visualization_files/figure-html/unnamed-chunk-25-1.png" width="576"></p>
<p><strong>Restricting statistical comparisons and using
t-test:</strong></p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/VlnPlot2.html">VlnPlot2</a></span><span class="op">(</span><span class="va">pbmc</span>, features <span class="op">=</span> <span class="va">genes</span>, group.by <span class="op">=</span> <span class="st">"cluster"</span>, cell <span class="op">=</span> <span class="va">cells</span>, </span>
<span>         stat.method <span class="op">=</span> <span class="st">"t.test"</span>, comparisons <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">3</span><span class="op">)</span><span class="op">)</span>, hide.ns <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p><img src="Visualization_files/figure-html/unnamed-chunk-26-1.png" width="576"></p>
</div>
<div class="section level4">
<h4 id="using-a-matrix">Using a Matrix<a class="anchor" aria-label="anchor" href="#using-a-matrix"></a>
</h4>
<p>For an example employing a matrix input, let’s consider you have
performed a Geneset Enrichment Analysis (GSEA) using the Hallmark 50
geneset to get the AUCell matrix:</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/GeneSetAnalysis.html">GeneSetAnalysis</a></span><span class="op">(</span><span class="va">pbmc</span>, genesets <span class="op">=</span> <span class="va">hall50</span><span class="op">$</span><span class="va">human</span><span class="op">)</span></span>
<span><span class="va">matr</span> <span class="op">&lt;-</span> <span class="va">pbmc</span><span class="op">@</span><span class="va">misc</span><span class="op">$</span><span class="va">AUCell</span><span class="op">$</span><span class="va">genesets</span></span>
<span></span>
<span><span class="co"># Plotting the first three pathways:</span></span>
<span><span class="fu"><a href="../reference/VlnPlot2.html">VlnPlot2</a></span><span class="op">(</span><span class="va">matr</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span>,<span class="op">]</span>, f <span class="op">=</span> <span class="va">pbmc</span><span class="op">$</span><span class="va">cluster</span>, ncol <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<p><img src="Visualization_files/figure-html/unnamed-chunk-27-1.png" width="576"></p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="visualize-cluster-distribution-in-samples">Visualize Cluster Distribution in Samples<a class="anchor" aria-label="anchor" href="#visualize-cluster-distribution-in-samples"></a>
</h2>
<div class="section level3">
<h3 id="introduction-3">Introduction<a class="anchor" aria-label="anchor" href="#introduction-3"></a>
</h3>
<p>The <code>ClusterDistrBar</code> function is designed to visualize
the distribution of clusters across different samples. It can show both
absolute counts and proportions, and it allows for various
customizations including axis reversal and normalization.</p>
</div>
<div class="section level3">
<h3 id="basic-usage-2">Basic Usage<a class="anchor" aria-label="anchor" href="#basic-usage-2"></a>
</h3>
<p>To create a basic bar plot showing the distribution of clusters
within samples, simply specify the origin (sample identifier) and
cluster variables from your dataset:</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/ClusterDistrBar.html">ClusterDistrBar</a></span><span class="op">(</span>origin <span class="op">=</span> <span class="va">pbmc</span><span class="op">$</span><span class="va">orig.ident</span>, cluster <span class="op">=</span> <span class="va">pbmc</span><span class="op">$</span><span class="va">cluster</span><span class="op">)</span></span></code></pre></div>
<p><img src="Visualization_files/figure-html/unnamed-chunk-28-1.png" width="672"></p>
</div>
<div class="section level3">
<h3 id="displaying-absolute-cell-counts">Displaying Absolute Cell Counts<a class="anchor" aria-label="anchor" href="#displaying-absolute-cell-counts"></a>
</h3>
<p>If you prefer to visualize the absolute cell count rather than
proportions, set the <code>percent</code> parameter to
<code>FALSE</code>:</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/ClusterDistrBar.html">ClusterDistrBar</a></span><span class="op">(</span>origin <span class="op">=</span> <span class="va">pbmc</span><span class="op">$</span><span class="va">orig.ident</span>, cluster <span class="op">=</span> <span class="va">pbmc</span><span class="op">$</span><span class="va">cluster</span>, percent <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p><img src="Visualization_files/figure-html/unnamed-chunk-29-1.png" width="672"></p>
</div>
<div class="section level3">
<h3 id="normalized-proportions-with-reversed-axes">Normalized Proportions with Reversed Axes<a class="anchor" aria-label="anchor" href="#normalized-proportions-with-reversed-axes"></a>
</h3>
<p>For a clearer view that normalizes the data by sample size and
reverses the x and y axes, use the <code>rev</code> and
<code>normalize</code> parameters:</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/ClusterDistrBar.html">ClusterDistrBar</a></span><span class="op">(</span>origin <span class="op">=</span> <span class="va">pbmc</span><span class="op">$</span><span class="va">orig.ident</span>, cluster <span class="op">=</span> <span class="va">pbmc</span><span class="op">$</span><span class="va">cluster</span>, rev <span class="op">=</span> <span class="cn">TRUE</span>, normalize <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="Visualization_files/figure-html/unnamed-chunk-30-1.png" width="672"></p>
</div>
<div class="section level3">
<h3 id="non-normalized-with-reversed-axes">Non-Normalized with Reversed Axes<a class="anchor" aria-label="anchor" href="#non-normalized-with-reversed-axes"></a>
</h3>
<p>To reverse the axes without normalizing by sample size:</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/ClusterDistrBar.html">ClusterDistrBar</a></span><span class="op">(</span>origin <span class="op">=</span> <span class="va">pbmc</span><span class="op">$</span><span class="va">orig.ident</span>, cluster <span class="op">=</span> <span class="va">pbmc</span><span class="op">$</span><span class="va">cluster</span>, rev <span class="op">=</span> <span class="cn">TRUE</span>, normalize <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p><img src="Visualization_files/figure-html/unnamed-chunk-31-1.png" width="672"></p>
</div>
<div class="section level3">
<h3 id="vertical-bar-plot">Vertical Bar Plot<a class="anchor" aria-label="anchor" href="#vertical-bar-plot"></a>
</h3>
<p>If a vertical orientation is preferred over the default horizontal
bars, set the <code>flip</code> parameter to <code>FALSE</code>:</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/ClusterDistrBar.html">ClusterDistrBar</a></span><span class="op">(</span>origin <span class="op">=</span> <span class="va">pbmc</span><span class="op">$</span><span class="va">orig.ident</span>, cluster <span class="op">=</span> <span class="va">pbmc</span><span class="op">$</span><span class="va">cluster</span>, flip <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p><img src="Visualization_files/figure-html/unnamed-chunk-32-1.png" width="384"></p>
</div>
<div class="section level3">
<h3 id="non-stacked-bar-plot">Non-Stacked Bar Plot<a class="anchor" aria-label="anchor" href="#non-stacked-bar-plot"></a>
</h3>
<p>If you prefer not to stack the bars, which can be useful for direct
comparisons of cluster sizes across samples, set the <code>stack</code>
parameter to <code>FALSE</code>:</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/ClusterDistrBar.html">ClusterDistrBar</a></span><span class="op">(</span>origin <span class="op">=</span> <span class="va">pbmc</span><span class="op">$</span><span class="va">orig.ident</span>, cluster <span class="op">=</span> <span class="va">pbmc</span><span class="op">$</span><span class="va">cluster</span>, flip <span class="op">=</span> <span class="cn">FALSE</span>, stack <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p><img src="Visualization_files/figure-html/unnamed-chunk-33-1.png" width="672"></p>
</div>
<div class="section level3">
<h3 id="exporting-data-matrix">Exporting Data Matrix<a class="anchor" aria-label="anchor" href="#exporting-data-matrix"></a>
</h3>
<p>In cases where a visual plot is not required and only the underlying
data matrix is needed, set the <code>plot</code> parameter to
<code>FALSE</code>:</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data_matrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ClusterDistrBar.html">ClusterDistrBar</a></span><span class="op">(</span>origin <span class="op">=</span> <span class="va">pbmc</span><span class="op">$</span><span class="va">orig.ident</span>, cluster <span class="op">=</span> <span class="va">pbmc</span><span class="op">$</span><span class="va">cluster</span>, plot <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="co"># View the matrix</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">data_matrix</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                sample1   sample2</span></span>
<span><span class="co">## B cell       16.071429 12.048193</span></span>
<span><span class="co">## CD4 T Memory 20.238095 16.566265</span></span>
<span><span class="co">## CD4 T Naive  25.000000 22.289157</span></span>
<span><span class="co">## CD8 T cell    4.166667 11.746988</span></span>
<span><span class="co">## DC            1.785714  3.614458</span></span>
<span><span class="co">## Mono CD14    19.642857 19.879518</span></span>
<span><span class="co">## Mono FCGR3A   5.952381  6.024096</span></span>
<span><span class="co">## NK cell       4.761905  4.819277</span></span>
<span><span class="co">## Platelet      2.380952  3.012048</span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="generate-a-waterfall-plot">Generate a Waterfall Plot<a class="anchor" aria-label="anchor" href="#generate-a-waterfall-plot"></a>
</h2>
<div class="section level3">
<h3 id="introduction-4">Introduction<a class="anchor" aria-label="anchor" href="#introduction-4"></a>
</h3>
<p>Waterfall plots are powerful visualization tools that can display
differences between two conditions, showing gene expression, gene set
enrichment, or other metrics. This function can handle inputs directly
from Seurat objects or pre-processed matrices.</p>
</div>
<div class="section level3">
<h3 id="preparing-the-data">Preparing the Data<a class="anchor" aria-label="anchor" href="#preparing-the-data"></a>
</h3>
<p>First, create a matrix to visualize using the
<code><a href="../reference/GeneSetAnalysis.html">GeneSetAnalysis()</a></code> function. In this example, rows represent
gene sets from the Hallmark 50, and columns represent individual cells.
If you have already created this matrix in the violin plot section, you
can skip this step.</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SeuratExtend</span><span class="op">)</span></span>
<span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/GeneSetAnalysis.html">GeneSetAnalysis</a></span><span class="op">(</span><span class="va">pbmc</span>, genesets <span class="op">=</span> <span class="va">hall50</span><span class="op">$</span><span class="va">human</span><span class="op">)</span></span>
<span><span class="va">matr</span> <span class="op">&lt;-</span> <span class="va">pbmc</span><span class="op">@</span><span class="va">misc</span><span class="op">$</span><span class="va">AUCell</span><span class="op">$</span><span class="va">genesets</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="basic-waterfall-plot">Basic Waterfall Plot<a class="anchor" aria-label="anchor" href="#basic-waterfall-plot"></a>
</h3>
<p>Generate a basic waterfall plot to compare two cell types, such as
‘CD14+ Mono’ with ‘CD8 T cells’:</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/WaterfallPlot.html">WaterfallPlot</a></span><span class="op">(</span><span class="va">matr</span>, f <span class="op">=</span> <span class="va">pbmc</span><span class="op">$</span><span class="va">cluster</span>, ident.1 <span class="op">=</span> <span class="st">"Mono CD14"</span>, ident.2 <span class="op">=</span> <span class="st">"CD8 T cell"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Attaching package: 'tidyr'</span></span></code></pre>
<pre><code><span><span class="co">## The following object is masked from 'package:S4Vectors':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     expand</span></span></code></pre>
<pre><code><span><span class="co">## The following object is masked from 'package:magrittr':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     extract</span></span></code></pre>
<pre><code><span><span class="co">## The following object is masked from 'package:reshape2':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     smiths</span></span></code></pre>
<pre><code><span><span class="co">## The following objects are masked from 'package:Matrix':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     expand, pack, unpack</span></span></code></pre>
<p><img src="Visualization_files/figure-html/unnamed-chunk-36-1.png" width="672"></p>
</div>
<div class="section level3">
<h3 id="filtering-by-bar-length">Filtering by Bar Length<a class="anchor" aria-label="anchor" href="#filtering-by-bar-length"></a>
</h3>
<p>To focus on significant differences, you can filter the plot to
include only bars exceeding a specific threshold. For instance, keeping
only bars with a length (t-score in this instance) greater than 1:</p>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/WaterfallPlot.html">WaterfallPlot</a></span><span class="op">(</span><span class="va">matr</span>, f <span class="op">=</span> <span class="va">pbmc</span><span class="op">$</span><span class="va">cluster</span>, ident.1 <span class="op">=</span> <span class="st">"Mono CD14"</span>, ident.2 <span class="op">=</span> <span class="st">"CD8 T cell"</span>, len.threshold <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<p><img src="Visualization_files/figure-html/unnamed-chunk-37-1.png" width="672"></p>
</div>
<div class="section level3">
<h3 id="comparing-gene-expression">Comparing Gene Expression<a class="anchor" aria-label="anchor" href="#comparing-gene-expression"></a>
</h3>
<p>You can also use the waterfall plot to compare expression levels of
genes directly from a Seurat object, using LogFC to determine the bar
length. Here’s how to do it for the top 100 variable features:</p>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/VariableFeatures.html" class="external-link">VariableFeatures</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">80</span><span class="op">]</span></span>
<span><span class="fu"><a href="../reference/WaterfallPlot.html">WaterfallPlot</a></span><span class="op">(</span></span>
<span>  <span class="va">pbmc</span>, group.by <span class="op">=</span> <span class="st">"cluster"</span>, features <span class="op">=</span> <span class="va">genes</span>,</span>
<span>  ident.1 <span class="op">=</span> <span class="st">"Mono CD14"</span>, ident.2 <span class="op">=</span> <span class="st">"CD8 T cell"</span>, length <span class="op">=</span> <span class="st">"logFC"</span><span class="op">)</span></span></code></pre></div>
<p><img src="Visualization_files/figure-html/unnamed-chunk-38-1.png" width="960"></p>
</div>
<div class="section level3">
<h3 id="focusing-on-extremes">Focusing on Extremes<a class="anchor" aria-label="anchor" href="#focusing-on-extremes"></a>
</h3>
<p>To further hone in on the most differentially expressed genes, you
might want to keep only the top and bottom 20 genes. This can highlight
the most critical differences between the two cell types:</p>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/WaterfallPlot.html">WaterfallPlot</a></span><span class="op">(</span></span>
<span>  <span class="va">pbmc</span>, group.by <span class="op">=</span> <span class="st">"cluster"</span>, features <span class="op">=</span> <span class="va">genes</span>,</span>
<span>  ident.1 <span class="op">=</span> <span class="st">"Mono CD14"</span>, ident.2 <span class="op">=</span> <span class="st">"CD8 T cell"</span>, length <span class="op">=</span> <span class="st">"logFC"</span>,</span>
<span>  top.n <span class="op">=</span> <span class="fl">20</span><span class="op">)</span></span></code></pre></div>
<p><img src="Visualization_files/figure-html/unnamed-chunk-39-1.png" width="672"></p>
</div>
</div>
<div class="section level2">
<h2 id="explore-color-functions">Explore Color Functions<a class="anchor" aria-label="anchor" href="#explore-color-functions"></a>
</h2>
<p>In this section, we will delve into the various color functions and
their applications within the <code>SeuratExtend</code> package. The
discussion is divided into three main parts:</p>
<ol style="list-style-type: decimal">
<li>Introduction to the discrete color palette generation functions
<code>color_pro</code> and <code>color_iwh</code>, which have presets
for 2-50 colors in different styles.</li>
<li>Usage of color-related parameters (such as <code>cols</code> or
<code>col_theme</code>) in visualization functions like
<code>DimPlot2</code>, <code>VlnPlot2</code>, <code>Heatmap</code>, and
<code>WaterfallPlot</code>.</li>
<li>Additional color-related functions, including a custom algorithm for
blending RYB colors.</li>
</ol>
<div class="section level3">
<h3 id="professional-discrete-color-presets-with-color_pro">Professional Discrete Color Presets with <code>color_pro</code><a class="anchor" aria-label="anchor" href="#professional-discrete-color-presets-with-color_pro"></a>
</h3>
<p>The <code>color_pro</code> function is designed to generate
professional discrete color presets, ideal for data science
visualizations, particularly in fields like scRNA-seq analysis where
aesthetics must not compromise the clarity and seriousness of scientific
communication.</p>
<div class="section level4">
<h4 id="the-philosophy-behind-color_pro">The Philosophy Behind <code>color_pro</code><a class="anchor" aria-label="anchor" href="#the-philosophy-behind-color_pro"></a>
</h4>
<p>Choosing the right colors for scientific visualizations is crucial.
Colors must be distinct enough to differentiate data points clearly but
coordinated and subdued enough to maintain professionalism and avoid
visual strain. Here are some examples of what to avoid in scientific
plotting:</p>
<ol style="list-style-type: decimal">
<li>
<p><strong>Coordinated but Indistinct Colors:</strong> Using
monochromatic schemes can reduce visual distinction, which might cause
data points to blend together.</p>
<p>Example of an inadvisable choice:</p>
</li>
</ol>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span>   <span class="fu"><a href="../reference/DimPlot2.html">DimPlot2</a></span><span class="op">(</span><span class="va">pbmc</span>, cols <span class="op">=</span> <span class="st">"Greens"</span><span class="op">)</span></span></code></pre></div>
<p><img src="Visualization_files/figure-html/unnamed-chunk-40-1.png" width="624"></p>
<ol start="2" style="list-style-type: decimal">
<li>
<p><strong>Sufficiently Distinct but Overly Saturated
Colors:</strong> High saturation can be visually aggressive and
distracting, detracting from the scientific message.</p>
<p>Example of overly saturated colors:</p>
</li>
</ol>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span>   <span class="fu"><a href="../reference/DimPlot2.html">DimPlot2</a></span><span class="op">(</span><span class="va">pbmc</span>, cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"#ccffaa"</span>,<span class="st">"#c00bff"</span>,<span class="st">"#cfdb00"</span>,<span class="st">"#0147ee"</span>,<span class="st">"#f67900"</span>,<span class="st">"#1b002c"</span>,<span class="st">"#00e748"</span>,<span class="st">"#e30146"</span>,<span class="st">"#ffb1e8"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="Visualization_files/figure-html/unnamed-chunk-41-1.png" width="624"></p>
<ol start="3" style="list-style-type: decimal">
<li>
<p><strong>Good Distinction and Coordination but Too
Lively:</strong> While certain vibrant schemes might be engaging in an
advertising context, they may be considered too informal for
professional journal standards.</p>
<p>Example of colors that might be too lively:</p>
</li>
</ol>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span>   <span class="fu"><a href="../reference/DimPlot2.html">DimPlot2</a></span><span class="op">(</span><span class="va">pbmc</span>, cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"#ff2026"</span>,<span class="st">"#cf5d00"</span>,<span class="st">"#ffd03f"</span>,<span class="st">"#649f00"</span>,<span class="st">"#a3f83d"</span>,<span class="st">"#82cc58"</span>,<span class="st">"#6645fe"</span>,<span class="st">"#d8009c"</span>,<span class="st">"#ff43a2"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="Visualization_files/figure-html/unnamed-chunk-42-1.png" width="624"></p>
<p>While the RColorBrewer package offers some good solutions, its
options are limited and support a maximum of only 12 colors. This can be
inadequate for visualizing data with a larger number of clusters. The
default ggplot color palette, derived from hue_pal(), can assign an
arbitrary number of colors, but similarly suffers from insufficient
distinction when many colors are used. This is because the default
palette differentiates colors only based on hue, without utilizing
luminance and saturation, which limits its effectiveness. To address
these limitations, <code>SeuratExtend</code> provides
<code>color_pro</code>, which includes seven color schemes: “default”,
“light”, “red”, “yellow”, “green”, “blue”, and “purple”. These presets
are generated using the algorithm from I Want Hue (<a href="http://medialab.github.io/iwanthue/" class="external-link uri">http://medialab.github.io/iwanthue/</a>) with adjusted
parameters, which is optimized for creating color palettes that are
visually pleasing and distinctly separable.</p>
</div>
<div class="section level4">
<h4 id="default-color-scheme">Default Color Scheme<a class="anchor" aria-label="anchor" href="#default-color-scheme"></a>
</h4>
<p>The “default” color scheme spans the entire hue domain but features
reduced brightness and saturation, supporting 2 to 50 colors with five
different presets per color. This scheme is ideal for general use where
distinctiveness and subtlety are equally important.</p>
<p>Example using the “default” color scheme:</p>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://wilkelab.org/cowplot/" class="external-link">cowplot</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SeuratExtend</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html" class="external-link">plot_grid</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="../reference/DimPlot2.html">DimPlot2</a></span><span class="op">(</span><span class="va">pbmc</span>, theme <span class="op">=</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SeuratTheme.html" class="external-link">NoAxes</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SeuratTheme.html" class="external-link">NoLegend</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  <span class="fu"><a href="../reference/ClusterDistrBar.html">ClusterDistrBar</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">$</span><span class="va">orig.ident</span>, <span class="va">pbmc</span><span class="op">$</span><span class="va">cluster</span>, flip <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">theme</span><span class="op">(</span>axis.title.x <span class="op">=</span> <span class="fu">element_blank</span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="Visualization_files/figure-html/unnamed-chunk-43-1.png" width="720"></p>
</div>
<div class="section level4">
<h4 id="light-color-scheme">Light Color Scheme<a class="anchor" aria-label="anchor" href="#light-color-scheme"></a>
</h4>
<p>The “light” color scheme also covers the entire hue range but with
increased brightness and reduced saturation, making it suitable when
using labels with darker texts which may require a lighter background
for visibility.</p>
<p>Example using the “light” color scheme:</p>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html" class="external-link">plot_grid</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="../reference/DimPlot2.html">DimPlot2</a></span><span class="op">(</span><span class="va">pbmc</span>, label <span class="op">=</span> <span class="cn">TRUE</span>, repel <span class="op">=</span> <span class="cn">TRUE</span>, theme <span class="op">=</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SeuratTheme.html" class="external-link">NoAxes</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SeuratTheme.html" class="external-link">NoLegend</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  <span class="fu"><a href="../reference/ClusterDistrBar.html">ClusterDistrBar</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">$</span><span class="va">orig.ident</span>, <span class="va">pbmc</span><span class="op">$</span><span class="va">cluster</span>, cols <span class="op">=</span> <span class="st">"light"</span>, flip <span class="op">=</span> <span class="cn">FALSE</span>, border <span class="op">=</span> <span class="st">"black"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">theme</span><span class="op">(</span>axis.title.x <span class="op">=</span> <span class="fu">element_blank</span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="Visualization_files/figure-html/unnamed-chunk-44-1.png" width="720"></p>
</div>
<div class="section level4">
<h4 id="specialized-color-schemes">Specialized Color Schemes<a class="anchor" aria-label="anchor" href="#specialized-color-schemes"></a>
</h4>
<p>For color coordination that reflects the biological or categorical
properties of the data, such as differentiating subtypes within a cell
lineage, the specialized color schemes like “red”, “yellow”, “green”,
“blue”, and “purple” offer hues confined to specific regions. These
schemes support 2 to 25 colors, providing options that are both vibrant
and harmonious without being overwhelming.</p>
<div class="section level5">
<h5 id="red-color-scheme">Red Color Scheme<a class="anchor" aria-label="anchor" href="#red-color-scheme"></a>
</h5>
<p>Example using the “red” color scheme:</p>
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html" class="external-link">plot_grid</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="../reference/DimPlot2.html">DimPlot2</a></span><span class="op">(</span><span class="va">pbmc</span>, cols <span class="op">=</span> <span class="st">"pro_red"</span>, theme <span class="op">=</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SeuratTheme.html" class="external-link">NoAxes</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SeuratTheme.html" class="external-link">NoLegend</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  <span class="fu"><a href="../reference/ClusterDistrBar.html">ClusterDistrBar</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">$</span><span class="va">orig.ident</span>, <span class="va">pbmc</span><span class="op">$</span><span class="va">cluster</span>, cols <span class="op">=</span> <span class="st">"pro_red"</span>, flip <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">theme</span><span class="op">(</span>axis.title.x <span class="op">=</span> <span class="fu">element_blank</span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="Visualization_files/figure-html/unnamed-chunk-45-1.png" width="720"></p>
</div>
<div class="section level5">
<h5 id="yellow-color-scheme">Yellow Color Scheme<a class="anchor" aria-label="anchor" href="#yellow-color-scheme"></a>
</h5>
<p>Example using the “yellow” color scheme:</p>
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html" class="external-link">plot_grid</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="../reference/DimPlot2.html">DimPlot2</a></span><span class="op">(</span><span class="va">pbmc</span>, cols <span class="op">=</span> <span class="st">"pro_yellow"</span>, theme <span class="op">=</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SeuratTheme.html" class="external-link">NoAxes</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SeuratTheme.html" class="external-link">NoLegend</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  <span class="fu"><a href="../reference/ClusterDistrBar.html">ClusterDistrBar</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">$</span><span class="va">orig.ident</span>, <span class="va">pbmc</span><span class="op">$</span><span class="va">cluster</span>, cols <span class="op">=</span> <span class="st">"pro_yellow"</span>, flip <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">theme</span><span class="op">(</span>axis.title.x <span class="op">=</span> <span class="fu">element_blank</span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="Visualization_files/figure-html/unnamed-chunk-46-1.png" width="720"></p>
</div>
<div class="section level5">
<h5 id="green-color-scheme">Green Color Scheme<a class="anchor" aria-label="anchor" href="#green-color-scheme"></a>
</h5>
<p>Example using the “green” color scheme:</p>
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html" class="external-link">plot_grid</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="../reference/DimPlot2.html">DimPlot2</a></span><span class="op">(</span><span class="va">pbmc</span>, cols <span class="op">=</span> <span class="st">"pro_green"</span>, theme <span class="op">=</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SeuratTheme.html" class="external-link">NoAxes</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SeuratTheme.html" class="external-link">NoLegend</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  <span class="fu"><a href="../reference/ClusterDistrBar.html">ClusterDistrBar</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">$</span><span class="va">orig.ident</span>, <span class="va">pbmc</span><span class="op">$</span><span class="va">cluster</span>, cols <span class="op">=</span> <span class="st">"pro_green"</span>, flip <span class="op">=</span> <span class="cn">FALSE</span>, border <span class="op">=</span> <span class="st">"black"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">theme</span><span class="op">(</span>axis.title.x <span class="op">=</span> <span class="fu">element_blank</span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="Visualization_files/figure-html/unnamed-chunk-47-1.png" width="720"></p>
</div>
<div class="section level5">
<h5 id="blue-color-scheme">Blue Color Scheme<a class="anchor" aria-label="anchor" href="#blue-color-scheme"></a>
</h5>
<p>Example using the “blue” color scheme:</p>
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html" class="external-link">plot_grid</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="../reference/DimPlot2.html">DimPlot2</a></span><span class="op">(</span><span class="va">pbmc</span>, cols <span class="op">=</span> <span class="st">"pro_blue"</span>, theme <span class="op">=</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SeuratTheme.html" class="external-link">NoAxes</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SeuratTheme.html" class="external-link">NoLegend</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  <span class="fu"><a href="../reference/ClusterDistrBar.html">ClusterDistrBar</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">$</span><span class="va">orig.ident</span>, <span class="va">pbmc</span><span class="op">$</span><span class="va">cluster</span>, cols <span class="op">=</span> <span class="st">"pro_blue"</span>, flip <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">theme</span><span class="op">(</span>axis.title.x <span class="op">=</span> <span class="fu">element_blank</span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="Visualization_files/figure-html/unnamed-chunk-48-1.png" width="720"></p>
</div>
<div class="section level5">
<h5 id="purple-color-scheme">Purple Color Scheme<a class="anchor" aria-label="anchor" href="#purple-color-scheme"></a>
</h5>
<p>Example using the “purple” color scheme:</p>
<div class="sourceCode" id="cb56"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html" class="external-link">plot_grid</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="../reference/DimPlot2.html">DimPlot2</a></span><span class="op">(</span><span class="va">pbmc</span>, cols <span class="op">=</span> <span class="st">"pro_purple"</span>, theme <span class="op">=</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SeuratTheme.html" class="external-link">NoAxes</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SeuratTheme.html" class="external-link">NoLegend</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  <span class="fu"><a href="../reference/ClusterDistrBar.html">ClusterDistrBar</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">$</span><span class="va">orig.ident</span>, <span class="va">pbmc</span><span class="op">$</span><span class="va">cluster</span>, cols <span class="op">=</span> <span class="st">"pro_purple"</span>, flip <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">theme</span><span class="op">(</span>axis.title.x <span class="op">=</span> <span class="fu">element_blank</span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="Visualization_files/figure-html/unnamed-chunk-49-1.png" width="720">
#### Generating and Customizing Colors with <code>color_pro</code></p>
<p>After showcasing <code>color_pro</code> color schemes through
practical plotting examples, let’s explore how you can directly generate
these color codes using the <code>color_pro</code> function. This allows
for greater flexibility in applying these colors beyond the integrated
visualization functions.</p>
</div>
<div class="section level5">
<h5 id="generating-color-codes">Generating Color Codes<a class="anchor" aria-label="anchor" href="#generating-color-codes"></a>
</h5>
<p>You can generate between 2 to 50 colors using the
<code>color_pro</code> function, which can be useful when you need a
custom color palette for your visualizations.</p>
<p>Example of generating different sets of colors:</p>
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SeuratExtend</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/color_pro.html">color_pro</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">2</span><span class="op">)</span>   <span class="co"># Example output: "#a05d49" "#6181a7"</span></span>
<span><span class="fu"><a href="../reference/color_pro.html">color_pro</a></span><span class="op">(</span><span class="fl">5</span><span class="op">)</span>       <span class="co"># Example output: "#996742" "#5e824b" "#5d7880" "#7169a7" "#9f516c"</span></span>
<span><span class="fu"><a href="../reference/color_pro.html">color_pro</a></span><span class="op">(</span><span class="fl">10</span><span class="op">)</span>      <span class="co"># Generates 10 colors</span></span>
<span><span class="fu"><a href="../reference/color_pro.html">color_pro</a></span><span class="op">(</span><span class="fl">20</span><span class="op">)</span>      <span class="co"># Generates 20 colors</span></span>
<span><span class="fu"><a href="../reference/color_pro.html">color_pro</a></span><span class="op">(</span><span class="fl">50</span><span class="op">)</span>      <span class="co"># Generates 50 colors</span></span></code></pre></div>
<p>The following plot demonstrates the visual impact of these palettes:
<img src="Visualization_files/figure-html/unnamed-chunk-51-1.png" width="720"></p>
</div>
<div class="section level5">
<h5 id="choosing-color-schemes">Choosing Color Schemes<a class="anchor" aria-label="anchor" href="#choosing-color-schemes"></a>
</h5>
<p><code>color_pro</code> allows the selection of up to seven different
color styles: “default”, “light”, “red”, “yellow”, “green”, “blue”,
“purple”. You can specify these styles by name or by their corresponding
numeric value.</p>
<p>Example of generating 10 colors from each style:</p>
<div class="sourceCode" id="cb58"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/color_pro.html">color_pro</a></span><span class="op">(</span><span class="fl">10</span>, col.space <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>  <span class="co"># default</span></span>
<span><span class="fu"><a href="../reference/color_pro.html">color_pro</a></span><span class="op">(</span><span class="fl">10</span>, <span class="fl">2</span><span class="op">)</span>              <span class="co"># light</span></span>
<span><span class="fu"><a href="../reference/color_pro.html">color_pro</a></span><span class="op">(</span><span class="fl">10</span>, <span class="fl">3</span><span class="op">)</span>              <span class="co"># red</span></span>
<span><span class="fu"><a href="../reference/color_pro.html">color_pro</a></span><span class="op">(</span><span class="fl">10</span>, <span class="fl">4</span><span class="op">)</span>              <span class="co"># yellow</span></span>
<span><span class="fu"><a href="../reference/color_pro.html">color_pro</a></span><span class="op">(</span><span class="fl">10</span>, <span class="fl">5</span><span class="op">)</span>              <span class="co"># green</span></span>
<span><span class="fu"><a href="../reference/color_pro.html">color_pro</a></span><span class="op">(</span><span class="fl">10</span>, <span class="fl">6</span><span class="op">)</span>              <span class="co"># blue</span></span>
<span><span class="fu"><a href="../reference/color_pro.html">color_pro</a></span><span class="op">(</span><span class="fl">10</span>, <span class="fl">7</span><span class="op">)</span>              <span class="co"># purple</span></span></code></pre></div>
<p>Visual comparison of these color schemes:
<img src="Visualization_files/figure-html/unnamed-chunk-53-1.png" width="720"></p>
</div>
<div class="section level5">
<h5 id="sorting-options">Sorting Options<a class="anchor" aria-label="anchor" href="#sorting-options"></a>
</h5>
<p><code>color_pro</code> supports sorting by “hue” (default) or by
“difference” for enhanced distinction among colors. This feature can be
specified by name or by numbers 1 or 2.</p>
<p>Example of sorting colors by hue and by difference:</p>
<div class="sourceCode" id="cb59"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/color_pro.html">color_pro</a></span><span class="op">(</span><span class="fl">10</span>, <span class="fl">1</span>, sort <span class="op">=</span> <span class="st">"hue"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/color_pro.html">color_pro</a></span><span class="op">(</span><span class="fl">10</span>, <span class="fl">1</span>, sort <span class="op">=</span> <span class="st">"diff"</span><span class="op">)</span></span></code></pre></div>
<p>Visualizing the effect of different sorting methods:
<img src="Visualization_files/figure-html/unnamed-chunk-55-1.png" width="720"></p>
</div>
<div class="section level5">
<h5 id="random-sequence-options">Random Sequence Options<a class="anchor" aria-label="anchor" href="#random-sequence-options"></a>
</h5>
<p>Each color scheme and number of colors have five different random
sequences available, providing variations even within the same
parameters.</p>
<p>Example of generating different sets from the default color
scheme:</p>
<div class="sourceCode" id="cb60"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/color_pro.html">color_pro</a></span><span class="op">(</span><span class="fl">10</span>, <span class="fl">1</span>, <span class="fl">1</span>, set <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/color_pro.html">color_pro</a></span><span class="op">(</span><span class="fl">10</span>, <span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/color_pro.html">color_pro</a></span><span class="op">(</span><span class="fl">10</span>, <span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">3</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/color_pro.html">color_pro</a></span><span class="op">(</span><span class="fl">10</span>, <span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">4</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/color_pro.html">color_pro</a></span><span class="op">(</span><span class="fl">10</span>, <span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">5</span><span class="op">)</span></span></code></pre></div>
<p>Visualizing different random sequences:
<img src="Visualization_files/figure-html/unnamed-chunk-57-1.png" width="720"></p>
</div>
</div>
</div>
<div class="section level3">
<h3 id="exploring-color_iwh-color-series">Exploring <code>color_iwh</code> Color Series<a class="anchor" aria-label="anchor" href="#exploring-color_iwh-color-series"></a>
</h3>
<p>In addition to <code>color_pro</code>, <code>SeuratExtend</code>
incorporates the <code>I Want Hue</code> algorithm to generate a series
of color palettes. These palettes, known as <code>color_iwh</code>,
include five default styles optimized for various visualization needs.
Unlike <code>color_pro</code>, <code>color_iwh</code> does not support
different sorting options and defaults to sorting by difference for
maximum color distinction.</p>
<div class="section level4">
<h4 id="color_iwh-color-series-overview">
<code>color_iwh</code> Color Series Overview<a class="anchor" aria-label="anchor" href="#color_iwh-color-series-overview"></a>
</h4>
<p>The <code>color_iwh</code> function provides the following predefined
color schemes: - <strong>default</strong>: Suitable for general use with
subtle color variations, supporting 2 to 20 colors. -
<strong>intense</strong>: Features vivid colors, supporting 2 to 30
colors, ideal for making impactful visual statements. -
<strong>pastel</strong>: Offers soft, soothing colors, supporting 2 to
18 colors, perfect for light-themed visualizations. -
<strong>all</strong>: Utilizes the full color spectrum with a soft
k-means clustering approach, supporting 2 to 50 colors. -
<strong>all_hard</strong>: Also covers the full color spectrum but uses
a hard force vector clustering method, supporting 30 to 50 colors.</p>
</div>
<div class="section level4">
<h4 id="generating-colors-with-color_iwh">Generating Colors with <code>color_iwh</code><a class="anchor" aria-label="anchor" href="#generating-colors-with-color_iwh"></a>
</h4>
<p>To generate colors using the <code>color_iwh</code> function, simply
specify the number of colors and the style index. Here are examples of
generating 10 colors from each predefined style:</p>
<p>Example of generating colors from each <code>color_iwh</code>
style:</p>
<div class="sourceCode" id="cb61"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/color_iwh.html">color_iwh</a></span><span class="op">(</span><span class="fl">10</span>, <span class="fl">1</span><span class="op">)</span>  <span class="co"># default</span></span>
<span><span class="fu"><a href="../reference/color_iwh.html">color_iwh</a></span><span class="op">(</span><span class="fl">10</span>, <span class="fl">2</span><span class="op">)</span>  <span class="co"># intense</span></span>
<span><span class="fu"><a href="../reference/color_iwh.html">color_iwh</a></span><span class="op">(</span><span class="fl">10</span>, <span class="fl">3</span><span class="op">)</span>  <span class="co"># pastel</span></span>
<span><span class="fu"><a href="../reference/color_iwh.html">color_iwh</a></span><span class="op">(</span><span class="fl">10</span>, <span class="fl">4</span><span class="op">)</span>  <span class="co"># all</span></span>
<span><span class="fu"><a href="../reference/color_iwh.html">color_iwh</a></span><span class="op">(</span><span class="fl">30</span>, <span class="fl">5</span><span class="op">)</span>  <span class="co"># all_hard</span></span></code></pre></div>
<p>Visual comparison of <code>color_iwh</code> palettes:
<img src="Visualization_files/figure-html/unnamed-chunk-59-1.png" width="720"></p>
</div>
</div>
<div class="section level3">
<h3 id="application-of-color-functions-in-visualization-tools">Application of Color Functions in Visualization Tools<a class="anchor" aria-label="anchor" href="#application-of-color-functions-in-visualization-tools"></a>
</h3>
<p>In the <code>SeuratExtend</code> package, functions such as
<code>DimPlot2</code>, <code>VlnPlot2</code>, <code>Heatmap</code>,
<code>WaterfallPlot</code>, and <code>ClusterDistrBar</code> allow easy
integration of color schemes directly through the <code>cols</code> or
<code>col_theme</code> parameters. This integration means that you do
not have to manually generate color codes using <code>color_pro</code>
or <code>color_iwh</code> unless customization beyond the presets is
needed. Below, we detail how to apply these parameters effectively in
various functions.</p>
<div class="section level4">
<h4 id="applying-colors-to-discrete-variables">Applying Colors to Discrete Variables<a class="anchor" aria-label="anchor" href="#applying-colors-to-discrete-variables"></a>
</h4>
<p>In <code>DimPlot2</code>, <code>VlnPlot2</code>, and
<code>ClusterDistrBar</code>, the <code>cols</code> parameter can accept
a variety of inputs to color discrete variables. These inputs
include:</p>
<ul>
<li>
<strong>color_pro styles</strong>: “default”, “light”, “pro_red”,
“pro_yellow”, “pro_green”, “pro_blue”, “pro_purple”.</li>
<li>
<strong>color_iwh styles</strong>: “iwh_default”, “iwh_intense”,
“iwh_pastel”, “iwh_all”, “iwh_all_hard”.</li>
<li>
<strong>Brewer color scales</strong>: Available through
<code>brewer.pal.info</code>, such as “Blues”, “Dark2”, etc.</li>
<li>
<strong>Manually specified colors</strong>: Direct input of color
codes.</li>
</ul>
<p>Example of using <code>color_pro</code> style in
<code>DimPlot2</code>:</p>
<div class="sourceCode" id="cb62"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/DimPlot2.html">DimPlot2</a></span><span class="op">(</span><span class="va">pbmc</span>, cols <span class="op">=</span> <span class="st">"light"</span><span class="op">)</span></span></code></pre></div>
<p><img src="Visualization_files/figure-html/unnamed-chunk-60-1.png" width="576"></p>
</div>
<div class="section level4">
<h4 id="applying-colors-to-continuous-variables">Applying Colors to Continuous Variables<a class="anchor" aria-label="anchor" href="#applying-colors-to-continuous-variables"></a>
</h4>
<p>In <code>DimPlot2</code>, <code>Heatmap</code>, and
<code>WaterfallPlot</code>, the <code>cols</code> or
<code>col_theme</code> parameters can also be used to assign colors to
continuous variables. Options for continuous variable coloration
include:</p>
<ul>
<li>
<strong>Predefined color schemes</strong>: Using “A”, “B”, “C”, “D”,
or “E” from the ‘viridis’ package.</li>
<li>
<strong>Three-point gradients</strong>: A named vector with keys
“low”, “mid”, and “high” for colors at specific data points, often used
for diverging color schemes. Example:
<code>c(low = "blue", mid = "white", high = "red")</code>
</li>
<li>
<strong>Two-point gradients</strong>: A simple two-color gradient
using keys “low” and “high”. Example:
<code>c(low = "blue", high = "red")</code>
</li>
<li>
<strong>Custom color gradients</strong>: A vector of colors to
generate a gradient across a range, suitable for complex data
visualizations.</li>
</ul>
<p>Example of applying a color gradient in <code>DimPlot2</code> for a
continuous variable:</p>
<div class="sourceCode" id="cb63"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/DimPlot2.html">DimPlot2</a></span><span class="op">(</span><span class="va">pbmc</span>, features <span class="op">=</span> <span class="st">"CD3D"</span>, cols <span class="op">=</span> <span class="st">"D"</span><span class="op">)</span></span></code></pre></div>
<p><img src="Visualization_files/figure-html/unnamed-chunk-61-1.png" width="480"></p>
</div>
</div>
<div class="section level3">
<h3 id="additional-color-related-functions">Additional Color-Related Functions<a class="anchor" aria-label="anchor" href="#additional-color-related-functions"></a>
</h3>
<p>In the discussion of the <code>FeaturePlot3</code> functionality, we
touched upon the RYB mixing system used in <code>SeuratExtend</code>.
The method for mixing these colors is a proprietary development of
<code>SeuratExtend</code>, designed as an approximation to the
traditional RYB color mixing. This approach includes specific
adjustments to the primary RYB colors to make them more suitable for
visualizing expression gradients:</p>
<ul>
<li>
<strong>Yellow</strong>: Since pure yellow can be too pale for
effective gradient display, it is deepened in
<code>SeuratExtend</code>.</li>
<li>
<strong>Red and Blue</strong>: Pure shades of these colors can be
harsh on the eyes, so they have been softened to enhance visual
comfort.</li>
</ul>
<p>These modifications ensure that the colors used in visualizations are
both effective in conveying information and easier on the eyes.</p>
<div class="section level4">
<h4 id="using-ryb2rgb-to-convert-ryb-to-rgb-hex-codes">Using <code>ryb2rgb()</code> to Convert RYB to RGB Hex Codes<a class="anchor" aria-label="anchor" href="#using-ryb2rgb-to-convert-ryb-to-rgb-hex-codes"></a>
</h4>
<p>The <code><a href="../reference/ryb2rgb.html">ryb2rgb()</a></code> function in <code>SeuratExtend</code>
translates RYB values into conventional RGB hex codes, which can then be
used in standard plotting functions. This function accepts a vector of
three numbers (ranging from 0 to 1), each representing the intensity of
red, yellow, and blue, respectively. Here is a simple example of how to
use <code><a href="../reference/ryb2rgb.html">ryb2rgb()</a></code>:</p>
<div class="sourceCode" id="cb64"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/ryb2rgb.html">ryb2rgb</a></span><span class="op">(</span>ryb <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>r <span class="op">=</span> <span class="fl">0.3</span>, y <span class="op">=</span> <span class="fl">0.5</span>, b <span class="op">=</span> <span class="fl">0.2</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "#CCAF80"</span></span></code></pre>
<div class="sourceCode" id="cb66"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Outputs: "#CCAF80"</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="visualizing-primary-and-secondary-colors">Visualizing Primary and Secondary Colors<a class="anchor" aria-label="anchor" href="#visualizing-primary-and-secondary-colors"></a>
</h4>
<p>To illustrate how <code><a href="../reference/ryb2rgb.html">ryb2rgb()</a></code> interprets different
combinations of primary and secondary colors, consider the following
example to create a visual palette:</p>
<div class="sourceCode" id="cb67"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://scales.r-lib.org" class="external-link">scales</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>  red <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">0</span>, <span class="fl">0</span><span class="op">)</span>,</span>
<span>  yellow <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span>, <span class="fl">0</span><span class="op">)</span>,</span>
<span>  blue <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">1</span><span class="op">)</span>,</span>
<span>  orange <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">0</span><span class="op">)</span>,</span>
<span>  purple <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">0</span>, <span class="fl">1</span><span class="op">)</span>,</span>
<span>  green <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span>, <span class="fl">1</span><span class="op">)</span>,</span>
<span>  black <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">1</span><span class="op">)</span>,</span>
<span>  grey <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.5</span>, <span class="fl">0.5</span>, <span class="fl">0.5</span><span class="op">)</span>,</span>
<span>  white <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span><span class="op">)</span></span>
<span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="fl">2</span>, <span class="va">ryb2rgb</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://scales.r-lib.org/reference/show_col.html" class="external-link">show_col</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="Visualization_files/figure-html/unnamed-chunk-63-1.png" width="480"></p>
<p>This section shows a palette derived from various RYB combinations,
demonstrating how <code><a href="../reference/ryb2rgb.html">ryb2rgb()</a></code> translates these combinations
into RGB hex codes. This functionality is particularly useful for
researchers and data scientists who need to customize their color
schemes beyond the standard options provided by most visualization
libraries.</p>
</div>
<div class="section level4">
<h4 id="using-save_colors-to-manage-color-settings">Using <code>save_colors</code> to Manage Color Settings<a class="anchor" aria-label="anchor" href="#using-save_colors-to-manage-color-settings"></a>
</h4>
<p>The <code>save_colors</code> function is designed to store custom
color settings within the <code>Seurat</code> object, facilitating their
reuse across various visualization functions. This approach allows for
consistent color usage across multiple plots and simplifies the
management of color settings within a project.</p>
<p>This function primarily serves to complement visualization functions
such as <code>DimPlot2</code> and <code>VlnPlot2</code>. By storing
color settings directly within the <code>Seurat</code> object,
<code>save_colors</code> enables these visualization tools to
automatically retrieve and apply the specified colors to variables such
as gene expressions or clustering results. This ensures consistency and
repeatability in the color schemes of your plots.</p>
<p>Here’s how you can use <code>save_colors</code> to specify and store
color settings for certain variables, which can then be automatically
utilized by functions like <code>DimPlot2</code>:</p>
<div class="sourceCode" id="cb68"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/save_colors.html">save_colors</a></span><span class="op">(</span><span class="va">pbmc</span>, col_list <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>  <span class="st">"cluster"</span> <span class="op">=</span> <span class="st">"pro_blue"</span>,</span>
<span>  <span class="st">"CD14"</span> <span class="op">=</span> <span class="st">"D"</span>,</span>
<span>  <span class="st">"CD3D"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"#EEEEEE"</span>, <span class="st">"black"</span><span class="op">)</span></span>
<span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Now, when using DimPlot2, the specified colors for 'cluster', 'CD14', and 'CD3D' are automatically applied</span></span>
<span><span class="fu"><a href="../reference/DimPlot2.html">DimPlot2</a></span><span class="op">(</span><span class="va">pbmc</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"cluster"</span>, <span class="st">"orig.ident"</span>, <span class="st">"CD14"</span>, <span class="st">"CD3D"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="Visualization_files/figure-html/unnamed-chunk-64-1.png" width="864"></p>
<p>This example demonstrates setting custom colors for the
<code>cluster</code>, <code>CD14</code>, and <code>CD3D</code> variables
and then using these colors in a dimension reduction plot without
needing to specify them again in the <code>DimPlot2</code> function. The
colors are stored in the <code>Seurat</code> object and retrieved
dynamically by the plotting function.</p>
<div class="sourceCode" id="cb69"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.4.0 (2024-04-24)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">## Running under: Ubuntu 20.04.6 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.9.0 </span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.9.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">##  [3] LC_TIME=de_BE.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">##  [5] LC_MONETARY=de_BE.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=de_BE.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">## [11] LC_MEASUREMENT=de_BE.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: Europe/Brussels</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">## [8] base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] hyc_0.1.3                 tidyr_1.3.1              </span></span>
<span><span class="co">##  [3] DelayedMatrixStats_1.26.0 DelayedArray_0.30.1      </span></span>
<span><span class="co">##  [5] SparseArray_1.4.8         S4Arrays_1.4.1           </span></span>
<span><span class="co">##  [7] abind_1.4-5               IRanges_2.38.0           </span></span>
<span><span class="co">##  [9] S4Vectors_0.42.0          MatrixGenerics_1.16.0    </span></span>
<span><span class="co">## [11] matrixStats_1.3.0         BiocGenerics_0.50.0      </span></span>
<span><span class="co">## [13] ggbeeswarm_0.7.2          ggtext_0.1.2             </span></span>
<span><span class="co">## [15] rlist_0.4.6.2             magrittr_2.0.3           </span></span>
<span><span class="co">## [17] ggpubr_0.6.0              ggrepel_0.9.5            </span></span>
<span><span class="co">## [19] cowplot_1.1.3             viridis_0.6.5            </span></span>
<span><span class="co">## [21] viridisLite_0.4.2         rlang_1.1.4              </span></span>
<span><span class="co">## [23] scales_1.3.0              reshape2_1.4.4           </span></span>
<span><span class="co">## [25] mosaic_1.9.1              mosaicData_0.20.4        </span></span>
<span><span class="co">## [27] ggformula_0.12.0          Matrix_1.7-0             </span></span>
<span><span class="co">## [29] ggplot2_3.5.1             lattice_0.22-6           </span></span>
<span><span class="co">## [31] dplyr_1.1.4               SeuratExtend_1.0.0       </span></span>
<span><span class="co">## [33] SeuratExtendData_0.2.0    Seurat_5.1.0             </span></span>
<span><span class="co">## [35] SeuratObject_5.0.2        sp_2.1-4                 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] RcppAnnoy_0.0.22         splines_4.4.0            later_1.3.2             </span></span>
<span><span class="co">##   [4] tibble_3.2.1             polyclip_1.10-6          fastDummies_1.7.3       </span></span>
<span><span class="co">##   [7] lifecycle_1.0.4          rstatix_0.7.2            globals_0.16.3          </span></span>
<span><span class="co">##  [10] MASS_7.3-60.2            backports_1.5.0          plotly_4.10.4           </span></span>
<span><span class="co">##  [13] sass_0.4.9               rmarkdown_2.27           jquerylib_0.1.4         </span></span>
<span><span class="co">##  [16] yaml_2.3.8               httpuv_1.6.15            sctransform_0.4.1       </span></span>
<span><span class="co">##  [19] spam_2.10-0              spatstat.sparse_3.0-3    reticulate_1.37.0       </span></span>
<span><span class="co">##  [22] pbapply_1.7-2            RColorBrewer_1.1-3       zlibbioc_1.50.0         </span></span>
<span><span class="co">##  [25] Rtsne_0.17               purrr_1.0.2              labelled_2.13.0         </span></span>
<span><span class="co">##  [28] irlba_2.3.5.1            listenv_0.9.1            spatstat.utils_3.0-4    </span></span>
<span><span class="co">##  [31] goftest_1.2-3            RSpectra_0.16-1          spatstat.random_3.2-3   </span></span>
<span><span class="co">##  [34] fitdistrplus_1.1-11      parallelly_1.37.1        pkgdown_2.0.9           </span></span>
<span><span class="co">##  [37] commonmark_1.9.1         leiden_0.4.3.1           codetools_0.2-20        </span></span>
<span><span class="co">##  [40] xml2_1.3.6               tidyselect_1.2.1         farver_2.1.2            </span></span>
<span><span class="co">##  [43] spatstat.explore_3.2-7   jsonlite_1.8.8           progressr_0.14.0        </span></span>
<span><span class="co">##  [46] ggridges_0.5.6           survival_3.7-0           systemfonts_1.1.0       </span></span>
<span><span class="co">##  [49] tools_4.4.0              ragg_1.3.2               ica_1.0-3               </span></span>
<span><span class="co">##  [52] Rcpp_1.0.12              glue_1.7.0               gridExtra_2.3           </span></span>
<span><span class="co">##  [55] xfun_0.44                withr_3.0.0              fastmap_1.2.0           </span></span>
<span><span class="co">##  [58] fansi_1.0.6              digest_0.6.35            R6_2.5.1                </span></span>
<span><span class="co">##  [61] mime_0.12                textshaping_0.4.0        colorspace_2.1-0        </span></span>
<span><span class="co">##  [64] scattermore_1.2          tensor_1.5               markdown_1.13           </span></span>
<span><span class="co">##  [67] spatstat.data_3.0-4      utf8_1.2.4               generics_0.1.3          </span></span>
<span><span class="co">##  [70] data.table_1.15.4        httr_1.4.7               htmlwidgets_1.6.4       </span></span>
<span><span class="co">##  [73] uwot_0.2.2               pkgconfig_2.0.3          gtable_0.3.5            </span></span>
<span><span class="co">##  [76] lmtest_0.9-40            XVector_0.44.0           htmltools_0.5.8.1       </span></span>
<span><span class="co">##  [79] carData_3.0-5            dotCall64_1.1-1          png_0.1-8               </span></span>
<span><span class="co">##  [82] knitr_1.47               rstudioapi_0.16.0        nlme_3.1-165            </span></span>
<span><span class="co">##  [85] cachem_1.1.0             zoo_1.8-12               stringr_1.5.1           </span></span>
<span><span class="co">##  [88] KernSmooth_2.23-24       parallel_4.4.0           miniUI_0.1.1.1          </span></span>
<span><span class="co">##  [91] vipor_0.4.7              desc_1.4.3               pillar_1.9.0            </span></span>
<span><span class="co">##  [94] grid_4.4.0               vctrs_0.6.5              RANN_2.6.1              </span></span>
<span><span class="co">##  [97] promises_1.3.0           car_3.1-2                xtable_1.8-4            </span></span>
<span><span class="co">## [100] cluster_2.1.6            beeswarm_0.4.0           evaluate_0.23           </span></span>
<span><span class="co">## [103] cli_3.6.2                compiler_4.4.0           crayon_1.5.2            </span></span>
<span><span class="co">## [106] future.apply_1.11.2      ggsignif_0.6.4           labeling_0.4.3          </span></span>
<span><span class="co">## [109] plyr_1.8.9               forcats_1.0.0            fs_1.6.4                </span></span>
<span><span class="co">## [112] stringi_1.8.4            deldir_2.0-4             munsell_0.5.1           </span></span>
<span><span class="co">## [115] lazyeval_0.2.2           spatstat.geom_3.2-9      mosaicCore_0.9.4.0      </span></span>
<span><span class="co">## [118] RcppHNSW_0.6.0           hms_1.1.3                patchwork_1.2.0         </span></span>
<span><span class="co">## [121] sparseMatrixStats_1.16.0 future_1.33.2            shiny_1.8.1.1           </span></span>
<span><span class="co">## [124] highr_0.11               haven_2.5.4              ROCR_1.0-11             </span></span>
<span><span class="co">## [127] gridtext_0.1.5           igraph_1.3.4             broom_1.0.6             </span></span>
<span><span class="co">## [130] memoise_2.0.1            bslib_0.7.0</span></span></code></pre>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Yichao Hua.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
