<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Lesson 3: Advanced Visualization Techniques • SeuratExtend</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../../bootstrap-toc.css">
<script src="../../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../../pkgdown.css" rel="stylesheet">
<script src="../../pkgdown.js"></script><meta property="og:title" content="Lesson 3: Advanced Visualization Techniques">
<meta property="og:description" content="SeuratExtend">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../../index.html">SeuratExtend</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../../articles/News.html">What's New in v1.1.0</a>
    </li>
    <li class="divider">
    </li>
<li>
      <a href="../../articles/Visualization.html">Enhanced Visualization</a>
    </li>
    <li>
      <a href="../../articles/GSEA.html">Geneset Enrichment Analysis (GSEA) </a>
    </li>
    <li>
      <a href="../../articles/Trajectory.html">Trajectory and Pseudotime Analysis </a>
    </li>
    <li>
      <a href="../../articles/SCENIC.html">SCENIC for Gene Regulatory Networks Analysis </a>
    </li>
    <li>
      <a href="../../articles/Utilities.html">Utility Tools and Functions</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    scRNA-seq Course
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../../articles/single-cell-course/1.basic-R.html">1. Introduction to R</a>
    </li>
    <li>
      <a href="../../articles/single-cell-course/2.Seurat.html">2. Basic Single-Cell Analysis Workflow</a>
    </li>
    <li>
      <a href="../../articles/single-cell-course/3.Visualization.html">3. Advanced Visualization Techniques</a>
    </li>
    <li>
      <a href="../../articles/single-cell-course/4.GSEA.html">4. Gene Set Enrichment Analysis</a>
    </li>
    <li>
      <a href="../../articles/single-cell-course/5.Core-Enhancement.html">5. Core scRNA-seq Workflow Enhancements</a>
    </li>
    <li>
      <a href="../../articles/single-cell-course/6.Advanced.html">6. Advanced Methods (Part 1)</a>
    </li>
    <li>
      <a href="../../articles/single-cell-course/6.Advanced-2.html">7. Advanced Methods (Part 2)</a>
    </li>
  </ul>
</li>
<li>
  <a href="../../articles/FAQ.html">FAQ</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Lesson 3: Advanced Visualization Techniques</h1>
                        <h4 data-toc-skip class="author">Yichao Hua</h4>
            
            <h4 data-toc-skip class="date">2024-11-29</h4>
      
      
      <div class="hidden name"><code>3.Visualization.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>In our previous lesson, we created a PBMC object, completed
clustering, and performed annotation. This serves as the foundation for
all our subsequent analyses. In this lesson, we’ll systematically
explore various visualization techniques using SeuratExtend to create
high-quality, publication-ready figures.</p>
<p>First, let’s set your working directory to where we’ll store all
course materials. If you followed Lesson 1, this would be:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Set working directory to the course folder created in Lesson 1</span></span>
<span><span class="co"># If you used a different location, replace this path with your chosen directory</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/getwd.html" class="external-link">setwd</a></span><span class="op">(</span><span class="st">"~/Documents/single-cell-course"</span><span class="op">)</span></span></code></pre></div>
<p>Let’s start by loading the necessary packages and our previously
saved PBMC object:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SeuratExtend</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Load the PBMC object we saved at the end of Lesson 2</span></span>
<span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"rds/pbmc_annotated.rds"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="create-an-enhanced-dimensional-reduction-plot">1. Create an Enhanced Dimensional Reduction Plot<a class="anchor" aria-label="anchor" href="#create-an-enhanced-dimensional-reduction-plot"></a>
</h2>
<p>In Seurat, dimension reduction plots such as UMAP are typically
created using <code>DimPlot</code> for discrete variables and
<code>FeaturePlot</code> for continuous variables.
<code>SeuratExtend</code> simplifies this process with
<code>DimPlot2</code>, which does not require differentiation between
variable types. This function automatically recognizes the type of input
parameters, whether discrete or continuous. Additionally,
<code>DimPlot2</code> introduces numerous extra parameters to enrich the
customization of the plots.</p>
<div class="section level3">
<h3 id="basic-usage">Basic Usage<a class="anchor" aria-label="anchor" href="#basic-usage"></a>
</h3>
<p>To generate a basic dimension reduction plot, simply call
<code>DimPlot2</code> with your Seurat object:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../../reference/DimPlot2.html">DimPlot2</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span></span></code></pre></div>
<p><img src="3.Visualization_files/figure-html/unnamed-chunk-2-1.png" width="624"></p>
</div>
<div class="section level3">
<h3 id="visualizing-different-variables">Visualizing Different Variables<a class="anchor" aria-label="anchor" href="#visualizing-different-variables"></a>
</h3>
<p><code>DimPlot2</code> can handle both discrete and continuous
variables seamlessly. Here’s how to input different variables into the
plot:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../../reference/DimPlot2.html">DimPlot2</a></span><span class="op">(</span><span class="va">pbmc</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"cluster"</span>, <span class="st">"MS4A1"</span>, <span class="st">"CD14"</span>, <span class="st">"CD3D"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="3.Visualization_files/figure-html/unnamed-chunk-3-1.png" width="864"></p>
</div>
<div class="section level3">
<h3 id="splitting-by-variable">Splitting by Variable<a class="anchor" aria-label="anchor" href="#splitting-by-variable"></a>
</h3>
<p>You can also split the visualization by a specific variable, which is
particularly useful for comparative analysis across conditions or
identities. In our current PBMC object, we only have one sample. Let’s
create a new column in the metadata to simulate a scenario with two
samples for demonstration purposes:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">42</span><span class="op">)</span>  <span class="co"># For reproducibility</span></span>
<span><span class="va">pbmc</span><span class="op">$</span><span class="va">sample</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"sample1"</span>, <span class="st">"sample1"</span>, <span class="st">"sample2"</span><span class="op">)</span>, size <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span>, replace <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">$</span><span class="va">sample</span><span class="op">)</span>  <span class="co"># Check the distribution</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## sample1 sample2 </span></span>
<span><span class="co">##    1796     904</span></span></code></pre>
<p>Now, let’s create a split plot:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../../reference/DimPlot2.html">DimPlot2</a></span><span class="op">(</span><span class="va">pbmc</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"cluster"</span>, <span class="st">"CD14"</span><span class="op">)</span>, split.by <span class="op">=</span> <span class="st">"sample"</span>, ncol <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<p><img src="3.Visualization_files/figure-html/unnamed-chunk-5-1.png" width="672"></p>
</div>
<div class="section level3">
<h3 id="highlighting-specific-cells">Highlighting Specific Cells<a class="anchor" aria-label="anchor" href="#highlighting-specific-cells"></a>
</h3>
<p>To highlight cells of interest, such as a specific cluster, you can
define the cells explicitly and use them in your plot:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">b_cells</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span><span class="op">[</span><span class="va">pbmc</span><span class="op">$</span><span class="va">cluster</span> <span class="op">==</span> <span class="st">"B cell"</span><span class="op">]</span></span>
<span><span class="fu"><a href="../../reference/DimPlot2.html">DimPlot2</a></span><span class="op">(</span><span class="va">pbmc</span>, cells.highlight <span class="op">=</span> <span class="va">b_cells</span><span class="op">)</span></span></code></pre></div>
<p><img src="3.Visualization_files/figure-html/unnamed-chunk-6-1.png" width="624"></p>
</div>
<div class="section level3">
<h3 id="advanced-customization">Advanced Customization<a class="anchor" aria-label="anchor" href="#advanced-customization"></a>
</h3>
<p>For each variable, you can specify custom colors, adjust themes, and
more. For detailed information on color customization, refer to the <a href="#explore-color-functions">Explore Color Functions</a> section:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../../reference/DimPlot2.html">DimPlot2</a></span><span class="op">(</span></span>
<span>  <span class="va">pbmc</span>,</span>
<span>  features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"cluster"</span>, <span class="st">"sample"</span>, <span class="st">"CD14"</span>, <span class="st">"CD3D"</span><span class="op">)</span>,</span>
<span>  cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>    <span class="st">"cluster"</span> <span class="op">=</span> <span class="st">"light"</span>,</span>
<span>    <span class="st">"CD14"</span> <span class="op">=</span> <span class="st">"D"</span>,</span>
<span>    <span class="st">"CD3D"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"#EEEEEE"</span>, <span class="st">"black"</span><span class="op">)</span></span>
<span>  <span class="op">)</span>,</span>
<span>  theme <span class="op">=</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SeuratTheme.html" class="external-link">NoAxes</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="3.Visualization_files/figure-html/unnamed-chunk-7-1.png" width="864"></p>
</div>
<div class="section level3">
<h3 id="adding-labels-and-boxes">Adding Labels and Boxes<a class="anchor" aria-label="anchor" href="#adding-labels-and-boxes"></a>
</h3>
<p>To further enhance the plot, you can add labels and bounding boxes to
clearly delineate different groups or points of interest:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../../reference/DimPlot2.html">DimPlot2</a></span><span class="op">(</span><span class="va">pbmc</span>, label <span class="op">=</span> <span class="cn">TRUE</span>, box <span class="op">=</span> <span class="cn">TRUE</span>, label.color <span class="op">=</span> <span class="st">"black"</span>, repel <span class="op">=</span> <span class="cn">TRUE</span>, theme <span class="op">=</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SeuratTheme.html" class="external-link">NoLegend</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="3.Visualization_files/figure-html/unnamed-chunk-8-1.png" width="480"></p>
</div>
<div class="section level3">
<h3 id="simplifying-labels-with-indices">Simplifying Labels with Indices<a class="anchor" aria-label="anchor" href="#simplifying-labels-with-indices"></a>
</h3>
<p>Sometimes, cluster names are too lengthy and can make the plot appear
cluttered when displayed with labels. To address this, consider using
indices to replace the cluster names in the plot, which helps make the
visualization cleaner. For instance, you can label clusters as ‘C1’,
‘C2’, etc., on the plot itself, while detailing what each index stands
for (e.g., ‘C1: B cell’, ‘C2: CD4 T Memory’) in the figure legend:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../../reference/DimPlot2.html">DimPlot2</a></span><span class="op">(</span><span class="va">pbmc</span>, index.title <span class="op">=</span> <span class="st">"C"</span>, box <span class="op">=</span> <span class="cn">TRUE</span>, label.color <span class="op">=</span> <span class="st">"black"</span><span class="op">)</span></span></code></pre></div>
<p><img src="3.Visualization_files/figure-html/unnamed-chunk-9-1.png" width="624"></p>
<p>This approach ensures that the plot remains legible and aesthetically
pleasing, even when dealing with numerous or complex labels.</p>
</div>
<div class="section level3">
<h3 id="adding-umap-arrows">Adding UMAP Arrows<a class="anchor" aria-label="anchor" href="#adding-umap-arrows"></a>
</h3>
<p>Starting from v1.1.0, you can add simplified axis indicators to your
dimension reduction plots using the <code>theme_umap_arrows</code>
function. This is particularly useful for UMAP or t-SNE plots where
traditional axes are often removed to reduce visual clutter:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Add simplified axis indicators</span></span>
<span><span class="fu"><a href="../../reference/DimPlot2.html">DimPlot2</a></span><span class="op">(</span></span>
<span>  <span class="va">pbmc</span>, </span>
<span>  features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"cluster"</span>, <span class="st">"sample"</span>, <span class="st">"CD14"</span>, <span class="st">"CD3D"</span><span class="op">)</span>,</span>
<span>  theme <span class="op">=</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SeuratTheme.html" class="external-link">NoAxes</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="../../reference/theme_umap_arrows.html">theme_umap_arrows</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="3.Visualization_files/figure-html/unnamed-chunk-10-1.png" width="864"></p>
</div>
</div>
<div class="section level2">
<h2 id="simultaneous-display-of-three-features-on-a-dimension-reduction-plot">2. Simultaneous Display of Three Features on a Dimension Reduction
Plot<a class="anchor" aria-label="anchor" href="#simultaneous-display-of-three-features-on-a-dimension-reduction-plot"></a>
</h2>
<p>In <code>SeuratExtend</code>, a unique visualization method allows
for the simultaneous display of three features on the same dimension
reduction plot. The functions <code>FeaturePlot3</code> and
<code>FeaturePlot3.grid</code> employ a color mixing system (either RYB
or RGB) to represent three different genes (or other continuous
variables). This method uses the principles of color mixing to
quantitatively display the expression levels or intensities of these
three features in each cell.</p>
<div class="section level3">
<h3 id="ryb-and-rgb-color-systems">RYB and RGB Color Systems<a class="anchor" aria-label="anchor" href="#ryb-and-rgb-color-systems"></a>
</h3>
<p>In the RGB system, black represents no or low expression, and
brighter colors indicate higher levels: <img src="images/RGB_mixing.png" width="500" alt="RGB System Example"></p>
<p>In the RYB system, white represents no expression, and deeper colors
indicate higher expression levels: <img src="images/RYB_mixing.png" width="500" alt="RYB System Example"></p>
</div>
<div class="section level3">
<h3 id="examples-using-ryb-and-rgb-systems">Examples Using RYB and RGB Systems<a class="anchor" aria-label="anchor" href="#examples-using-ryb-and-rgb-systems"></a>
</h3>
<p>Here’s how to display three markers using the RYB system, with red
for CD3D, yellow for CD14, and blue for CD79A:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../../reference/FeaturePlot3.html">FeaturePlot3</a></span><span class="op">(</span><span class="va">pbmc</span>, color <span class="op">=</span> <span class="st">"ryb"</span>, feature.1 <span class="op">=</span> <span class="st">"CD3D"</span>, feature.2 <span class="op">=</span> <span class="st">"CD14"</span>, feature.3 <span class="op">=</span> <span class="st">"CD79A"</span>, pt.size <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></span></code></pre></div>
<p><img src="3.Visualization_files/figure-html/unnamed-chunk-11-1.png" width="576"></p>
<p>For the RGB system, with red for CD3D, green for CD14, and blue for
CD79A:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../../reference/FeaturePlot3.html">FeaturePlot3</a></span><span class="op">(</span><span class="va">pbmc</span>, color <span class="op">=</span> <span class="st">"rgb"</span>, feature.1 <span class="op">=</span> <span class="st">"CD3D"</span>, feature.2 <span class="op">=</span> <span class="st">"CD14"</span>, feature.3 <span class="op">=</span> <span class="st">"CD79A"</span>, pt.size <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<p><img src="3.Visualization_files/figure-html/unnamed-chunk-12-1.png" width="576"></p>
</div>
<div class="section level3">
<h3 id="batch-visualization-with-featureplot3-grid">Batch Visualization with <code>FeaturePlot3.grid</code><a class="anchor" aria-label="anchor" href="#batch-visualization-with-featureplot3-grid"></a>
</h3>
<p><code>FeaturePlot3.grid</code> extends <code>FeaturePlot3</code> by
allowing multiple plots to be generated in one go. The
<code>features</code> parameter requires a vector where every three
values are assigned a color (RYB or RGB) and placed together in one
plot. If you wish to skip a color, use <code>NA</code> as a
placeholder.</p>
<p>For instance, to place the following five genes into two plots using
the RYB system, and skip yellow in the second plot:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../../reference/FeaturePlot3-grid.html">FeaturePlot3.grid</a></span><span class="op">(</span><span class="va">pbmc</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD3D"</span>, <span class="st">"CD14"</span>, <span class="st">"CD79A"</span>, <span class="st">"FCGR3A"</span>, <span class="cn">NA</span>, <span class="st">"LYZ"</span><span class="op">)</span>, pt.size <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></span></code></pre></div>
<p><img src="3.Visualization_files/figure-html/unnamed-chunk-13-1.png" width="624"></p>
</div>
<div class="section level3">
<h3 id="tips-on-point-size">Tips on Point Size<a class="anchor" aria-label="anchor" href="#tips-on-point-size"></a>
</h3>
<p>The background is usually white, so the choice of color system and
point size can significantly affect visual perception. In the RYB
system, where higher expression results in darker colors, a smaller
<code>pt.size</code> is preferable to prevent overlapping points. In
contrast, in the RGB system, higher expressions result in lighter
colors, potentially leading to visibility issues for highly expressed
cells that may blend into the white background. Here, a larger
<code>pt.size</code> is recommended so that the darker, low-expression
points can form a “background” to highlight the lighter, high-expression
points.</p>
</div>
</div>
<div class="section level2">
<h2 id="visualize-cluster-distribution-in-samples">3. Visualize Cluster Distribution in Samples<a class="anchor" aria-label="anchor" href="#visualize-cluster-distribution-in-samples"></a>
</h2>
<div class="section level3">
<h3 id="introduction-1">Introduction<a class="anchor" aria-label="anchor" href="#introduction-1"></a>
</h3>
<p>The <code>ClusterDistrBar</code> function is designed to visualize
the distribution of clusters across different samples. It can show both
absolute counts and proportions, and it allows for various
customizations including axis reversal and normalization.</p>
</div>
<div class="section level3">
<h3 id="basic-usage-1">Basic Usage<a class="anchor" aria-label="anchor" href="#basic-usage-1"></a>
</h3>
<p>To create a basic bar plot showing the distribution of clusters
within samples, simply specify the origin (sample identifier) and
cluster variables from your dataset:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../../reference/ClusterDistrBar.html">ClusterDistrBar</a></span><span class="op">(</span>origin <span class="op">=</span> <span class="va">pbmc</span><span class="op">$</span><span class="va">sample</span>, cluster <span class="op">=</span> <span class="va">pbmc</span><span class="op">$</span><span class="va">cluster</span><span class="op">)</span></span></code></pre></div>
<p><img src="3.Visualization_files/figure-html/unnamed-chunk-14-1.png" width="672"></p>
</div>
<div class="section level3">
<h3 id="displaying-absolute-cell-counts">Displaying Absolute Cell Counts<a class="anchor" aria-label="anchor" href="#displaying-absolute-cell-counts"></a>
</h3>
<p>If you prefer to visualize the absolute cell count rather than
proportions, set the <code>percent</code> parameter to
<code>FALSE</code>:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../../reference/ClusterDistrBar.html">ClusterDistrBar</a></span><span class="op">(</span>origin <span class="op">=</span> <span class="va">pbmc</span><span class="op">$</span><span class="va">sample</span>, cluster <span class="op">=</span> <span class="va">pbmc</span><span class="op">$</span><span class="va">cluster</span>, percent <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p><img src="3.Visualization_files/figure-html/unnamed-chunk-15-1.png" width="672"></p>
</div>
<div class="section level3">
<h3 id="vertical-bar-plot">Vertical Bar Plot<a class="anchor" aria-label="anchor" href="#vertical-bar-plot"></a>
</h3>
<p>If a vertical orientation is preferred over the default horizontal
bars, set the <code>flip</code> parameter to <code>FALSE</code>:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../../reference/ClusterDistrBar.html">ClusterDistrBar</a></span><span class="op">(</span>origin <span class="op">=</span> <span class="va">pbmc</span><span class="op">$</span><span class="va">sample</span>, cluster <span class="op">=</span> <span class="va">pbmc</span><span class="op">$</span><span class="va">cluster</span>, flip <span class="op">=</span> <span class="cn">FALSE</span>, reverse_order <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="3.Visualization_files/figure-html/unnamed-chunk-16-1.png" width="384"></p>
</div>
<div class="section level3">
<h3 id="non-stacked-bar-plot">Non-Stacked Bar Plot<a class="anchor" aria-label="anchor" href="#non-stacked-bar-plot"></a>
</h3>
<p>If you prefer not to stack the bars, which can be useful for direct
comparisons of cluster sizes across samples, set the <code>stack</code>
parameter to <code>FALSE</code>:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../../reference/ClusterDistrBar.html">ClusterDistrBar</a></span><span class="op">(</span>origin <span class="op">=</span> <span class="va">pbmc</span><span class="op">$</span><span class="va">sample</span>, cluster <span class="op">=</span> <span class="va">pbmc</span><span class="op">$</span><span class="va">cluster</span>, flip <span class="op">=</span> <span class="cn">FALSE</span>, stack <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p><img src="3.Visualization_files/figure-html/unnamed-chunk-17-1.png" width="672"></p>
</div>
<div class="section level3">
<h3 id="exporting-data-matrix">Exporting Data Matrix<a class="anchor" aria-label="anchor" href="#exporting-data-matrix"></a>
</h3>
<p>In cases where a visual plot is not required and only the underlying
data matrix is needed, set the <code>plot</code> parameter to
<code>FALSE</code>:</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data_matrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/ClusterDistrBar.html">ClusterDistrBar</a></span><span class="op">(</span>origin <span class="op">=</span> <span class="va">pbmc</span><span class="op">$</span><span class="va">sample</span>, cluster <span class="op">=</span> <span class="va">pbmc</span><span class="op">$</span><span class="va">cluster</span>, plot <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="co"># View the matrix</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">data_matrix</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                sample1    sample2</span></span>
<span><span class="co">## B cell       12.973274 12.8318584</span></span>
<span><span class="co">## CD4 T Memory 19.710468 17.4778761</span></span>
<span><span class="co">## CD4 T Naive  23.608018 25.7743363</span></span>
<span><span class="co">## CD8 T cell   12.416481 11.5044248</span></span>
<span><span class="co">## DC            1.113586  1.7699115</span></span>
<span><span class="co">## Mono CD14    17.984410 18.4734513</span></span>
<span><span class="co">## Mono FCGR3A   5.790646  6.3053097</span></span>
<span><span class="co">## NK cell       6.013363  5.0884956</span></span>
<span><span class="co">## Platelet      0.389755  0.7743363</span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="create-an-enhanced-violin-plot">4. Create an Enhanced Violin Plot<a class="anchor" aria-label="anchor" href="#create-an-enhanced-violin-plot"></a>
</h2>
<p>While DimPlot uses color gradients to represent expression levels,
violin plots offer a more quantitative view of expression distribution.
The enhanced <code>VlnPlot2</code> function in SeuratExtend integrates
functionalities to superimpose boxplots, easily add statistical
annotations, and offers greater flexibility in plot presentation
compared to the original <code>VlnPlot</code> in Seurat.</p>
<p>The method to employ <code>VlnPlot2</code> will differ depending on
whether you’re using a Seurat object or a matrix. In this lesson, we’ll
focus on using a Seurat object.</p>
<div class="section level3">
<h3 id="basic-violin-plot-with-box-plot-and-points">Basic Violin Plot with Box Plot and Points<a class="anchor" aria-label="anchor" href="#basic-violin-plot-with-box-plot-and-points"></a>
</h3>
<p>Let’s start by selecting the genes we want to analyze. Here’s an
example using three genes:</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD3D"</span>, <span class="st">"CD14"</span>, <span class="st">"CD79A"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../../reference/VlnPlot2.html">VlnPlot2</a></span><span class="op">(</span><span class="va">pbmc</span>, features <span class="op">=</span> <span class="va">genes</span>, ncol <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<p><img src="3.Visualization_files/figure-html/unnamed-chunk-19-1.png" width="480"></p>
</div>
<div class="section level3">
<h3 id="hide-points-and-outliers-for-a-cleaner-appearance">Hide Points and Outliers for a Cleaner Appearance<a class="anchor" aria-label="anchor" href="#hide-points-and-outliers-for-a-cleaner-appearance"></a>
</h3>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../../reference/VlnPlot2.html">VlnPlot2</a></span><span class="op">(</span><span class="va">pbmc</span>, features <span class="op">=</span> <span class="va">genes</span>, pt <span class="op">=</span> <span class="cn">FALSE</span>, hide.outlier <span class="op">=</span> <span class="cn">TRUE</span>, ncol <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<p><img src="3.Visualization_files/figure-html/unnamed-chunk-20-1.png" width="480"></p>
</div>
<div class="section level3">
<h3 id="grouping-by-cluster-and-splitting-each-cluster-by-samples">Grouping by Cluster and Splitting Each Cluster by Samples<a class="anchor" aria-label="anchor" href="#grouping-by-cluster-and-splitting-each-cluster-by-samples"></a>
</h3>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../../reference/VlnPlot2.html">VlnPlot2</a></span><span class="op">(</span><span class="va">pbmc</span>, features <span class="op">=</span> <span class="va">genes</span>, group.by <span class="op">=</span> <span class="st">"cluster"</span>, split.by <span class="op">=</span> <span class="st">"sample"</span><span class="op">)</span></span></code></pre></div>
<p><img src="3.Visualization_files/figure-html/unnamed-chunk-21-1.png" width="864"></p>
</div>
<div class="section level3">
<h3 id="filtering-for-certain-subtypes-and-arranging-plots-in-columns">Filtering for Certain Subtypes and Arranging Plots in Columns<a class="anchor" aria-label="anchor" href="#filtering-for-certain-subtypes-and-arranging-plots-in-columns"></a>
</h3>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cells</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span><span class="op">[</span><span class="va">pbmc</span><span class="op">$</span><span class="va">cluster</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"B cell"</span>, <span class="st">"Mono CD14"</span>, <span class="st">"CD8 T cell"</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="fu"><a href="../../reference/VlnPlot2.html">VlnPlot2</a></span><span class="op">(</span><span class="va">pbmc</span>, features <span class="op">=</span> <span class="va">genes</span>, group.by <span class="op">=</span> <span class="st">"cluster"</span>, cells <span class="op">=</span> <span class="va">cells</span><span class="op">)</span></span></code></pre></div>
<p><img src="3.Visualization_files/figure-html/unnamed-chunk-22-1.png" width="576"></p>
</div>
<div class="section level3">
<h3 id="adding-statistical-annotations-using-the-wilcoxon-test">Adding Statistical Annotations Using the Wilcoxon Test<a class="anchor" aria-label="anchor" href="#adding-statistical-annotations-using-the-wilcoxon-test"></a>
</h3>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../../reference/VlnPlot2.html">VlnPlot2</a></span><span class="op">(</span><span class="va">pbmc</span>, features <span class="op">=</span> <span class="va">genes</span>, group.by <span class="op">=</span> <span class="st">"cluster"</span>, cells <span class="op">=</span> <span class="va">cells</span>, </span>
<span>         stat.method <span class="op">=</span> <span class="st">"wilcox.test"</span>, hide.ns <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="3.Visualization_files/figure-html/unnamed-chunk-23-1.png" width="576"></p>
</div>
<div class="section level3">
<h3 id="restricting-statistical-comparisons-and-using-t-test">Restricting Statistical Comparisons and Using t-test<a class="anchor" aria-label="anchor" href="#restricting-statistical-comparisons-and-using-t-test"></a>
</h3>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../../reference/VlnPlot2.html">VlnPlot2</a></span><span class="op">(</span><span class="va">pbmc</span>, features <span class="op">=</span> <span class="va">genes</span>, group.by <span class="op">=</span> <span class="st">"cluster"</span>, cells <span class="op">=</span> <span class="va">cells</span>, </span>
<span>         stat.method <span class="op">=</span> <span class="st">"t.test"</span>, comparisons <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">3</span><span class="op">)</span><span class="op">)</span>, hide.ns <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p><img src="3.Visualization_files/figure-html/unnamed-chunk-24-1.png" width="576"></p>
</div>
<div class="section level3">
<h3 id="adding-mean-and-median-lines">Adding Mean and Median Lines<a class="anchor" aria-label="anchor" href="#adding-mean-and-median-lines"></a>
</h3>
<p>For genes with low expression where boxplot might be hard to
interpret (e.g., when median and quartiles overlap at zero), you can add
mean/median lines for better visualization:</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">lowExprGenes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CCR7"</span>, <span class="st">"IL7R"</span>, <span class="st">"TCF7"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../../reference/VlnPlot2.html">VlnPlot2</a></span><span class="op">(</span><span class="va">pbmc</span>, </span>
<span>         features <span class="op">=</span> <span class="va">lowExprGenes</span>, </span>
<span>         show.mean <span class="op">=</span> <span class="cn">TRUE</span>,      <span class="co"># Show mean and median lines</span></span>
<span>         mean_colors <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"red"</span>, <span class="st">"blue"</span><span class="op">)</span>,  <span class="co"># Colors for mean and median</span></span>
<span>         cols <span class="op">=</span> <span class="st">"light"</span>,        <span class="co"># Light color scheme for better visibility</span></span>
<span>         ncol <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<p><img src="3.Visualization_files/figure-html/unnamed-chunk-25-1.png" width="576"></p>
</div>
</div>
<div class="section level2">
<h2 id="generate-a-heatmap-plot">5. Generate a Heatmap Plot<a class="anchor" aria-label="anchor" href="#generate-a-heatmap-plot"></a>
</h2>
<div class="section level3">
<h3 id="introduction-2">Introduction<a class="anchor" aria-label="anchor" href="#introduction-2"></a>
</h3>
<p>The <code>Heatmap</code> function provides a flexible and
comprehensive way to visualize matrices, especially those produced by
the <code>CalcStats</code> function.</p>
</div>
<div class="section level3">
<h3 id="basic-usage-2">Basic Usage<a class="anchor" aria-label="anchor" href="#basic-usage-2"></a>
</h3>
<p>First, let’s generate a sample matrix using the
<code>CalcStats</code> function:</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Assuming pbmc data and VariableFeatures function are available</span></span>
<span><span class="va">genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/VariableFeatures.html" class="external-link">VariableFeatures</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span></span>
<span><span class="va">toplot</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/CalcStats.html">CalcStats</a></span><span class="op">(</span><span class="va">pbmc</span>, features <span class="op">=</span> <span class="va">genes</span>, method <span class="op">=</span> <span class="st">"zscore"</span>, order <span class="op">=</span> <span class="st">"p"</span>, n <span class="op">=</span> <span class="fl">4</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">toplot</span>, <span class="fl">10</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##             B cell CD4 T Memory CD4 T Naive CD8 T cell          DC  Mono CD14</span></span>
<span><span class="co">## CD79A    2.6618127   -0.3562737  -0.3290077 -0.3786881 -0.17755727 -0.3557219</span></span>
<span><span class="co">## CD79B    2.4985291   -0.5208352  -0.4673972 -0.3908471 -0.41913798 -0.5194812</span></span>
<span><span class="co">## MS4A1    2.6641850   -0.3764259  -0.3706329 -0.3598127 -0.25991613 -0.3741869</span></span>
<span><span class="co">## TCL1A    2.6529418   -0.3558571  -0.3687355 -0.3450457 -0.07084613 -0.3691535</span></span>
<span><span class="co">## MALAT1   0.5332295    0.8934148   0.5249507  0.8180707 -0.46598087 -0.5652067</span></span>
<span><span class="co">## TRABD2A -0.2134085    2.0559473   1.3735467 -0.4193964 -0.45593495 -0.4890393</span></span>
<span><span class="co">## MYC      0.6612194    1.7263843   1.1618128  0.2927673 -0.56974739 -0.7537375</span></span>
<span><span class="co">## BEX2     0.2726716    1.4948863   1.4176286  0.5119763 -0.81115423 -0.9065074</span></span>
<span><span class="co">## LTB      1.3384952    0.9219028   1.4451552  0.1225591 -0.67675218 -0.9194058</span></span>
<span><span class="co">## CD2     -0.8215673    0.7197231   1.5507200  1.4481432 -0.38941270 -0.8493816</span></span>
<span><span class="co">##         Mono FCGR3A     NK cell   Platelet</span></span>
<span><span class="co">## CD79A   -0.35728720 -0.34074387 -0.3665329</span></span>
<span><span class="co">## CD79B    0.56426973 -0.18106015 -0.5640398</span></span>
<span><span class="co">## MS4A1   -0.31198084 -0.33629645 -0.2749332</span></span>
<span><span class="co">## TCL1A   -0.34902797 -0.36804994 -0.4262259</span></span>
<span><span class="co">## MALAT1  -0.07311915  0.58929272 -2.2546517</span></span>
<span><span class="co">## TRABD2A -0.39852095 -0.68614635 -0.7670475</span></span>
<span><span class="co">## MYC     -0.70035646 -0.67417336 -1.1441692</span></span>
<span><span class="co">## BEX2    -0.91577760 -0.01463581 -1.0490877</span></span>
<span><span class="co">## LTB     -0.31365215 -0.87726190 -1.0410403</span></span>
<span><span class="co">## CD2     -0.70178407  0.01279555 -0.9692362</span></span></code></pre>
<p>Now, we can produce a basic heatmap:</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../../reference/Heatmap.html">Heatmap</a></span><span class="op">(</span><span class="va">toplot</span>, lab_fill <span class="op">=</span> <span class="st">"zscore"</span><span class="op">)</span></span></code></pre></div>
<p><img src="3.Visualization_files/figure-html/unnamed-chunk-27-1.png" width="480"></p>
<p>Note that this provides a convenient method to display marker genes
(or differentially expressed genes, DEGs) for each cluster, which is
very practical. However, this is not the only method. Let’s also learn
about Seurat’s built-in method for finding DEGs.</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Find all markers of B cells</span></span>
<span><span class="va">bcell.markers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindMarkers.html" class="external-link">FindMarkers</a></span><span class="op">(</span><span class="va">pbmc</span>, ident.1 <span class="op">=</span> <span class="st">"B cell"</span>, logfc.threshold <span class="op">=</span> <span class="fl">1</span>, only.pos <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">bcell.markers</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                   p_val avg_log2FC pct.1 pct.2     p_val_adj</span></span>
<span><span class="co">## CD79A      0.000000e+00   6.783307 0.934 0.043  0.000000e+00</span></span>
<span><span class="co">## MS4A1      0.000000e+00   5.737084 0.860 0.052  0.000000e+00</span></span>
<span><span class="co">## LINC00926 2.387605e-276   7.258980 0.562 0.009 3.971542e-272</span></span>
<span><span class="co">## CD79B     2.983575e-275   4.595655 0.911 0.143 4.962879e-271</span></span>
<span><span class="co">## TCL1A     8.485045e-274   6.612804 0.619 0.022 1.411402e-269</span></span>
<span><span class="co">## HLA-DQA1  4.817482e-269   3.984138 0.885 0.118 8.013400e-265</span></span></code></pre>
<p>This compares all B cells to non-B cells to find DEGs. If you want to
compare B cells with specific clusters, you can set the
<code>ident.2</code> parameter.</p>
<p>We can also use <code>FindAllMarkers</code> to conveniently generate
marker genes for all clusters:</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc.markers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindAllMarkers.html" class="external-link">FindAllMarkers</a></span><span class="op">(</span><span class="va">pbmc</span>, logfc.threshold <span class="op">=</span> <span class="fl">2</span>, only.pos <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Calculating cluster B cell</span></span></code></pre>
<pre><code><span><span class="co">## Calculating cluster CD4 T Memory</span></span></code></pre>
<pre><code><span><span class="co">## Calculating cluster CD4 T Naive</span></span></code></pre>
<pre><code><span><span class="co">## Calculating cluster CD8 T cell</span></span></code></pre>
<pre><code><span><span class="co">## Calculating cluster DC</span></span></code></pre>
<pre><code><span><span class="co">## Calculating cluster Mono CD14</span></span></code></pre>
<pre><code><span><span class="co">## Calculating cluster Mono FCGR3A</span></span></code></pre>
<pre><code><span><span class="co">## Calculating cluster NK cell</span></span></code></pre>
<pre><code><span><span class="co">## Calculating cluster Platelet</span></span></code></pre>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">pbmc.markers</span>, <span class="fl">10</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                   p_val avg_log2FC pct.1 pct.2     p_val_adj cluster      gene</span></span>
<span><span class="co">## CD79A      0.000000e+00   6.783307 0.934 0.043  0.000000e+00  B cell     CD79A</span></span>
<span><span class="co">## MS4A1      0.000000e+00   5.737084 0.860 0.052  0.000000e+00  B cell     MS4A1</span></span>
<span><span class="co">## LINC00926 2.387605e-276   7.258980 0.562 0.009 3.971542e-272  B cell LINC00926</span></span>
<span><span class="co">## CD79B     2.983575e-275   4.595655 0.911 0.143 4.962879e-271  B cell     CD79B</span></span>
<span><span class="co">## TCL1A     8.485045e-274   6.612804 0.619 0.022 1.411402e-269  B cell     TCL1A</span></span>
<span><span class="co">## HLA-DQA1  4.817482e-269   3.984138 0.885 0.118 8.013400e-265  B cell  HLA-DQA1</span></span>
<span><span class="co">## VPREB3    6.983605e-238   6.957038 0.484 0.007 1.161653e-233  B cell    VPREB3</span></span>
<span><span class="co">## HLA-DQB1  3.947592e-234   4.054111 0.862 0.146 6.566424e-230  B cell  HLA-DQB1</span></span>
<span><span class="co">## CD74      1.691378e-188   3.004657 1.000 0.818 2.813439e-184  B cell      CD74</span></span>
<span><span class="co">## HLA-DRA   1.834090e-187   2.864834 1.000 0.490 3.050826e-183  B cell   HLA-DRA</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="customizing-the-heatmap">Customizing the Heatmap<a class="anchor" aria-label="anchor" href="#customizing-the-heatmap"></a>
</h3>
<div class="section level4">
<h4 id="modifying-axis-and-labels">Modifying Axis and Labels<a class="anchor" aria-label="anchor" href="#modifying-axis-and-labels"></a>
</h4>
<p>Sometimes, the first name on the x-axis might be too long and exceed
the left boundary of the plot. To prevent this issue and ensure all
labels are fully visible, you can increase the space on the left side of
the plot by adjusting the <code>plot.margin</code> parameter:</p>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../../reference/Heatmap.html">Heatmap</a></span><span class="op">(</span><span class="va">toplot</span>, lab_fill <span class="op">=</span> <span class="st">"zscore"</span>, plot.margin <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">margin</a></span><span class="op">(</span>l <span class="op">=</span> <span class="fl">30</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="3.Visualization_files/figure-html/unnamed-chunk-30-1.png" width="480"></p>
<p>For denser matrices, you may wish to only show a subset of gene
names:</p>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/VariableFeatures.html" class="external-link">VariableFeatures</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">500</span><span class="op">]</span></span>
<span><span class="va">toplot2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/CalcStats.html">CalcStats</a></span><span class="op">(</span><span class="va">pbmc</span>, features <span class="op">=</span> <span class="va">genes</span>, method <span class="op">=</span> <span class="st">"zscore"</span>, order <span class="op">=</span> <span class="st">"p"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../../reference/Heatmap.html">Heatmap</a></span><span class="op">(</span><span class="va">toplot2</span>, lab_fill <span class="op">=</span> <span class="st">"zscore"</span>, feature_text_subset <span class="op">=</span> <span class="va">genes</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">20</span><span class="op">]</span>, expand_limits_x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">0.5</span>, <span class="fl">12</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="3.Visualization_files/figure-html/unnamed-chunk-31-1.png" width="480"></p>
</div>
<div class="section level4">
<h4 id="faceting-the-heatmap">Faceting the Heatmap<a class="anchor" aria-label="anchor" href="#faceting-the-heatmap"></a>
</h4>
<p>You can also split the heatmap based on gene groups:</p>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gene_groups</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"group1"</span>, <span class="st">"group2"</span>, <span class="st">"group3"</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">toplot2</span><span class="op">)</span>, replace <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="../../reference/Heatmap.html">Heatmap</a></span><span class="op">(</span><span class="va">toplot2</span>, lab_fill <span class="op">=</span> <span class="st">"zscore"</span>, facet_row <span class="op">=</span> <span class="va">gene_groups</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.text.y <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="3.Visualization_files/figure-html/unnamed-chunk-32-1.png" width="480"></p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="create-enhanced-dot-plots">6. Create Enhanced Dot Plots<a class="anchor" aria-label="anchor" href="#create-enhanced-dot-plots"></a>
</h2>
<div class="section level3">
<h3 id="introduction-3">Introduction<a class="anchor" aria-label="anchor" href="#introduction-3"></a>
</h3>
<p>While heatmaps are excellent for visualizing overall patterns, dot
plots offer a complementary visualization that combines two key pieces
of information: 1. The percentage of cells expressing each feature
(represented by dot size) 2. The average expression level in expressing
cells (represented by color intensity)</p>
<p>This dual representation can provide more detailed insights than a
standard heatmap, particularly when expression patterns vary between
cell populations.</p>
</div>
<div class="section level3">
<h3 id="basic-usage-3">Basic Usage<a class="anchor" aria-label="anchor" href="#basic-usage-3"></a>
</h3>
<p>Let’s create a basic dot plot for some variable features:</p>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Select some variable features</span></span>
<span><span class="va">genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/VariableFeatures.html" class="external-link">VariableFeatures</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span></span>
<span><span class="fu"><a href="../../reference/DotPlot2.html">DotPlot2</a></span><span class="op">(</span><span class="va">pbmc</span>, features <span class="op">=</span> <span class="va">genes</span><span class="op">)</span></span></code></pre></div>
<p><img src="3.Visualization_files/figure-html/unnamed-chunk-33-1.png" width="576"></p>
</div>
<div class="section level3">
<h3 id="grouped-features">Grouped Features<a class="anchor" aria-label="anchor" href="#grouped-features"></a>
</h3>
<p>You can organize features into groups for clearer visualization:</p>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Create grouped features</span></span>
<span><span class="va">grouped_features</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>  <span class="st">"B_cell_markers"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"MS4A1"</span>, <span class="st">"CD79A"</span><span class="op">)</span>,</span>
<span>  <span class="st">"T_cell_markers"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD3D"</span>, <span class="st">"CD8A"</span>, <span class="st">"IL7R"</span><span class="op">)</span>,</span>
<span>  <span class="st">"Myeloid_markers"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD14"</span>, <span class="st">"FCGR3A"</span>, <span class="st">"S100A8"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../../reference/DotPlot2.html">DotPlot2</a></span><span class="op">(</span><span class="va">pbmc</span>, features <span class="op">=</span> <span class="va">grouped_features</span><span class="op">)</span></span></code></pre></div>
<p><img src="3.Visualization_files/figure-html/unnamed-chunk-34-1.png" width="624"></p>
</div>
<div class="section level3">
<h3 id="split-visualization">Split Visualization<a class="anchor" aria-label="anchor" href="#split-visualization"></a>
</h3>
<p>You can split the visualization by a metadata column to compare
expression patterns across different conditions:</p>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Basic split visualization</span></span>
<span><span class="fu"><a href="../../reference/DotPlot2.html">DotPlot2</a></span><span class="op">(</span><span class="va">pbmc</span>, features <span class="op">=</span> <span class="va">genes</span>, group.by <span class="op">=</span> <span class="st">"cluster"</span>, split.by <span class="op">=</span> <span class="st">"orig.ident"</span>, show_grid <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p><img src="3.Visualization_files/figure-html/unnamed-chunk-35-1.png" width="768"></p>
</div>
<div class="section level3">
<h3 id="alternative-split-visualization">Alternative Split Visualization<a class="anchor" aria-label="anchor" href="#alternative-split-visualization"></a>
</h3>
<p>The split groups can be represented either by border colors (default)
or by dot colors:</p>
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Using colors instead of borders for split groups</span></span>
<span><span class="fu"><a href="../../reference/DotPlot2.html">DotPlot2</a></span><span class="op">(</span><span class="va">pbmc</span>, </span>
<span>         features <span class="op">=</span> <span class="va">genes</span>, </span>
<span>         group.by <span class="op">=</span> <span class="st">"cluster"</span>, </span>
<span>         split.by <span class="op">=</span> <span class="st">"orig.ident"</span>, </span>
<span>         split.by.method <span class="op">=</span> <span class="st">"color"</span>, </span>
<span>         show_grid <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p><img src="3.Visualization_files/figure-html/unnamed-chunk-36-1.png" width="768"></p>
</div>
<div class="section level3">
<h3 id="customizing-appearance">Customizing Appearance<a class="anchor" aria-label="anchor" href="#customizing-appearance"></a>
</h3>
<p>You can customize various aspects of the plot, including color
schemes, dot sizes, and grid lines:</p>
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../../reference/DotPlot2.html">DotPlot2</a></span><span class="op">(</span><span class="va">pbmc</span>, </span>
<span>         features <span class="op">=</span> <span class="va">grouped_features</span>, </span>
<span>         color_scheme <span class="op">=</span> <span class="st">"BuRd"</span>, </span>
<span>         border <span class="op">=</span> <span class="cn">FALSE</span>,        <span class="co"># Remove dot borders</span></span>
<span>         show_grid <span class="op">=</span> <span class="cn">FALSE</span>,     <span class="co"># Remove grid lines</span></span>
<span>         flip <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>          <span class="co"># Flip coordinates</span></span></code></pre></div>
<p><img src="3.Visualization_files/figure-html/unnamed-chunk-37-1.png" width="672"></p>
</div>
</div>
<div class="section level2">
<h2 id="generate-a-waterfall-plot">7. Generate a Waterfall Plot<a class="anchor" aria-label="anchor" href="#generate-a-waterfall-plot"></a>
</h2>
<div class="section level3">
<h3 id="introduction-4">Introduction<a class="anchor" aria-label="anchor" href="#introduction-4"></a>
</h3>
<p>Waterfall plots are powerful visualization tools that can display
differences between two conditions, showing gene expression, gene set
enrichment, or other metrics. This function can handle inputs directly
from Seurat objects or pre-processed matrices.</p>
<p>You can use the waterfall plot to compare expression levels of genes
directly from a Seurat object, using LogFC to determine the bar length.
Here’s how to do it for the top 80 variable features:</p>
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/VariableFeatures.html" class="external-link">VariableFeatures</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">80</span><span class="op">]</span></span>
<span><span class="fu"><a href="../../reference/WaterfallPlot.html">WaterfallPlot</a></span><span class="op">(</span></span>
<span>  <span class="va">pbmc</span>, group.by <span class="op">=</span> <span class="st">"cluster"</span>, features <span class="op">=</span> <span class="va">genes</span>,</span>
<span>  ident.1 <span class="op">=</span> <span class="st">"Mono CD14"</span>, ident.2 <span class="op">=</span> <span class="st">"CD8 T cell"</span>, length <span class="op">=</span> <span class="st">"logFC"</span><span class="op">)</span></span></code></pre></div>
<p><img src="3.Visualization_files/figure-html/unnamed-chunk-38-1.png" width="960"></p>
</div>
<div class="section level3">
<h3 id="focusing-on-extremes">Focusing on Extremes<a class="anchor" aria-label="anchor" href="#focusing-on-extremes"></a>
</h3>
<p>To further hone in on the most differentially expressed genes, you
might want to keep only the top and bottom 20 genes. This can highlight
the most critical differences between the two cell types:</p>
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../../reference/WaterfallPlot.html">WaterfallPlot</a></span><span class="op">(</span></span>
<span>  <span class="va">pbmc</span>, group.by <span class="op">=</span> <span class="st">"cluster"</span>, features <span class="op">=</span> <span class="va">genes</span>,</span>
<span>  ident.1 <span class="op">=</span> <span class="st">"Mono CD14"</span>, ident.2 <span class="op">=</span> <span class="st">"CD8 T cell"</span>, length <span class="op">=</span> <span class="st">"logFC"</span>,</span>
<span>  top.n <span class="op">=</span> <span class="fl">20</span><span class="op">)</span></span></code></pre></div>
<p><img src="3.Visualization_files/figure-html/unnamed-chunk-39-1.png" width="672"></p>
<p>Additionally, you can set parameters to modify the plot, such as
filtering features based on a specific threshold, rotating the plot 90
degrees (by setting <code>flip</code>), and more. For detailed options,
refer to the function’s help documentation.</p>
</div>
</div>
<div class="section level2">
<h2 id="create-volcano-plots">8. Create Volcano Plots<a class="anchor" aria-label="anchor" href="#create-volcano-plots"></a>
</h2>
<div class="section level3">
<h3 id="introduction-5">Introduction<a class="anchor" aria-label="anchor" href="#introduction-5"></a>
</h3>
<p>The Volcano plot is a specialized variation of the scatter plot that
combines statistical significance (-log10 p-value) with magnitude of
change (typically log fold change). While Waterfall plots focus on
ordered visualization of differences, Volcano plots provide a
comprehensive view of both the magnitude and statistical significance of
changes.</p>
</div>
<div class="section level3">
<h3 id="basic-usage-4">Basic Usage<a class="anchor" aria-label="anchor" href="#basic-usage-4"></a>
</h3>
<p>Create a basic volcano plot comparing two cell types:</p>
<div class="sourceCode" id="cb56"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../../reference/VolcanoPlot.html">VolcanoPlot</a></span><span class="op">(</span><span class="va">pbmc</span>, </span>
<span>            ident.1 <span class="op">=</span> <span class="st">"B cell"</span>,</span>
<span>            ident.2 <span class="op">=</span> <span class="st">"CD8 T cell"</span><span class="op">)</span></span></code></pre></div>
<p><img src="3.Visualization_files/figure-html/unnamed-chunk-40-1.png" width="480"></p>
</div>
<div class="section level3">
<h3 id="customizing-thresholds">Customizing Thresholds<a class="anchor" aria-label="anchor" href="#customizing-thresholds"></a>
</h3>
<p>You can set custom thresholds for both axes:</p>
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../../reference/VolcanoPlot.html">VolcanoPlot</a></span><span class="op">(</span></span>
<span>  <span class="va">pbmc</span>,</span>
<span>  ident.1 <span class="op">=</span> <span class="st">"B cell"</span>,</span>
<span>  ident.2 <span class="op">=</span> <span class="st">"CD8 T cell"</span>,</span>
<span>  x.threshold <span class="op">=</span> <span class="fl">0.5</span>,    <span class="co"># Log fold change threshold</span></span>
<span>  y.threshold <span class="op">=</span> <span class="fl">2</span>     <span class="co"># -log10(p-value) threshold</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="3.Visualization_files/figure-html/unnamed-chunk-41-1.png" width="480"></p>
</div>
</div>
<div class="section level2">
<h2 id="tips-choosing-the-right-visualization">9. Tips: Choosing the Right Visualization<a class="anchor" aria-label="anchor" href="#tips-choosing-the-right-visualization"></a>
</h2>
<p>When visualizing gene expression or pathway activity data, different
plot types serve different purposes. Here’s a guide to help you
choose:</p>
<ol style="list-style-type: decimal">
<li>Overview Visualizations:
<ul>
<li>
<strong>Heatmaps</strong>:
<ul>
<li>Best for: Comparing many features across multiple groups using color
intensity</li>
<li>Advantage: Compact visualization of large-scale patterns</li>
<li>Limitation: Color intensity alone may not show detailed expression
patterns</li>
</ul>
</li>
<li>
<strong>Dot Plots</strong>:
<ul>
<li>Best for: Similar to heatmaps but with additional information</li>
<li>Advantage: Shows both expression level (color) and percent
expressing cells (size)</li>
<li>Limitation: Takes more space than heatmaps</li>
</ul>
</li>
</ul>
</li>
<li>Two-Group Comparisons:
<ul>
<li>
<strong>Waterfall Plots</strong>:
<ul>
<li>Best for: Ordered visualization of differences between two
groups</li>
<li>Advantage: Clear ranking of most differential features</li>
<li>Limitation: Limited to two-group comparisons</li>
</ul>
</li>
<li>
<strong>Volcano Plots</strong>:
<ul>
<li>Best for: Combining fold change with statistical significance</li>
<li>Advantage: Shows both effect size and significance</li>
<li>Limitation: Also limited to two-group comparisons</li>
</ul>
</li>
</ul>
</li>
<li>Detailed Expression Analysis:
<ul>
<li>
<strong>Violin Plots</strong>:
<ul>
<li>Best for: Detailed view of expression distribution</li>
<li>Advantage: Shows full distribution shape and statistical
comparisons</li>
<li>Limitation: Space-intensive, limiting the number of features that
can be shown</li>
</ul>
</li>
</ul>
</li>
</ol>
</div>
<div class="section level2">
<h2 id="explore-color-functions">10. Explore Color Functions<a class="anchor" aria-label="anchor" href="#explore-color-functions"></a>
</h2>
<p>The <code>SeuratExtend</code> package provides comprehensive color
management across all visualization functions. Each function type
handles colors slightly differently depending on whether it’s dealing
with discrete or continuous variables.</p>
<p>For more detailed information, refer to the online tutorial: <a href="https://huayc09.github.io/SeuratExtend/articles/Visualization.html#explore-color-functions" class="external-link uri">https://huayc09.github.io/SeuratExtend/articles/Visualization.html#explore-color-functions</a></p>
<div class="section level3">
<h3 id="recommended-color-schemes">Recommended Color Schemes<a class="anchor" aria-label="anchor" href="#recommended-color-schemes"></a>
</h3>
<p>Here are some recommended color schemes for different visualization
needs in single-cell analysis:</p>
<div class="section level4">
<h4 id="for-discrete-variables-e-g--clusters-conditions">For Discrete Variables (e.g., clusters, conditions)<a class="anchor" aria-label="anchor" href="#for-discrete-variables-e-g--clusters-conditions"></a>
</h4>
<p>The “light” color scheme from SeuratExtend’s <code>color_pro</code>
is particularly effective for labeled visualizations:</p>
<div class="sourceCode" id="cb58"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://wilkelab.org/cowplot/" class="external-link">cowplot</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html" class="external-link">plot_grid</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="../../reference/DimPlot2.html">DimPlot2</a></span><span class="op">(</span><span class="va">pbmc</span>, cols <span class="op">=</span> <span class="st">"light"</span>, theme <span class="op">=</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SeuratTheme.html" class="external-link">NoAxes</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SeuratTheme.html" class="external-link">NoLegend</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  <span class="fu"><a href="../../reference/ClusterDistrBar.html">ClusterDistrBar</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">$</span><span class="va">orig.ident</span>, <span class="va">pbmc</span><span class="op">$</span><span class="va">cluster</span>, cols <span class="op">=</span> <span class="st">"light"</span>, flip <span class="op">=</span> <span class="cn">FALSE</span>, border <span class="op">=</span> <span class="st">"black"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.title.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="3.Visualization_files/figure-html/unnamed-chunk-42-1.png" width="720"></p>
</div>
<div class="section level4">
<h4 id="for-continuous-variables-e-g--gene-expression">For Continuous Variables (e.g., gene expression)<a class="anchor" aria-label="anchor" href="#for-continuous-variables-e-g--gene-expression"></a>
</h4>
<p>the “A” or “D” color schemes from viridis:</p>
<div class="sourceCode" id="cb59"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../../reference/Heatmap.html">Heatmap</a></span><span class="op">(</span><span class="va">toplot</span>, lab_fill <span class="op">=</span> <span class="st">"zscore"</span>, color_scheme <span class="op">=</span> <span class="st">"A"</span><span class="op">)</span></span></code></pre></div>
<p><img src="3.Visualization_files/figure-html/unnamed-chunk-43-1.png" width="480"></p>
<div class="sourceCode" id="cb60"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../../reference/WaterfallPlot.html">WaterfallPlot</a></span><span class="op">(</span></span>
<span>  <span class="va">pbmc</span>, group.by <span class="op">=</span> <span class="st">"cluster"</span>, features <span class="op">=</span> <span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/VariableFeatures.html" class="external-link">VariableFeatures</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">80</span><span class="op">]</span>,</span>
<span>  ident.1 <span class="op">=</span> <span class="st">"Mono CD14"</span>, ident.2 <span class="op">=</span> <span class="st">"CD8 T cell"</span>, length <span class="op">=</span> <span class="st">"logFC"</span>,</span>
<span>  top.n <span class="op">=</span> <span class="fl">20</span>, color_theme <span class="op">=</span> <span class="st">"D"</span><span class="op">)</span></span></code></pre></div>
<p><img src="3.Visualization_files/figure-html/unnamed-chunk-44-1.png" width="672"></p>
<p>You can also use various color palettes from RColorBrewer (<a href="https://r-graph-gallery.com/38-rcolorbrewers-palettes.html" class="external-link uri">https://r-graph-gallery.com/38-rcolorbrewers-palettes.html</a>):</p>
<div class="sourceCode" id="cb61"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">RColorBrewer</span><span class="op">)</span></span>
<span><span class="va">markers_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>  <span class="st">"MS4A1"</span>, <span class="st">"GNLY"</span>, <span class="st">"CD3E"</span>,</span>
<span>  <span class="st">"CD8A"</span>, <span class="st">"CCR7"</span>, <span class="st">"CD14"</span>,</span>
<span>  <span class="st">"FCER1A"</span>, <span class="st">"FCGR3A"</span>, <span class="st">"PPBP"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../../reference/DimPlot2.html">DimPlot2</a></span><span class="op">(</span></span>
<span>  <span class="va">pbmc</span>,</span>
<span>  features <span class="op">=</span> <span class="va">markers_genes</span>,</span>
<span>  cols <span class="op">=</span> <span class="st">"OrRd"</span>,</span>
<span>  theme <span class="op">=</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SeuratTheme.html" class="external-link">NoAxes</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="3.Visualization_files/figure-html/unnamed-chunk-45-1.png" width="768"></p>
<div class="sourceCode" id="cb62"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../../reference/DimPlot2.html">DimPlot2</a></span><span class="op">(</span></span>
<span>  <span class="va">pbmc</span>,</span>
<span>  features <span class="op">=</span> <span class="va">markers_genes</span>,</span>
<span>  cols <span class="op">=</span> <span class="st">"Spectral-rev"</span>,</span>
<span>  theme <span class="op">=</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SeuratTheme.html" class="external-link">NoAxes</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="3.Visualization_files/figure-html/unnamed-chunk-46-1.png" width="768"></p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="mastering-ggplot2-for-custom-visualizations">11. Mastering ggplot2 for Custom Visualizations<a class="anchor" aria-label="anchor" href="#mastering-ggplot2-for-custom-visualizations"></a>
</h2>
<div class="section level3">
<h3 id="introduction-to-ggplot2">Introduction to ggplot2<a class="anchor" aria-label="anchor" href="#introduction-to-ggplot2"></a>
</h3>
<p>While <code>SeuratExtend</code> provides many convenient functions
for visualization, all of these are built on top of
<code>ggplot2.</code> Understanding <code>ggplot2</code> can give you
more flexibility in customizing your plots and creating entirely new
visualizations. In this section, we’ll cover the basics of
<code>ggplot2</code> and how to use it with Seurat data.</p>
</div>
<div class="section level3">
<h3 id="the-grammar-of-graphics">The Grammar of Graphics<a class="anchor" aria-label="anchor" href="#the-grammar-of-graphics"></a>
</h3>
<p>ggplot2 is based on the Grammar of Graphics, a layered approach to
creating visualizations. The basic components are:</p>
<ol style="list-style-type: decimal">
<li>Data: The dataset you want to visualize</li>
<li>Aesthetics: How your data maps to visual properties (position,
color, size, etc.)</li>
<li>Geometries: The actual shapes used to represent your data (points,
lines, bars, etc.)</li>
<li>Facets: Optional subdivision of the plot into multiple plots</li>
<li>Statistics: Optional statistical transformations of the data</li>
<li>Coordinates: The space on which the data will be plotted</li>
<li>Themes: All non-data ink</li>
</ol>
</div>
<div class="section level3">
<h3 id="creating-a-basic-plot-from-seurat-data">Creating a Basic Plot from Seurat Data<a class="anchor" aria-label="anchor" href="#creating-a-basic-plot-from-seurat-data"></a>
</h3>
<p>Let’s start by creating the most basic UMAP plot using ggplot2:</p>
<div class="sourceCode" id="cb63"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Extract data from the Seurat object</span></span>
<span><span class="va">umap_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/FetchData.html" class="external-link">FetchData</a></span><span class="op">(</span><span class="va">pbmc</span>, vars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"umap_1"</span>, <span class="st">"umap_2"</span>, <span class="st">"cluster"</span>, <span class="st">"CD3D"</span>, <span class="st">"sample"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">umap_data</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                      umap_1     umap_2     cluster     CD3D  sample</span></span>
<span><span class="co">## AAACATACAACCAC-1   4.075107   3.848914 CD4 T Naive 2.863463 sample1</span></span>
<span><span class="co">## AAACATTGAGCTAC-1   4.548936 -12.860142      B cell 0.000000 sample1</span></span>
<span><span class="co">## AAACATTGATCAGC-1   7.082211   5.789762 CD4 T Naive 3.489090 sample1</span></span>
<span><span class="co">## AAACCGTGCTTCCG-1 -11.915672  -1.153419 Mono FCGR3A 0.000000 sample1</span></span>
<span><span class="co">## AAACCGTGTATGCG-1   4.301518  -2.344230     NK cell 0.000000 sample1</span></span>
<span><span class="co">## AAACGCACTGGTAC-1   3.169024   6.136975 CD4 T Naive 1.726522 sample1</span></span></code></pre>
<div class="sourceCode" id="cb65"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Create the most basic plot</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">umap_data</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">umap_1</span>, y <span class="op">=</span> <span class="va">umap_2</span>, color <span class="op">=</span> <span class="va">CD3D</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="3.Visualization_files/figure-html/unnamed-chunk-47-1.png" width="576"></p>
<p>This is the most basic plot we can create. It shows the UMAP
coordinates of our cells, colored by CD3D expression. However, it’s not
very visually appealing or informative in its current state. Let’s
improve it step by step:</p>
<div class="sourceCode" id="cb66"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Create an improved plot</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://sjmgarnier.github.io/viridis/" class="external-link">viridis</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">umap_data</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">umap_1</span>, y <span class="op">=</span> <span class="va">umap_2</span>, color <span class="op">=</span> <span class="va">CD3D</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">0.5</span>, alpha <span class="op">=</span> <span class="fl">0.7</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://sjmgarnier.github.io/viridis/reference/scale_viridis.html" class="external-link">scale_color_viridis</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_minimal</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"UMAP colored by CD3D expression"</span>,</span>
<span>       x <span class="op">=</span> <span class="st">"UMAP_1"</span>, y <span class="op">=</span> <span class="st">"UMAP_2"</span><span class="op">)</span></span></code></pre></div>
<p><img src="3.Visualization_files/figure-html/unnamed-chunk-48-1.png" width="576"></p>
<p>Let’s break down the improvements we made:</p>
<ol style="list-style-type: decimal">
<li>We adjusted the size and transparency of the points using
<code>size = 0.5, alpha = 0.7</code> in <code><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point()</a></code>. This
helps to visualize dense areas better.</li>
<li>We used <code><a href="https://sjmgarnier.github.io/viridis/reference/scale_viridis.html" class="external-link">scale_color_viridis()</a></code> to apply a
colorblind-friendly, perceptually uniform color scale.</li>
<li>We applied <code><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_minimal()</a></code> to remove unnecessary plot
elements and give it a cleaner look.</li>
<li>We added a title and axis labels using <code><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs()</a></code>.</li>
</ol>
<p>These small changes significantly improve the readability and
aesthetics of our plot. In the following sections, we’ll explore even
more ways to customize and enhance our visualizations.</p>
</div>
<div class="section level3">
<h3 id="customizing-the-plot">Customizing the Plot<a class="anchor" aria-label="anchor" href="#customizing-the-plot"></a>
</h3>
<p>Now let’s customize our plot further:</p>
<div class="sourceCode" id="cb67"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">umap_data</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">umap_1</span>, y <span class="op">=</span> <span class="va">umap_2</span>, color <span class="op">=</span> <span class="va">CD3D</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">0.5</span>, alpha <span class="op">=</span> <span class="fl">0.7</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://sjmgarnier.github.io/viridis/reference/scale_viridis.html" class="external-link">scale_color_viridis</a></span><span class="op">(</span>option <span class="op">=</span> <span class="st">"A"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html" class="external-link">facet_wrap</a></span><span class="op">(</span><span class="op">~</span><span class="va">sample</span>, ncol <span class="op">=</span> <span class="fl">3</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.text <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>        axis.ticks <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>        panel.grid <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>        strip.background <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_rect</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="st">"white"</span><span class="op">)</span>,</span>
<span>        strip.text <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>face <span class="op">=</span> <span class="st">"bold"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"UMAP by cluster, colored by CD3D expression"</span>,</span>
<span>       x <span class="op">=</span> <span class="st">"UMAP_1"</span>, y <span class="op">=</span> <span class="st">"UMAP_2"</span><span class="op">)</span></span></code></pre></div>
<p><img src="3.Visualization_files/figure-html/unnamed-chunk-49-1.png" width="672"></p>
<p>In this version:</p>
<ol style="list-style-type: decimal">
<li>We’ve used <code><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html" class="external-link">facet_wrap()</a></code> to create separate plots for
each cluster.</li>
<li>We’ve customized the theme further using <code><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme()</a></code> to
remove axis text, ticks, and grid lines.</li>
<li>We’ve styled the facet labels with a white background and bold
text.</li>
</ol>
</div>
<div class="section level3">
<h3 id="combining-multiple-plots">Combining Multiple Plots<a class="anchor" aria-label="anchor" href="#combining-multiple-plots"></a>
</h3>
<p>Often, we want to combine multiple plots into a single figure. The
<code>cowplot</code> package is great for this:</p>
<div class="sourceCode" id="cb68"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://wilkelab.org/cowplot/" class="external-link">cowplot</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Create two plots</span></span>
<span><span class="va">plot1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">umap_data</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">umap_1</span>, y <span class="op">=</span> <span class="va">umap_2</span>, color <span class="op">=</span> <span class="va">cluster</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">0.5</span>, alpha <span class="op">=</span> <span class="fl">0.7</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="../../reference/color_pro.html">color_pro</a></span><span class="op">(</span><span class="fl">9</span>,<span class="fl">2</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_minimal</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Clusters"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">plot2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">umap_data</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">umap_1</span>, y <span class="op">=</span> <span class="va">umap_2</span>, color <span class="op">=</span> <span class="va">CD3D</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">0.5</span>, alpha <span class="op">=</span> <span class="fl">0.7</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_viridis.html" class="external-link">scale_color_viridis_c</a></span><span class="op">(</span>option <span class="op">=</span> <span class="st">"A"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_minimal</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"CD3D Expression"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Combine plots</span></span>
<span><span class="va">combined_plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html" class="external-link">plot_grid</a></span><span class="op">(</span><span class="va">plot1</span>, <span class="va">plot2</span>, labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"A"</span>, <span class="st">"B"</span><span class="op">)</span>, ncol <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Add a title to the combined plot</span></span>
<span><span class="va">title</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://wilkelab.org/cowplot/reference/ggdraw.html" class="external-link">ggdraw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://wilkelab.org/cowplot/reference/draw_label.html" class="external-link">draw_label</a></span><span class="op">(</span><span class="st">"UMAP Visualization of PBMC Data"</span>, fontface <span class="op">=</span> <span class="st">"bold"</span>, x <span class="op">=</span> <span class="fl">0</span>, hjust <span class="op">=</span> <span class="fl">0</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>plot.margin <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">margin</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">7</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html" class="external-link">plot_grid</a></span><span class="op">(</span><span class="va">title</span>, <span class="va">combined_plot</span>, ncol <span class="op">=</span> <span class="fl">1</span>, rel_heights <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.1</span>, <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="3.Visualization_files/figure-html/unnamed-chunk-50-1.png" width="768"></p>
</div>
<div class="section level3">
<h3 id="saving-your-plots">Saving Your Plots<a class="anchor" aria-label="anchor" href="#saving-your-plots"></a>
</h3>
<p>Finally, let’s save our plot using <code><a href="https://ggplot2.tidyverse.org/reference/ggsave.html" class="external-link">ggsave()</a></code>:</p>
<div class="sourceCode" id="cb69"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="st">"results"</span>, showWarnings <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggsave.html" class="external-link">ggsave</a></span><span class="op">(</span><span class="st">"results/pbmc_umap_plot.png"</span>, <span class="va">combined_plot</span>, width <span class="op">=</span> <span class="fl">8</span>, height <span class="op">=</span> <span class="fl">4.5</span>, dpi <span class="op">=</span> <span class="fl">300</span><span class="op">)</span></span></code></pre></div>
<p>This will save the plot as a high-resolution PNG file.</p>
</div>
<div class="section level3">
<h3 id="tips-for-further-customization">Tips for Further Customization<a class="anchor" aria-label="anchor" href="#tips-for-further-customization"></a>
</h3>
<p>Remember, while ggplot2 offers extensive customization options, it
can sometimes be complex. Don’t hesitate to consult the ggplot2
documentation, online resources, or AI assistants like Claude or ChatGPT
for help with specific customizations.</p>
<p>This concludes our lesson on advanced visualization techniques. You
now have a solid foundation in creating and customizing plots for your
single-cell RNA-seq data analysis. In the next lesson, we’ll explore
more advanced analytical techniques to further your understanding of
single-cell data.</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Yichao Hua.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
