<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Lesson 4: Gene Set Enrichment Analysis and Utility Functions • SeuratExtend</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../../bootstrap-toc.css">
<script src="../../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../../pkgdown.css" rel="stylesheet">
<script src="../../pkgdown.js"></script><meta property="og:title" content="Lesson 4: Gene Set Enrichment Analysis and Utility Functions">
<meta property="og:description" content="SeuratExtend">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../../index.html">SeuratExtend</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../../articles/News.html">What's New in v1.1.0</a>
    </li>
    <li class="divider">
    </li>
<li>
      <a href="../../articles/Visualization.html">Enhanced Visualization</a>
    </li>
    <li>
      <a href="../../articles/GSEA.html">Geneset Enrichment Analysis (GSEA) </a>
    </li>
    <li>
      <a href="../../articles/Trajectory.html">Trajectory and Pseudotime Analysis </a>
    </li>
    <li>
      <a href="../../articles/SCENIC.html">SCENIC for Gene Regulatory Networks Analysis </a>
    </li>
    <li>
      <a href="../../articles/Utilities.html">Utility Tools and Functions</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    scRNA-seq Course
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../../articles/single-cell-course/1.basic-R.html">1. Introduction to R</a>
    </li>
    <li>
      <a href="../../articles/single-cell-course/2.Seurat.html">2. Basic Single-Cell Analysis Workflow</a>
    </li>
    <li>
      <a href="../../articles/single-cell-course/3.Visualization.html">3. Advanced Visualization Techniques</a>
    </li>
    <li>
      <a href="../../articles/single-cell-course/4.GSEA.html">4. Gene Set Enrichment Analysis</a>
    </li>
    <li>
      <a href="../../articles/single-cell-course/5.Core-Enhancement.html">5. Core scRNA-seq Workflow Enhancements</a>
    </li>
    <li>
      <a href="../../articles/single-cell-course/6.Advanced.html">6. Advanced Methods (Part 1)</a>
    </li>
    <li>
      <a href="../../articles/single-cell-course/6.Advanced-2.html">7. Advanced Methods (Part 2)</a>
    </li>
  </ul>
</li>
<li>
  <a href="../../articles/FAQ.html">FAQ</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Lesson 4: Gene Set Enrichment Analysis and
Utility Functions</h1>
                        <h4 data-toc-skip class="author">Yichao Hua</h4>
            
            <h4 data-toc-skip class="date">2024-11-29</h4>
      
      
      <div class="hidden name"><code>4.GSEA.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="introduction-to-gene-set-enrichment-analysis-gsea">Introduction to Gene Set Enrichment Analysis (GSEA)<a class="anchor" aria-label="anchor" href="#introduction-to-gene-set-enrichment-analysis-gsea"></a>
</h2>
<p>In our previous lessons, we primarily focused on gene-level analysis.
However, it’s often necessary to analyze data at the gene set level to
gain broader biological insights. A common scenario involves examining
the enrichment of gene sets from databases like Gene Ontology (GO) or
from literature in different cell clusters within our data. This is
where Gene Set Enrichment Analysis (GSEA) comes into play.</p>
<p>Traditionally, functional enrichment analysis for bulk RNA-seq data
involves identifying differentially expressed genes (DEGs) based on a
predetermined cut-off and then comparing the resulting gene list against
pathway databases to calculate enrichment scores. However, this approach
has several limitations:</p>
<ol style="list-style-type: decimal">
<li>It fails to capture the nuances of single-cell data.</li>
<li>It relies on arbitrary cut-offs, potentially missing important
biological signals.</li>
</ol>
<p>For single-cell analysis, we have specialized tools designed to
calculate GSEA more accurately. In the <code>SeuratExtend</code>
package, we use the AUCell method to compute pathway enrichment scores.
This method calculates an enrichment score for each cell, allowing us to
compare differences between clusters at the single-cell level.</p>
<p>In this lesson, we’ll cover:</p>
<ul>
<li>GSEA calculation methods</li>
<li>Common gene set / pathway databases (GO, Reactome, Hallmark50,
etc.)</li>
<li>How to use these databases in your analysis</li>
</ul>
</div>
<div class="section level2">
<h2 id="conducting-gsea-using-go-or-reactome-databases">Conducting GSEA using GO or Reactome Databases<a class="anchor" aria-label="anchor" href="#conducting-gsea-using-go-or-reactome-databases"></a>
</h2>
<p>Before we dive into the analysis, let’s briefly introduce two key
databases we’ll be using:</p>
<ol style="list-style-type: decimal">
<li>
<p><strong>Gene Ontology (GO)</strong>: GO is a major bioinformatics
initiative to unify the representation of gene and gene product
attributes across all species. It provides a controlled vocabulary of
terms for describing gene product characteristics and gene product
annotation data. The GO project classifies gene functions into three
aspects:</p>
<ul>
<li>Biological Process (BP)</li>
<li>Molecular Function (MF)</li>
<li>Cellular Component (CC)</li>
</ul>
<p>For more information, visit the official GO website: <a href="https://geneontology.org/" class="external-link uri">https://geneontology.org/</a></p>
</li>
<li><p><strong>Reactome</strong>: Reactome is an open-source, curated
and peer-reviewed pathway database. It provides intuitive bioinformatics
tools for the visualization, interpretation and analysis of pathway
knowledge to support basic and clinical research, genome analysis,
modeling, systems biology and education. You can explore Reactome’s
extensive resources at their official website: <a href="https://reactome.org/" class="external-link uri">https://reactome.org/</a></p></li>
</ol>
<p>The <code>SeuratExtend</code> package integrates both the GO and
Reactome databases, streamlining the GSEA analysis process. This is
primarily facilitated through the <code>GeneSetAnalysisGO</code> and
<code>GeneSetAnalysisReactome</code> functions, among other
supplementary functions. In this section, we’ll delve into the usage and
features of these functions.</p>
<div class="section level3">
<h3 id="setting-up-the-environment">Setting Up the Environment<a class="anchor" aria-label="anchor" href="#setting-up-the-environment"></a>
</h3>
<p>First, let’s set your working directory to where we’ll store all
course materials. If you followed Lesson 1, this would be:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Set working directory to the course folder created in Lesson 1</span></span>
<span><span class="co"># If you used a different location, replace this path with your chosen directory</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/getwd.html" class="external-link">setwd</a></span><span class="op">(</span><span class="st">"~/Documents/single-cell-course"</span><span class="op">)</span></span></code></pre></div>
<p>Load the necessary libraries and our previously saved PBMC
object:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SeuratExtend</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">options</a></span><span class="op">(</span>max.print <span class="op">=</span> <span class="fl">15</span>, spe <span class="op">=</span> <span class="st">"human"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Load the PBMC object we saved at the end of Lesson 2</span></span>
<span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"rds/pbmc_annotated.rds"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="gene-ontology-go-database-analysis">Gene Ontology (GO) Database Analysis<a class="anchor" aria-label="anchor" href="#gene-ontology-go-database-analysis"></a>
</h3>
<p>Performing GSEA using the GO database can be resource-intensive due
to its extensive size. To make the analysis more feasible, you might
consider evaluating pathways under specific categories. For instance, in
the example below, we’ll focus on pathways under the
“immune_system_process” category.</p>
<p>Note: When running <code>GeneSetAnalysisGO</code> for the first time,
you may be prompted to install some required packages such as
<code>DelayedArray</code>, <code>DelayedMatrixStats</code>, and
<code>doMC</code> (for multi-threading computations). Simply confirm the
installation of these packages when prompted.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/GeneSetAnalysis.html">GeneSetAnalysisGO</a></span><span class="op">(</span><span class="va">pbmc</span>, parent <span class="op">=</span> <span class="st">"immune_system_process"</span>, nCores <span class="op">=</span> <span class="fl">4</span><span class="op">)</span> <span class="co"># calculating with 4 cores</span></span>
<span><span class="va">matr</span> <span class="op">&lt;-</span> <span class="va">pbmc</span><span class="op">@</span><span class="va">misc</span><span class="op">$</span><span class="va">AUCell</span><span class="op">$</span><span class="va">GO</span><span class="op">$</span><span class="va">immune_system_process</span></span>
<span><span class="va">matr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/RenameGO.html">RenameGO</a></span><span class="op">(</span><span class="va">matr</span>, add_n_gene <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="co"># Convert GO IDs to pathway names</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">matr</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">5</span>,<span class="fl">3</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                                                         cells</span></span>
<span><span class="co">## gene sets                                                AAACATACAACCAC-1</span></span>
<span><span class="co">##   GO:0002376 immune system process                             0.23253379</span></span>
<span><span class="co">##   GO:0001776 leukocyte homeostasis                             0.05306271</span></span>
<span><span class="co">##   GO:0002200 somatic diversification of immune receptors       0.03449422</span></span>
<span><span class="co">##   GO:0002252 immune effector process                           0.10151694</span></span>
<span><span class="co">##   GO:0002253 activation of immune response                     0.08329149</span></span>
<span><span class="co">##                                                         cells</span></span>
<span><span class="co">## gene sets                                                AAACATTGAGCTAC-1</span></span>
<span><span class="co">##   GO:0002376 immune system process                             0.27669108</span></span>
<span><span class="co">##   GO:0001776 leukocyte homeostasis                             0.07098796</span></span>
<span><span class="co">##   GO:0002200 somatic diversification of immune receptors       0.03453330</span></span>
<span><span class="co">##   GO:0002252 immune effector process                           0.14171700</span></span>
<span><span class="co">##   GO:0002253 activation of immune response                     0.12190542</span></span>
<span><span class="co">##                                                         cells</span></span>
<span><span class="co">## gene sets                                                AAACATTGATCAGC-1</span></span>
<span><span class="co">##   GO:0002376 immune system process                             0.28803920</span></span>
<span><span class="co">##   GO:0001776 leukocyte homeostasis                             0.08491692</span></span>
<span><span class="co">##   GO:0002200 somatic diversification of immune receptors       0.04000547</span></span>
<span><span class="co">##   GO:0002252 immune effector process                           0.13312274</span></span>
<span><span class="co">##   GO:0002253 activation of immune response                     0.09172829</span></span></code></pre>
<p>Note that the generated matrix is similar in structure to a gene
count matrix, with cell names as column names. The key difference is
that the row names now represent pathway names instead of gene
names.</p>
<p>For the “parent” argument in <code><a href="../../reference/GeneSetAnalysis.html">GeneSetAnalysisGO()</a></code>, you
can input any term from the GO database, be it a GO ID or a pathway
name. To get a glimpse of commonly used GO categories, you can run
<code><a href="../../reference/GeneSetAnalysis.html">GeneSetAnalysisGO()</a></code> without any arguments:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../../reference/GeneSetAnalysis.html">GeneSetAnalysisGO</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                 immune_system_process                  response_to_stimulus </span></span>
<span><span class="co">##                          "GO:0002376"                          "GO:0050896" </span></span>
<span><span class="co">##                             signaling                     metabolic_process </span></span>
<span><span class="co">##                          "GO:0023052"                          "GO:0008152" </span></span>
<span><span class="co">## regulation_of_vasculature_development                   signal_transduction </span></span>
<span><span class="co">##                          "GO:1901342"                          "GO:0007165"</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="visualizing-go-gsea-results">Visualizing GO GSEA Results<a class="anchor" aria-label="anchor" href="#visualizing-go-gsea-results"></a>
</h3>
<p>There are several ways to visualize the results of our GSEA analysis.
Here are some suggested methods:</p>
<ol style="list-style-type: decimal">
<li>
<strong>Heatmap</strong>: Useful for comparing multiple groups with
numerous features but provides a less detailed representation.</li>
</ol>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">toplot</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/CalcStats.html">CalcStats</a></span><span class="op">(</span><span class="va">matr</span>, f <span class="op">=</span> <span class="va">pbmc</span><span class="op">$</span><span class="va">cluster</span>, order <span class="op">=</span> <span class="st">"p"</span>, n <span class="op">=</span> <span class="fl">4</span><span class="op">)</span></span>
<span><span class="fu"><a href="../../reference/Heatmap.html">Heatmap</a></span><span class="op">(</span><span class="va">toplot</span>, lab_fill <span class="op">=</span> <span class="st">"zscore"</span><span class="op">)</span></span></code></pre></div>
<p><img src="4.GSEA_files/figure-html/unnamed-chunk-5-1.png" width="1104"></p>
<p>Note that we’re using the <code>CalcStats</code> function here. In
Lesson 3, we learned to use this function with a Seurat object as the
first input parameter. However, <code>CalcStats</code> and some
visualization functions like <code>VlnPlot2</code> and
<code>WaterfallPlot</code> also support matrices as input, increasing
flexibility. The main difference when using a matrix is that you need to
provide an <code>f</code> parameter, which is a factor or vector
indicating cell identity. Typically, this can be directly referenced
from a column in the <code>meta.data</code>.</p>
<ol start="2" style="list-style-type: decimal">
<li>
<strong>Violin Plot</strong>: Best suited for comparing multiple
groups with fewer features while presenting more details for individual
data points.</li>
</ol>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../../reference/VlnPlot2.html">VlnPlot2</a></span><span class="op">(</span><span class="va">matr</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span>,<span class="op">]</span>, f <span class="op">=</span> <span class="va">pbmc</span><span class="op">$</span><span class="va">cluster</span>, ncol <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<p><img src="4.GSEA_files/figure-html/unnamed-chunk-6-1.png" width="480"></p>
<ol start="3" style="list-style-type: decimal">
<li>
<strong>Waterfall Plot</strong>: Ideal for comparing two
groups.</li>
</ol>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../../reference/WaterfallPlot.html">WaterfallPlot</a></span><span class="op">(</span><span class="va">matr</span>, f <span class="op">=</span> <span class="va">pbmc</span><span class="op">$</span><span class="va">cluster</span>, ident.1 <span class="op">=</span> <span class="st">"B cell"</span>, ident.2 <span class="op">=</span> <span class="st">"CD8 T cell"</span>, top.n <span class="op">=</span> <span class="fl">20</span><span class="op">)</span></span></code></pre></div>
<p><img src="4.GSEA_files/figure-html/unnamed-chunk-7-1.png" width="1008"></p>
</div>
<div class="section level3">
<h3 id="reactome-database-analysis">Reactome Database Analysis<a class="anchor" aria-label="anchor" href="#reactome-database-analysis"></a>
</h3>
<p>For GSEA using the Reactome database, we’ll follow a similar process.
Let’s assess pathways under the “Immune System” category:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/GeneSetAnalysis.html">GeneSetAnalysisReactome</a></span><span class="op">(</span><span class="va">pbmc</span>, parent <span class="op">=</span> <span class="st">"Immune System"</span><span class="op">)</span></span>
<span><span class="va">matr</span> <span class="op">&lt;-</span> <span class="va">pbmc</span><span class="op">@</span><span class="va">misc</span><span class="op">$</span><span class="va">AUCell</span><span class="op">$</span><span class="va">Reactome</span><span class="op">$</span><span class="va">`Immune System`</span></span>
<span><span class="va">matr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/RenameGO.html">RenameReactome</a></span><span class="op">(</span><span class="va">matr</span><span class="op">)</span></span>
<span><span class="fu"><a href="../../reference/Heatmap.html">Heatmap</a></span><span class="op">(</span><span class="fu"><a href="../../reference/CalcStats.html">CalcStats</a></span><span class="op">(</span><span class="va">matr</span>, f <span class="op">=</span> <span class="va">pbmc</span><span class="op">$</span><span class="va">cluster</span>, order <span class="op">=</span> <span class="st">"p"</span>, n <span class="op">=</span> <span class="fl">4</span><span class="op">)</span>, lab_fill <span class="op">=</span> <span class="st">"zscore"</span><span class="op">)</span></span></code></pre></div>
<p><img src="4.GSEA_files/figure-html/unnamed-chunk-8-1.png" width="1008"></p>
<p>Similar to the GO database, running
<code><a href="../../reference/GeneSetAnalysis.html">GeneSetAnalysisReactome()</a></code> without any arguments lets you
view commonly used categories in the Reactome database:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../../reference/GeneSetAnalysis.html">GeneSetAnalysisReactome</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                           R-HSA-109582                           R-HSA-112316 </span></span>
<span><span class="co">##                           "Hemostasis"                      "Neuronal System" </span></span>
<span><span class="co">##                          R-HSA-1266738                          R-HSA-1430728 </span></span>
<span><span class="co">##                "Developmental Biology"                           "Metabolism" </span></span>
<span><span class="co">##                          R-HSA-1474165                          R-HSA-1474244 </span></span>
<span><span class="co">##                         "Reproduction"    "Extracellular matrix organization" </span></span>
<span><span class="co">##                          R-HSA-1500931                           R-HSA-162582 </span></span>
<span><span class="co">##              "Cell-Cell communication"                  "Signal Transduction" </span></span>
<span><span class="co">##                          R-HSA-1640170                          R-HSA-1643685 </span></span>
<span><span class="co">##                           "Cell Cycle"                              "Disease" </span></span>
<span><span class="co">##                           R-HSA-168256                          R-HSA-1852241 </span></span>
<span><span class="co">##                        "Immune System" "Organelle biogenesis and maintenance" </span></span>
<span><span class="co">##                           R-HSA-382551                           R-HSA-392499 </span></span>
<span><span class="co">##         "Transport of small molecules"               "Metabolism of proteins" </span></span>
<span><span class="co">##                           R-HSA-397014 </span></span>
<span><span class="co">##                   "Muscle contraction" </span></span>
<span><span class="co">##  [ reached getOption("max.print") -- omitted 12 entries ]</span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="perform-gsea-using-customized-gene-sets">Perform GSEA using Customized Gene Sets<a class="anchor" aria-label="anchor" href="#perform-gsea-using-customized-gene-sets"></a>
</h2>
<p>While GO and Reactome databases provide comprehensive pathway
information, sometimes researchers need to work with specific,
customized gene sets. In this section, we’ll explore how to perform GSEA
using custom gene sets and access pre-defined gene set collections.</p>
<div class="section level3">
<h3 id="using-the-hallmark-50-gene-set">Using the Hallmark 50 Gene Set<a class="anchor" aria-label="anchor" href="#using-the-hallmark-50-gene-set"></a>
</h3>
<p>The <a href="https://www.gsea-msigdb.org/gsea/msigdb/human/genesets.jsp?collection=H" class="external-link">Hallmark
50</a> gene set is commonly used for general screening. When using the
<code>GeneSetAnalysis</code> function, the <code>genesets</code>
parameter requires a list in a specific format. Each list item should be
a gene set containing all the gene names in that set. For example, the
structure of <code>hall50$human</code> looks like this:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">hall50</span><span class="op">$</span><span class="va">human</span>, list.len <span class="op">=</span> <span class="fl">10</span>, max.level <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## List of 50</span></span>
<span><span class="co">##  $ HALLMARK_TNFA_SIGNALING_VIA_NFKB          : chr [1:200] "JUNB" "CXCL2" "ATF3" "NFKBIA" ...</span></span>
<span><span class="co">##  $ HALLMARK_HYPOXIA                          : chr [1:200] "PGK1" "PDK1" "GBE1" "PFKL" ...</span></span>
<span><span class="co">##  $ HALLMARK_CHOLESTEROL_HOMEOSTASIS          : chr [1:74] "FDPS" "CYP51A1" "IDI1" "FDFT1" ...</span></span>
<span><span class="co">##  $ HALLMARK_MITOTIC_SPINDLE                  : chr [1:199] "ARHGEF2" "CLASP1" "KIF11" "AC027237.1" ...</span></span>
<span><span class="co">##  $ HALLMARK_WNT_BETA_CATENIN_SIGNALING       : chr [1:42] "MYC" "CTNNB1" "JAG2" "NOTCH1" ...</span></span>
<span><span class="co">##  $ HALLMARK_TGF_BETA_SIGNALING               : chr [1:54] "TGFBR1" "SMAD7" "TGFB1" "SMURF2" ...</span></span>
<span><span class="co">##  $ HALLMARK_IL6_JAK_STAT3_SIGNALING          : chr [1:87] "IL4R" "IL6ST" "STAT1" "IL1R1" ...</span></span>
<span><span class="co">##  $ HALLMARK_DNA_REPAIR                       : chr [1:150] "POLR2H" "POLR2A" "POLR2G" "POLR2E" ...</span></span>
<span><span class="co">##  $ HALLMARK_G2M_CHECKPOINT                   : chr [1:200] "AURKA" "CCNA2" "TOP2A" "CCNB2" ...</span></span>
<span><span class="co">##  $ HALLMARK_APOPTOSIS                        : chr [1:161] "CASP3" "CASP9" "DFFA" "CASP7" ...</span></span>
<span><span class="co">##   [list output truncated]</span></span></code></pre>
<p>Each list item is a character vector containing the gene names
belonging to that gene set. Now, let’s use this gene set for our
analysis:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/GeneSetAnalysis.html">GeneSetAnalysis</a></span><span class="op">(</span><span class="va">pbmc</span>, genesets <span class="op">=</span> <span class="va">hall50</span><span class="op">$</span><span class="va">human</span><span class="op">)</span></span>
<span><span class="va">matr</span> <span class="op">&lt;-</span> <span class="va">pbmc</span><span class="op">@</span><span class="va">misc</span><span class="op">$</span><span class="va">AUCell</span><span class="op">$</span><span class="va">genesets</span></span>
<span><span class="fu"><a href="../../reference/Heatmap.html">Heatmap</a></span><span class="op">(</span><span class="fu"><a href="../../reference/CalcStats.html">CalcStats</a></span><span class="op">(</span><span class="va">matr</span>, f <span class="op">=</span> <span class="va">pbmc</span><span class="op">$</span><span class="va">cluster</span><span class="op">)</span>, lab_fill <span class="op">=</span> <span class="st">"zscore"</span><span class="op">)</span></span></code></pre></div>
<p><img src="4.GSEA_files/figure-html/unnamed-chunk-11-1.png" width="720"></p>
<p>The resulting AUCell matrix is stored in
<code>seu@misc$AUCell[[title]]</code>.</p>
</div>
<div class="section level3">
<h3 id="exploring-additional-gene-sets">Exploring Additional Gene Sets<a class="anchor" aria-label="anchor" href="#exploring-additional-gene-sets"></a>
</h3>
<p>The <code>SeuratExtendData::Genesets_data</code> object offers an
extensive collection of gene sets from the <a href="https://www.gsea-msigdb.org/gsea/msigdb/human/collections.jsp" class="external-link">GSEA
MSigDB website</a>. You can view available collections:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">Genesets_data</span><span class="op">$</span><span class="va">human</span><span class="op">$</span><span class="va">GSEA</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] "positional gene sets"                </span></span>
<span><span class="co">##  [2] "all curated gene sets"               </span></span>
<span><span class="co">##  [3] "chemical and genetic perturbations"  </span></span>
<span><span class="co">##  [4] "BioCarta gene sets"                  </span></span>
<span><span class="co">##  [5] "KEGG gene sets"                      </span></span>
<span><span class="co">##  [6] "PID gene sets"                       </span></span>
<span><span class="co">##  [7] "all canonical pathways"              </span></span>
<span><span class="co">##  [8] "all motif gene sets"                 </span></span>
<span><span class="co">##  [9] "transcription factor targets"        </span></span>
<span><span class="co">## [10] "all computational gene sets"         </span></span>
<span><span class="co">## [11] "all immunologic signatures gene sets"</span></span>
<span><span class="co">## [12] "hallmark gene sets"</span></span></code></pre>
<p>For cell type annotations, <code>PanglaoDB_data</code> provides
marker lists for 178 distinct cell types from <a href="https://panglaodb.se/markers.html" class="external-link">PanglaoDB</a>:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">PanglaoDB_data</span><span class="op">$</span><span class="va">marker_list_human</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] "Acinar cells"                   "Adipocyte progenitor cells"    </span></span>
<span><span class="co">##  [3] "Adipocytes"                     "Adrenergic neurons"            </span></span>
<span><span class="co">##  [5] "Airway epithelial cells"        "Airway goblet cells"           </span></span>
<span><span class="co">##  [7] "Airway smooth muscle cells"     "Alpha cells"                   </span></span>
<span><span class="co">##  [9] "Alveolar macrophages"           "Anterior pituitary gland cells"</span></span>
<span><span class="co">## [11] "Astrocytes"                     "B cells"                       </span></span>
<span><span class="co">## [13] "B cells memory"                 "B cells naive"                 </span></span>
<span><span class="co">## [15] "Basal cells"                   </span></span>
<span><span class="co">##  [ reached getOption("max.print") -- omitted 163 entries ]</span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="converting-pathway-ids-to-names">Converting Pathway IDs to Names<a class="anchor" aria-label="anchor" href="#converting-pathway-ids-to-names"></a>
</h2>
<p>Pathway IDs can be cryptic. The <code>RenameGO</code> and
<code>RenameReactome</code> functions help convert these IDs to more
descriptive names:</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Convert GO IDs</span></span>
<span><span class="fu"><a href="../../reference/RenameGO.html">RenameGO</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"GO:0002376"</span>,<span class="st">"GO:0050896"</span><span class="op">)</span>, spe <span class="op">=</span> <span class="st">"human"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "GO:0002376 immune system process (3213g)"</span></span>
<span><span class="co">## [2] "GO:0050896 response to stimulus (9342g)"</span></span></code></pre>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Convert Reactome IDs</span></span>
<span><span class="fu"><a href="../../reference/RenameGO.html">RenameReactome</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"R-HSA-109582"</span>,<span class="st">"R-HSA-112316"</span><span class="op">)</span>, spe <span class="op">=</span> <span class="st">"human"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "R-HSA-109582 Hemostasis (679g)"      "R-HSA-112316 Neuronal System (411g)"</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="filtering-pathway-lists">Filtering Pathway Lists<a class="anchor" aria-label="anchor" href="#filtering-pathway-lists"></a>
</h2>
<p>To focus on relevant pathways, use <code>FilterGOTerms</code> and
<code>FilterReactomeTerms</code>:</p>
<div class="section level3">
<h3 id="filtering-go-pathways">Filtering GO Pathways<a class="anchor" aria-label="anchor" href="#filtering-go-pathways"></a>
</h3>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># By parent term</span></span>
<span><span class="va">terms</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/FilterGOTerms.html">FilterGOTerms</a></span><span class="op">(</span>parent <span class="op">=</span> <span class="st">"GO:0002376"</span><span class="op">)</span> <span class="co"># immune system process (3213g)</span></span>
<span><span class="fu"><a href="../../reference/RenameGO.html">RenameGO</a></span><span class="op">(</span><span class="va">terms</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] "GO:0001773 myeloid dendritic cell activation (28g)"      </span></span>
<span><span class="co">##  [2] "GO:0001774 microglial cell activation (45g)"             </span></span>
<span><span class="co">##  [3] "GO:0001776 leukocyte homeostasis (84g)"                  </span></span>
<span><span class="co">##  [4] "GO:0001777 T cell homeostatic proliferation (4g)"        </span></span>
<span><span class="co">##  [5] "GO:0001779 natural killer cell differentiation (21g)"    </span></span>
<span><span class="co">##  [6] "GO:0001780 neutrophil homeostasis (16g)"                 </span></span>
<span><span class="co">##  [7] "GO:0001782 B cell homeostasis (27g)"                     </span></span>
<span><span class="co">##  [8] "GO:0001787 natural killer cell proliferation (11g)"      </span></span>
<span><span class="co">##  [9] "GO:0001788 antibody-dependent cellular cytotoxicity (2g)"</span></span>
<span><span class="co">## [10] "GO:0001794 type IIa hypersensitivity (4g)"               </span></span>
<span><span class="co">## [11] "GO:0001802 type III hypersensitivity (2g)"               </span></span>
<span><span class="co">## [12] "GO:0001806 type IV hypersensitivity (4g)"                </span></span>
<span><span class="co">## [13] "GO:0001865 NK T cell differentiation (10g)"              </span></span>
<span><span class="co">## [14] "GO:0001866 NK T cell proliferation (7g)"                 </span></span>
<span><span class="co">## [15] "GO:0001867 complement activation, lectin pathway (11g)"  </span></span>
<span><span class="co">##  [ reached getOption("max.print") -- omitted 392 entries ]</span></span></code></pre>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># By gene count</span></span>
<span><span class="va">terms2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/FilterGOTerms.html">FilterGOTerms</a></span><span class="op">(</span>term <span class="op">=</span> <span class="va">terms</span>, n.min <span class="op">=</span> <span class="fl">10</span>, n.max <span class="op">=</span> <span class="fl">100</span><span class="op">)</span></span>
<span><span class="fu"><a href="../../reference/RenameGO.html">RenameGO</a></span><span class="op">(</span><span class="va">terms2</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] "GO:0001773 myeloid dendritic cell activation (28g)"                                        </span></span>
<span><span class="co">##  [2] "GO:0001774 microglial cell activation (45g)"                                               </span></span>
<span><span class="co">##  [3] "GO:0001776 leukocyte homeostasis (84g)"                                                    </span></span>
<span><span class="co">##  [4] "GO:0001779 natural killer cell differentiation (21g)"                                      </span></span>
<span><span class="co">##  [5] "GO:0001780 neutrophil homeostasis (16g)"                                                   </span></span>
<span><span class="co">##  [6] "GO:0001782 B cell homeostasis (27g)"                                                       </span></span>
<span><span class="co">##  [7] "GO:0001787 natural killer cell proliferation (11g)"                                        </span></span>
<span><span class="co">##  [8] "GO:0001865 NK T cell differentiation (10g)"                                                </span></span>
<span><span class="co">##  [9] "GO:0001867 complement activation, lectin pathway (11g)"                                    </span></span>
<span><span class="co">## [10] "GO:0001913 T cell mediated cytotoxicity (46g)"                                             </span></span>
<span><span class="co">## [11] "GO:0002200 somatic diversification of immune receptors (77g)"                              </span></span>
<span><span class="co">## [12] "GO:0002204 somatic recombination of immunoglobulin genes involved in immune response (50g)"</span></span>
<span><span class="co">## [13] "GO:0002208 somatic diversification of immunoglobulins involved in immune response (50g)"   </span></span>
<span><span class="co">## [14] "GO:0002227 innate immune response in mucosa (19g)"                                         </span></span>
<span><span class="co">## [15] "GO:0002228 natural killer cell mediated immunity (64g)"                                    </span></span>
<span><span class="co">##  [ reached getOption("max.print") -- omitted 128 entries ]</span></span></code></pre>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># End-level pathways</span></span>
<span><span class="va">terms3</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/FilterGOTerms.html">FilterGOTerms</a></span><span class="op">(</span>term <span class="op">=</span> <span class="va">terms</span>, only.end.terms <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="../../reference/RenameGO.html">RenameGO</a></span><span class="op">(</span><span class="va">terms3</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] "GO:0001777 T cell homeostatic proliferation (4g)"                       </span></span>
<span><span class="co">##  [2] "GO:0001780 neutrophil homeostasis (16g)"                                </span></span>
<span><span class="co">##  [3] "GO:0001787 natural killer cell proliferation (11g)"                     </span></span>
<span><span class="co">##  [4] "GO:0001788 antibody-dependent cellular cytotoxicity (2g)"               </span></span>
<span><span class="co">##  [5] "GO:0001802 type III hypersensitivity (2g)"                              </span></span>
<span><span class="co">##  [6] "GO:0001806 type IV hypersensitivity (4g)"                               </span></span>
<span><span class="co">##  [7] "GO:0001865 NK T cell differentiation (10g)"                             </span></span>
<span><span class="co">##  [8] "GO:0001866 NK T cell proliferation (7g)"                                </span></span>
<span><span class="co">##  [9] "GO:0001867 complement activation, lectin pathway (11g)"                 </span></span>
<span><span class="co">## [10] "GO:0001905 activation of membrane attack complex (3g)"                  </span></span>
<span><span class="co">## [11] "GO:0001922 B-1 B cell homeostasis (3g)"                                 </span></span>
<span><span class="co">## [12] "GO:0002223 stimulatory C-type lectin receptor signaling pathway (112g)" </span></span>
<span><span class="co">## [13] "GO:0002227 innate immune response in mucosa (19g)"                      </span></span>
<span><span class="co">## [14] "GO:0002232 leukocyte chemotaxis involved in inflammatory response (5g)" </span></span>
<span><span class="co">## [15] "GO:0002264 endothelial cell activation involved in immune response (1g)"</span></span>
<span><span class="co">##  [ reached getOption("max.print") -- omitted 185 entries ]</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="filtering-reactome-pathways">Filtering Reactome Pathways<a class="anchor" aria-label="anchor" href="#filtering-reactome-pathways"></a>
</h3>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">terms</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/FilterGOTerms.html">FilterReactomeTerms</a></span><span class="op">(</span>parent <span class="op">=</span> <span class="st">"R-HSA-168256"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../../reference/RenameGO.html">RenameReactome</a></span><span class="op">(</span><span class="va">terms</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] "R-HSA-1059683 Interleukin-6 signaling (11g)"                                         </span></span>
<span><span class="co">##  [2] "R-HSA-109703 PKB-mediated events (2g)"                                               </span></span>
<span><span class="co">##  [3] "R-HSA-109704 PI3K Cascade (44g)"                                                     </span></span>
<span><span class="co">##  [4] "R-HSA-1168372 Downstream signaling events of B Cell Receptor (BCR) (81g)"            </span></span>
<span><span class="co">##  [5] "R-HSA-1169091 Activation of NF-kappaB in B cells (67g)"                              </span></span>
<span><span class="co">##  [6] "R-HSA-1169092 Activation of RAS in B cells (5g)"                                     </span></span>
<span><span class="co">##  [7] "R-HSA-1169408 ISG15 antiviral mechanism (74g)"                                       </span></span>
<span><span class="co">##  [8] "R-HSA-1169410 Antiviral mechanism by IFN-stimulated genes (82g)"                     </span></span>
<span><span class="co">##  [9] "R-HSA-1170546 Prolactin receptor signaling (15g)"                                    </span></span>
<span><span class="co">## [10] "R-HSA-1222556 ROS and RNS production in phagocytes (36g)"                            </span></span>
<span><span class="co">## [11] "R-HSA-1236973 Cross-presentation of particulate exogenous antigens (phagosomes) (8g)"</span></span>
<span><span class="co">## [12] "R-HSA-1236974 ER-Phagosome pathway (83g)"                                            </span></span>
<span><span class="co">## [13] "R-HSA-1236975 Antigen processing-Cross presentation (99g)"                           </span></span>
<span><span class="co">## [14] "R-HSA-1236977 Endosomal/Vacuolar pathway (11g)"                                      </span></span>
<span><span class="co">## [15] "R-HSA-1236978 Cross-presentation of soluble exogenous antigens (endosomes) (50g)"    </span></span>
<span><span class="co">##  [ reached getOption("max.print") -- omitted 182 entries ]</span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="finding-pathways-in-databases">Finding Pathways in Databases<a class="anchor" aria-label="anchor" href="#finding-pathways-in-databases"></a>
</h2>
<p>Navigating large databases like GO and Reactome can be challenging.
The <code>SearchDatabase</code> function simplifies this process with
various search parameters.</p>
<div class="section level3">
<h3 id="general-search">General Search<a class="anchor" aria-label="anchor" href="#general-search"></a>
</h3>
<p>You can search by gene name, pathway ID, or keywords:</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Search for pathways that either contain the gene CD3D in their gene list,</span></span>
<span><span class="co"># or have 'metabolic' in their pathway name</span></span>
<span><span class="va">result</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/SearchDatabase.html">SearchDatabase</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD3D"</span>, <span class="st">"metabolic"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">result</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] "GO:0000023 maltose metabolic process (3g)"                                  </span></span>
<span><span class="co">##  [2] "GO:0000038 very long-chain fatty acid metabolic process (31g)"              </span></span>
<span><span class="co">##  [3] "GO:0000052 citrulline metabolic process (8g)"                               </span></span>
<span><span class="co">##  [4] "GO:0000053 argininosuccinate metabolic process (1g)"                        </span></span>
<span><span class="co">##  [5] "GO:0000096 sulfur amino acid metabolic process (43g)"                       </span></span>
<span><span class="co">##  [6] "GO:0000255 allantoin metabolic process (2g)"                                </span></span>
<span><span class="co">##  [7] "GO:0000270 peptidoglycan metabolic process (6g)"                            </span></span>
<span><span class="co">##  [8] "GO:0000820 regulation of glutamine family amino acid metabolic process (6g)"</span></span>
<span><span class="co">##  [9] "GO:0000821 regulation of arginine metabolic process (2g)"                   </span></span>
<span><span class="co">## [10] "GO:0000959 mitochondrial RNA metabolic process (48g)"                       </span></span>
<span><span class="co">## [11] "GO:0001523 retinoid metabolic process (103g)"                               </span></span>
<span><span class="co">## [12] "GO:0001573 ganglioside metabolic process (21g)"                             </span></span>
<span><span class="co">## [13] "GO:0001575 globoside metabolic process (1g)"                                </span></span>
<span><span class="co">## [14] "GO:0001676 long-chain fatty acid metabolic process (110g)"                  </span></span>
<span><span class="co">## [15] "GO:0001692 histamine metabolic process (4g)"                                </span></span>
<span><span class="co">##  [ reached getOption("max.print") -- omitted 1026 entries ]</span></span></code></pre>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://pillar.r-lib.org/reference/glimpse.html" class="external-link">glimpse</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">result</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## List of 6</span></span>
<span><span class="co">##  $ GO:0000023 maltose metabolic process (3g)                    :List of 3</span></span>
<span><span class="co">##   ..$ SetID  : chr "GO:0000023"</span></span>
<span><span class="co">##   ..$ SetName: chr "maltose metabolic process"</span></span>
<span><span class="co">##   ..$ Genes  : chr [1:3] "MGAM" "GAA" "GANC"</span></span>
<span><span class="co">##  $ GO:0000038 very long-chain fatty acid metabolic process (31g):List of 3</span></span>
<span><span class="co">##   ..$ SetID  : chr "GO:0000038"</span></span>
<span><span class="co">##   ..$ SetName: chr "very long-chain fatty acid metabolic process"</span></span>
<span><span class="co">##   ..$ Genes  : chr [1:31] "ELOVL7" "HACD1" "SLC27A2" "ACAA1" ...</span></span>
<span><span class="co">##  $ GO:0000052 citrulline metabolic process (8g)                 :List of 3</span></span>
<span><span class="co">##   ..$ SetID  : chr "GO:0000052"</span></span>
<span><span class="co">##   ..$ SetName: chr "citrulline metabolic process"</span></span>
<span><span class="co">##   ..$ Genes  : chr [1:8] "DDAH1" "DDAH2" "OTC" "ASS1" ...</span></span>
<span><span class="co">##  $ GO:0000053 argininosuccinate metabolic process (1g)          :List of 3</span></span>
<span><span class="co">##   ..$ SetID  : chr "GO:0000053"</span></span>
<span><span class="co">##   ..$ SetName: chr "argininosuccinate metabolic process"</span></span>
<span><span class="co">##   ..$ Genes  : chr "ASS1"</span></span>
<span><span class="co">##  $ GO:0000096 sulfur amino acid metabolic process (43g)         :List of 3</span></span>
<span><span class="co">##   ..$ SetID  : chr "GO:0000096"</span></span>
<span><span class="co">##   ..$ SetName: chr "sulfur amino acid metabolic process"</span></span>
<span><span class="co">##   ..$ Genes  : chr [1:43] "F2Z2F3" "AHCYL1" "CBSL" "MTHFD1" ...</span></span>
<span><span class="co">##  $ GO:0000255 allantoin metabolic process (2g)                  :List of 3</span></span>
<span><span class="co">##   ..$ SetID  : chr "GO:0000255"</span></span>
<span><span class="co">##   ..$ SetName: chr "allantoin metabolic process"</span></span>
<span><span class="co">##   ..$ Genes  : chr [1:2] "URAD" "ALLC"</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="species-specific-search">Species-Specific Search<a class="anchor" aria-label="anchor" href="#species-specific-search"></a>
</h3>
<p>Specify ‘human’ or ‘mouse’ using the ‘spe’ parameter:</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">result</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/SearchDatabase.html">SearchDatabase</a></span><span class="op">(</span><span class="st">"antigen pr"</span>, spe <span class="op">=</span> <span class="st">"mouse"</span><span class="op">)</span> <span class="co"># Pathways related to antigen presentation</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">result</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] "GO:0002291 T cell activation via T cell receptor contact with antigen bound to MHC molecule on antigen presenting cell (7g)"</span></span>
<span><span class="co">##  [2] "GO:0002428 antigen processing and presentation of peptide antigen via MHC class Ib (40g)"                                   </span></span>
<span><span class="co">##  [3] "GO:0002457 T cell antigen processing and presentation (7g)"                                                                 </span></span>
<span><span class="co">##  [4] "GO:0002468 dendritic cell antigen processing and presentation (11g)"                                                        </span></span>
<span><span class="co">##  [5] "GO:0002470 plasmacytoid dendritic cell antigen processing and presentation (1g)"                                            </span></span>
<span><span class="co">##  [6] "GO:0002472 macrophage antigen processing and presentation (1g)"                                                             </span></span>
<span><span class="co">##  [7] "GO:0002474 antigen processing and presentation of peptide antigen via MHC class I (54g)"                                    </span></span>
<span><span class="co">##  [8] "GO:0002475 antigen processing and presentation via MHC class Ib (44g)"                                                      </span></span>
<span><span class="co">##  [9] "GO:0002476 antigen processing and presentation of endogenous peptide antigen via MHC class Ib (39g)"                        </span></span>
<span><span class="co">## [10] "GO:0002477 antigen processing and presentation of exogenous peptide antigen via MHC class Ib (4g)"                          </span></span>
<span><span class="co">## [11] "GO:0002478 antigen processing and presentation of exogenous peptide antigen (25g)"                                          </span></span>
<span><span class="co">## [12] "GO:0002479 antigen processing and presentation of exogenous peptide antigen via MHC class I, TAP-dependent (2g)"            </span></span>
<span><span class="co">## [13] "GO:0002481 antigen processing and presentation of exogenous protein antigen via MHC class Ib, TAP-dependent (3g)"           </span></span>
<span><span class="co">## [14] "GO:0002483 antigen processing and presentation of endogenous peptide antigen (44g)"                                         </span></span>
<span><span class="co">## [15] "GO:0002484 antigen processing and presentation of endogenous peptide antigen via MHC class I via ER pathway (39g)"          </span></span>
<span><span class="co">##  [ reached getOption("max.print") -- omitted 48 entries ]</span></span></code></pre>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://pillar.r-lib.org/reference/glimpse.html" class="external-link">glimpse</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">result</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## List of 6</span></span>
<span><span class="co">##  $ GO:0002291 T cell activation via T cell receptor contact with antigen bound to MHC molecule on antigen presenting cell (7g):List of 3</span></span>
<span><span class="co">##   ..$ SetID  : chr "GO:0002291"</span></span>
<span><span class="co">##   ..$ SetName: chr "T cell activation via T cell receptor contact with antigen bound to MHC molecule on antigen presenting cell"</span></span>
<span><span class="co">##   ..$ Genes  : chr [1:7] "Fgl2" "Cd81" "Apbb1ip" "Havcr2" ...</span></span>
<span><span class="co">##  $ GO:0002428 antigen processing and presentation of peptide antigen via MHC class Ib (40g)                                   :List of 3</span></span>
<span><span class="co">##   ..$ SetID  : chr "GO:0002428"</span></span>
<span><span class="co">##   ..$ SetName: chr "antigen processing and presentation of peptide antigen via MHC class Ib"</span></span>
<span><span class="co">##   ..$ Genes  : chr [1:40] "Fcgrt" "Azgp1" "Hfe" "H2-M10.1" ...</span></span>
<span><span class="co">##  $ GO:0002457 T cell antigen processing and presentation (7g)                                                                 :List of 3</span></span>
<span><span class="co">##   ..$ SetID  : chr "GO:0002457"</span></span>
<span><span class="co">##   ..$ SetName: chr "T cell antigen processing and presentation"</span></span>
<span><span class="co">##   ..$ Genes  : chr [1:7] "Was" "Hfe" "Trex1" "Treml4" ...</span></span>
<span><span class="co">##  $ GO:0002468 dendritic cell antigen processing and presentation (11g)                                                        :List of 3</span></span>
<span><span class="co">##   ..$ SetID  : chr "GO:0002468"</span></span>
<span><span class="co">##   ..$ SetName: chr "dendritic cell antigen processing and presentation"</span></span>
<span><span class="co">##   ..$ Genes  : chr [1:11] "Ccr7" "Fgl2" "Nod1" "Slc11a1" ...</span></span>
<span><span class="co">##  $ GO:0002470 plasmacytoid dendritic cell antigen processing and presentation (1g)                                            :List of 3</span></span>
<span><span class="co">##   ..$ SetID  : chr "GO:0002470"</span></span>
<span><span class="co">##   ..$ SetName: chr "plasmacytoid dendritic cell antigen processing and presentation"</span></span>
<span><span class="co">##   ..$ Genes  : chr "Clec4a2"</span></span>
<span><span class="co">##  $ GO:0002472 macrophage antigen processing and presentation (1g)                                                             :List of 3</span></span>
<span><span class="co">##   ..$ SetID  : chr "GO:0002472"</span></span>
<span><span class="co">##   ..$ SetName: chr "macrophage antigen processing and presentation"</span></span>
<span><span class="co">##   ..$ Genes  : chr "Fgl2"</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="customizing-return-types">Customizing Return Types<a class="anchor" aria-label="anchor" href="#customizing-return-types"></a>
</h3>
<p>The <code>SearchDatabase</code> function offers various return types
to suit different analytical needs:</p>
<ol style="list-style-type: decimal">
<li>When <code>return = "ID"</code>, the function returns matching
pathway IDs. These IDs can be directly used with
<code>FilterGOTerms</code> and <code>FilterReactomeTerms</code>
functions for further filtering and analysis.</li>
</ol>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Return pathway IDs</span></span>
<span><span class="va">result</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/SearchDatabase.html">SearchDatabase</a></span><span class="op">(</span><span class="st">"metabolic"</span>, return <span class="op">=</span> <span class="st">"ID"</span><span class="op">)</span></span>
<span><span class="va">result</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] "GO:0000023" "GO:0000038" "GO:0000052" "GO:0000053" "GO:0000096"</span></span>
<span><span class="co">##  [6] "GO:0000255" "GO:0000270" "GO:0000820" "GO:0000821" "GO:0000959"</span></span>
<span><span class="co">## [11] "GO:0001523" "GO:0001573" "GO:0001575" "GO:0001676" "GO:0001692"</span></span>
<span><span class="co">##  [ reached getOption("max.print") -- omitted 873 entries ]</span></span></code></pre>
<ol start="2" style="list-style-type: decimal">
<li>When <code>return = "genelist"</code>, the function returns a list
of genes formatted to match the input required by the
<code>GeneSetAnalysis</code> function. This allows you to directly use
the search results for enrichment analysis of specific pathways, without
needing to call <code>GeneSetAnalysisGO</code> or
<code>GeneSetAnalysisReactome</code> (which are typically used for more
general screening).</li>
</ol>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Return as a gene list for GeneSetAnalysis</span></span>
<span><span class="va">result</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/SearchDatabase.html">SearchDatabase</a></span><span class="op">(</span><span class="st">"metabolic"</span>, return <span class="op">=</span> <span class="st">"genelist"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">result</span>, list.len <span class="op">=</span> <span class="fl">10</span>, max.level <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## List of 888</span></span>
<span><span class="co">##  $ GO:0000023   : chr [1:3] "MGAM" "GAA" "GANC"</span></span>
<span><span class="co">##  $ GO:0000038   : chr [1:31] "ELOVL7" "HACD1" "SLC27A2" "ACAA1" ...</span></span>
<span><span class="co">##  $ GO:0000052   : chr [1:8] "DDAH1" "DDAH2" "OTC" "ASS1" ...</span></span>
<span><span class="co">##  $ GO:0000053   : chr "ASS1"</span></span>
<span><span class="co">##  $ GO:0000096   : chr [1:43] "F2Z2F3" "AHCYL1" "CBSL" "MTHFD1" ...</span></span>
<span><span class="co">##  $ GO:0000255   : chr [1:2] "URAD" "ALLC"</span></span>
<span><span class="co">##  $ GO:0000270   : chr [1:6] "PGLYRP1" "LYG2" "LYG1" "PGLYRP4" ...</span></span>
<span><span class="co">##  $ GO:0000820   : chr [1:6] "ATP2B4" "ATCAY" "NR1H4" "SLC7A7" ...</span></span>
<span><span class="co">##  $ GO:0000821   : chr [1:2] "ATP2B4" "SLC7A7"</span></span>
<span><span class="co">##  $ GO:0000959   : chr [1:48] "M0R2C6" "POLRMT" "PRORP" "FOXO3" ...</span></span>
<span><span class="co">##   [list output truncated]</span></span></code></pre>
<p>This flexibility enables you to choose the most appropriate output
format based on your specific needs, thereby streamlining subsequent
analysis steps.</p>
</div>
</div>
<div class="section level2">
<h2 id="creating-gsea-plots">Creating GSEA Plots<a class="anchor" aria-label="anchor" href="#creating-gsea-plots"></a>
</h2>
<p>The <code>GSEAplot</code> function generates visualizations similar
to those from the Broad Institute’s GSEA:</p>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../../reference/GSEAplot.html">GSEAplot</a></span><span class="op">(</span></span>
<span>  <span class="va">pbmc</span>, </span>
<span>  ident.1 <span class="op">=</span> <span class="st">"CD4 T Naive"</span>, </span>
<span>  title <span class="op">=</span> <span class="st">"INTERFERON_GAMMA_RESPONSE"</span>,</span>
<span>  geneset <span class="op">=</span> <span class="va">hall50</span><span class="op">$</span><span class="va">human</span><span class="op">$</span><span class="va">HALLMARK_INTERFERON_GAMMA_RESPONSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="4.GSEA_files/figure-html/unnamed-chunk-26-1.png" width="528"></p>
<p>This plot shows the negative enrichment of the
“INTERFERON_GAMMA_RESPONSE” gene set in the “CD4 T Naive” cell
population compared to non-CD4 T Naive cells. The negative Enrichment
Score and the downward trend of the enrichment plot indicate that this
pathway is downregulated in CD4 T Naive cells. The highly significant
p-value suggests that this downregulation is statistically
meaningful.</p>
</div>
<div class="section level2">
<h2 id="utility-functions-for-gene-analysis">Utility Functions for Gene Analysis<a class="anchor" aria-label="anchor" href="#utility-functions-for-gene-analysis"></a>
</h2>
<p>In addition to GSEA, there are several utility functions in
<code>SeuratExtend</code> that can greatly facilitate your scRNA-seq
analysis. Let’s explore some of these useful tools.</p>
<div class="section level3">
<h3 id="gene-naming-conversions">Gene Naming Conversions<a class="anchor" aria-label="anchor" href="#gene-naming-conversions"></a>
</h3>
<p>Gene naming conventions often vary between organisms and databases.
<code>SeuratExtend</code> provides functions to convert between human
and mouse gene symbols, as well as Ensembl IDs. These functions use a
localized version of the <code>biomaRt</code> database, improving
reliability and performance.</p>
<div class="section level4">
<h4 id="available-conversion-functions">Available Conversion Functions:<a class="anchor" aria-label="anchor" href="#available-conversion-functions"></a>
</h4>
<ul>
<li><code>HumanToMouseGenesymbol</code></li>
<li><code>MouseToHumanGenesymbol</code></li>
<li><code>EnsemblToGenesymbol</code></li>
<li><code>GenesymbolToEnsembl</code></li>
</ul>
<p>These functions share a similar usage pattern, as detailed below
using <code>HumanToMouseGenesymbol</code> as an example.</p>
</div>
<div class="section level4">
<h4 id="getting-started-with-examples">Getting Started with Examples<a class="anchor" aria-label="anchor" href="#getting-started-with-examples"></a>
</h4>
<p>First, let’s retrieve a few human gene symbols from a dataset as an
example:</p>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">human_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/VariableFeatures.html" class="external-link">VariableFeatures</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">human_genes</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "PPBP"   "LYZ"    "S100A9" "IGLL5"  "GNLY"   "FTL"</span></span></code></pre>
</div>
<div class="section level4">
<h4 id="default-usage">Default Usage<a class="anchor" aria-label="anchor" href="#default-usage"></a>
</h4>
<p>By default, <code>HumanToMouseGenesymbol</code> returns a data frame
showing how human gene symbols (HGNC) match with mouse gene symbols
(MGI):</p>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../../reference/gene-naming-conversions.html">HumanToMouseGenesymbol</a></span><span class="op">(</span><span class="va">human_genes</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##          MGI.symbol HGNC.symbol</span></span>
<span><span class="co">## 227   9530003J23Rik         LYZ</span></span>
<span><span class="co">## 6304           Ftl1         FTL</span></span>
<span><span class="co">## 6305       Ftl1-ps1         FTL</span></span>
<span><span class="co">## 8244         Gm5849      S100A9</span></span>
<span><span class="co">## 15496          Ppbp        PPBP</span></span></code></pre>
<p>This table indicates that not all human genes have direct mouse
homologs, and some human genes may correspond to multiple mouse
genes.</p>
</div>
<div class="section level4">
<h4 id="simplified-output">Simplified Output<a class="anchor" aria-label="anchor" href="#simplified-output"></a>
</h4>
<p>If you prefer a simpler vector output without the matching
details:</p>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../../reference/gene-naming-conversions.html">HumanToMouseGenesymbol</a></span><span class="op">(</span><span class="va">human_genes</span>, match <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "9530003J23Rik" "Ftl1"          "Ftl1-ps1"      "Gm5849"       </span></span>
<span><span class="co">## [5] "Ppbp"</span></span></code></pre>
</div>
<div class="section level4">
<h4 id="converting-gene-expression-matrices">Converting Gene Expression Matrices<a class="anchor" aria-label="anchor" href="#converting-gene-expression-matrices"></a>
</h4>
<p>These functions can also directly convert human gene expression
matrices to their mouse counterparts:</p>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Create an example gene expression matrix</span></span>
<span><span class="va">human_matr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/AssayData.html" class="external-link">GetAssayData</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span><span class="op">[</span><span class="va">human_genes</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">human_matr</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 6 x 4 sparse Matrix of class "dgCMatrix"</span></span>
<span><span class="co">##        AAACATACAACCAC-1 AAACATTGAGCTAC-1 AAACATTGATCAGC-1 AAACCGTGCTTCCG-1</span></span>
<span><span class="co">## PPBP           .                .                .                1.566387</span></span>
<span><span class="co">## LYZ            1.635208         1.962726         1.994867         4.521175</span></span>
<span><span class="co">## S100A9         .                .                .                3.838845</span></span>
<span><span class="co">## IGLL5          .                .                .                .       </span></span>
<span><span class="co">## GNLY           .                .                1.429261         .       </span></span>
<span><span class="co">## FTL            3.923285         3.063189         3.273549         5.692271</span></span></code></pre>
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Convert to a mouse gene expression matrix</span></span>
<span><span class="fu"><a href="../../reference/gene-naming-conversions.html">HumanToMouseGenesymbol</a></span><span class="op">(</span><span class="va">human_matr</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 4 x 4 sparse Matrix of class "dgCMatrix"</span></span>
<span><span class="co">##               AAACATACAACCAC-1 AAACATTGAGCTAC-1 AAACATTGATCAGC-1</span></span>
<span><span class="co">## Ppbp                  .                .                .       </span></span>
<span><span class="co">## 9530003J23Rik         1.635208         1.962726         1.994867</span></span>
<span><span class="co">## Gm5849                .                .                .       </span></span>
<span><span class="co">## Ftl1                  3.923285         3.063189         3.273549</span></span>
<span><span class="co">##               AAACCGTGCTTCCG-1</span></span>
<span><span class="co">## Ppbp                  1.566387</span></span>
<span><span class="co">## 9530003J23Rik         4.521175</span></span>
<span><span class="co">## Gm5849                3.838845</span></span>
<span><span class="co">## Ftl1                  5.692271</span></span></code></pre>
</div>
<div class="section level4">
<h4 id="other-gene-naming-conversion-functions">Other Gene Naming Conversion Functions<a class="anchor" aria-label="anchor" href="#other-gene-naming-conversion-functions"></a>
</h4>
<p>The usage patterns for the other conversion functions in
<code>SeuratExtend</code>, such as <code>MouseToHumanGenesymbol</code>,
<code>GenesymbolToEnsembl</code>, and <code>EnsemblToGenesymbol</code>,
are similar to those already discussed. These functions also leverage
local databases to enhance performance and reliability but provide
options to use online databases via <code>biomaRt</code> if
necessary.</p>
<p>Here are some examples demonstrating the use of other gene naming
conversion functions:</p>
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Converting mouse gene symbols to human</span></span>
<span><span class="va">mouse_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Cd14"</span>, <span class="st">"Cd3d"</span>, <span class="st">"Cd79a"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../../reference/gene-naming-conversions.html">MouseToHumanGenesymbol</a></span><span class="op">(</span><span class="va">mouse_genes</span>, match <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "CD14"  "CD3D"  "CD79A"</span></span></code></pre>
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Converting human gene symbols to Ensembl IDs</span></span>
<span><span class="va">human_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"PPBP"</span>, <span class="st">"LYZ"</span>, <span class="st">"S100A9"</span>, <span class="st">"IGLL5"</span>, <span class="st">"GNLY"</span>, <span class="st">"FTL"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../../reference/gene-naming-conversions.html">GenesymbolToEnsembl</a></span><span class="op">(</span><span class="va">human_genes</span>, spe <span class="op">=</span> <span class="st">"human"</span>, keep.seq <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##              PPBP               LYZ            S100A9             IGLL5 </span></span>
<span><span class="co">## "ENSG00000163736" "ENSG00000090382" "ENSG00000163220" "ENSG00000254709" </span></span>
<span><span class="co">##              GNLY               FTL </span></span>
<span><span class="co">## "ENSG00000115523" "ENSG00000087086"</span></span></code></pre>
<div class="sourceCode" id="cb59"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Converting mouse gene symbols to Ensembl IDs</span></span>
<span><span class="fu"><a href="../../reference/gene-naming-conversions.html">GenesymbolToEnsembl</a></span><span class="op">(</span><span class="va">mouse_genes</span>, spe <span class="op">=</span> <span class="st">"mouse"</span>, keep.seq <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                 Cd14                 Cd3d                Cd79a </span></span>
<span><span class="co">## "ENSMUSG00000051439" "ENSMUSG00000032094" "ENSMUSG00000003379"</span></span></code></pre>
<div class="sourceCode" id="cb61"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Converting Ensembl IDs to human gene symbols</span></span>
<span><span class="fu"><a href="../../reference/gene-naming-conversions.html">EnsemblToGenesymbol</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ENSG00000163736"</span>, <span class="st">"ENSG00000090382"</span><span class="op">)</span>, spe <span class="op">=</span> <span class="st">"human"</span>, keep.seq <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## ENSG00000163736 ENSG00000090382 </span></span>
<span><span class="co">##          "PPBP"           "LYZ"</span></span></code></pre>
<div class="sourceCode" id="cb63"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Converting Ensembl IDs to mouse gene symbols</span></span>
<span><span class="fu"><a href="../../reference/gene-naming-conversions.html">EnsemblToGenesymbol</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ENSMUSG00000051439"</span>, <span class="st">"ENSMUSG00000032094"</span><span class="op">)</span>, spe <span class="op">=</span> <span class="st">"mouse"</span>, keep.seq <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## ENSMUSG00000051439 ENSMUSG00000032094 </span></span>
<span><span class="co">##             "Cd14"             "Cd3d"</span></span></code></pre>
</div>
</div>
</div>
<div class="section level2">
<h2 id="assess-proportion-of-positive-cells-in-clusters">Assess Proportion of Positive Cells in Clusters<a class="anchor" aria-label="anchor" href="#assess-proportion-of-positive-cells-in-clusters"></a>
</h2>
<p>This section describes how to utilize the
<code>feature_percent</code> function in the <code>SeuratExtend</code>
package to determine the proportion of positive cells within specified
clusters or groups based on defined criteria. This function is
particularly useful for identifying the expression levels of genes or
other features within subpopulations of cells in scRNA-seq datasets.</p>
<div class="section level3">
<h3 id="basic-usage">Basic Usage<a class="anchor" aria-label="anchor" href="#basic-usage"></a>
</h3>
<p>To calculate the proportion of positive cells for the top 10 variable
features in a Seurat object:</p>
<div class="sourceCode" id="cb65"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">options</a></span><span class="op">(</span>max.print <span class="op">=</span> <span class="fl">50</span><span class="op">)</span></span>
<span></span>
<span><span class="va">genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/VariableFeatures.html" class="external-link">VariableFeatures</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># Default usage</span></span>
<span><span class="va">proportions</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/feature_percent.html">feature_percent</a></span><span class="op">(</span><span class="va">pbmc</span>, feature <span class="op">=</span> <span class="va">genes</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">proportions</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##            B cell CD4 T Memory CD4 T Naive  CD8 T cell         DC   Mono CD14</span></span>
<span><span class="co">## PPBP   0.01432665  0.007812500  0.01826484 0.012232416 0.02777778 0.053061224</span></span>
<span><span class="co">## LYZ    0.42406877  0.464843750  0.53272451 0.412844037 0.97222222 0.997959184</span></span>
<span><span class="co">## S100A9 0.13467049  0.130859375  0.18112633 0.159021407 0.66666667 0.993877551</span></span>
<span><span class="co">## IGLL5  0.23495702  0.009765625  0.01065449 0.006116208 0.05555556 0.004081633</span></span>
<span><span class="co">## GNLY   0.08022923  0.076171875  0.11872146 0.311926606 0.22222222 0.126530612</span></span>
<span><span class="co">##        Mono FCGR3A     NK cell   Platelet</span></span>
<span><span class="co">## PPBP    0.07453416 0.006493506 1.00000000</span></span>
<span><span class="co">## LYZ     0.98757764 0.454545455 0.50000000</span></span>
<span><span class="co">## S100A9  0.86335404 0.149350649 0.28571429</span></span>
<span><span class="co">## IGLL5   0.03105590 0.032467532 0.07142857</span></span>
<span><span class="co">## GNLY    0.13043478 0.941558442 0.07142857</span></span>
<span><span class="co">##  [ reached getOption("max.print") -- omitted 5 rows ]</span></span></code></pre>
<div class="sourceCode" id="cb67"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">RColorBrewer</span><span class="op">)</span></span>
<span><span class="fu"><a href="../../reference/Heatmap.html">Heatmap</a></span><span class="op">(</span><span class="va">proportions</span>, lab_fill <span class="op">=</span> <span class="st">"Proportion of\npositive cell"</span>, color_scheme <span class="op">=</span> <span class="st">"OrRd"</span><span class="op">)</span></span></code></pre></div>
<p><img src="4.GSEA_files/figure-html/unnamed-chunk-37-1.png" width="480"></p>
<p>This will return a matrix where rows are features and columns are
clusters, showing the proportion of cells in each cluster where the
feature’s expression is above the default threshold (0).</p>
</div>
<div class="section level3">
<h3 id="adjusting-the-expression-threshold">Adjusting the Expression Threshold<a class="anchor" aria-label="anchor" href="#adjusting-the-expression-threshold"></a>
</h3>
<p>To count a cell as positive only if its expression is above a value
of 2:</p>
<div class="sourceCode" id="cb68"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">proportions_above_2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/feature_percent.html">feature_percent</a></span><span class="op">(</span><span class="va">pbmc</span>, feature <span class="op">=</span> <span class="va">genes</span>, above <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">proportions_above_2</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##             B cell CD4 T Memory CD4 T Naive  CD8 T cell        DC   Mono CD14</span></span>
<span><span class="co">## PPBP   0.005730659  0.001953125  0.00000000 0.006116208 0.0000000 0.014285714</span></span>
<span><span class="co">## LYZ    0.209169054  0.166015625  0.17351598 0.155963303 0.9444444 0.997959184</span></span>
<span><span class="co">## S100A9 0.048710602  0.011718750  0.01065449 0.024464832 0.3611111 0.977551020</span></span>
<span><span class="co">## IGLL5  0.157593123  0.000000000  0.00000000 0.000000000 0.0000000 0.002040816</span></span>
<span><span class="co">## GNLY   0.025787966  0.009765625  0.00608828 0.250764526 0.0000000 0.028571429</span></span>
<span><span class="co">##        Mono FCGR3A     NK cell   Platelet</span></span>
<span><span class="co">## PPBP    0.01242236 0.000000000 1.00000000</span></span>
<span><span class="co">## LYZ     0.92546584 0.142857143 0.42857143</span></span>
<span><span class="co">## S100A9  0.65217391 0.006493506 0.14285714</span></span>
<span><span class="co">## IGLL5   0.00621118 0.006493506 0.00000000</span></span>
<span><span class="co">## GNLY    0.00000000 0.922077922 0.07142857</span></span>
<span><span class="co">##  [ reached getOption("max.print") -- omitted 5 rows ]</span></span></code></pre>
<div class="sourceCode" id="cb70"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../../reference/Heatmap.html">Heatmap</a></span><span class="op">(</span><span class="va">proportions_above_2</span>, lab_fill <span class="op">=</span> <span class="st">"Proportion of\nexpression &gt; 2"</span>, color_scheme <span class="op">=</span> <span class="st">"OrRd"</span><span class="op">)</span></span></code></pre></div>
<p><img src="4.GSEA_files/figure-html/unnamed-chunk-39-1.png" width="518.4"></p>
</div>
<div class="section level3">
<h3 id="targeting-specific-clusters">Targeting Specific Clusters<a class="anchor" aria-label="anchor" href="#targeting-specific-clusters"></a>
</h3>
<p>To calculate proportions for only a subset of clusters:</p>
<div class="sourceCode" id="cb71"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">proportions_subset</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/feature_percent.html">feature_percent</a></span><span class="op">(</span><span class="va">pbmc</span>, feature <span class="op">=</span> <span class="va">genes</span>, ident <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"B cell"</span>, <span class="st">"CD8 T cell"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">proportions_subset</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##            B cell  CD8 T cell</span></span>
<span><span class="co">## PPBP   0.01432665 0.012232416</span></span>
<span><span class="co">## LYZ    0.42406877 0.412844037</span></span>
<span><span class="co">## S100A9 0.13467049 0.159021407</span></span>
<span><span class="co">## IGLL5  0.23495702 0.006116208</span></span>
<span><span class="co">## GNLY   0.08022923 0.311926606</span></span>
<span><span class="co">## FTL    0.99426934 0.984709480</span></span>
<span><span class="co">## PF4    0.01432665 0.012232416</span></span>
<span><span class="co">## FTH1   0.98280802 0.981651376</span></span>
<span><span class="co">## GNG11  0.00286533 0.018348624</span></span>
<span><span class="co">## FCER1A 0.00286533 0.009174312</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="proportions-relative-to-total-cell-numbers">Proportions Relative to Total Cell Numbers<a class="anchor" aria-label="anchor" href="#proportions-relative-to-total-cell-numbers"></a>
</h3>
<p>To also check the proportion of expressed cells in total across
selected clusters:</p>
<div class="sourceCode" id="cb73"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">proportions_total</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/feature_percent.html">feature_percent</a></span><span class="op">(</span><span class="va">pbmc</span>, feature <span class="op">=</span> <span class="va">genes</span>, total <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">proportions_total</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##            B cell CD4 T Memory CD4 T Naive  CD8 T cell         DC   Mono CD14</span></span>
<span><span class="co">## PPBP   0.01432665  0.007812500  0.01826484 0.012232416 0.02777778 0.053061224</span></span>
<span><span class="co">## LYZ    0.42406877  0.464843750  0.53272451 0.412844037 0.97222222 0.997959184</span></span>
<span><span class="co">## S100A9 0.13467049  0.130859375  0.18112633 0.159021407 0.66666667 0.993877551</span></span>
<span><span class="co">## IGLL5  0.23495702  0.009765625  0.01065449 0.006116208 0.05555556 0.004081633</span></span>
<span><span class="co">## GNLY   0.08022923  0.076171875  0.11872146 0.311926606 0.22222222 0.126530612</span></span>
<span><span class="co">##        Mono FCGR3A     NK cell   Platelet      total</span></span>
<span><span class="co">## PPBP    0.07453416 0.006493506 1.00000000 0.02925926</span></span>
<span><span class="co">## LYZ     0.98757764 0.454545455 0.50000000 0.60407407</span></span>
<span><span class="co">## S100A9  0.86335404 0.149350649 0.28571429 0.35629630</span></span>
<span><span class="co">## IGLL5   0.03105590 0.032467532 0.07142857 0.04111111</span></span>
<span><span class="co">## GNLY    0.13043478 0.941558442 0.07142857 0.17925926</span></span>
<span><span class="co">##  [ reached getOption("max.print") -- omitted 5 rows ]</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="logical-output-for-expression">Logical Output for Expression<a class="anchor" aria-label="anchor" href="#logical-output-for-expression"></a>
</h3>
<p>For scenarios where you need a logical output indicating whether a
significant proportion of cells are expressing the feature above a
certain level (e.g., 20%):</p>
<div class="sourceCode" id="cb75"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">expressed_logical</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/feature_percent.html">feature_percent</a></span><span class="op">(</span><span class="va">pbmc</span>, feature <span class="op">=</span> <span class="va">genes</span>, if.expressed <span class="op">=</span> <span class="cn">TRUE</span>, min.pct <span class="op">=</span> <span class="fl">0.2</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">expressed_logical</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##        B cell CD4 T Memory CD4 T Naive CD8 T cell    DC Mono CD14 Mono FCGR3A</span></span>
<span><span class="co">## PPBP    FALSE        FALSE       FALSE      FALSE FALSE     FALSE       FALSE</span></span>
<span><span class="co">## LYZ      TRUE         TRUE        TRUE       TRUE  TRUE      TRUE        TRUE</span></span>
<span><span class="co">## S100A9  FALSE        FALSE       FALSE      FALSE  TRUE      TRUE        TRUE</span></span>
<span><span class="co">## IGLL5    TRUE        FALSE       FALSE      FALSE FALSE     FALSE       FALSE</span></span>
<span><span class="co">## GNLY    FALSE        FALSE       FALSE       TRUE  TRUE     FALSE       FALSE</span></span>
<span><span class="co">##        NK cell Platelet</span></span>
<span><span class="co">## PPBP     FALSE     TRUE</span></span>
<span><span class="co">## LYZ       TRUE     TRUE</span></span>
<span><span class="co">## S100A9   FALSE     TRUE</span></span>
<span><span class="co">## IGLL5    FALSE    FALSE</span></span>
<span><span class="co">## GNLY      TRUE    FALSE</span></span>
<span><span class="co">##  [ reached getOption("max.print") -- omitted 5 rows ]</span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="conclusion">Conclusion<a class="anchor" aria-label="anchor" href="#conclusion"></a>
</h2>
<p>In this lesson, we’ve delved deep into Gene Set Enrichment Analysis
(GSEA) and its application in single-cell RNA sequencing data analysis.
We’ve learned how to perform GSEA using GO and Reactome databases, how
to handle custom gene sets, and how to leverage various utility
functions to enhance the flexibility and depth of our analysis.</p>
<p>These skills will enable you to gain deeper insights into the
biological significance revealed by single-cell data, helping you
uncover new insights and hypotheses in your research. In the upcoming
lessons, we will continue to explore more advanced analytical
techniques, further expanding your capabilities in single-cell data
analysis.</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Yichao Hua.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
