<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Lesson 5: Core scRNA-seq Workflow Enhancements • SeuratExtend</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../../bootstrap-toc.css">
<script src="../../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../../pkgdown.css" rel="stylesheet">
<script src="../../pkgdown.js"></script><meta property="og:title" content="Lesson 5: Core scRNA-seq Workflow Enhancements">
<meta property="og:description" content="SeuratExtend">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../../index.html">SeuratExtend</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../../articles/News.html">What's New in v1.1.0</a>
    </li>
    <li class="divider">
    </li>
<li>
      <a href="../../articles/Visualization.html">Enhanced Visualization</a>
    </li>
    <li>
      <a href="../../articles/GSEA.html">Geneset Enrichment Analysis (GSEA) </a>
    </li>
    <li>
      <a href="../../articles/Trajectory.html">Trajectory and Pseudotime Analysis </a>
    </li>
    <li>
      <a href="../../articles/SCENIC.html">SCENIC for Gene Regulatory Networks Analysis </a>
    </li>
    <li>
      <a href="../../articles/Utilities.html">Utility Tools and Functions</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    scRNA-seq Course
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../../articles/single-cell-course/1.basic-R.html">1. Introduction to R</a>
    </li>
    <li>
      <a href="../../articles/single-cell-course/2.Seurat.html">2. Basic Single-Cell Analysis Workflow</a>
    </li>
    <li>
      <a href="../../articles/single-cell-course/3.Visualization.html">3. Advanced Visualization Techniques</a>
    </li>
    <li>
      <a href="../../articles/single-cell-course/4.GSEA.html">4. Gene Set Enrichment Analysis</a>
    </li>
    <li>
      <a href="../../articles/single-cell-course/5.Core-Enhancement.html">5. Core scRNA-seq Workflow Enhancements</a>
    </li>
    <li>
      <a href="../../articles/single-cell-course/6.Advanced.html">6. Advanced Methods (Part 1)</a>
    </li>
    <li>
      <a href="../../articles/single-cell-course/6.Advanced-2.html">7. Advanced Methods (Part 2)</a>
    </li>
  </ul>
</li>
<li>
  <a href="../../articles/FAQ.html">FAQ</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Lesson 5: Core scRNA-seq Workflow
Enhancements</h1>
                        <h4 data-toc-skip class="author">Yichao Hua</h4>
            
            <h4 data-toc-skip class="date">2024-11-29</h4>
      
      
      <div class="hidden name"><code>5.Core-Enhancement.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>In Lesson 2, we explored the basic workflow of single-cell analysis
using a small PBMC dataset. While this served as an excellent
introduction to the field, it only scratched the surface of what’s
possible with single-cell RNA sequencing (scRNA-seq) analysis. In this
lesson, we’ll delve deeper into various aspects of the workflow,
exploring more advanced techniques and considerations.</p>
<p>We’ll cover the following topics:</p>
<ol style="list-style-type: decimal">
<li>Merging multiple datasets</li>
<li>Detailed quality control (nGene, percent.mt)</li>
<li>Doublet removal</li>
<li>Data integration</li>
<li>Cell-Cycle Scoring and Regression</li>
<li>Alternative normalization methods (SCTransform)</li>
</ol>
</div>
<div class="section level2">
<h2 id="merging-multiple-datasets">1. Merging Multiple Datasets<a class="anchor" aria-label="anchor" href="#merging-multiple-datasets"></a>
</h2>
<p>In real-world scenarios, we often need to analyze multiple datasets
together. This could be due to various reasons, such as comparing
different conditions, time points, or even different technologies used
on the same sample.</p>
<p>For this exercise, we’ll use public datasets from the 10X Genomics
website. These datasets are from the same sample but were processed
using different technologies: the 3’ kit and the 5’ kit. These two
approaches capture genes differently, which can lead to some variation
in the results.</p>
<p>First, let’s set your working directory to where we’ll store all
course materials. If you followed Lesson 1, this would be:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Set working directory to the course folder created in Lesson 1</span></span>
<span><span class="co"># If you used a different location, replace this path with your chosen directory</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/getwd.html" class="external-link">setwd</a></span><span class="op">(</span><span class="st">"~/Documents/single-cell-course"</span><span class="op">)</span></span></code></pre></div>
<p>Load the necessary libraries and read in our data:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SeuratExtend</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Load 3' dataset</span></span>
<span><span class="va">pbmc_3p</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/Read10X.html" class="external-link">Read10X</a></span><span class="op">(</span><span class="st">"data/pbmc_ACDA/3p_ACDA/filtered_feature_bc_matrix/"</span><span class="op">)</span></span>
<span><span class="va">pbmc_3p</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/CreateSeuratObject.html" class="external-link">CreateSeuratObject</a></span><span class="op">(</span><span class="va">pbmc_3p</span>, project <span class="op">=</span> <span class="st">"pbmc_3p"</span>, min.features <span class="op">=</span> <span class="fl">200</span>, min.cells <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Load 5' dataset</span></span>
<span><span class="va">pbmc_5p</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/Read10X.html" class="external-link">Read10X</a></span><span class="op">(</span><span class="st">"data/pbmc_ACDA/5p_ACDA/filtered_feature_bc_matrix/"</span><span class="op">)</span></span>
<span><span class="va">pbmc_5p</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/CreateSeuratObject.html" class="external-link">CreateSeuratObject</a></span><span class="op">(</span><span class="va">pbmc_5p</span>, project <span class="op">=</span> <span class="st">"pbmc_5p"</span>, min.features <span class="op">=</span> <span class="fl">200</span>, min.cells <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Merge the datasets</span></span>
<span><span class="va">pbmc_merge</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/merge.html" class="external-link">merge</a></span><span class="op">(</span><span class="va">pbmc_3p</span>, <span class="va">pbmc_5p</span>, add.cell.ids <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"pbmc_3p"</span>,<span class="st">"pbmc_5p"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">pbmc_merge</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/SplitLayers.html" class="external-link">JoinLayers</a></span><span class="op">(</span><span class="va">pbmc_merge</span><span class="op">)</span> <span class="co"># This step is required for Seurat v5</span></span>
<span><span class="va">pbmc_merge</span></span></code></pre></div>
<pre><code><span><span class="co">## An object of class Seurat </span></span>
<span><span class="co">## 27910 features across 8845 samples within 1 assay </span></span>
<span><span class="co">## Active assay: RNA (27910 features, 0 variable features)</span></span>
<span><span class="co">##  1 layer present: counts</span></span></code></pre>
<p>By merging these datasets, we can compare and analyze them together.
The <code>add.cell.ids</code> parameter adds a prefix to cell names to
distinguish their origin after merging.</p>
</div>
<div class="section level2">
<h2 id="detailed-quality-control-ngene-percent-mt">2. Detailed Quality Control (nGene, percent.mt)<a class="anchor" aria-label="anchor" href="#detailed-quality-control-ngene-percent-mt"></a>
</h2>
<p>In our previous lesson, we used a basic quality control criterion:
<code>min.features = 200</code>, which retains cells with more than 200
detected genes. However, this is quite a low threshold, chosen because
the previous dataset had a low sequencing depth. In practice, these
thresholds need to be adjusted based on the specific characteristics of
your data.</p>
<p>Let’s start by examining the distribution of detected genes (nGene)
in our merged dataset:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../../reference/VlnPlot2.html">VlnPlot2</a></span><span class="op">(</span><span class="va">pbmc_merge</span>, features <span class="op">=</span> <span class="st">"nFeature_RNA"</span>, group.by <span class="op">=</span> <span class="st">"orig.ident"</span><span class="op">)</span></span></code></pre></div>
<p><img src="5.Core-Enhancement_files/figure-html/unnamed-chunk-3-1.png" width="576"></p>
<p>We can observe that the 3’ and 5’ technologies detect different
numbers of genes. This difference is due to the nature of these
technologies:</p>
<ul>
<li>3’ Technology: This approach sequences the 3’ end of mRNA
transcripts. It typically detects more genes and is more efficient, but
it cannot perform TCR/BCR sequencing.</li>
<li>5’ Technology: This method sequences from the 5’ end of mRNA
transcripts. It allows for TCR/BCR sequencing, making it particularly
useful when studying different lymphocyte clones.</li>
</ul>
<div class="section level3">
<h3 id="setting-ngene-thresholds">Setting nGene Thresholds<a class="anchor" aria-label="anchor" href="#setting-ngene-thresholds"></a>
</h3>
<p>Cells with very low nGene are considered low quality and should be
removed. The exact threshold depends on various factors including the
sample type, cell type, and sequencing technology. Here are some general
suggestions:</p>
<ul>
<li>For 3’ sequencing: Usually set above 1000</li>
<li>For 5’ sequencing: Slightly lower, around 800-1000</li>
<li>For single nuclei sequencing: Even lower, typically 500-800</li>
</ul>
<p>However, these are just general references. It’s crucial to examine
the distribution in your data (using plots like the one above) to make
an informed decision. For our current dataset, we’ll use 1000 as the
lower threshold.</p>
<p>We also need to set an upper limit for nGene. While a higher nGene
generally indicates better sequencing quality, extremely high values
often represent anomalies such as doublets (two cells captured and
sequenced as one). A common upper threshold is between 7500 and 8000
genes.</p>
<p>Note that this upper limit may need adjustment for certain
technologies (e.g., Smart-seq2) that typically have higher sequencing
depth, or for cell types (like fibroblasts or certain tumor cells) that
naturally express more genes.</p>
</div>
<div class="section level3">
<h3 id="mitochondrial-gene-percentage-percent-mt">Mitochondrial Gene Percentage (percent.mt)<a class="anchor" aria-label="anchor" href="#mitochondrial-gene-percentage-percent-mt"></a>
</h3>
<p>Another important quality control metric is the percentage of
mitochondrial genes (percent.mt). A high percentage of mitochondrial
reads often indicates low-quality or dying cells. This is because as
cells die, their cytoplasmic mRNA degrades, but mitochondrial RNA is
more stable, leading to a higher proportion of mitochondrial reads.</p>
<p>Let’s calculate and visualize the percent.mt for our dataset:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc_merge</span><span class="op">[[</span><span class="st">"percent.mt"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/PercentageFeatureSet.html" class="external-link">PercentageFeatureSet</a></span><span class="op">(</span><span class="va">pbmc_merge</span>, pattern <span class="op">=</span> <span class="st">"^MT-"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../../reference/VlnPlot2.html">VlnPlot2</a></span><span class="op">(</span><span class="va">pbmc_merge</span>, features <span class="op">=</span> <span class="st">"percent.mt"</span>, group.by <span class="op">=</span> <span class="st">"orig.ident"</span><span class="op">)</span></span></code></pre></div>
<p><img src="5.Core-Enhancement_files/figure-html/unnamed-chunk-4-1.png" width="576"></p>
<p>In this code, <code>"^MT-"</code> indicates genes starting with “MT-”
(for human data). For mouse data, you would use <code>"^mt-"</code>.</p>
<p>Typically, we set a threshold of percent.mt &lt; 20%. For single
nuclei data, which shouldn’t contain mitochondria, a stricter threshold
like percent.mt &lt; 5% is often used.</p>
</div>
<div class="section level3">
<h3 id="applying-quality-control-filters">Applying Quality Control Filters<a class="anchor" aria-label="anchor" href="#applying-quality-control-filters"></a>
</h3>
<p>Now that we’ve discussed our quality control metrics, let’s apply
these filters to our data:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Before filtering</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"Before filtering:"</span>, <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">pbmc_merge</span><span class="op">)</span>, <span class="st">"cells"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "Before filtering: 8845 cells"</span></span></code></pre>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Apply filters</span></span>
<span><span class="va">pbmc_merge</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span></span>
<span>  <span class="va">pbmc_merge</span>,</span>
<span>  subset <span class="op">=</span> <span class="va">nFeature_RNA</span> <span class="op">&gt;</span> <span class="fl">1000</span> <span class="op">&amp;</span></span>
<span>    <span class="va">nFeature_RNA</span> <span class="op">&lt;</span> <span class="fl">7500</span> <span class="op">&amp;</span></span>
<span>    <span class="va">percent.mt</span> <span class="op">&lt;</span> <span class="fl">20</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># After filtering</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"After filtering:"</span>, <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">pbmc_merge</span><span class="op">)</span>, <span class="st">"cells"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "After filtering: 8256 cells"</span></span></code></pre>
<p>By applying these filters, we’ve removed low-quality cells and
potential doublets from our dataset, providing a cleaner foundation for
downstream analysis.</p>
</div>
</div>
<div class="section level2">
<h2 id="doublet-removal">3. Doublet Removal<a class="anchor" aria-label="anchor" href="#doublet-removal"></a>
</h2>
<p>In our previous section, we used a simple upper limit on nGene to
remove potential doublets. However, this method is relatively crude.
More sophisticated algorithms have been developed to identify and remove
doublets with greater precision. One such tool is
<code>scDblFinder</code>.</p>
<p><code>scDblFinder</code> is a machine learning-based tool that
identifies doublets in single-cell RNA sequencing data. It works by
simulating artificial doublets from the existing data and then using
these simulations to train a model that can distinguish real cells from
doublets based on their gene expression profiles.</p>
<p>Before we begin, let’s ensure we have the necessary packages
installed:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://bioconductor.github.io/BiocManager/" class="external-link">"BiocManager"</a></span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://github.com/plger/scDblFinder" class="external-link">"scDblFinder"</a></span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"scDblFinder"</span><span class="op">)</span></span></code></pre></div>
<p>Here’s how we can use <code>scDblFinder</code> in our analysis:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/plger/scDblFinder" class="external-link">scDblFinder</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Convert Seurat object to SingleCellExperiment</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/as.SingleCellExperiment.html" class="external-link">as.SingleCellExperiment</a></span><span class="op">(</span><span class="va">pbmc_merge</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Run scDblFinder</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://plger.github.io/scDblFinder/reference/scDblFinder.html" class="external-link">scDblFinder</a></span><span class="op">(</span><span class="va">sce</span>, samples <span class="op">=</span> <span class="st">"orig.ident"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Add scDblFinder results to our Seurat object</span></span>
<span><span class="va">pbmc_merge</span><span class="op">$</span><span class="va">scDblFinder.class</span> <span class="op">&lt;-</span> <span class="va">sce</span><span class="op">$</span><span class="va">scDblFinder.class</span></span>
<span></span>
<span><span class="co"># View doublet predictions</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">pbmc_merge</span><span class="op">$</span><span class="va">scDblFinder.class</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## singlet doublet </span></span>
<span><span class="co">##    7908     348</span></span></code></pre>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Remove predicted doublets</span></span>
<span><span class="va">pbmc_merge</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">pbmc_merge</span>, subset <span class="op">=</span> <span class="va">scDblFinder.class</span> <span class="op">==</span> <span class="st">"singlet"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Check the number of cells after doublet removal</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"After doublet removal:"</span>, <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">pbmc_merge</span><span class="op">)</span>, <span class="st">"cells"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "After doublet removal: 7908 cells"</span></span></code></pre>
<p>For most standard single-cell RNA-seq analyses, using
<code>scDblFinder</code> for doublet removal is a recommended practice
that will improve your data quality. However, it’s important to keep in
mind that like any computational method, it’s not 100% accurate. There
are a few considerations to keep in mind:</p>
<ol style="list-style-type: decimal">
<li>
<p>False positives: Some cell types might be mistakenly identified
as doublets. For example:</p>
<ul>
<li>Proliferating cells, which naturally have higher RNA content</li>
<li>Rare cell types (like plasmacytoid dendritic cells, or pDCs) that
may have unique gene expression profiles</li>
</ul>
</li>
<li><p>Biological relevance: In some cases, what appears computationally
as a “doublet” might be a biologically relevant state, such as cell-cell
interactions or cells undergoing cell fusion.</p></li>
<li><p>Dataset specificity: The accuracy of doublet detection can vary
depending on the characteristics of your dataset, including the cell
types present and the sequencing technology used.</p></li>
</ol>
<p>If you’re particularly interested in cell types that might be at risk
of being misclassified as doublets, it’s advisable to carefully review
the results before removing cells. You might consider:</p>
<ul>
<li>Checking which cell types are disproportionately classified as
doublets</li>
<li>Comparing the gene expression profiles of cells classified as
doublets vs. singlets</li>
</ul>
<p>Remember, the goal of doublet removal is to improve data quality, but
it shouldn’t come at the cost of losing biologically relevant
information. Always interpret the results in the context of your
biological question and experimental design.</p>
</div>
<div class="section level2">
<h2 id="data-integration">4. Data Integration<a class="anchor" aria-label="anchor" href="#data-integration"></a>
</h2>
<p>Now that we’ve performed quality control and doublet removal, let’s
proceed with the traditional Seurat workflow on our merged PBMC
dataset:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc_merge</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/NormalizeData.html" class="external-link">NormalizeData</a></span><span class="op">(</span><span class="va">pbmc_merge</span><span class="op">)</span></span>
<span><span class="va">pbmc_merge</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindVariableFeatures.html" class="external-link">FindVariableFeatures</a></span><span class="op">(</span><span class="va">pbmc_merge</span><span class="op">)</span></span>
<span><span class="va">pbmc_merge</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/ScaleData.html" class="external-link">ScaleData</a></span><span class="op">(</span><span class="va">pbmc_merge</span><span class="op">)</span></span>
<span><span class="va">pbmc_merge</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunPCA.html" class="external-link">RunPCA</a></span><span class="op">(</span><span class="va">pbmc_merge</span><span class="op">)</span></span>
<span><span class="va">pbmc_merge</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunUMAP.html" class="external-link">RunUMAP</a></span><span class="op">(</span><span class="va">pbmc_merge</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">)</span></span>
<span><span class="va">pbmc_merge</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindNeighbors.html" class="external-link">FindNeighbors</a></span><span class="op">(</span><span class="va">pbmc_merge</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">)</span></span>
<span><span class="va">pbmc_merge</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindClusters.html" class="external-link">FindClusters</a></span><span class="op">(</span><span class="va">pbmc_merge</span>, resolution <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Number of nodes: 7908</span></span>
<span><span class="co">## Number of edges: 272296</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Running Louvain algorithm...</span></span>
<span><span class="co">## Maximum modularity in 10 random starts: 0.9334</span></span>
<span><span class="co">## Number of communities: 17</span></span>
<span><span class="co">## Elapsed time: 0 seconds</span></span></code></pre>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../../reference/DimPlot2.html">DimPlot2</a></span><span class="op">(</span><span class="va">pbmc_merge</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"seurat_clusters"</span>, <span class="st">"orig.ident"</span>, <span class="st">"CD3D"</span>, <span class="st">"CD14"</span><span class="op">)</span>, theme <span class="op">=</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SeuratTheme.html" class="external-link">NoAxes</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="5.Core-Enhancement_files/figure-html/unnamed-chunk-8-1.png" width="864"></p>
<p>Upon examining the results, we notice that the clustering is
significantly influenced by the different techniques used (3’ vs 5’
sequencing). This effect is known as batch effect. Ideally, we want our
clustering to reflect the genuine cell type differences across all
samples, rather than technical differences between batches.</p>
<p>To address this issue, we can use a method called Harmony for batch
effect removal (or data integration). Harmony is an algorithm that
adjusts for batch effects in single-cell data while preserving
biological variation.</p>
<p>The principle behind Harmony is relatively straightforward: After
running PCA, some principal components (PCs) reflect differences between
cell clusters, while others represent differences between samples or
batches. Harmony corrects for this by removing the PCs associated with
batch differences while retaining those that represent genuine
biological variation between cell types. This way, when we subsequently
run UMAP and find clusters, the results will be based primarily on cell
type differences rather than batch effects.</p>
<p>Let’s see how to implement Harmony:</p>
<p>First, if you haven’t installed Harmony, you can do so with the
following code:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st">"harmony"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"harmony"</span><span class="op">)</span></span></code></pre></div>
<p>Now, let’s run Harmony (note that all steps up to and including
RunPCA remain the same as before):</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">harmony</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Loading required package: Rcpp</span></span></code></pre>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc_merge</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/harmony/man/RunHarmony.html" class="external-link">RunHarmony</a></span><span class="op">(</span><span class="va">pbmc_merge</span>, group.by.vars <span class="op">=</span> <span class="st">"orig.ident"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Transposing data matrix</span></span></code></pre>
<pre><code><span><span class="co">## Initializing state using k-means centroids initialization</span></span></code></pre>
<pre><code><span><span class="co">## Harmony 1/10</span></span></code></pre>
<pre><code><span><span class="co">## Harmony 2/10</span></span></code></pre>
<pre><code><span><span class="co">## Harmony 3/10</span></span></code></pre>
<pre><code><span><span class="co">## Harmony 4/10</span></span></code></pre>
<pre><code><span><span class="co">## Harmony 5/10</span></span></code></pre>
<pre><code><span><span class="co">## Harmony converged after 5 iterations</span></span></code></pre>
<p>After running Harmony, we proceed with the usual workflow, but now
using the Harmony-corrected dimensions:</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc_merge</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunUMAP.html" class="external-link">RunUMAP</a></span><span class="op">(</span><span class="va">pbmc_merge</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span>, reduction <span class="op">=</span> <span class="st">"harmony"</span><span class="op">)</span></span>
<span><span class="va">pbmc_merge</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindNeighbors.html" class="external-link">FindNeighbors</a></span><span class="op">(</span><span class="va">pbmc_merge</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span>, reduction <span class="op">=</span> <span class="st">"harmony"</span><span class="op">)</span></span>
<span><span class="va">pbmc_merge</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindClusters.html" class="external-link">FindClusters</a></span><span class="op">(</span><span class="va">pbmc_merge</span>, resolution <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Number of nodes: 7908</span></span>
<span><span class="co">## Number of edges: 276741</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Running Louvain algorithm...</span></span>
<span><span class="co">## Maximum modularity in 10 random starts: 0.8982</span></span>
<span><span class="co">## Number of communities: 12</span></span>
<span><span class="co">## Elapsed time: 0 seconds</span></span></code></pre>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../../reference/DimPlot2.html">DimPlot2</a></span><span class="op">(</span><span class="va">pbmc_merge</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"seurat_clusters"</span>, <span class="st">"orig.ident"</span>, <span class="st">"CD3D"</span>, <span class="st">"CD14"</span><span class="op">)</span>, theme <span class="op">=</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SeuratTheme.html" class="external-link">NoAxes</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="5.Core-Enhancement_files/figure-html/unnamed-chunk-11-1.png" width="864"></p>
<p>Now we can see that the clustering is based on cell types, and the
differences between the two samples (3’ and 5’ sequencing) have been
largely eliminated.</p>
<p>To better understand what Harmony has done, let’s compare the outputs
of PCA and Harmony:</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/Embeddings.html" class="external-link">Embeddings</a></span><span class="op">(</span><span class="va">pbmc_merge</span>, reduction <span class="op">=</span> <span class="st">"pca"</span><span class="op">)</span>, <span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                                PC_1       PC_2      PC_3       PC_4       PC_5</span></span>
<span><span class="co">## pbmc_3p_AAACCCAAGTGGTGGT-1 8.026468 -0.5569947 -9.762277  2.5274767  0.8005453</span></span>
<span><span class="co">## pbmc_3p_AAACGAACACGTCGTG-1 5.573758  2.1691034 -5.162184 -0.4959971  0.2222917</span></span>
<span><span class="co">## pbmc_3p_AAACGAAGTACCGGCT-1 8.495268 -0.5732680 -7.473360  2.3103625 -0.1895022</span></span>
<span><span class="co">##                                  PC_6      PC_7       PC_8      PC_9      PC_10</span></span>
<span><span class="co">## pbmc_3p_AAACCCAAGTGGTGGT-1 -2.0585799  3.759818 -3.4910666 1.7315773 -0.7347548</span></span>
<span><span class="co">## pbmc_3p_AAACGAACACGTCGTG-1  0.4529012 -1.762161  2.7161359 0.5657139  0.7199774</span></span>
<span><span class="co">## pbmc_3p_AAACGAAGTACCGGCT-1 -1.6952750  2.065426 -0.5903696 1.7199383  0.3588536</span></span>
<span><span class="co">##                                 PC_11     PC_12      PC_13      PC_14</span></span>
<span><span class="co">## pbmc_3p_AAACCCAAGTGGTGGT-1 -1.0875434 0.1700239  0.4207566 0.03125937</span></span>
<span><span class="co">## pbmc_3p_AAACGAACACGTCGTG-1 -0.3842285 2.3215834 -2.2742613 0.82687791</span></span>
<span><span class="co">## pbmc_3p_AAACGAAGTACCGGCT-1 -1.5259350 2.1005151 -0.8343805 1.95330595</span></span>
<span><span class="co">##                                 PC_15      PC_16       PC_17    PC_18</span></span>
<span><span class="co">## pbmc_3p_AAACCCAAGTGGTGGT-1 -0.5672855 -0.4333132  0.08478484 2.323176</span></span>
<span><span class="co">## pbmc_3p_AAACGAACACGTCGTG-1 -0.9649320 -2.1877979 -0.12708595 1.235451</span></span>
<span><span class="co">## pbmc_3p_AAACGAAGTACCGGCT-1 -0.4316831  0.1932790  0.25442805 1.317927</span></span>
<span><span class="co">##                                 PC_19       PC_20      PC_21     PC_22</span></span>
<span><span class="co">## pbmc_3p_AAACCCAAGTGGTGGT-1 -0.6686046  0.03036024 -0.0941112 0.1672879</span></span>
<span><span class="co">## pbmc_3p_AAACGAACACGTCGTG-1 -0.8574119  0.03821086  1.2165595 0.8631614</span></span>
<span><span class="co">## pbmc_3p_AAACGAAGTACCGGCT-1 -0.9668927 -1.37593441 -0.7858357 0.2565052</span></span>
<span><span class="co">##                                 PC_23      PC_24      PC_25      PC_26</span></span>
<span><span class="co">## pbmc_3p_AAACCCAAGTGGTGGT-1 -0.9431425 -1.0961104  1.5384187 -2.2628330</span></span>
<span><span class="co">## pbmc_3p_AAACGAACACGTCGTG-1  0.4129460  0.4193254  0.3886109  0.7913914</span></span>
<span><span class="co">## pbmc_3p_AAACGAAGTACCGGCT-1 -1.1181363 -0.4406358 -0.4530322  0.3490909</span></span>
<span><span class="co">##                                 PC_27      PC_28      PC_29      PC_30</span></span>
<span><span class="co">## pbmc_3p_AAACCCAAGTGGTGGT-1 -0.5962907 -1.2608390 -0.3187917  0.9369346</span></span>
<span><span class="co">## pbmc_3p_AAACGAACACGTCGTG-1  0.2735535  0.4259994 -0.2572480 -0.8599530</span></span>
<span><span class="co">## pbmc_3p_AAACGAAGTACCGGCT-1 -0.2951472 -0.5082128 -0.4150948  0.9005109</span></span>
<span><span class="co">##                                PC_31      PC_32      PC_33      PC_34</span></span>
<span><span class="co">## pbmc_3p_AAACCCAAGTGGTGGT-1 1.2749214  3.4092358  0.8348811 -0.0346970</span></span>
<span><span class="co">## pbmc_3p_AAACGAACACGTCGTG-1 0.1665506 -0.8067949 -0.1263073 -0.5833165</span></span>
<span><span class="co">## pbmc_3p_AAACGAAGTACCGGCT-1 1.3315582  0.3988518  0.2921252 -0.9007669</span></span>
<span><span class="co">##                                 PC_35     PC_36      PC_37       PC_38</span></span>
<span><span class="co">## pbmc_3p_AAACCCAAGTGGTGGT-1  2.5667212 2.3171880  0.5490577 -1.13369946</span></span>
<span><span class="co">## pbmc_3p_AAACGAACACGTCGTG-1  2.3531437 0.8391040  1.0686152 -1.89676424</span></span>
<span><span class="co">## pbmc_3p_AAACGAAGTACCGGCT-1 -0.8700297 0.4757511 -0.3616716  0.01687375</span></span>
<span><span class="co">##                                 PC_39        PC_40      PC_41      PC_42</span></span>
<span><span class="co">## pbmc_3p_AAACCCAAGTGGTGGT-1 -0.2095914  0.480441908 -2.7389293  0.2525975</span></span>
<span><span class="co">## pbmc_3p_AAACGAACACGTCGTG-1  0.7896557 -2.035066291 -0.3943168 -0.5208215</span></span>
<span><span class="co">## pbmc_3p_AAACGAAGTACCGGCT-1  0.9659524 -0.001937676  1.4530170  1.0012862</span></span>
<span><span class="co">##                                PC_43       PC_44     PC_45      PC_46</span></span>
<span><span class="co">## pbmc_3p_AAACCCAAGTGGTGGT-1 0.1412820 -1.73131972 -2.527521 -1.4119483</span></span>
<span><span class="co">## pbmc_3p_AAACGAACACGTCGTG-1 0.5976390  0.38622614 -0.149614  1.5306139</span></span>
<span><span class="co">## pbmc_3p_AAACGAAGTACCGGCT-1 0.3023768  0.08900624 -1.403962  0.0547492</span></span>
<span><span class="co">##                                  PC_47      PC_48      PC_49       PC_50</span></span>
<span><span class="co">## pbmc_3p_AAACCCAAGTGGTGGT-1 -0.01225440 -0.2481923 -0.6904021 -0.07385215</span></span>
<span><span class="co">## pbmc_3p_AAACGAACACGTCGTG-1 -1.68094773  0.6444774 -0.4056492  0.37459072</span></span>
<span><span class="co">## pbmc_3p_AAACGAAGTACCGGCT-1  0.04912378  1.5050713 -1.5763721  0.47545033</span></span></code></pre>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/Embeddings.html" class="external-link">Embeddings</a></span><span class="op">(</span><span class="va">pbmc_merge</span>, reduction <span class="op">=</span> <span class="st">"harmony"</span><span class="op">)</span>, <span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                            harmony_1 harmony_2  harmony_3 harmony_4   harmony_5</span></span>
<span><span class="co">## pbmc_3p_AAACCCAAGTGGTGGT-1  8.388242  1.335736 -4.8668592  4.084444  0.41571811</span></span>
<span><span class="co">## pbmc_3p_AAACGAACACGTCGTG-1  6.167808  4.105549 -0.2162408  1.838270 -0.04784189</span></span>
<span><span class="co">## pbmc_3p_AAACGAAGTACCGGCT-1  8.820353  1.359890 -2.4252238  3.893617 -0.54624121</span></span>
<span><span class="co">##                             harmony_6  harmony_7  harmony_8  harmony_9</span></span>
<span><span class="co">## pbmc_3p_AAACCCAAGTGGTGGT-1 -1.4299355  2.1941987 -3.4002057  0.4277696</span></span>
<span><span class="co">## pbmc_3p_AAACGAACACGTCGTG-1  0.4370867 -2.3376343  1.5153617 -0.4987599</span></span>
<span><span class="co">## pbmc_3p_AAACGAAGTACCGGCT-1 -1.0283993  0.5093963 -0.5792729  0.3888035</span></span>
<span><span class="co">##                            harmony_10 harmony_11 harmony_12 harmony_13</span></span>
<span><span class="co">## pbmc_3p_AAACCCAAGTGGTGGT-1 -0.9714052 -0.7872133 -0.2519855  0.3722271</span></span>
<span><span class="co">## pbmc_3p_AAACGAACACGTCGTG-1  0.7019349 -0.3863685  1.3163239 -2.0239064</span></span>
<span><span class="co">## pbmc_3p_AAACGAAGTACCGGCT-1  0.1175696 -1.2006624  1.5922227 -0.7882126</span></span>
<span><span class="co">##                            harmony_14 harmony_15 harmony_16 harmony_17</span></span>
<span><span class="co">## pbmc_3p_AAACCCAAGTGGTGGT-1 -0.2408347 -0.7151454  0.3595161  0.2865837</span></span>
<span><span class="co">## pbmc_3p_AAACGAACACGTCGTG-1  0.2982565 -0.7589811 -1.2856362  0.3511975</span></span>
<span><span class="co">## pbmc_3p_AAACGAAGTACCGGCT-1  1.6059979 -0.5914752  1.0340767  0.4834976</span></span>
<span><span class="co">##                            harmony_18 harmony_19  harmony_20 harmony_21</span></span>
<span><span class="co">## pbmc_3p_AAACCCAAGTGGTGGT-1   2.636532 -0.4950449  0.32679250  0.3099240</span></span>
<span><span class="co">## pbmc_3p_AAACGAACACGTCGTG-1   1.543012 -0.5290196  0.02989942  1.3705860</span></span>
<span><span class="co">## pbmc_3p_AAACGAAGTACCGGCT-1   1.599963 -0.7014958 -1.07391170 -0.3530896</span></span>
<span><span class="co">##                            harmony_22 harmony_23  harmony_24 harmony_25</span></span>
<span><span class="co">## pbmc_3p_AAACCCAAGTGGTGGT-1  0.2147109 -0.6313436 -1.00809759  1.3825846</span></span>
<span><span class="co">## pbmc_3p_AAACGAACACGTCGTG-1  0.6656945  0.2965954  0.09918623  0.4108313</span></span>
<span><span class="co">## pbmc_3p_AAACGAAGTACCGGCT-1  0.2676497 -0.8437084 -0.43827684 -0.5232299</span></span>
<span><span class="co">##                            harmony_26 harmony_27  harmony_28  harmony_29</span></span>
<span><span class="co">## pbmc_3p_AAACCCAAGTGGTGGT-1 -1.8557766 -0.5722115 -0.98298000 -0.16484030</span></span>
<span><span class="co">## pbmc_3p_AAACGAACACGTCGTG-1  0.6545480  0.2271289 -0.08663837 -0.03217491</span></span>
<span><span class="co">## pbmc_3p_AAACGAAGTACCGGCT-1  0.6936615 -0.3396854 -0.31663298 -0.30187817</span></span>
<span><span class="co">##                            harmony_30 harmony_31 harmony_32 harmony_33</span></span>
<span><span class="co">## pbmc_3p_AAACCCAAGTGGTGGT-1  1.0043036  1.1697774  2.9939500  0.9802589</span></span>
<span><span class="co">## pbmc_3p_AAACGAACACGTCGTG-1 -0.7063055  0.4585552 -0.5126894 -0.3914611</span></span>
<span><span class="co">## pbmc_3p_AAACGAAGTACCGGCT-1  0.8235596  1.2546143  0.1458695  0.4756327</span></span>
<span><span class="co">##                             harmony_34 harmony_35 harmony_36 harmony_37</span></span>
<span><span class="co">## pbmc_3p_AAACCCAAGTGGTGGT-1  0.06597315  2.4718136  2.3226729  0.5578287</span></span>
<span><span class="co">## pbmc_3p_AAACGAACACGTCGTG-1 -0.52521668  2.1049645  0.8880769  1.0138088</span></span>
<span><span class="co">## pbmc_3p_AAACGAAGTACCGGCT-1 -0.83771104 -0.8171414  0.4976160 -0.4034560</span></span>
<span><span class="co">##                             harmony_38 harmony_39  harmony_40 harmony_41</span></span>
<span><span class="co">## pbmc_3p_AAACCCAAGTGGTGGT-1 -1.19440293 -0.4690139  0.42072567 -2.7500936</span></span>
<span><span class="co">## pbmc_3p_AAACGAACACGTCGTG-1 -1.78737338  0.9218403 -1.98124213 -0.3375522</span></span>
<span><span class="co">## pbmc_3p_AAACGAAGTACCGGCT-1  0.05253346  0.6830896 -0.01472767  1.4352334</span></span>
<span><span class="co">##                            harmony_42 harmony_43 harmony_44  harmony_45</span></span>
<span><span class="co">## pbmc_3p_AAACCCAAGTGGTGGT-1  0.1544133  0.1317209 -1.6108667 -2.55868404</span></span>
<span><span class="co">## pbmc_3p_AAACGAACACGTCGTG-1 -0.4857095  0.4854964  0.3541641 -0.03231285</span></span>
<span><span class="co">## pbmc_3p_AAACGAAGTACCGGCT-1  0.9027680  0.2622781  0.1967017 -1.48418850</span></span>
<span><span class="co">##                             harmony_46  harmony_47 harmony_48 harmony_49</span></span>
<span><span class="co">## pbmc_3p_AAACCCAAGTGGTGGT-1 -1.46082294 -0.10673104 -0.1552405 -0.7432192</span></span>
<span><span class="co">## pbmc_3p_AAACGAACACGTCGTG-1  1.54188352 -1.64739830  0.6040680 -0.4226623</span></span>
<span><span class="co">## pbmc_3p_AAACGAAGTACCGGCT-1 -0.01742599 -0.04189954  1.6258599 -1.6174858</span></span>
<span><span class="co">##                             harmony_50</span></span>
<span><span class="co">## pbmc_3p_AAACCCAAGTGGTGGT-1 -0.01808169</span></span>
<span><span class="co">## pbmc_3p_AAACGAACACGTCGTG-1  0.24020904</span></span>
<span><span class="co">## pbmc_3p_AAACGAAGTACCGGCT-1  0.51006150</span></span></code></pre>
<p>You’ll notice that they have the same shape. This is because Harmony
essentially corrects the original PC matrix. To visualize this
correction more clearly, let’s use Violin Plots to compare the first few
PCA and Harmony dimensions:</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../../reference/VlnPlot2.html">VlnPlot2</a></span><span class="op">(</span><span class="va">pbmc_merge</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"PC_"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">9</span><span class="op">)</span>, group.by <span class="op">=</span> <span class="st">"orig.ident"</span><span class="op">)</span></span></code></pre></div>
<p><img src="5.Core-Enhancement_files/figure-html/unnamed-chunk-14-1.png" width="864"></p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../../reference/VlnPlot2.html">VlnPlot2</a></span><span class="op">(</span><span class="va">pbmc_merge</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"harmony_"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">9</span><span class="op">)</span>, group.by <span class="op">=</span> <span class="st">"orig.ident"</span><span class="op">)</span></span></code></pre></div>
<p><img src="5.Core-Enhancement_files/figure-html/unnamed-chunk-15-1.png" width="864"></p>
<p>As we can see, the PCs that were previously correlated with the
<code>orig.ident</code> (i.e., showed clear differences between 3’ and
5’ samples) have been corrected. The Harmony-corrected PCs now primarily
retain information about differences between cell types, rather than
differences between sequencing techniques.</p>
<p>This integration step is crucial when working with multiple datasets
or batches, as it allows us to focus on the biological differences of
interest rather than technical variations between samples or batches.
However, it’s important to use this technique judiciously and always
consider whether the differences between your samples might be
biologically relevant rather than just technical noise.</p>
<p>Let’s save the current Seurat object as an RDS file:</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">saveRDS</a></span><span class="op">(</span><span class="va">pbmc_merge</span>, file <span class="op">=</span> <span class="st">"rds/pbmc_merge.rds"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="cell-cycle-scoring-and-regression">5. Cell-Cycle Scoring and Regression<a class="anchor" aria-label="anchor" href="#cell-cycle-scoring-and-regression"></a>
</h2>
<p>When cells are in a state of division, they highly express cell
cycle-related genes. These genes have such a strong signal that in
unsupervised clustering, highly proliferating cells often form their own
cluster, regardless of their original cell type. For example, in
lymphocyte analysis, unsupervised clustering might identify various CD8,
CD4 subgroups and NK cells, along with a cluster of mitotic cells. While
this is often acceptable, there might be cases where you’re particularly
interested in understanding which cell types these mitotic cells
originally belonged to. In such situations, we need methods to mitigate
the confounding effects of cell cycle genes.</p>
<p>This part of our tutorial is based on the Seurat vignette on cell
cycle analysis: <a href="https://satijalab.org/seurat/articles/cell_cycle_vignette" class="external-link uri">https://satijalab.org/seurat/articles/cell_cycle_vignette</a></p>
<p>Let’s start by calculating cell cycle scores:</p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load cell cycle genes</span></span>
<span><span class="va">s.genes</span> <span class="op">&lt;-</span> <span class="va">cc.genes</span><span class="op">$</span><span class="va">s.genes</span></span>
<span><span class="va">g2m.genes</span> <span class="op">&lt;-</span> <span class="va">cc.genes</span><span class="op">$</span><span class="va">g2m.genes</span></span>
<span></span>
<span><span class="co"># Calculate cell cycle scores</span></span>
<span><span class="va">pbmc_merge</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/CellCycleScoring.html" class="external-link">CellCycleScoring</a></span><span class="op">(</span><span class="va">pbmc_merge</span>, s.features <span class="op">=</span> <span class="va">s.genes</span>, g2m.features <span class="op">=</span> <span class="va">g2m.genes</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># View the results</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">pbmc_merge</span><span class="op">@</span><span class="va">meta.data</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                            orig.ident nCount_RNA nFeature_RNA percent.mt</span></span>
<span><span class="co">## pbmc_3p_AAACCCAAGTGGTGGT-1    pbmc_3p       6764         2301   7.140745</span></span>
<span><span class="co">## pbmc_3p_AAACGAACACGTCGTG-1    pbmc_3p       5860         2212   6.621160</span></span>
<span><span class="co">## pbmc_3p_AAACGAAGTACCGGCT-1    pbmc_3p       5616         2049   5.359687</span></span>
<span><span class="co">## pbmc_3p_AAACGAATCGGAGTGA-1    pbmc_3p       6736         2879   8.283848</span></span>
<span><span class="co">## pbmc_3p_AAACGCTAGATTGGGC-1    pbmc_3p       8285         2840   7.314424</span></span>
<span><span class="co">## pbmc_3p_AAACGCTCAAAGCGTG-1    pbmc_3p       6333         2292   5.337123</span></span>
<span><span class="co">##                            scDblFinder.class RNA_snn_res.0.5 seurat_clusters</span></span>
<span><span class="co">## pbmc_3p_AAACCCAAGTGGTGGT-1           singlet               0               0</span></span>
<span><span class="co">## pbmc_3p_AAACGAACACGTCGTG-1           singlet               1               1</span></span>
<span><span class="co">## pbmc_3p_AAACGAAGTACCGGCT-1           singlet               3               3</span></span>
<span><span class="co">## pbmc_3p_AAACGAATCGGAGTGA-1           singlet               2               2</span></span>
<span><span class="co">## pbmc_3p_AAACGCTAGATTGGGC-1           singlet               0               0</span></span>
<span><span class="co">## pbmc_3p_AAACGCTCAAAGCGTG-1           singlet               0               0</span></span>
<span><span class="co">##                                 S.Score   G2M.Score Phase</span></span>
<span><span class="co">## pbmc_3p_AAACCCAAGTGGTGGT-1  0.012127603  0.00916068     S</span></span>
<span><span class="co">## pbmc_3p_AAACGAACACGTCGTG-1 -0.004794661  0.02908853   G2M</span></span>
<span><span class="co">## pbmc_3p_AAACGAAGTACCGGCT-1  0.006306741  0.03890038   G2M</span></span>
<span><span class="co">## pbmc_3p_AAACGAATCGGAGTGA-1 -0.059615105 -0.04554034    G1</span></span>
<span><span class="co">## pbmc_3p_AAACGCTAGATTGGGC-1  0.034862991  0.03874900   G2M</span></span>
<span><span class="co">## pbmc_3p_AAACGCTCAAAGCGTG-1 -0.007030940 -0.02781424    G1</span></span></code></pre>
<p>After running <code>CellCycleScoring</code>, you’ll notice three new
columns in the <code>meta.data</code>: - <code>S.Score</code>: The score
for S phase - <code>G2M.Score</code>: The score for G2/M phase -
<code>Phase</code>: The determined cell cycle phase for each cell</p>
<p>Once we’ve calculated these scores, we can regress out their effects
during the data scaling step:</p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc_merge</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/ScaleData.html" class="external-link">ScaleData</a></span><span class="op">(</span><span class="va">pbmc_merge</span>, vars.to.regress <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"S.Score"</span>, <span class="st">"G2M.Score"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>The rest of the workflow remains unchanged (i.e., you would proceed
with <code>RunPCA</code> and subsequent steps as before).</p>
<p>Similarly, you can also regress out other confounding factors, such
as the percentage of mitochondrial genes:</p>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc_merge</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/ScaleData.html" class="external-link">ScaleData</a></span><span class="op">(</span><span class="va">pbmc_merge</span>, vars.to.regress <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"S.Score"</span>, <span class="st">"G2M.Score"</span>, <span class="st">"percent.mt"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>It’s important to note that while we’re introducing this concept in
class, we won’t be running these steps in our current analysis. The
reason for introducing Cell-Cycle Regression is that some tutorials
recommend doing this routinely. However, based on personal experience,
this regression isn’t always effective, and whether to perform this step
depends on your specific research goals.</p>
<p>Here are some considerations when deciding whether to perform cell
cycle regression:</p>
<ol style="list-style-type: decimal">
<li><p>Research Question: If your research focuses on cell cycle-related
processes or if you’re studying tissues with varying proliferation
rates, you might want to retain this information rather than regress it
out.</p></li>
<li><p>Biological Relevance: In some cases, the cell cycle state might
be biologically relevant to your study. For instance, if you’re studying
a developmental process or cancer, differences in proliferation rates
between cell types or conditions could be important.</p></li>
<li><p>Effect on Other Genes: Regressing out cell cycle effects can
sometimes impact the expression patterns of other genes, potentially
obscuring other biological signals of interest.</p></li>
<li><p>Visualization: Sometimes, having a distinct cluster of mitotic
cells can be informative and acceptable in your analysis, providing a
quick visual cue about proliferating cells in your sample.</p></li>
<li><p>Downstream Analysis: Consider how cell cycle regression might
affect any downstream analyses you plan to perform, such as trajectory
inference or differential expression analysis.</p></li>
</ol>
<p>In my personal opinion, retaining a cluster of mitotic cells in your
cell clustering results is often acceptable and can be informative.
However, the decision to regress out cell cycle effects should be made
based on your specific research questions and the nature of your
biological system.</p>
<p>If you do decide to perform cell cycle regression, it’s a good
practice to compare your results before and after regression to
understand how it’s affecting your data. This can help you make an
informed decision about whether to include this step in your final
analysis pipeline.</p>
</div>
<div class="section level2">
<h2 id="alternative-normalization-methods-sctransform">6. Alternative Normalization Methods (SCTransform)<a class="anchor" aria-label="anchor" href="#alternative-normalization-methods-sctransform"></a>
</h2>
<p>So far, we’ve been using the <code>NormalizeData</code> function for
data normalization. However, Seurat offers an alternative normalization
method called <code>SCTransform</code> (for more details, see <a href="https://satijalab.org/seurat/articles/sctransform_vignette" class="external-link uri">https://satijalab.org/seurat/articles/sctransform_vignette</a>).</p>
<p><code>SCTransform</code> is a powerful normalization method that aims
to remove technical artifacts (e.g. Sequencing depth variations, Dropout
events) from scRNA-seq data while preserving biological heterogeneity.
It uses a regularized negative binomial regression model to normalize
and transform the data, accounting for sequencing depth and other
technical factors.</p>
<p>If you decide to use SCTransform, it replaces the combination of
<code>NormalizeData</code>, <code>FindVariableFeatures</code>, and
<code>ScaleData</code> steps in the traditional Seurat workflow. The
rest of the workflow remains largely unchanged. Here’s an example of how
to implement SCTransform:</p>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"glmGamPoi"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> </span>
<span>    <span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"glmGamPoi"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">options</a></span><span class="op">(</span>future.globals.maxSize <span class="op">=</span> <span class="fl">1000</span> <span class="op">*</span> <span class="fl">1024</span><span class="op">^</span><span class="fl">2</span><span class="op">)</span>  <span class="co"># Increase to 1000 MiB</span></span>
<span></span>
<span><span class="co"># Assuming pbmc_merge has been filtered for nGene and percent.mt, but not yet normalized</span></span>
<span></span>
<span><span class="va">pbmc_merge</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SCTransform.html" class="external-link">SCTransform</a></span><span class="op">(</span><span class="va">pbmc_merge</span><span class="op">)</span></span>
<span><span class="va">pbmc_merge</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunPCA.html" class="external-link">RunPCA</a></span><span class="op">(</span><span class="va">pbmc_merge</span><span class="op">)</span></span>
<span><span class="va">pbmc_merge</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/harmony/man/RunHarmony.html" class="external-link">RunHarmony</a></span><span class="op">(</span><span class="va">pbmc_merge</span>, group.by.vars <span class="op">=</span> <span class="st">"orig.ident"</span>, theta <span class="op">=</span> <span class="fl">5</span><span class="op">)</span> <span class="co"># Increase theta for stronger batch correction between 3' and 5' data</span></span>
<span><span class="va">pbmc_merge</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunUMAP.html" class="external-link">RunUMAP</a></span><span class="op">(</span><span class="va">pbmc_merge</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">20</span>, reduction <span class="op">=</span> <span class="st">"harmony"</span><span class="op">)</span> <span class="co"># SCTransform often benefits from using more dimensions</span></span>
<span><span class="va">pbmc_merge</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindNeighbors.html" class="external-link">FindNeighbors</a></span><span class="op">(</span><span class="va">pbmc_merge</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">20</span>, reduction <span class="op">=</span> <span class="st">"harmony"</span><span class="op">)</span></span>
<span><span class="va">pbmc_merge</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindClusters.html" class="external-link">FindClusters</a></span><span class="op">(</span><span class="va">pbmc_merge</span>, resolution <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Number of nodes: 7908</span></span>
<span><span class="co">## Number of edges: 279097</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Running Louvain algorithm...</span></span>
<span><span class="co">## Maximum modularity in 10 random starts: 0.9110</span></span>
<span><span class="co">## Number of communities: 15</span></span>
<span><span class="co">## Elapsed time: 0 seconds</span></span></code></pre>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../../reference/DimPlot2.html">DimPlot2</a></span><span class="op">(</span><span class="va">pbmc_merge</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"seurat_clusters"</span>, <span class="st">"orig.ident"</span>, <span class="st">"CD3D"</span>, <span class="st">"CD14"</span><span class="op">)</span>, theme <span class="op">=</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SeuratTheme.html" class="external-link">NoAxes</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="5.Core-Enhancement_files/figure-html/unnamed-chunk-20-1.png" width="864"></p>
<p>Both SCTransform and traditional normalization methods are widely
accepted in the scRNA-seq analysis community. The authors of SCTransform
argue that their method more effectively recovers true gene expression
patterns and results in more refined clustering. However, there are some
considerations to keep in mind:</p>
<ol style="list-style-type: decimal">
<li><p>Computational Resources: SCTransform typically requires more
computational resources and storage space, especially for larger
datasets.</p></li>
<li><p>Complex Microenvironments: For data from complex
microenvironments (such as tumors), SCTransform may not always
outperform traditional normalization methods.</p></li>
<li><p>Interpretability: The output of SCTransform can sometimes be less
intuitive to interpret compared to traditional normalization
methods.</p></li>
<li><p>Compatibility: Some downstream analysis tools might be optimized
for traditionally normalized data, so it’s worth checking compatibility
if you plan to use specialized analysis methods.</p></li>
</ol>
<p>In my personal practice, I often run both methods and compare the
results, choosing the one that provides clearer clustering, more
distinct biological signatures, or better alignment with expected cell
type distributions. This comparison can be particularly useful when
working with a new dataset or biological system.</p>
<p>Here are some tips for comparing SCTransform with traditional
normalization:</p>
<ol style="list-style-type: decimal">
<li><p>Visual Comparison: Compare UMAP plots from both methods. Look for
clearer separation between cell types and more distinct
clustering.</p></li>
<li><p>Marker Gene Expression: Check the expression patterns of known
marker genes. The method that shows clearer and more biologically
sensible patterns might be preferable.</p></li>
<li><p>Differential Expression Analysis: Perform differential expression
analysis between clusters using both methods. Compare the results for
biological relevance and consistency with prior knowledge.</p></li>
<li><p>Batch Effect Correction: If you’re working with multiple batches,
compare how well each method handles batch effects.</p></li>
<li><p>Rare Cell Types: Pay attention to whether rare cell populations
are better preserved or more clearly defined in one method versus the
other.</p></li>
</ol>
<p>Remember, the choice between SCTransform and traditional
normalization should be based on your specific dataset, research
questions, and computational resources. It’s always a good idea to
critically evaluate the results and ensure they align with biological
expectations and prior knowledge about your system.</p>
</div>
<div class="section level2">
<h2 id="complete-workflow-summary">7. Complete Workflow Summary<a class="anchor" aria-label="anchor" href="#complete-workflow-summary"></a>
</h2>
<p>To conclude this lesson, let’s provide a comprehensive code block
that encompasses all the steps we’ve covered. We’ll present two main
workflows: one using the traditional <code>NormalizeData</code>
approach, and another using <code>SCTransform.</code> Both workflows
will include calculation of mitotic scores and regression of percent.mt,
with an option to regress out mitotic scores if desired.</p>
<div class="section level3">
<h3 id="traditional-workflow">Traditional Workflow<a class="anchor" aria-label="anchor" href="#traditional-workflow"></a>
</h3>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load necessary libraries</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SeuratExtend</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">harmony</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/plger/scDblFinder" class="external-link">scDblFinder</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Load and merge datasets</span></span>
<span><span class="va">pbmc_3p</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/Read10X.html" class="external-link">Read10X</a></span><span class="op">(</span><span class="st">"data/pbmc_ACDA/3p_ACDA/filtered_feature_bc_matrix/"</span><span class="op">)</span></span>
<span><span class="va">pbmc_3p</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/CreateSeuratObject.html" class="external-link">CreateSeuratObject</a></span><span class="op">(</span><span class="va">pbmc_3p</span>, project <span class="op">=</span> <span class="st">"pbmc_3p"</span>, min.features <span class="op">=</span> <span class="fl">500</span>, min.cells <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span></span>
<span><span class="va">pbmc_5p</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/Read10X.html" class="external-link">Read10X</a></span><span class="op">(</span><span class="st">"data/pbmc_ACDA/5p_ACDA/filtered_feature_bc_matrix/"</span><span class="op">)</span></span>
<span><span class="va">pbmc_5p</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/CreateSeuratObject.html" class="external-link">CreateSeuratObject</a></span><span class="op">(</span><span class="va">pbmc_5p</span>, project <span class="op">=</span> <span class="st">"pbmc_5p"</span>, min.features <span class="op">=</span> <span class="fl">500</span>, min.cells <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span></span>
<span><span class="va">pbmc_merge</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/merge.html" class="external-link">merge</a></span><span class="op">(</span><span class="va">pbmc_3p</span>, <span class="va">pbmc_5p</span>, add.cell.ids <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"pbmc_3p"</span>,<span class="st">"pbmc_5p"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">pbmc_merge</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/SplitLayers.html" class="external-link">JoinLayers</a></span><span class="op">(</span><span class="va">pbmc_merge</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Quality control</span></span>
<span><span class="va">pbmc_merge</span><span class="op">[[</span><span class="st">"percent.mt"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/PercentageFeatureSet.html" class="external-link">PercentageFeatureSet</a></span><span class="op">(</span><span class="va">pbmc_merge</span>, pattern <span class="op">=</span> <span class="st">"^MT-"</span><span class="op">)</span></span>
<span><span class="va">pbmc_merge</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span></span>
<span>  <span class="va">pbmc_merge</span>,</span>
<span>  subset <span class="op">=</span> <span class="va">nFeature_RNA</span> <span class="op">&gt;</span> <span class="fl">1000</span> <span class="op">&amp;</span></span>
<span>    <span class="va">nFeature_RNA</span> <span class="op">&lt;</span> <span class="fl">7500</span> <span class="op">&amp;</span></span>
<span>    <span class="va">percent.mt</span> <span class="op">&lt;</span> <span class="fl">20</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Doublet removal</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/as.SingleCellExperiment.html" class="external-link">as.SingleCellExperiment</a></span><span class="op">(</span><span class="va">pbmc_merge</span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://plger.github.io/scDblFinder/reference/scDblFinder.html" class="external-link">scDblFinder</a></span><span class="op">(</span><span class="va">sce</span>, samples <span class="op">=</span> <span class="st">"orig.ident"</span><span class="op">)</span></span>
<span><span class="va">pbmc_merge</span><span class="op">$</span><span class="va">scDblFinder.class</span> <span class="op">&lt;-</span> <span class="va">sce</span><span class="op">$</span><span class="va">scDblFinder.class</span></span>
<span><span class="va">pbmc_merge</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">pbmc_merge</span>, subset <span class="op">=</span> <span class="va">scDblFinder.class</span> <span class="op">==</span> <span class="st">"singlet"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Traditional normalization and scaling</span></span>
<span><span class="va">pbmc_merge</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/NormalizeData.html" class="external-link">NormalizeData</a></span><span class="op">(</span><span class="va">pbmc_merge</span><span class="op">)</span></span>
<span><span class="va">pbmc_merge</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindVariableFeatures.html" class="external-link">FindVariableFeatures</a></span><span class="op">(</span><span class="va">pbmc_merge</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Calculate cell cycle scores</span></span>
<span><span class="va">s.genes</span> <span class="op">&lt;-</span> <span class="va">cc.genes</span><span class="op">$</span><span class="va">s.genes</span></span>
<span><span class="va">g2m.genes</span> <span class="op">&lt;-</span> <span class="va">cc.genes</span><span class="op">$</span><span class="va">g2m.genes</span></span>
<span><span class="va">pbmc_merge</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/CellCycleScoring.html" class="external-link">CellCycleScoring</a></span><span class="op">(</span><span class="va">pbmc_merge</span>, s.features <span class="op">=</span> <span class="va">s.genes</span>, g2m.features <span class="op">=</span> <span class="va">g2m.genes</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Scale data, regressing out percent.mt</span></span>
<span><span class="co"># To regress out mitotic score, add "S.Score" and "G2M.Score" to vars.to.regress</span></span>
<span><span class="va">pbmc_merge</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/ScaleData.html" class="external-link">ScaleData</a></span><span class="op">(</span><span class="va">pbmc_merge</span>, vars.to.regress <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"percent.mt"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Dimension reduction and clustering</span></span>
<span><span class="va">pbmc_merge</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunPCA.html" class="external-link">RunPCA</a></span><span class="op">(</span><span class="va">pbmc_merge</span><span class="op">)</span></span>
<span><span class="va">pbmc_merge</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/harmony/man/RunHarmony.html" class="external-link">RunHarmony</a></span><span class="op">(</span><span class="va">pbmc_merge</span>, group.by.vars <span class="op">=</span> <span class="st">"orig.ident"</span><span class="op">)</span></span>
<span><span class="va">pbmc_merge</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunUMAP.html" class="external-link">RunUMAP</a></span><span class="op">(</span><span class="va">pbmc_merge</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span>, reduction <span class="op">=</span> <span class="st">"harmony"</span><span class="op">)</span></span>
<span><span class="va">pbmc_merge</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindNeighbors.html" class="external-link">FindNeighbors</a></span><span class="op">(</span><span class="va">pbmc_merge</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span>, reduction <span class="op">=</span> <span class="st">"harmony"</span><span class="op">)</span></span>
<span><span class="va">pbmc_merge</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindClusters.html" class="external-link">FindClusters</a></span><span class="op">(</span><span class="va">pbmc_merge</span>, resolution <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Visualization</span></span>
<span><span class="fu"><a href="../../reference/DimPlot2.html">DimPlot2</a></span><span class="op">(</span><span class="va">pbmc_merge</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"seurat_clusters"</span>, <span class="st">"orig.ident"</span>, <span class="st">"CD3D"</span>, <span class="st">"CD14"</span><span class="op">)</span>, theme <span class="op">=</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SeuratTheme.html" class="external-link">NoAxes</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>It’s worth mentioning that <code>SeuratExtend</code> provides a
convenient function called <code>RunBasicSeurat</code>. This function
integrates multiple steps including cell filtering, calculating
percent.mt, as well as <code>NormalizeData</code>,
<code>FindVariableFeatures</code>, <code>ScaleData</code>,
<code>RunPCA</code>, <code>RunHarmony</code>, <code>RunUMAP</code>,
<code>FindNeighbors</code>, and <code>FindClusters</code> into a single
step, making it easier to use. Here’s an example of how to use it:</p>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load necessary libraries</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SeuratExtend</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/plger/scDblFinder" class="external-link">scDblFinder</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Load and merge datasets</span></span>
<span><span class="va">pbmc_3p</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/Read10X.html" class="external-link">Read10X</a></span><span class="op">(</span><span class="st">"data/pbmc_ACDA/3p_ACDA/filtered_feature_bc_matrix/"</span><span class="op">)</span></span>
<span><span class="va">pbmc_3p</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/CreateSeuratObject.html" class="external-link">CreateSeuratObject</a></span><span class="op">(</span><span class="va">pbmc_3p</span>, project <span class="op">=</span> <span class="st">"pbmc_3p"</span>, min.features <span class="op">=</span> <span class="fl">500</span>, min.cells <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span></span>
<span><span class="va">pbmc_5p</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/Read10X.html" class="external-link">Read10X</a></span><span class="op">(</span><span class="st">"data/pbmc_ACDA/5p_ACDA/filtered_feature_bc_matrix/"</span><span class="op">)</span></span>
<span><span class="va">pbmc_5p</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/CreateSeuratObject.html" class="external-link">CreateSeuratObject</a></span><span class="op">(</span><span class="va">pbmc_5p</span>, project <span class="op">=</span> <span class="st">"pbmc_5p"</span>, min.features <span class="op">=</span> <span class="fl">500</span>, min.cells <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span></span>
<span><span class="va">pbmc_merge</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/merge.html" class="external-link">merge</a></span><span class="op">(</span><span class="va">pbmc_3p</span>, <span class="va">pbmc_5p</span>, add.cell.ids <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"pbmc_3p"</span>,<span class="st">"pbmc_5p"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">pbmc_merge</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/SplitLayers.html" class="external-link">JoinLayers</a></span><span class="op">(</span><span class="va">pbmc_merge</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Doublet removal</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/as.SingleCellExperiment.html" class="external-link">as.SingleCellExperiment</a></span><span class="op">(</span><span class="va">pbmc_merge</span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://plger.github.io/scDblFinder/reference/scDblFinder.html" class="external-link">scDblFinder</a></span><span class="op">(</span><span class="va">sce</span>, samples <span class="op">=</span> <span class="st">"orig.ident"</span><span class="op">)</span></span>
<span><span class="va">pbmc_merge</span><span class="op">$</span><span class="va">scDblFinder.class</span> <span class="op">&lt;-</span> <span class="va">sce</span><span class="op">$</span><span class="va">scDblFinder.class</span></span>
<span><span class="va">pbmc_merge</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">pbmc_merge</span>, subset <span class="op">=</span> <span class="va">scDblFinder.class</span> <span class="op">==</span> <span class="st">"singlet"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Run the integrated workflow</span></span>
<span><span class="va">pbmc_merge</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/RunBasicSeurat.html">RunBasicSeurat</a></span><span class="op">(</span></span>
<span>  <span class="va">pbmc_merge</span>,</span>
<span>  spe <span class="op">=</span> <span class="st">"human"</span>,</span>
<span>  nFeature_RNA.min <span class="op">=</span> <span class="fl">1000</span>,</span>
<span>  nFeature_RNA.max <span class="op">=</span> <span class="fl">7500</span>,</span>
<span>  percent.mt.max <span class="op">=</span> <span class="fl">20</span>,</span>
<span>  dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span>,</span>
<span>  resolution <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  vars.to.regress <span class="op">=</span> <span class="st">"percent.mt"</span>,</span>
<span>  harmony.by <span class="op">=</span> <span class="st">"orig.ident"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Visualization</span></span>
<span><span class="fu"><a href="../../reference/DimPlot2.html">DimPlot2</a></span><span class="op">(</span><span class="va">pbmc_merge</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"seurat_clusters"</span>, <span class="st">"orig.ident"</span>, <span class="st">"CD3D"</span>, <span class="st">"CD14"</span><span class="op">)</span>, theme <span class="op">=</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SeuratTheme.html" class="external-link">NoAxes</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>This approach simplifies the workflow significantly, allowing you to
perform multiple analysis steps with a single function call. It’s
particularly useful for standard analyses or when you want to quickly
process your data using default parameters.</p>
</div>
<div class="section level3">
<h3 id="sctransform-workflow">SCTransform Workflow<a class="anchor" aria-label="anchor" href="#sctransform-workflow"></a>
</h3>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load necessary libraries</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SeuratExtend</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">harmony</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/plger/scDblFinder" class="external-link">scDblFinder</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Load and merge datasets</span></span>
<span><span class="va">pbmc_3p</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/Read10X.html" class="external-link">Read10X</a></span><span class="op">(</span><span class="st">"data/pbmc_ACDA/3p_ACDA/filtered_feature_bc_matrix/"</span><span class="op">)</span></span>
<span><span class="va">pbmc_3p</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/CreateSeuratObject.html" class="external-link">CreateSeuratObject</a></span><span class="op">(</span><span class="va">pbmc_3p</span>, project <span class="op">=</span> <span class="st">"pbmc_3p"</span>, min.features <span class="op">=</span> <span class="fl">500</span>, min.cells <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span></span>
<span><span class="va">pbmc_5p</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/Read10X.html" class="external-link">Read10X</a></span><span class="op">(</span><span class="st">"data/pbmc_ACDA/5p_ACDA/filtered_feature_bc_matrix/"</span><span class="op">)</span></span>
<span><span class="va">pbmc_5p</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/CreateSeuratObject.html" class="external-link">CreateSeuratObject</a></span><span class="op">(</span><span class="va">pbmc_5p</span>, project <span class="op">=</span> <span class="st">"pbmc_5p"</span>, min.features <span class="op">=</span> <span class="fl">500</span>, min.cells <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span></span>
<span><span class="va">pbmc_merge</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/merge.html" class="external-link">merge</a></span><span class="op">(</span><span class="va">pbmc_3p</span>, <span class="va">pbmc_5p</span>, add.cell.ids <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"pbmc_3p"</span>,<span class="st">"pbmc_5p"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">pbmc_merge</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/SplitLayers.html" class="external-link">JoinLayers</a></span><span class="op">(</span><span class="va">pbmc_merge</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Quality control</span></span>
<span><span class="va">pbmc_merge</span><span class="op">[[</span><span class="st">"percent.mt"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/PercentageFeatureSet.html" class="external-link">PercentageFeatureSet</a></span><span class="op">(</span><span class="va">pbmc_merge</span>, pattern <span class="op">=</span> <span class="st">"^MT-"</span><span class="op">)</span></span>
<span><span class="va">pbmc_merge</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span></span>
<span>  <span class="va">pbmc_merge</span>,</span>
<span>  subset <span class="op">=</span> <span class="va">nFeature_RNA</span> <span class="op">&gt;</span> <span class="fl">1000</span> <span class="op">&amp;</span></span>
<span>    <span class="va">nFeature_RNA</span> <span class="op">&lt;</span> <span class="fl">7500</span> <span class="op">&amp;</span></span>
<span>    <span class="va">percent.mt</span> <span class="op">&lt;</span> <span class="fl">20</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Doublet removal</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/as.SingleCellExperiment.html" class="external-link">as.SingleCellExperiment</a></span><span class="op">(</span><span class="va">pbmc_merge</span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://plger.github.io/scDblFinder/reference/scDblFinder.html" class="external-link">scDblFinder</a></span><span class="op">(</span><span class="va">sce</span>, samples <span class="op">=</span> <span class="st">"orig.ident"</span><span class="op">)</span></span>
<span><span class="va">pbmc_merge</span><span class="op">$</span><span class="va">scDblFinder.class</span> <span class="op">&lt;-</span> <span class="va">sce</span><span class="op">$</span><span class="va">scDblFinder.class</span></span>
<span><span class="va">pbmc_merge</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">pbmc_merge</span>, subset <span class="op">=</span> <span class="va">scDblFinder.class</span> <span class="op">==</span> <span class="st">"singlet"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Calculate cell cycle scores</span></span>
<span><span class="va">s.genes</span> <span class="op">&lt;-</span> <span class="va">cc.genes</span><span class="op">$</span><span class="va">s.genes</span></span>
<span><span class="va">g2m.genes</span> <span class="op">&lt;-</span> <span class="va">cc.genes</span><span class="op">$</span><span class="va">g2m.genes</span></span>
<span><span class="va">pbmc_merge</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/CellCycleScoring.html" class="external-link">CellCycleScoring</a></span><span class="op">(</span><span class="va">pbmc_merge</span>, s.features <span class="op">=</span> <span class="va">s.genes</span>, g2m.features <span class="op">=</span> <span class="va">g2m.genes</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># SCTransform normalization</span></span>
<span><span class="co"># To regress out mitotic score, add "S.Score" and "G2M.Score" to vars.to.regress</span></span>
<span><span class="va">pbmc_merge</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SCTransform.html" class="external-link">SCTransform</a></span><span class="op">(</span><span class="va">pbmc_merge</span>, vars.to.regress <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"percent.mt"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Dimension reduction and clustering</span></span>
<span><span class="va">pbmc_merge</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunPCA.html" class="external-link">RunPCA</a></span><span class="op">(</span><span class="va">pbmc_merge</span><span class="op">)</span></span>
<span><span class="va">pbmc_merge</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/harmony/man/RunHarmony.html" class="external-link">RunHarmony</a></span><span class="op">(</span><span class="va">pbmc_merge</span>, group.by.vars <span class="op">=</span> <span class="st">"orig.ident"</span>, theta <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span>
<span><span class="va">pbmc_merge</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunUMAP.html" class="external-link">RunUMAP</a></span><span class="op">(</span><span class="va">pbmc_merge</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">20</span>, reduction <span class="op">=</span> <span class="st">"harmony"</span><span class="op">)</span></span>
<span><span class="va">pbmc_merge</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindNeighbors.html" class="external-link">FindNeighbors</a></span><span class="op">(</span><span class="va">pbmc_merge</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">20</span>, reduction <span class="op">=</span> <span class="st">"harmony"</span><span class="op">)</span></span>
<span><span class="va">pbmc_merge</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindClusters.html" class="external-link">FindClusters</a></span><span class="op">(</span><span class="va">pbmc_merge</span>, resolution <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Visualization</span></span>
<span><span class="fu"><a href="../../reference/DimPlot2.html">DimPlot2</a></span><span class="op">(</span><span class="va">pbmc_merge</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"seurat_clusters"</span>, <span class="st">"orig.ident"</span>, <span class="st">"CD3D"</span>, <span class="st">"CD14"</span><span class="op">)</span>, theme <span class="op">=</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SeuratTheme.html" class="external-link">NoAxes</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>These comprehensive workflows incorporate all the key steps we’ve
discussed in this lesson, including data loading, quality control,
doublet removal, normalization (using both traditional and
<code>SCTransform</code> methods), cell cycle scoring, batch correction
with Harmony, and clustering.</p>
<p>Remember that the choice between the traditional workflow and
<code>SCTransform</code> depends on your specific dataset and research
questions. It’s often beneficial to try both methods and compare the
results. Additionally, the decision to regress out mitotic scores should
be based on your biological question and the nature of your data.</p>
<p>By providing these complete workflows, you can easily reproduce the
entire analysis process and modify it as needed for your own datasets.
Happy analyzing!</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Yichao Hua.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
