<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Lesson 6: Advanced scRNA-seq Analytical Methods (Part 2) • SeuratExtend</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../../bootstrap-toc.css">
<script src="../../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../../pkgdown.css" rel="stylesheet">
<script src="../../pkgdown.js"></script><meta property="og:title" content="Lesson 6: Advanced scRNA-seq Analytical Methods (Part 2)">
<meta property="og:description" content="SeuratExtend">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../../index.html">SeuratExtend</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../../articles/News.html">What's New in v1.1.0</a>
    </li>
    <li class="divider">
    </li>
<li>
      <a href="../../articles/Visualization.html">Enhanced Visualization</a>
    </li>
    <li>
      <a href="../../articles/GSEA.html">Geneset Enrichment Analysis (GSEA) </a>
    </li>
    <li>
      <a href="../../articles/Trajectory.html">Trajectory and Pseudotime Analysis </a>
    </li>
    <li>
      <a href="../../articles/SCENIC.html">SCENIC for Gene Regulatory Networks Analysis </a>
    </li>
    <li>
      <a href="../../articles/Utilities.html">Utility Tools and Functions</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    scRNA-seq Course
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../../articles/single-cell-course/1.basic-R.html">1. Introduction to R</a>
    </li>
    <li>
      <a href="../../articles/single-cell-course/2.Seurat.html">2. Basic Single-Cell Analysis Workflow</a>
    </li>
    <li>
      <a href="../../articles/single-cell-course/3.Visualization.html">3. Advanced Visualization Techniques</a>
    </li>
    <li>
      <a href="../../articles/single-cell-course/4.GSEA.html">4. Gene Set Enrichment Analysis</a>
    </li>
    <li>
      <a href="../../articles/single-cell-course/5.Core-Enhancement.html">5. Core scRNA-seq Workflow Enhancements</a>
    </li>
    <li>
      <a href="../../articles/single-cell-course/6.Advanced.html">6. Advanced Methods (Part 1)</a>
    </li>
    <li>
      <a href="../../articles/single-cell-course/6.Advanced-2.html">7. Advanced Methods (Part 2)</a>
    </li>
  </ul>
</li>
<li>
  <a href="../../articles/FAQ.html">FAQ</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Lesson 6: Advanced scRNA-seq Analytical Methods
(Part 2)</h1>
                        <h4 data-toc-skip class="author">Yichao Hua</h4>
            
            <h4 data-toc-skip class="date">2024-11-29</h4>
      
      
      <div class="hidden name"><code>6.Advanced-2.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="label-transfer-with-singlecellnet">4. Label Transfer with singleCellNet<a class="anchor" aria-label="anchor" href="#label-transfer-with-singlecellnet"></a>
</h2>
<div class="section level3">
<h3 id="introduction-to-label-transfer">Introduction to Label Transfer<a class="anchor" aria-label="anchor" href="#introduction-to-label-transfer"></a>
</h3>
<p>Label transfer is a crucial technique in single-cell RNA sequencing
analysis. It allows us to use information from well-annotated datasets
(reference data) to classify cells in new, unannotated datasets (query
data). This is particularly useful when working with novel or poorly
characterized cell populations.</p>
</div>
<div class="section level3">
<h3 id="singlecellnet-a-tool-for-label-transfer">singleCellNet: A Tool for Label Transfer<a class="anchor" aria-label="anchor" href="#singlecellnet-a-tool-for-label-transfer"></a>
</h3>
<p><a href="https://github.com/pcahan1/singleCellNet" class="external-link">singleCellNet</a>
is a machine learning-based tool designed for classifying cells in
single-cell RNA sequencing data. It offers robust performance across
different platforms and species and provides interpretable feature
importance scores.</p>
</div>
<div class="section level3">
<h3 id="setting-up-singlecellnet">Setting Up singleCellNet<a class="anchor" aria-label="anchor" href="#setting-up-singlecellnet"></a>
</h3>
<p>First, let’s install and load the necessary libraries:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"devtools"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"devtools"</span><span class="op">)</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"singleCellNet"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"pcahan1/singleCellNet"</span><span class="op">)</span></span>
<span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SeuratExtend</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://pcahan1.github.io/singleCellNet" class="external-link">singleCellNet</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SingleCellExperiment</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="preparing-the-data">Preparing the Data<a class="anchor" aria-label="anchor" href="#preparing-the-data"></a>
</h3>
<p>For this example, we’ll use two datasets: one as a reference
(training) set and another as a query (test) set. Let’s assume we have
these as Seurat objects:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load your reference and query Seurat objects</span></span>
<span><span class="co"># In a real scenario, you would load your data here</span></span>
<span><span class="va">reference_seurat</span> <span class="op">&lt;-</span> <span class="va">pbmc</span></span>
<span><span class="va">query_seurat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"rds/pbmc_merge.rds"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Convert Seurat objects to SingleCellExperiment objects</span></span>
<span><span class="va">sce_reference</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/as.SingleCellExperiment.html" class="external-link">as.SingleCellExperiment</a></span><span class="op">(</span><span class="va">reference_seurat</span><span class="op">)</span></span>
<span><span class="va">sce_query</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/as.SingleCellExperiment.html" class="external-link">as.SingleCellExperiment</a></span><span class="op">(</span><span class="va">query_seurat</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Find common genes between the two datasets</span></span>
<span><span class="va">commonGenes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://generics.r-lib.org/reference/setops.html" class="external-link">intersect</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">sce_reference</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">sce_query</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Subset both datasets to only include common genes</span></span>
<span><span class="va">sce_reference</span> <span class="op">&lt;-</span> <span class="va">sce_reference</span><span class="op">[</span><span class="va">commonGenes</span>, <span class="op">]</span></span>
<span><span class="va">sce_query</span> <span class="op">&lt;-</span> <span class="va">sce_query</span><span class="op">[</span><span class="va">commonGenes</span>, <span class="op">]</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="training-the-classifier">Training the Classifier<a class="anchor" aria-label="anchor" href="#training-the-classifier"></a>
</h3>
<p>Now, let’s train the singleCellNet classifier using our reference
dataset:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Prepare training data</span></span>
<span><span class="va">stTrain</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>  cell_name <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sce_reference</span><span class="op">)</span><span class="op">)</span>, </span>
<span>  cell_type <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sce_reference</span><span class="op">)</span><span class="op">$</span><span class="va">cluster</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">expTrain</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assays</a></span><span class="op">(</span><span class="va">sce_reference</span><span class="op">)</span><span class="op">$</span><span class="va">counts</span></span>
<span></span>
<span><span class="co"># Train the classifier</span></span>
<span><span class="va">class_info</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/singleCellNet/man/scn_train.html" class="external-link">scn_train</a></span><span class="op">(</span></span>
<span>  stTrain <span class="op">=</span> <span class="va">stTrain</span>, </span>
<span>  expTrain <span class="op">=</span> <span class="va">expTrain</span>, </span>
<span>  nTopGenes <span class="op">=</span> <span class="fl">10</span>, </span>
<span>  nRand <span class="op">=</span> <span class="fl">70</span>, </span>
<span>  nTrees <span class="op">=</span> <span class="fl">1000</span>, </span>
<span>  nTopGenePairs <span class="op">=</span> <span class="fl">25</span>, </span>
<span>  dLevel <span class="op">=</span> <span class="st">"cell_type"</span>, </span>
<span>  colName_samp <span class="op">=</span> <span class="st">"cell_name"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Sample table has been prepared</span></span>
<span><span class="co">## Expression data has been normalized</span></span>
<span><span class="co">## Finding classification genes</span></span>
<span><span class="co">## Done testing</span></span>
<span><span class="co">## There are  136  classification genes</span></span>
<span><span class="co">## Finding top pairs</span></span>
<span><span class="co">## nPairs =  190  for  B cell </span></span>
<span><span class="co">## nPairs =  190  for  NK cell </span></span>
<span><span class="co">## nPairs =  190  for  Mono CD14 </span></span>
<span><span class="co">## nPairs =  190  for  CD4 T Memory </span></span>
<span><span class="co">## nPairs =  190  for  CD4 T Naive </span></span>
<span><span class="co">## nPairs =  190  for  Mono FCGR3A </span></span>
<span><span class="co">## nPairs =  190  for  CD8 T cell </span></span>
<span><span class="co">## nPairs =  190  for  DC </span></span>
<span><span class="co">## nPairs =  190  for  Platelet </span></span>
<span><span class="co">## There are 224 top gene pairs</span></span>
<span><span class="co">## Finished pair transforming the data</span></span>
<span><span class="co">## Number of missing genes  0 </span></span>
<span><span class="co">## All Done</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="classifying-cells-in-the-query-dataset">Classifying Cells in the Query Dataset<a class="anchor" aria-label="anchor" href="#classifying-cells-in-the-query-dataset"></a>
</h3>
<p>With our trained classifier, we can now classify cells in our query
dataset:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">classRes_query</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/singleCellNet/man/scn_predict.html" class="external-link">scn_predict</a></span><span class="op">(</span></span>
<span>  cnProc <span class="op">=</span> <span class="va">class_info</span><span class="op">[[</span><span class="st">'cnProc'</span><span class="op">]</span><span class="op">]</span>, </span>
<span>  expDat <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assays</a></span><span class="op">(</span><span class="va">sce_query</span><span class="op">)</span><span class="op">$</span><span class="va">counts</span>, </span>
<span>  nrand <span class="op">=</span> <span class="fl">50</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Loaded in the cnProc</span></span>
<span><span class="co">## All Done</span></span></code></pre>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Add classification results to the query Seurat object</span></span>
<span><span class="va">query_seurat</span><span class="op">$</span><span class="va">Predicted_Labels</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">classRes_query</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="va">classRes_query</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">query_seurat</span><span class="op">)</span><span class="op">]</span>, <span class="fl">2</span>, <span class="va">which.max</span><span class="op">)</span><span class="op">]</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="visualizing-the-results">Visualizing the Results<a class="anchor" aria-label="anchor" href="#visualizing-the-results"></a>
</h3>
<p>Let’s visualize the classification results:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Create a UMAP plot colored by the singleCellNet classifications</span></span>
<span><span class="fu"><a href="../../reference/DimPlot2.html">DimPlot2</a></span><span class="op">(</span><span class="va">query_seurat</span>, group.by <span class="op">=</span> <span class="st">"Predicted_Labels"</span>, label <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## The 'I want hue' color presets were generated from: https://medialab.github.io/iwanthue/</span></span>
<span><span class="co">## This message is shown once per session</span></span></code></pre>
<p><img src="6.Advanced-2_files/figure-html/visualize_results-1.png" width="720"></p>
</div>
<div class="section level3">
<h3 id="interpreting-the-results">Interpreting the Results<a class="anchor" aria-label="anchor" href="#interpreting-the-results"></a>
</h3>
<p>This plot shows how cells in our query dataset have been classified
based on the reference dataset. Each color represents a different cell
type predicted by singleCellNet.</p>
<p>Key points to consider when interpreting these results: 1. Clusters
of cells with the same classification suggest consistent cell types. 2.
Cells classified as multiple types or with low confidence scores may
represent novel cell types or states not present in the reference data.
3. The accuracy of classification depends on the quality and relevance
of the reference dataset.</p>
</div>
<div class="section level3">
<h3 id="conclusion">Conclusion<a class="anchor" aria-label="anchor" href="#conclusion"></a>
</h3>
<p>singleCellNet provides a powerful method for transferring cell type
labels from a reference dataset to a new query dataset. This can greatly
aid in the annotation of novel single-cell RNA sequencing data,
especially when dealing with complex or poorly characterized
tissues.</p>
<p>However, it’s important to remember that computational predictions
should always be validated with biological knowledge and, when possible,
experimental validation.</p>
</div>
</div>
<div class="section level2">
<h2 id="magic-for-denoising-and-smoothing-gene-expression">5. MAGIC for Denoising and Smoothing Gene Expression<a class="anchor" aria-label="anchor" href="#magic-for-denoising-and-smoothing-gene-expression"></a>
</h2>
<p><a href="https://github.com/KrishnaswamyLab/MAGIC" class="external-link">MAGIC (Markov
Affinity-based Graph Imputation of Cells)</a> is another powerful tool
in the single-cell RNA sequencing analysis toolkit. It’s designed to
address one of the major challenges in scRNA-seq data: the high levels
of noise and data sparsity due to technical limitations in the
sequencing process.</p>
<div class="section level3">
<h3 id="what-is-magic-and-why-do-we-need-it">What is MAGIC and Why Do We Need It?<a class="anchor" aria-label="anchor" href="#what-is-magic-and-why-do-we-need-it"></a>
</h3>
<p>Single-cell RNA sequencing data often suffers from:</p>
<ol style="list-style-type: decimal">
<li>Dropout events: Where genes are not detected even though they are
actually expressed.</li>
<li>Technical noise: Random fluctuations in gene expression
measurements.</li>
</ol>
<p>These issues can make it difficult to identify true biological
signals and patterns in the data. MAGIC helps to address these problems
by:</p>
<ul>
<li>Denoising the data: Reducing random technical noise.</li>
<li>Imputing missing values: Filling in likely gene expression values
for dropout events.</li>
<li>Smoothing gene expression: Creating a more continuous representation
of gene expression across cells.</li>
</ul>
<p>MAGIC works by constructing a graph representing the similarity
between cells, and then using this graph to share information between
similar cells. This process helps to recover gene-gene relationships and
reveal cellular trajectories that might be obscured in the raw data.</p>
</div>
<div class="section level3">
<h3 id="running-magic-with-seuratextend">Running MAGIC with SeuratExtend<a class="anchor" aria-label="anchor" href="#running-magic-with-seuratextend"></a>
</h3>
<p>The <code>SeuratExtend</code> package makes it easy to run MAGIC on
your Seurat object. Here’s how you can do it:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mye_small</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/connections.html" class="external-link">url</a></span><span class="op">(</span><span class="st">"https://zenodo.org/records/10944066/files/pbmc10k_mye_small_velocyto.rds"</span>, <span class="st">"rb"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">mye_small</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/Palantir-Magic.html">Palantir.Magic</a></span><span class="op">(</span><span class="va">mye_small</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Determing nearest neighbor graph...</span></span></code></pre>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># MAGIC creates a new assay called "magic" in the Seurat object</span></span>
<span><span class="co"># Let's normalize this new assay</span></span>
<span><span class="va">mye_small</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/NormalizeData.html" class="external-link">NormalizeData</a></span><span class="op">(</span><span class="va">mye_small</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="visualizing-the-effects-of-magic">Visualizing the Effects of MAGIC<a class="anchor" aria-label="anchor" href="#visualizing-the-effects-of-magic"></a>
</h3>
<p>To see the impact of MAGIC, we can compare the original gene
expression with the MAGIC-imputed expression:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Compare original and MAGIC-imputed expression for CD14 and FLT3</span></span>
<span><span class="fu"><a href="../../reference/DimPlot2.html">DimPlot2</a></span><span class="op">(</span><span class="va">mye_small</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD14"</span>, <span class="st">"magic_CD14"</span>, <span class="st">"FLT3"</span>, <span class="st">"magic_FLT3"</span><span class="op">)</span>, </span>
<span>         theme <span class="op">=</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SeuratTheme.html" class="external-link">NoAxes</a></span><span class="op">(</span><span class="op">)</span>, cols <span class="op">=</span> <span class="st">"A"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Loading required package: viridis</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: viridisLite</span></span></code></pre>
<p><img src="6.Advanced-2_files/figure-html/visualize-magic-1.png" width="768"></p>
<p>In these plots: - “CD14” and “FLT3” show the original gene
expression. - “magic_CD14” and “magic_FLT3” show the MAGIC-imputed
expression.</p>
<p>You should notice that the MAGIC-imputed plots appear smoother and
have fewer cells with zero expression. This is because MAGIC has filled
in likely expression values for dropout events and smoothed the overall
expression pattern.</p>
</div>
<div class="section level3">
<h3 id="interpreting-magic-results">Interpreting MAGIC Results<a class="anchor" aria-label="anchor" href="#interpreting-magic-results"></a>
</h3>
<p>When interpreting MAGIC results, keep in mind:</p>
<ol style="list-style-type: decimal">
<li>Smoothed data can reveal patterns that were obscured in the original
data.</li>
<li>However, MAGIC may also remove some of the biological noise that
could be important in certain analyses.</li>
<li>MAGIC-imputed data is great for visualization and exploratory
analysis, but you should use the original data for statistical tests and
differential expression analysis.</li>
</ol>
</div>
<div class="section level3">
<h3 id="using-magic-in-your-analysis-workflow">Using MAGIC in Your Analysis Workflow<a class="anchor" aria-label="anchor" href="#using-magic-in-your-analysis-workflow"></a>
</h3>
<p>MAGIC can be particularly useful for:</p>
<ul>
<li>Visualizing gene expression patterns, especially for genes with low
expression.</li>
<li>Exploring gene-gene relationships.</li>
<li>Enhancing trajectory inference analyses.</li>
</ul>
<p>However, always compare MAGIC results with your original data to
ensure you’re not over-interpreting imputed patterns.</p>
</div>
</div>
<div class="section level2">
<h2 id="copy-number-variation-analysis-copykat">6. Copy Number Variation Analysis: CopyKat<a class="anchor" aria-label="anchor" href="#copy-number-variation-analysis-copykat"></a>
</h2>
<p><a href="https://github.com/navinlabcode/copykat" class="external-link">CopyKat</a>
(Copynumber Karyotyping of Tumors) is a computational tool designed to
infer genomic copy number and subclonal structure of human tumors from
single-cell RNA sequencing data. It’s particularly useful in cancer
research for identifying aneuploid cells and understanding tumor
heterogeneity.</p>
<div class="section level3">
<h3 id="what-is-copykat-and-why-is-it-important">What is CopyKat and Why is it Important?<a class="anchor" aria-label="anchor" href="#what-is-copykat-and-why-is-it-important"></a>
</h3>
<p>CopyKat addresses several key challenges in single-cell RNA
sequencing of human tumors:</p>
<ol style="list-style-type: decimal">
<li>Distinguishing cancer cells from stromal cell types</li>
<li>Identifying the presence of multiple tumor subclones</li>
<li>Inferring genome-wide aneuploidy at 5MB resolution in single
cells</li>
</ol>
<p>The tool is based on the principle that gene expression levels of
many adjacent genes can be influenced by genomic DNA copy numbers in
that region. CopyKat’s estimated copy numbers can achieve a high
concordance (80%) with actual DNA copy numbers obtained by whole genome
DNA sequencing.</p>
</div>
<div class="section level3">
<h3 id="installing-copykat">Installing CopyKat<a class="anchor" aria-label="anchor" href="#installing-copykat"></a>
</h3>
<p>You can install CopyKat from GitHub using the following R code:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st">"copykat"</span><span class="op">)</span><span class="op">)</span> <span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"navinlabcode/copykat"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="running-copykat">Running CopyKat<a class="anchor" aria-label="anchor" href="#running-copykat"></a>
</h3>
<p>To run CopyKat, you need a raw gene expression matrix, with gene IDs
in rows and cell names in columns. The matrix values are typically the
count of unique molecular identifiers (UMI) from high-throughput
single-cell RNA-seq data.</p>
<p>Here’s a basic example of running CopyKat:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">copykat</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Assuming 'exp.rawdata' is your raw UMI count matrix</span></span>
<span><span class="va">copykat.test</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/copykat/man/copykat.html" class="external-link">copykat</a></span><span class="op">(</span></span>
<span>  rawmat<span class="op">=</span><span class="va">exp.rawdata</span>, </span>
<span>  id.type<span class="op">=</span><span class="st">"S"</span>, </span>
<span>  cell.line<span class="op">=</span><span class="st">"no"</span>, </span>
<span>  ngene.chr<span class="op">=</span><span class="fl">5</span>, </span>
<span>  win.size<span class="op">=</span><span class="fl">25</span>, </span>
<span>  KS.cut<span class="op">=</span><span class="fl">0.1</span>, </span>
<span>  sam.name<span class="op">=</span><span class="st">"test"</span>, </span>
<span>  distance<span class="op">=</span><span class="st">"euclidean"</span>, </span>
<span>  n.cores<span class="op">=</span><span class="fl">1</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="interpreting-copykat-results">Interpreting CopyKat Results<a class="anchor" aria-label="anchor" href="#interpreting-copykat-results"></a>
</h3>
<p>CopyKat provides several outputs:</p>
<ol style="list-style-type: decimal">
<li>Prediction of aneuploid (tumor) and diploid (normal) cells</li>
<li>A copy number matrix for each cell</li>
<li>A heatmap visualization of copy number variations</li>
</ol>
<p>Here’s an example of what the prediction results might look like:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">copykat.test</span><span class="op">$</span><span class="va">prediction</span><span class="op">)</span></span></code></pre></div>
<p>And here’s how you might access the copy number matrix:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">copykat.test</span><span class="op">$</span><span class="va">CNAmat</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="visualizing-copykat-results">Visualizing CopyKat Results<a class="anchor" aria-label="anchor" href="#visualizing-copykat-results"></a>
</h3>
<p>CopyKat generates a heatmap of estimated copy numbers. Here’s an
example of what this heatmap might look like:</p>
<div class="float">
<img src="images/test_copykat_heatmap.jpeg" alt="CopyKat Heatmap"><div class="figcaption">CopyKat Heatmap</div>
</div>
<p>In this heatmap: - Rows represent single cells - Columns represent
220kb bins in genomic order - Colors indicate copy number state (red for
amplifications, blue for deletions)</p>
</div>
<div class="section level3">
<h3 id="subpopulation-analysis">Subpopulation Analysis<a class="anchor" aria-label="anchor" href="#subpopulation-analysis"></a>
</h3>
<p>After identifying aneuploid cells, CopyKat can be used to define
subpopulations of tumor cells based on their copy number profiles. This
can provide insights into tumor heterogeneity and evolution.</p>
</div>
<div class="section level3">
<h3 id="conclusion-1">Conclusion<a class="anchor" aria-label="anchor" href="#conclusion-1"></a>
</h3>
<p>CopyKat is a powerful tool for analyzing copy number variations in
single-cell RNA sequencing data. It can help researchers distinguish
tumor cells from normal cells, identify subclones within tumors, and
gain insights into tumor heterogeneity. However, as with all
computational predictions, it’s important to validate key findings with
orthogonal methods when possible.</p>
</div>
</div>
<div class="section level2">
<h2 id="tcrbcr-analysis-screpertoire">7. TCR/BCR Analysis: scRepertoire<a class="anchor" aria-label="anchor" href="#tcrbcr-analysis-screpertoire"></a>
</h2>
<p><a href="https://github.com/ncborcherding/scRepertoire" class="external-link">scRepertoire</a> is
a toolkit designed for single-cell immune profiling analysis,
specifically focusing on T-cell receptor (TCR) and B-cell receptor (BCR)
data. It provides a suite of functions to analyze and visualize immune
repertoire data in conjunction with single-cell RNA sequencing data.</p>
<div class="section level3">
<h3 id="understanding-tcr-and-bcr">Understanding TCR and BCR<a class="anchor" aria-label="anchor" href="#understanding-tcr-and-bcr"></a>
</h3>
<p>T-cell receptors (TCRs) and B-cell receptors (BCRs) are crucial
components of the adaptive immune system:</p>
<ol style="list-style-type: decimal">
<li>TCRs are found on T cells and recognize peptide antigens presented
by MHC molecules.</li>
<li>BCRs are found on B cells and can recognize a wide variety of
antigens directly.</li>
</ol>
<p>These receptors are highly diverse due to a process called V(D)J
recombination, which allows for the generation of millions of unique
receptors. This diversity is key to the immune system’s ability to
recognize a vast array of potential pathogens.</p>
</div>
<div class="section level3">
<h3 id="why-is-tcrbcr-analysis-important">Why is TCR/BCR Analysis Important?<a class="anchor" aria-label="anchor" href="#why-is-tcrbcr-analysis-important"></a>
</h3>
<p>Analyzing TCR and BCR repertoires can provide insights into:</p>
<ol style="list-style-type: decimal">
<li>The specificity and diversity of immune responses</li>
<li>Clonal expansion in response to antigens</li>
<li>Potential targets for immunotherapy</li>
<li>The development of the immune repertoire over time or in different
conditions</li>
</ol>
</div>
<div class="section level3">
<h3 id="introduction-to-screpertoire">Introduction to scRepertoire<a class="anchor" aria-label="anchor" href="#introduction-to-screpertoire"></a>
</h3>
<p><strong>scRepertoire</strong> is an R package that allows for easy
integration of single-cell RNA sequencing data with TCR/BCR repertoire
data. It supports various single-cell clonal formats including 10x
Genomics, AIRR, BD, MiXCR, Omniscope, TRUST4, and WAT3R.</p>
<p>Key features of scRepertoire include:</p>
<ol style="list-style-type: decimal">
<li>Loading and combining TCR/BCR data from various sources</li>
<li>Analyzing clonotype frequencies and distributions</li>
<li>Visualizing repertoire diversity</li>
<li>Integration with popular single-cell analysis frameworks like
Seurat</li>
</ol>
</div>
<div class="section level3">
<h3 id="getting-started-with-screpertoire">Getting Started with scRepertoire<a class="anchor" aria-label="anchor" href="#getting-started-with-screpertoire"></a>
</h3>
<p>To install scRepertoire, you can use the following commands:</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"scRepertoire"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="reading-cellranger-output">Reading CellRanger Output<a class="anchor" aria-label="anchor" href="#reading-cellranger-output"></a>
</h3>
<p>scRepertoire can directly read the output from 10x Genomics Cell
Ranger pipeline. The main file it uses is the
<code>filtered_contig_annotations.csv</code>, which is typically located
in the <code>./outs/</code> directory of the VDJ alignment folder.</p>
<p>Here’s how you might load this data:</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://www.borch.dev/uploads/scRepertoire/" class="external-link">scRepertoire</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Read the filtered_contig_annotations.csv files</span></span>
<span><span class="va">S1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="st">"path/to/Sample1/outs/filtered_contig_annotations.csv"</span><span class="op">)</span></span>
<span><span class="va">S2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="st">"path/to/Sample2/outs/filtered_contig_annotations.csv"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Create a list of the contig data</span></span>
<span><span class="va">contig_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">S1</span>, <span class="va">S2</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Process the contig data</span></span>
<span><span class="va">combined</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://www.borch.dev/uploads/scRepertoire/reference/combineTCR.html" class="external-link">combineTCR</a></span><span class="op">(</span><span class="va">contig_list</span>, samples <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"sample1"</span>, <span class="st">"sample2"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>This <code>combined</code> object now contains detailed information
about the TCR/BCR sequences, including:</p>
<ul>
<li>Clonotype frequencies</li>
<li>V, D, and J gene usage</li>
<li>CDR3 sequences</li>
<li>Chain pairing information</li>
</ul>
</div>
<div class="section level3">
<h3 id="further-resources">Further Resources<a class="anchor" aria-label="anchor" href="#further-resources"></a>
</h3>
<p>scRepertoire offers a wide range of functions for in-depth analysis
of TCR/BCR data. For detailed tutorials and function information, users
are encouraged to visit the <a href="https://www.borch.dev/uploads/screpertoire/" class="external-link">comprehensive
documentation website</a>.</p>
</div>
</div>
<div class="section level2">
<h2 id="conclusion-2">Conclusion<a class="anchor" aria-label="anchor" href="#conclusion-2"></a>
</h2>
<p>This concludes our comprehensive journey through advanced single-cell
RNA sequencing analysis techniques. We’ve explored a wide range of
sophisticated methods, from regulatory network inference with SCENIC to
trajectory analysis with scVelo and Palantir, cell-cell communication
analysis with CellChat and NicheNet, label transfer with singleCellNet,
data imputation with MAGIC, copy number variation analysis with CopyKat,
and TCR/BCR analysis with scRepertoire.</p>
<p>These advanced techniques provide powerful tools to extract deeper
biological insights from your single-cell data. However, it’s important
to remember that each of these methods comes with its own assumptions
and limitations. Always interpret results in the context of your
biological system and research questions, and when possible, validate
key findings with orthogonal experimental approaches.</p>
<p>As the field of single-cell genomics continues to evolve rapidly, new
tools and methods are constantly being developed. We encourage you to
stay curious, keep learning, and explore the latest developments in this
exciting field. Your journey in single-cell analysis is just beginning,
and we hope this course has provided you with a solid foundation to
tackle complex biological questions using these cutting-edge
techniques.</p>
<p>Remember, the key to successful analysis often lies in combining
multiple approaches and integrating different types of data. Don’t
hesitate to revisit earlier lessons as you apply these advanced
techniques to your own research projects. Good luck with your future
single-cell analyses!</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Yichao Hua.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
