<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Lesson 6: Advanced scRNA-seq Analytical Methods (Part 1) • SeuratExtend</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../../bootstrap-toc.css">
<script src="../../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../../pkgdown.css" rel="stylesheet">
<script src="../../pkgdown.js"></script><meta property="og:title" content="Lesson 6: Advanced scRNA-seq Analytical Methods (Part 1)">
<meta property="og:description" content="SeuratExtend">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../../index.html">SeuratExtend</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../../articles/News.html">What's New in v1.1.0</a>
    </li>
    <li class="divider">
    </li>
<li>
      <a href="../../articles/Visualization.html">Enhanced Visualization</a>
    </li>
    <li>
      <a href="../../articles/GSEA.html">Geneset Enrichment Analysis (GSEA) </a>
    </li>
    <li>
      <a href="../../articles/Trajectory.html">Trajectory and Pseudotime Analysis </a>
    </li>
    <li>
      <a href="../../articles/SCENIC.html">SCENIC for Gene Regulatory Networks Analysis </a>
    </li>
    <li>
      <a href="../../articles/Utilities.html">Utility Tools and Functions</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    scRNA-seq Course
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../../articles/single-cell-course/1.basic-R.html">1. Introduction to R</a>
    </li>
    <li>
      <a href="../../articles/single-cell-course/2.Seurat.html">2. Basic Single-Cell Analysis Workflow</a>
    </li>
    <li>
      <a href="../../articles/single-cell-course/3.Visualization.html">3. Advanced Visualization Techniques</a>
    </li>
    <li>
      <a href="../../articles/single-cell-course/4.GSEA.html">4. Gene Set Enrichment Analysis</a>
    </li>
    <li>
      <a href="../../articles/single-cell-course/5.Core-Enhancement.html">5. Core scRNA-seq Workflow Enhancements</a>
    </li>
    <li>
      <a href="../../articles/single-cell-course/6.Advanced.html">6. Advanced Methods (Part 1)</a>
    </li>
    <li>
      <a href="../../articles/single-cell-course/6.Advanced-2.html">7. Advanced Methods (Part 2)</a>
    </li>
  </ul>
</li>
<li>
  <a href="../../articles/FAQ.html">FAQ</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Lesson 6: Advanced scRNA-seq Analytical Methods
(Part 1)</h1>
                        <h4 data-toc-skip class="author">Yichao Hua</h4>
            
            <h4 data-toc-skip class="date">2024-11-29</h4>
      
      
      <div class="hidden name"><code>6.Advanced.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>In this lesson, we’ll explore advanced analytical methods for
single-cell RNA sequencing data. These techniques provide deeper
insights into cellular processes, regulatory networks, and cell-cell
interactions. While some of these operations may be time-consuming and
might not be performed in real-time during the class, we provide
tutorials and explanations for each method to facilitate your
independent exploration after the course.</p>
<p>This lesson will cover the following topics:</p>
<ol style="list-style-type: decimal">
<li>Gene Regulatory Network Analysis: SCENIC</li>
<li>Trajectory Analysis: scVelo and Palantir</li>
<li>Label Transfer: singleCellNet</li>
<li>Cell-Cell Communication: CellChat and NicheNet</li>
<li>Data Imputation: MAGIC</li>
<li>Copy Number Variation Analysis: CopyKat</li>
<li>TCR/BCR Analysis: scRepertoire</li>
</ol>
<p>First, let’s set your working directory to where we’ll store all
course materials. If you followed Lesson 1, this would be:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Set working directory to the course folder created in Lesson 1</span></span>
<span><span class="co"># If you used a different location, replace this path with your chosen directory</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/getwd.html" class="external-link">setwd</a></span><span class="op">(</span><span class="st">"~/Documents/single-cell-course"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="scenic-single-cell-regulatory-network-inference-and-clustering">1. SCENIC: Single-Cell Regulatory Network Inference and
Clustering<a class="anchor" aria-label="anchor" href="#scenic-single-cell-regulatory-network-inference-and-clustering"></a>
</h2>
<p>SCENIC (Single-Cell Regulatory Network Inference and Clustering) is a
powerful computational method that provides insights into the regulatory
networks governing gene expression in single cells. It identifies
transcription factors (TFs), their target genes, and their activity in
individual cells.</p>
<div class="section level3">
<h3 id="background">Background<a class="anchor" aria-label="anchor" href="#background"></a>
</h3>
<p>SCENIC was introduced in a <a href="https://www.nature.com/articles/nmeth.4463" class="external-link">Nature Methods
paper</a> and has become a valuable tool in single-cell analysis. The
method consists of three main steps:</p>
<ol style="list-style-type: decimal">
<li>Identification of co-expression modules using GENIE3 or
GRNBoost</li>
<li>Refinement of these modules based on enrichment of TF binding motifs
(RcisTarget)</li>
<li>Evaluation of the activity of each module in each cell (AUCell)</li>
</ol>
<p>For optimal performance, it’s recommended to use the Nextflow
pipeline to run SCENIC, which can be found <a href="https://github.com/aertslab/SCENICprotocol/" class="external-link">here</a>.</p>
</div>
<div class="section level3">
<h3 id="importing-scenic-results-into-seurat">Importing SCENIC Results into Seurat<a class="anchor" aria-label="anchor" href="#importing-scenic-results-into-seurat"></a>
</h3>
<p>While <code>SeuratExtend</code> doesn’t currently include a
<code>RunScenic</code> function, it does provide tools to import and
analyze SCENIC results within the Seurat framework. Let’s start by
importing a pre-computed SCENIC loom file:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SeuratExtend</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Download the pre-computed SCENIC loom file</span></span>
<span><span class="va">scenic_loom_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">"pyscenic_integrated-output.loom"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/download.file.html" class="external-link">download.file</a></span><span class="op">(</span><span class="st">"https://zenodo.org/records/10944066/files/pbmc3k_small_pyscenic_integrated-output.loom"</span>,</span>
<span>              <span class="va">scenic_loom_path</span>, mode <span class="op">=</span> <span class="st">"wb"</span><span class="op">)</span>  <span class="co"># Use binary mode for Windows compatibility</span></span>
<span></span>
<span><span class="co"># Use the example PBMC dataset from SeuratExtend package</span></span>
<span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu">SeuratExtend</span><span class="fu">::</span><span class="va">pbmc</span></span>
<span></span>
<span><span class="co"># Import SCENIC results into our Seurat object</span></span>
<span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/ImportPyscenicLoom.html">ImportPyscenicLoom</a></span><span class="op">(</span><span class="va">scenic_loom_path</span>, seu <span class="op">=</span> <span class="va">pbmc</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="examining-scenic-outputs">Examining SCENIC Outputs<a class="anchor" aria-label="anchor" href="#examining-scenic-outputs"></a>
</h3>
<p>SCENIC results are stored in <code>seu@misc$SCENIC</code>. Let’s take
a look at some of the key components:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># View the AUCell matrix (TF activity scores)</span></span>
<span><span class="va">tf_auc</span> <span class="op">&lt;-</span> <span class="va">pbmc</span><span class="op">@</span><span class="va">misc</span><span class="op">$</span><span class="va">SCENIC</span><span class="op">$</span><span class="va">RegulonsAUC</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">tf_auc</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                         AHR     ARID3A        ARNT     ARNTL        ATF1</span></span>
<span><span class="co">## CTATAAGATCGTTT-1 0.01406902 0.03347861 0.000000000 0.1144568 0.016730526</span></span>
<span><span class="co">## GTGATTCTGGTTCA-1 0.00000000 0.00000000 0.000000000 0.1730939 0.017191920</span></span>
<span><span class="co">## ACGTTGGACCGTAA-1 0.00000000 0.03189382 0.000000000 0.1463286 0.003671087</span></span>
<span><span class="co">## GGATACTGCAGCTA-1 0.02483910 0.01347068 0.006867406 0.0945589 0.016670344</span></span>
<span><span class="co">## GATTCGGACAGGAG-1 0.00000000 0.03413893 0.000000000 0.1889417 0.008686232</span></span>
<span><span class="co">## GCTCCATGCCGAAT-1 0.02917300 0.01637612 0.069730586 0.1855961 0.022648398</span></span></code></pre>
<p>The AUCell scores represent the enrichment of target genes for each
TF in each cell, indicating regulon activity.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># View the list of TFs and their target genes</span></span>
<span><span class="va">tf_gene_list</span> <span class="op">&lt;-</span> <span class="va">pbmc</span><span class="op">@</span><span class="va">misc</span><span class="op">$</span><span class="va">SCENIC</span><span class="op">$</span><span class="va">Regulons</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">tf_gene_list</span>, list.len <span class="op">=</span> <span class="fl">10</span>, max.level <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## List of 298</span></span>
<span><span class="co">##  $ AHR      : chr [1:49] "NFYC-AS1" "MYSM1" "ZZZ3" "FUBP1" ...</span></span>
<span><span class="co">##  $ ARID3A   : chr [1:24] "CDC7" "IL6R" "LAMC1" "LINC01136" ...</span></span>
<span><span class="co">##  $ ARNT     : chr [1:6] "EPHB3" "SMIM14" "GPR68" "ANXA2" ...</span></span>
<span><span class="co">##  $ ARNTL    : chr [1:9] "SLC4A10" "AP2M1" "EMC2" "PHF20L1" ...</span></span>
<span><span class="co">##  $ ATF1     : chr [1:79] "CAMK2N1" "PAFAH2" "BCAS2" "S100A13" ...</span></span>
<span><span class="co">##  $ ATF2     : chr [1:136] "SLC25A33" "LUZP1" "THRAP3" "FUBP1" ...</span></span>
<span><span class="co">##  $ ATF3     : chr [1:625] "AURKAIP1" "C1orf233" "DFFB" "MTOR" ...</span></span>
<span><span class="co">##  $ ATF4     : chr [1:73] "STX12" "MECR" "MAP7D1" "AKIRIN1" ...</span></span>
<span><span class="co">##  $ ATF6     : chr [1:24] "PER3" "PRKD3" "BMPR2" "ZFYVE28" ...</span></span>
<span><span class="co">##  $ BACH1    : chr [1:32] "SLC1A7" "SNX27" "QPCT" "DYSF" ...</span></span>
<span><span class="co">##   [list output truncated]</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="visualizing-scenic-results">Visualizing SCENIC Results<a class="anchor" aria-label="anchor" href="#visualizing-scenic-results"></a>
</h3>
<p>Now that we have integrated SCENIC data into our Seurat object, we
can use various visualization tools to explore these regulatory
networks.</p>
<div class="section level4">
<h4 id="identifying-top-activated-tfs-in-each-cluster">Identifying Top Activated TFs in Each Cluster<a class="anchor" aria-label="anchor" href="#identifying-top-activated-tfs-in-each-cluster"></a>
</h4>
<p>Let’s create a heatmap to visualize the top activated TFs in each
cluster:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tf_zscore</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/CalcStats.html">CalcStats</a></span><span class="op">(</span><span class="va">tf_auc</span>, f <span class="op">=</span> <span class="va">pbmc</span><span class="op">$</span><span class="va">cluster</span>, order <span class="op">=</span> <span class="st">"p"</span>, n <span class="op">=</span> <span class="fl">4</span>, t <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="../../reference/Heatmap.html">Heatmap</a></span><span class="op">(</span><span class="va">tf_zscore</span>, lab_fill <span class="op">=</span> <span class="st">"zscore"</span><span class="op">)</span></span></code></pre></div>
<p><img src="6.Advanced_files/figure-html/unnamed-chunk-6-1.png" width="576"></p>
<p>This heatmap shows the relative activity (z-score) of top TFs across
different cell clusters. Darker colors indicate higher activity.</p>
</div>
<div class="section level4">
<h4 id="comparing-tf-gene-expression-levels-and-regulon-activity">Comparing TF Gene Expression Levels and Regulon Activity<a class="anchor" aria-label="anchor" href="#comparing-tf-gene-expression-levels-and-regulon-activity"></a>
</h4>
<p>We can compare the expression levels of TFs with their regulon
activity:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../../reference/DimPlot2.html">DimPlot2</a></span><span class="op">(</span></span>
<span>  <span class="va">pbmc</span>,</span>
<span>  features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ETS1"</span>, <span class="st">"ATF3"</span>, <span class="st">"tf_ETS1"</span>, <span class="st">"tf_ATF3"</span><span class="op">)</span>,</span>
<span>  cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">"tf_ETS1"</span> <span class="op">=</span> <span class="st">"D"</span>, <span class="st">"tf_ATF3"</span> <span class="op">=</span> <span class="st">"D"</span><span class="op">)</span>,</span>
<span>  theme <span class="op">=</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SeuratTheme.html" class="external-link">NoAxes</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="6.Advanced_files/figure-html/unnamed-chunk-7-1.png" width="768"></p>
<p>In these plots, “ETS1” and “ATF3” show the gene expression levels,
while “tf_ETS1” and “tf_ATF3” show the regulon activity (AUCell
scores).</p>
</div>
<div class="section level4">
<h4 id="comparing-regulon-activity-between-cell-types">Comparing Regulon Activity Between Cell Types<a class="anchor" aria-label="anchor" href="#comparing-regulon-activity-between-cell-types"></a>
</h4>
<p>Let’s create a waterfall plot to compare regulon activity between two
cell types:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Set the default assay to "TF" for easier access to regulon activity</span></span>
<span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/DefaultAssay.html" class="external-link">DefaultAssay</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"TF"</span></span>
<span></span>
<span><span class="co"># Create a waterfall plot</span></span>
<span><span class="fu"><a href="../../reference/WaterfallPlot.html">WaterfallPlot</a></span><span class="op">(</span></span>
<span>  <span class="va">pbmc</span>,</span>
<span>  features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span>,  <span class="co"># Use all available TFs in the "TF" assay</span></span>
<span>  ident.1 <span class="op">=</span> <span class="st">"Mono CD14"</span>,      <span class="co"># First group of cells</span></span>
<span>  ident.2 <span class="op">=</span> <span class="st">"CD8 T cell"</span>,     <span class="co"># Second group of cells</span></span>
<span>  exp.transform <span class="op">=</span> <span class="cn">FALSE</span>,      <span class="co"># Disable transformation of expression data</span></span>
<span>  top.n <span class="op">=</span> <span class="fl">20</span>                  <span class="co"># Display the top 20 most differentially active TFs</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Attaching package: 'tidyr'</span></span></code></pre>
<pre><code><span><span class="co">## The following object is masked from 'package:reshape2':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     smiths</span></span></code></pre>
<pre><code><span><span class="co">## The following objects are masked from 'package:Matrix':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     expand, pack, unpack</span></span></code></pre>
<p><img src="6.Advanced_files/figure-html/unnamed-chunk-8-1.png" width="768"></p>
<p>This plot shows the differential activity of TFs between monocytes
and CD8 T cells, highlighting key regulators that distinguish these cell
types.</p>
</div>
</div>
<div class="section level3">
<h3 id="advanced-visualization-with-cytoscape">Advanced Visualization with Cytoscape<a class="anchor" aria-label="anchor" href="#advanced-visualization-with-cytoscape"></a>
</h3>
<p>SCENIC outputs can be further visualized using tools like Cytoscape
to create gene regulatory network diagrams. These visualizations can
help in understanding the complex relationships between transcription
factors and their target genes.</p>
<div class="float">
<img src="images/scenic_gene_regulatory_networks.png" alt="Gene Regulatory Networks"><div class="figcaption">Gene Regulatory Networks</div>
</div>
<p>Figure: Gene regulatory networks predicted by SCENIC for Mono CD14
and Mono FCGR3A cell populations, with nodes colored by relative gene
expression (round nodes) or regulon activity (square nodes).</p>
</div>
<div class="section level3">
<h3 id="conclusion">Conclusion<a class="anchor" aria-label="anchor" href="#conclusion"></a>
</h3>
<p>SCENIC provides valuable insights into the regulatory networks
governing gene expression in single cells. By integrating SCENIC results
with Seurat, we can leverage powerful visualization tools to explore
these networks in the context of our single-cell data.</p>
</div>
</div>
<div class="section level2">
<h2 id="trajectory-analysis-scvelo-and-palantir">2. Trajectory Analysis: scVelo and Palantir<a class="anchor" aria-label="anchor" href="#trajectory-analysis-scvelo-and-palantir"></a>
</h2>
<div class="section level3">
<h3 id="introduction-to-trajectory-analysis">Introduction to Trajectory Analysis<a class="anchor" aria-label="anchor" href="#introduction-to-trajectory-analysis"></a>
</h3>
<p>In single-cell RNA sequencing, trajectory analysis is a powerful
approach to understand how cells change over time or during biological
processes like differentiation. Instead of viewing cells as static
entities, trajectory analysis allows us to capture the dynamic nature of
cellular processes, revealing the paths cells take as they transition
between different states.</p>
<p>Trajectory analysis can help answer questions such as: - How do stem
cells differentiate into specialized cell types? - What are the
intermediate states in a biological process? - Which genes drive the
progression from one cell state to another?</p>
<p>In this lesson, we’ll explore two popular tools for trajectory
analysis: <strong>scVelo</strong> and <strong>Palantir</strong>. These
methods use different approaches to infer cellular trajectories and
provide insights into cell fate decisions.</p>
</div>
<div class="section level3">
<h3 id="setting-up-the-python-environment">Setting Up the Python Environment<a class="anchor" aria-label="anchor" href="#setting-up-the-python-environment"></a>
</h3>
<p>Before we begin, it’s important to note that while we’re working in
R, some of the tools we’ll use (scVelo and Palantir) are Python-based.
Normally, this would require switching between R and Python
environments, which can be challenging for beginners. However, the
<code>SeuratExtend</code> package has been designed to seamlessly
integrate these Python tools into our R workflow.</p>
<p>To make this integration possible, we use a tool called
<code>reticulate</code>, which allows R to interact with Python. We also
use <code>conda</code>, a package management system, to create a
specialized Python environment with all the necessary packages.</p>
<p>If you’ve never used <code>conda</code> or <code>reticulate</code>
before, you might need to install miniconda first. You can do this in R
with the following command:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">reticulate</span><span class="fu">::</span><span class="fu"><a href="https://rstudio.github.io/reticulate/reference/install_miniconda.html" class="external-link">install_miniconda</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>Next, we need to create a conda environment with all the required
Python packages.</p>
<p><strong>Note:</strong> Before creating the conda environment, ensure
that Git is installed on your system, as it’s required for installing
certain Python packages:</p>
<ul>
<li>
<strong>Windows</strong>: Download and install from <a href="https://gitforwindows.org/" class="external-link">Git for Windows</a>
</li>
<li>
<strong>Mac</strong>: Install via Homebrew
(<code>brew install git</code>) or Xcode Command Line Tools
(<code>xcode-select --install</code>)</li>
<li>
<strong>Linux</strong>: Install using your package manager:
<ul>
<li>Ubuntu/Debian: <code>sudo apt-get install git</code>
</li>
<li>Fedora: <code>sudo dnf install git</code>
</li>
<li>CentOS/RHEL: <code>sudo yum install git</code>
</li>
</ul>
</li>
</ul>
<p><code>SeuratExtend</code> provides a function to do this
automatically:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">SeuratExtend</span><span class="fu">::</span><span class="fu"><a href="../../reference/create_condaenv_seuratextend.html">create_condaenv_seuratextend</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>This function will set up a conda environment named “seuratextend”
with all the necessary Python packages. You only need to run this once,
and <code>SeuratExtend</code> will automatically use this environment in
future sessions.</p>
<p><strong>Note:</strong> This function automatically detects your
operating system and sets up the environment accordingly. It is
currently supported and tested on Windows, Intel-based macOS (not Apple
Silicon/M1/M2), and Linux (Ubuntu 20.04) systems. Users with Apple
Silicon devices who are interested in contributing to the development of
M1/M2 support are welcome to reach out via GitHub Issues.</p>
</div>
<div class="section level3">
<h3 id="scvelo-rna-velocity-based-trajectory-analysis">scVelo: RNA Velocity-based Trajectory Analysis<a class="anchor" aria-label="anchor" href="#scvelo-rna-velocity-based-trajectory-analysis"></a>
</h3>
<p><a href="https://scvelo.readthedocs.io/en/stable/" class="external-link">scVelo</a> is a
powerful tool for trajectory analysis that leverages RNA velocity
information to predict cell fate and differentiation trajectories. RNA
velocity is based on the ratio of unspliced to spliced mRNA, which can
indicate the future state of cells.</p>
<p>Before running scVelo, we typically need to process our data with a
tool called <strong>Velocyto</strong>, which quantifies spliced and
unspliced mRNA. Velocyto usually needs to be run on a Linux server,
which is beyond the scope of our in-class exercise. For this tutorial,
we’ll use a pre-computed Velocyto output (a .loom file) to demonstrate
the workflow.</p>
<div class="section level4">
<h4 id="preparing-data-for-scvelo">Preparing Data for scVelo<a class="anchor" aria-label="anchor" href="#preparing-data-for-scvelo"></a>
</h4>
<p>First, we need to prepare our data for scVelo analysis. This involves
creating an <code>AnnData</code> object (a Python object used by many
single-cell analysis tools) that contains both the gene expression data
and the splicing information.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SeuratExtend</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Download the example Seurat Object and loom file</span></span>
<span><span class="va">mye_small</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/connections.html" class="external-link">url</a></span><span class="op">(</span><span class="st">"https://zenodo.org/records/10944066/files/pbmc10k_mye_small_velocyto.rds"</span>, <span class="st">"rb"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">loom_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">"pbmc10k_mye_small.loom"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/download.file.html" class="external-link">download.file</a></span><span class="op">(</span><span class="st">"https://zenodo.org/records/10944066/files/pbmc10k_mye_small.loom"</span>, </span>
<span>              <span class="va">loom_path</span>,</span>
<span>              mode <span class="op">=</span> <span class="st">"wb"</span><span class="op">)</span>  <span class="co"># Use binary mode for Windows compatibility</span></span>
<span></span>
<span><span class="co"># Set up the path for saving the AnnData object in the HDF5 (h5ad) format</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="va">.Platform</span><span class="op">$</span><span class="va">OS.type</span> <span class="op">==</span> <span class="st">"windows"</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">adata_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/normalizePath.html" class="external-link">normalizePath</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">"mye_small.h5ad"</span><span class="op">)</span>, winslash <span class="op">=</span> <span class="st">"/"</span><span class="op">)</span></span>
<span><span class="op">}</span> <span class="kw">else</span> <span class="op">{</span></span>
<span>    <span class="va">adata_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">"mye_small.h5ad"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="fu"><a href="../../reference/scVelo.SeuratToAnndata.html">scVelo.SeuratToAnndata</a></span><span class="op">(</span></span>
<span>  <span class="va">mye_small</span>,</span>
<span>  filename <span class="op">=</span> <span class="va">adata_path</span>,</span>
<span>  velocyto.loompath <span class="op">=</span> <span class="va">loom_path</span>,</span>
<span>  prefix <span class="op">=</span> <span class="st">"sample1_"</span>,</span>
<span>  postfix <span class="op">=</span> <span class="st">"-1"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Running scvelo 0.3.0 (python 3.10.12) on 2024-11-29 15:28.</span></span>
<span><span class="co">## Filtered out 10891 genes that are detected 20 counts (shared).</span></span>
<span><span class="co">## Normalized count data: X, spliced, unspliced.</span></span>
<span><span class="co">## Extracted 2000 highly variable genes.</span></span>
<span><span class="co">## Logarithmized X.</span></span>
<span><span class="co">## computing neighbors</span></span>
<span><span class="co">##     finished (0:00:04) --&gt; added </span></span>
<span><span class="co">##     'distances' and 'connectivities', weighted adjacency matrices (adata.obsp)</span></span>
<span><span class="co">## computing moments based on connectivities</span></span>
<span><span class="co">##     finished (0:00:00) --&gt; added </span></span>
<span><span class="co">##     'Ms' and 'Mu', moments of un/spliced abundances (adata.layers)</span></span>
<span><span class="co">## computing velocities</span></span>
<span><span class="co">##     finished (0:00:00) --&gt; added </span></span>
<span><span class="co">##     'velocity', velocity vectors for each individual cell (adata.layers)</span></span>
<span><span class="co">## computing velocity graph (using 1/28 cores)</span></span>
<span><span class="co">## WARNING: Unable to create progress bar. Consider installing `tqdm` as `pip install tqdm` and `ipywidgets` as `pip install ipywidgets`,</span></span>
<span><span class="co">## or disable the progress bar using `show_progress_bar=False`.</span></span>
<span><span class="co">##     finished (0:00:01) --&gt; added </span></span>
<span><span class="co">##     'velocity_graph', sparse matrix with cosine correlations (adata.uns)</span></span></code></pre>
<p>This step combines our Seurat object with the Velocyto output,
creating an AnnData object that scVelo can use.
<code>SeuratExtend</code> handles all the Python interactions behind the
scenes, allowing us to work entirely in R.</p>
</div>
<div class="section level4">
<h4 id="generating-scvelo-plots">Generating scVelo Plots<a class="anchor" aria-label="anchor" href="#generating-scvelo-plots"></a>
</h4>
<p>Now that we have prepared our data, we can generate scVelo plots to
visualize the predicted cell trajectories.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Generate basic scVelo plot</span></span>
<span><span class="fu"><a href="../../reference/scVelo.SeuratToAnndata.html">scVelo.Plot</a></span><span class="op">(</span>color <span class="op">=</span> <span class="st">"cluster"</span>, save <span class="op">=</span> <span class="st">"scvelo_basic.png"</span>, figsize <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">5</span>,<span class="fl">4</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="float">
<img src="figures/scvelo_scvelo_basic.png" alt="Basic scVelo Plot"><div class="figcaption">Basic scVelo Plot</div>
</div>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Generate a more customized scVelo plot</span></span>
<span><span class="fu"><a href="../../reference/scVelo.SeuratToAnndata.html">scVelo.Plot</a></span><span class="op">(</span></span>
<span>  style <span class="op">=</span> <span class="st">"scatter"</span>,</span>
<span>  color <span class="op">=</span> <span class="st">"cluster"</span>,</span>
<span>  groups <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"DC"</span>, <span class="st">"Mono CD14"</span><span class="op">)</span>,</span>
<span>  palette <span class="op">=</span> <span class="fu"><a href="../../reference/color_pro.html">color_pro</a></span><span class="op">(</span><span class="fl">3</span>, <span class="st">"light"</span><span class="op">)</span>,</span>
<span>  xlim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">5</span><span class="op">)</span>, ylim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">10</span><span class="op">)</span>,</span>
<span>  save <span class="op">=</span> <span class="st">"scvelo_custom.png"</span>,</span>
<span>  figsize <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">5</span>,<span class="fl">4</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="float">
<img src="figures/scvelo_scvelo_custom.png" alt="Customized scVelo Plot"><div class="figcaption">Customized scVelo Plot</div>
</div>
<p>These plots show the predicted trajectories of cells based on RNA
velocity. The arrows indicate the likely future state of cells,
providing insights into differentiation paths. In the basic plot, we can
see how different cell clusters are related and the direction of their
differentiation. The customized plot focuses on specific cell types (DC
and Mono CD14), allowing us to examine their relationships more
closely.</p>
</div>
</div>
<div class="section level3">
<h3 id="palantir-pseudotime-and-cell-fate-analysis">Palantir: Pseudotime and Cell Fate Analysis<a class="anchor" aria-label="anchor" href="#palantir-pseudotime-and-cell-fate-analysis"></a>
</h3>
<p><a href="https://github.com/dpeerlab/Palantir" class="external-link">Palantir</a> is
another powerful tool for trajectory analysis that focuses on
determining cell fates and calculating pseudotime.</p>
<div class="section level4">
<h4 id="understanding-pseudotime-and-cell-fate">Understanding Pseudotime and Cell Fate<a class="anchor" aria-label="anchor" href="#understanding-pseudotime-and-cell-fate"></a>
</h4>
<p>Before we dive into the analysis, let’s clarify two key concepts:</p>
<ol style="list-style-type: decimal">
<li><p><strong>Pseudotime</strong>: This is a measure of how far along a
cell is in a biological process, such as differentiation. Unlike real
time, which we can’t directly measure in a single-cell experiment,
pseudotime orders cells based on their gene expression profiles. Cells
with similar gene expression are assumed to be at similar stages in the
process.</p></li>
<li><p><strong>Cell Fate</strong>: This refers to the ultimate cell type
or state that a cell will become. In a differentiation process, for
example, a stem cell might have multiple possible fates (it could become
different types of specialized cells). Palantir helps predict these
potential fates based on the cell’s current state and the overall
structure of the data.</p></li>
</ol>
</div>
<div class="section level4">
<h4 id="running-diffusion-map">Running Diffusion Map<a class="anchor" aria-label="anchor" href="#running-diffusion-map"></a>
</h4>
<p>Palantir starts by calculating a diffusion map, which helps to align
cells along differentiation trajectories. This is a way of representing
the data that captures the underlying structure of the differentiation
process.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mye_small</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/Palantir-Methods.html">Palantir.RunDM</a></span><span class="op">(</span><span class="va">mye_small</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Determing nearest neighbor graph...</span></span></code></pre>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Visualize the multiscale space (ms) embedding</span></span>
<span><span class="fu"><a href="../../reference/DimPlot2.html">DimPlot2</a></span><span class="op">(</span><span class="va">mye_small</span>, reduction <span class="op">=</span> <span class="st">"ms"</span>, group.by <span class="op">=</span> <span class="st">"cluster"</span>, label <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="6.Advanced_files/figure-html/palantir_dm-1.png" width="700"></p>
<p>This plot shows the cells arranged in the multiscale space calculated
by Palantir. Cells that are close together in this space are likely to
be at similar stages in the differentiation process.</p>
</div>
<div class="section level4">
<h4 id="calculating-pseudotime-and-cell-fates">Calculating Pseudotime and Cell Fates<a class="anchor" aria-label="anchor" href="#calculating-pseudotime-and-cell-fates"></a>
</h4>
<p>Next, we’ll use Palantir to calculate pseudotime and predict cell
fates. To do this, we need to specify a starting point - typically, this
would be a cell that we believe to be at the beginning of the process
(e.g., a stem cell).</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Select a start cell (this would typically be done interactively)</span></span>
<span><span class="va">start_cell</span> <span class="op">&lt;-</span> <span class="st">"sample1_GAGAGGTAGCAGTACG-1"</span></span>
<span></span>
<span><span class="co"># Calculate pseudotime</span></span>
<span><span class="va">mye_small</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/Palantir-Methods.html">Palantir.Pseudotime</a></span><span class="op">(</span><span class="va">mye_small</span>, start_cell <span class="op">=</span> <span class="va">start_cell</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Sampling and flocking waypoints...</span></span>
<span><span class="co">## Time for determining waypoints: 0.0011181831359863281 minutes</span></span>
<span><span class="co">## Determining pseudotime...</span></span>
<span><span class="co">## Shortest path distances using 30-nearest neighbor graph...</span></span>
<span><span class="co">## Time for shortest paths: 0.10287365118662516 minutes</span></span>
<span><span class="co">## Iteratively refining the pseudotime...</span></span>
<span><span class="co">## Correlation at iteration 1: 1.0000</span></span>
<span><span class="co">## Entropy and branch probabilities...</span></span>
<span><span class="co">## Markov chain construction...</span></span>
<span><span class="co">## Identification of terminal states...</span></span>
<span><span class="co">## Computing fundamental matrix and absorption probabilities...</span></span>
<span><span class="co">## Project results to all cells...</span></span></code></pre>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ps</span> <span class="op">&lt;-</span> <span class="va">mye_small</span><span class="op">@</span><span class="va">misc</span><span class="op">$</span><span class="va">Palantir</span><span class="op">$</span><span class="va">Pseudotime</span></span>
<span></span>
<span><span class="co"># Visualize pseudotime and cell fates</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">ps</span><span class="op">)</span><span class="op">[</span><span class="fl">3</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"fate1"</span>, <span class="st">"fate2"</span><span class="op">)</span></span>
<span><span class="va">mye_small</span><span class="op">@</span><span class="va">meta.data</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">ps</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">ps</span></span>
<span><span class="fu"><a href="../../reference/DimPlot2.html">DimPlot2</a></span><span class="op">(</span><span class="va">mye_small</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">ps</span><span class="op">)</span>, reduction <span class="op">=</span> <span class="st">"ms"</span>, </span>
<span>         cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>Entropy <span class="op">=</span> <span class="st">"D"</span><span class="op">)</span>, theme <span class="op">=</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SeuratTheme.html" class="external-link">NoAxes</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="6.Advanced_files/figure-html/palantir_pseudotime-1.png" width="768"></p>
<p>These plots show several important pieces of information:</p>
<ul>
<li>
<strong>Pseudotime</strong>: This shows the ordering of cells from
the start to the end of the process.</li>
<li>
<strong>Entropy</strong>: This represents the uncertainty in cell
fate. Higher entropy (lighter color) suggests that a cell has multiple
potential fates.</li>
<li>
<strong>Fate1 and Fate2</strong>: These show the probability of
cells progressing towards specific fates.</li>
</ul>
</div>
</div>
<div class="section level3">
<h3 id="comparing-gene-expression-along-trajectories">Comparing Gene Expression Along Trajectories<a class="anchor" aria-label="anchor" href="#comparing-gene-expression-along-trajectories"></a>
</h3>
<p>To gain more biological insights, we can visualize how gene
expression changes along the calculated trajectories.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Generate gene trend curves</span></span>
<span><span class="fu"><a href="../../reference/GeneTrendCurve-Methods.html">GeneTrendCurve.Palantir</a></span><span class="op">(</span><span class="va">mye_small</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD14"</span>, <span class="st">"FCGR3A"</span><span class="op">)</span>, pseudotime.data <span class="op">=</span> <span class="va">ps</span><span class="op">)</span></span></code></pre></div>
<p><img src="6.Advanced_files/figure-html/gene_trends-1.png" width="672"></p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Generate gene trend heatmap</span></span>
<span><span class="fu"><a href="../../reference/GeneTrendHeatmap-Methods.html">GeneTrendHeatmap.Palantir</a></span><span class="op">(</span></span>
<span>  <span class="va">mye_small</span>, </span>
<span>  features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD14"</span>, <span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/VariableFeatures.html" class="external-link">VariableFeatures</a></span><span class="op">(</span><span class="va">mye_small</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span><span class="op">)</span>, </span>
<span>  pseudotime.data <span class="op">=</span> <span class="va">ps</span>, </span>
<span>  magic <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>  lineage <span class="op">=</span> <span class="st">"fate1"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="6.Advanced_files/figure-html/gene_trends-2.png" width="672"></p>
<p>The gene trend curves show how the expression of specific genes (in
this case, CD14 and FCGR3A) changes over pseudotime. This can help
identify genes that are important at different stages of the
process.</p>
<p>The heatmap provides a broader view, showing how multiple genes
change their expression along the trajectory. This can help identify
groups of genes that are co-regulated or that mark specific stages of
the process.</p>
</div>
<div class="section level3">
<h3 id="conclusion-1">Conclusion<a class="anchor" aria-label="anchor" href="#conclusion-1"></a>
</h3>
<p>Trajectory analysis tools like scVelo and Palantir provide powerful
means to understand the dynamics of cell differentiation in single-cell
RNA-seq data. By visualizing cell trajectories, calculating pseudotime,
and examining gene expression trends, we can gain deep insights into the
processes of cellular development and lineage commitment.</p>
<p>These methods allow us to go beyond static snapshots of cellular
states and start to understand the dynamic processes that drive
biological phenomena. However, it’s important to remember that these are
computational predictions based on the data we provide. The biological
interpretation of these results should always be validated with
additional experiments when possible.</p>
</div>
</div>
<div class="section level2">
<h2 id="cell-cell-communication-cellchat-and-nichenet">3. Cell-Cell Communication: CellChat and NicheNet<a class="anchor" aria-label="anchor" href="#cell-cell-communication-cellchat-and-nichenet"></a>
</h2>
<div class="section level3">
<h3 id="introduction-to-cell-cell-communication-analysis">Introduction to Cell-Cell Communication Analysis<a class="anchor" aria-label="anchor" href="#introduction-to-cell-cell-communication-analysis"></a>
</h3>
<p>In complex biological systems, cells don’t function in isolation.
They constantly communicate with each other through various signaling
molecules. Understanding these communication patterns is crucial for
deciphering how different cell types work together in tissues and
organs. In single-cell RNA sequencing data, we can infer these
communication patterns by analyzing the expression of ligands (signaling
molecules) and their corresponding receptors in different cell
types.</p>
<p>In this section, we’ll explore two popular tools for analyzing
cell-cell communication: <strong>CellChat</strong> and
<strong>NicheNet</strong>. These tools use different approaches to infer
and visualize cell-cell communication networks from single-cell RNA
sequencing data.</p>
</div>
<div class="section level3">
<h3 id="cellchat-quantitative-inference-of-cell-cell-communication">CellChat: Quantitative Inference of Cell-Cell Communication<a class="anchor" aria-label="anchor" href="#cellchat-quantitative-inference-of-cell-cell-communication"></a>
</h3>
<p><a href="https://github.com/sqjin/CellChat" class="external-link">CellChat</a> is an
R-based tool for inferring, visualizing and analyzing cell-cell
communication networks from single-cell RNA sequencing data. It provides
a systematic and quantitative approach to predict major signaling inputs
and outputs for cells and how these signals are coordinated across cell
types.</p>
<div class="section level4">
<h4 id="installing-and-loading-cellchat">Installing and Loading CellChat<a class="anchor" aria-label="anchor" href="#installing-and-loading-cellchat"></a>
</h4>
<p>First, let’s install and load the CellChat package:</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st">"CellChat"</span><span class="op">)</span><span class="op">)</span> <span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"sqjin/CellChat"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">CellChat</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com" class="external-link">patchwork</a></span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="preparing-data-for-cellchat">Preparing Data for CellChat<a class="anchor" aria-label="anchor" href="#preparing-data-for-cellchat"></a>
</h4>
<p>CellChat works with a Seurat object, so we can use our existing
<code>pbmc</code> object. We need to extract the necessary data and
create a CellChat object:</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data.input</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/AssayData.html" class="external-link">GetAssayData</a></span><span class="op">(</span><span class="va">pbmc</span>, assay <span class="op">=</span> <span class="st">"RNA"</span>, slot <span class="op">=</span> <span class="st">"data"</span><span class="op">)</span> <span class="co"># normalized data matrix</span></span>
<span><span class="va">labels</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/Idents.html" class="external-link">Idents</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span></span>
<span><span class="va">meta</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>group <span class="op">=</span> <span class="va">labels</span>, row.names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">labels</span><span class="op">)</span><span class="op">)</span> <span class="co"># create a dataframe of cell labels</span></span>
<span></span>
<span><span class="va">cellchat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/CellChat/man/createCellChat.html" class="external-link">createCellChat</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">data.input</span>, meta <span class="op">=</span> <span class="va">meta</span>, group.by <span class="op">=</span> <span class="st">"group"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "Create a CellChat object from a data matrix"</span></span>
<span><span class="co">## Set cell identities for the new CellChat object </span></span>
<span><span class="co">## The cell groups used for CellChat analysis are  B cell, CD4 T Memory, CD4 T Naive, CD8 T cell, DC, Mono CD14, Mono FCGR3A, NK cell, Platelet</span></span></code></pre>
</div>
<div class="section level4">
<h4 id="running-cellchat-analysis">Running CellChat Analysis<a class="anchor" aria-label="anchor" href="#running-cellchat-analysis"></a>
</h4>
<p>Now we can run the CellChat analysis:</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cellchat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/CellChat/man/addMeta.html" class="external-link">addMeta</a></span><span class="op">(</span><span class="va">cellchat</span>, meta <span class="op">=</span> <span class="va">meta</span><span class="op">)</span></span>
<span><span class="va">cellchat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/CellChat/man/setIdent.html" class="external-link">setIdent</a></span><span class="op">(</span><span class="va">cellchat</span>, ident.use <span class="op">=</span> <span class="st">"group"</span><span class="op">)</span></span>
<span><span class="va">cellchat</span><span class="op">@</span><span class="va">DB</span> <span class="op">&lt;-</span> <span class="va">CellChatDB.human</span>  <span class="co"># use human database</span></span>
<span><span class="va">cellchat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/CellChat/man/subsetData.html" class="external-link">subsetData</a></span><span class="op">(</span><span class="va">cellchat</span><span class="op">)</span>  <span class="co"># subset the expression data of signaling genes for saving computation cost</span></span>
<span><span class="va">cellchat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/CellChat/man/identifyOverExpressedGenes.html" class="external-link">identifyOverExpressedGenes</a></span><span class="op">(</span><span class="va">cellchat</span><span class="op">)</span></span>
<span><span class="va">cellchat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/CellChat/man/identifyOverExpressedInteractions.html" class="external-link">identifyOverExpressedInteractions</a></span><span class="op">(</span><span class="va">cellchat</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## The number of highly variable ligand-receptor pairs used for signaling inference is 249</span></span></code></pre>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cellchat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/CellChat/man/projectData.html" class="external-link">projectData</a></span><span class="op">(</span><span class="va">cellchat</span>, <span class="va">PPI.human</span><span class="op">)</span></span>
<span><span class="va">cellchat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/CellChat/man/computeCommunProb.html" class="external-link">computeCommunProb</a></span><span class="op">(</span><span class="va">cellchat</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## triMean is used for calculating the average gene expression per cell group. </span></span>
<span><span class="co">## [1] "&gt;&gt;&gt; Run CellChat on sc/snRNA-seq data &lt;&lt;&lt; [2024-11-29 15:29:12.414443]"</span></span>
<span><span class="co">## [1] "&gt;&gt;&gt; CellChat inference is done. Parameter values are stored in `object@options$parameter` &lt;&lt;&lt; [2024-11-29 15:30:03.118037]"</span></span></code></pre>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cellchat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/CellChat/man/filterCommunication.html" class="external-link">filterCommunication</a></span><span class="op">(</span><span class="va">cellchat</span>, min.cells <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span>
<span><span class="va">cellchat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/CellChat/man/computeCommunProbPathway.html" class="external-link">computeCommunProbPathway</a></span><span class="op">(</span><span class="va">cellchat</span><span class="op">)</span></span>
<span><span class="va">cellchat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/CellChat/man/aggregateNet.html" class="external-link">aggregateNet</a></span><span class="op">(</span><span class="va">cellchat</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="visualizing-cellchat-results">Visualizing CellChat Results<a class="anchor" aria-label="anchor" href="#visualizing-cellchat-results"></a>
</h4>
<p>CellChat provides various functions to visualize the inferred
cell-cell communication networks:</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Visualize the overall communication network</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/CellChat/man/netVisual_circle.html" class="external-link">netVisual_circle</a></span><span class="op">(</span><span class="va">cellchat</span><span class="op">@</span><span class="va">net</span><span class="op">$</span><span class="va">count</span>, vertex.weight <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">cellchat</span><span class="op">@</span><span class="va">idents</span><span class="op">)</span>, weight.scale <span class="op">=</span> <span class="cn">T</span>, label.edge<span class="op">=</span> <span class="cn">F</span>, title.name <span class="op">=</span> <span class="st">"Number of interactions"</span><span class="op">)</span></span></code></pre></div>
<p><img src="6.Advanced_files/figure-html/visualize_cellchat-1.png" width="768"></p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Heatmap of communication probability</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/CellChat/man/netVisual_heatmap.html" class="external-link">netVisual_heatmap</a></span><span class="op">(</span><span class="va">cellchat</span><span class="op">)</span></span></code></pre></div>
<p><img src="6.Advanced_files/figure-html/visualize_cellchat-2.png" width="768"></p>
<p>These visualizations help us understand which cell types are
communicating with each other and through which signaling pathways.</p>
</div>
</div>
<div class="section level3">
<h3 id="nichenet-ligand-target-prediction-in-cell-cell-communication">NicheNet: Ligand-Target Prediction in Cell-Cell Communication<a class="anchor" aria-label="anchor" href="#nichenet-ligand-target-prediction-in-cell-cell-communication"></a>
</h3>
<p><a href="https://github.com/saeyslab/nichenetr" class="external-link">NicheNet</a> is a
powerful tool for studying cell-cell communication in single-cell RNA
sequencing data. Unlike other methods that focus solely on
ligand-receptor interactions, NicheNet goes a step further by predicting
which genes in the receiver cells are targeted by the ligands expressed
by sender cells.</p>
<p>NicheNet’s main strengths are: 1. It predicts which ligands from
sender cells are most likely to influence gene expression in receiver
cells. 2. It infers which genes in the receiver cells are affected by
these ligands. 3. It provides context-specific predictions by
integrating prior knowledge with your specific dataset.</p>
<div class="section level4">
<h4 id="installing-and-loading-nichenet">Installing and Loading NicheNet<a class="anchor" aria-label="anchor" href="#installing-and-loading-nichenet"></a>
</h4>
<p>Let’s install and load the NicheNet package:</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://devtools.r-lib.org/" class="external-link">"devtools"</a></span><span class="op">)</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"devtools"</span><span class="op">)</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://github.com/saeyslab/nichenetr" class="external-link">"nichenetr"</a></span><span class="op">)</span><span class="op">)</span> <span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"saeyslab/nichenetr"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="preparing-for-nichenet-analysis">Preparing for NicheNet Analysis<a class="anchor" aria-label="anchor" href="#preparing-for-nichenet-analysis"></a>
</h4>
<p>For this example, we’ll use a dataset from a study on T cell
responses in lymph nodes before and after viral infection. We’ll focus
on how other immune cells might be influencing CD8 T cells after
infection.</p>
<p>First, let’s load the necessary libraries and data:</p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/saeyslab/nichenetr" class="external-link">nichenetr</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Load pre-processed Seurat object</span></span>
<span><span class="va">seuratObj</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/connections.html" class="external-link">url</a></span><span class="op">(</span><span class="st">"https://zenodo.org/record/3531889/files/seuratObj.rds"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">seuratObj</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/UpdateSeuratObject.html" class="external-link">UpdateSeuratObject</a></span><span class="op">(</span><span class="va">seuratObj</span><span class="op">)</span></span>
<span></span>
<span><span class="va">organism</span> <span class="op">&lt;-</span> <span class="st">"mouse"</span></span>
<span></span>
<span><span class="kw">if</span><span class="op">(</span><span class="va">organism</span> <span class="op">==</span> <span class="st">"human"</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="va">lr_network</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/connections.html" class="external-link">url</a></span><span class="op">(</span><span class="st">"https://zenodo.org/record/7074291/files/lr_network_human_21122021.rds"</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="va">ligand_target_matrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/connections.html" class="external-link">url</a></span><span class="op">(</span><span class="st">"https://zenodo.org/record/7074291/files/ligand_target_matrix_nsga2r_final.rds"</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="va">weighted_networks</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/connections.html" class="external-link">url</a></span><span class="op">(</span><span class="st">"https://zenodo.org/record/7074291/files/weighted_networks_nsga2r_final.rds"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span> <span class="kw">else</span> <span class="kw">if</span><span class="op">(</span><span class="va">organism</span> <span class="op">==</span> <span class="st">"mouse"</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="va">lr_network</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/connections.html" class="external-link">url</a></span><span class="op">(</span><span class="st">"https://zenodo.org/record/7074291/files/lr_network_mouse_21122021.rds"</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="va">ligand_target_matrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/connections.html" class="external-link">url</a></span><span class="op">(</span><span class="st">"https://zenodo.org/record/7074291/files/ligand_target_matrix_nsga2r_final_mouse.rds"</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="va">weighted_networks</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/connections.html" class="external-link">url</a></span><span class="op">(</span><span class="st">"https://zenodo.org/record/7074291/files/weighted_networks_nsga2r_final_mouse.rds"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="running-nichenet-analysis">Running NicheNet Analysis<a class="anchor" aria-label="anchor" href="#running-nichenet-analysis"></a>
</h4>
<p>Now, let’s perform the NicheNet analysis:</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Define sender and receiver cells</span></span>
<span><span class="va">sender_celltypes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD4 T"</span>, <span class="st">"Treg"</span>, <span class="st">"Mono"</span>, <span class="st">"NK"</span>, <span class="st">"B"</span>, <span class="st">"DC"</span><span class="op">)</span></span>
<span><span class="va">receiver_celltype</span> <span class="op">&lt;-</span> <span class="st">"CD8 T"</span></span>
<span></span>
<span><span class="co"># Run NicheNet analysis</span></span>
<span><span class="va">nichenet_output</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/nichenetr/man/nichenet_seuratobj_aggregate.html" class="external-link">nichenet_seuratobj_aggregate</a></span><span class="op">(</span></span>
<span>  seurat_obj <span class="op">=</span> <span class="va">seuratObj</span>, </span>
<span>  sender <span class="op">=</span> <span class="va">sender_celltypes</span>,</span>
<span>  receiver <span class="op">=</span> <span class="va">receiver_celltype</span>, </span>
<span>  condition_colname <span class="op">=</span> <span class="st">"aggregate"</span>,</span>
<span>  condition_oi <span class="op">=</span> <span class="st">"LCMV"</span>,</span>
<span>  condition_reference <span class="op">=</span> <span class="st">"SS"</span>,</span>
<span>  ligand_target_matrix <span class="op">=</span> <span class="va">ligand_target_matrix</span>,</span>
<span>  lr_network <span class="op">=</span> <span class="va">lr_network</span>,</span>
<span>  weighted_networks <span class="op">=</span> <span class="va">weighted_networks</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "The RNA assay will be used for the analysis."</span></span>
<span><span class="co">## [1] "Read in and process NicheNet's networks"</span></span>
<span><span class="co">## [1] "Define expressed ligands and receptors in receiver and sender cells"</span></span>
<span><span class="co">## [1] "Perform DE analysis in receiver cell"</span></span>
<span><span class="co">## [1] "Perform NicheNet ligand activity analysis"</span></span>
<span><span class="co">## [1] "Infer active target genes of the prioritized ligands"</span></span>
<span><span class="co">## [1] "Infer receptors of the prioritized ligands"</span></span>
<span><span class="co">## [1] "Perform DE analysis in sender cells"</span></span></code></pre>
</div>
<div class="section level4">
<h4 id="interpreting-nichenet-results">Interpreting NicheNet Results<a class="anchor" aria-label="anchor" href="#interpreting-nichenet-results"></a>
</h4>
<p>Let’s look at the key outputs from our NicheNet analysis:</p>
<ol style="list-style-type: decimal">
<li>Top predicted ligands:</li>
</ol>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">nichenet_output</span><span class="op">$</span><span class="va">top_ligands</span>, <span class="fl">10</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] "Ebi3"   "Ptprc"  "H2-M3"  "H2-M2"  "H2-T10" "H2-T22" "H2-T23" "H2-K1" </span></span>
<span><span class="co">##  [9] "H2-Q4"  "H2-Q6"</span></span></code></pre>
<p>These are the ligands that NicheNet predicts are most likely to be
influencing gene expression in CD8 T cells after viral infection.</p>
<ol start="2" style="list-style-type: decimal">
<li>Ligand activity and target genes:</li>
</ol>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">nichenet_output</span><span class="op">$</span><span class="va">ligand_activity_target_heatmap</span></span></code></pre></div>
<p><img src="6.Advanced_files/figure-html/ligand_activity-1.png" width="1440"></p>
<p>This heatmap is a key summary of the NicheNet analysis. It shows: -
The top-ranked ligands (rows) - Their predicted target genes in CD8 T
cells (columns) - The ligand activity scores (left heatmap) - The
expression of ligands in sender cells (middle heatmap) - The regulatory
potential scores between ligands and target genes (right heatmap)</p>
<ol start="3" style="list-style-type: decimal">
<li>Ligand expression in sender cells:</li>
</ol>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">nichenet_output</span><span class="op">$</span><span class="va">ligand_expression_dotplot</span></span></code></pre></div>
<p><img src="6.Advanced_files/figure-html/ligand_expression-1.png" width="960"></p>
<p>This plot shows which sender cell types are expressing the top
predicted ligands.</p>
</div>
<div class="section level4">
<h4 id="conclusion-2">Conclusion<a class="anchor" aria-label="anchor" href="#conclusion-2"></a>
</h4>
<p>NicheNet provides a powerful way to generate hypotheses about
cell-cell communication in your single-cell data. By integrating prior
knowledge about ligand-target interactions with your specific dataset,
it can predict which ligands from sender cells are most likely
influencing gene expression changes in receiver cells.</p>
<p>Key points to remember when interpreting NicheNet results:</p>
<ol style="list-style-type: decimal">
<li>The top-ranked ligands are predictions and should be validated
experimentally.</li>
<li>Consider both the ligand activity scores and the expression of
ligands in sender cells.</li>
<li>Look for biological coherence in the predicted target genes.</li>
<li>These predictions can guide further experiments to understand the
mechanisms of cell-cell communication in your biological system.</li>
</ol>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Yichao Hua.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
